// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/00f48457337a3cf416df5c18c5435ac4.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../node_modules/@slyte/core/index.js";
import { Component,LyteUiComponentComponentRegistry } from "../../component.js";

/**
 * Renders a tour hint component
 * @component lyte-tour-hint
 * @version 3.1.0
 */


class LyteTourHintComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({

			/**
			 * @componentProperty {boolean} ltPropBackgroundAnimation
			 * @default false
			 * 
			 */

			'ltPropBackgroundAnimation' : prop('boolean' , { default : false }),
			'ltPropTargetClass': prop("string", { default: 'lyteTourTarget' }),
			'ltPropBindToBody' : prop('boolean' , { default : false }),
			'tourStepIndex'			: prop('number' , { default : 0 }),
			'ltPropArrowPosition'	: prop('string' , { default : 'start' }),
			'ltPropAppendBackground' : prop('boolean' , { default : false }),
			'ltPropLabel'	: prop('string' , { default : '' })
		}), arg1);
	}

    init() {

	}

    static methods(arg1) {
        return Object.assign(super.methods({
            beforeWormholeAppend : function(arg){
                if(this.childComp){
                    delete this.childComp;
                }
                if(this.actualTourDiv){
                    delete this.actualTourDiv;
                }
                this.childComp = arg;
                this.actualTourDiv = this.childComp.querySelector(".lyteTourWrap");
            }
        }), arg1);
    }

    didDestroy() {
		if(this.childComp){
			this.childComp.remove();
			delete this.childComp;
		}
		if(this.actualTourDiv){
			delete this.actualTourDiv;
		}
	}

    _() {
        _;
    }
}

LyteTourHintComponent._template = "<template tag-name=\"lyte-tour-hint\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropBindToBody}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-wormhole case=\"true\" on-before-append=\"{{method(&quot;beforeWormholeAppend&quot;)}}\"> <template is=\"registerYield\" yield-name=\"lyte-content\"> <div class=\"lyteTourTargetBackground {{ltPropTargetClass}}\"></div> </template> </lyte-wormhole> </template></template></template>";;
LyteTourHintComponent._dynamicNodes = [{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"r","p":[1,1],"dN":[{"t":"a","p":[1]}],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;

LyteTourHintComponent._observedAttributes = [
    "ltPropBackgroundAnimation",
    "ltPropTargetClass",
    "ltPropBindToBody",
    "tourStepIndex",
    "ltPropArrowPosition",
    "ltPropAppendBackground",
    "ltPropLabel"
];

export { LyteTourHintComponent };
LyteTourHintComponent.register("lyte-tour-hint", {
    hash: "LyteTourHintComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
