// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/05a624b0da9e781763942b3f25159b68.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component, LyteUiComponentComponentRegistry } from "../component.js";
import $L from "../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";

/**
 * Renders a rating component
 * @component lyte-rating
 * @version 3.1.0
 * @methods onRender,onClick,onHover,onOut
 */

class LyteRatingComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({

			/**
             * @componentProperty {number} ltPropValue
             * @version 3.1.0
             * @default 0
			 * @minValue 0
             */
			"ltPropValue" : prop("number",{"default":0}),

			/**
             * @componentProperty {number} ltPropCount
             * @version 3.1.0
             * @default 5
			 * @condition ltPropType svg,class
             */
			"ltPropCount" : prop("number",{"default":5}),

			/**
             * @componentProperty {string} ltPropWrapperClass
             * @version 3.1.0
             */
			"ltPropWrapperClass" : prop("string"),

			/**
             * @componentProperty {svg|multiple|switch|toggle|class} ltPropType
             * @version 3.1.0
             * @default svg
             */
			"ltPropType" : prop("string",{"default":"svg"}),	//svg,multiple,switch,toggle,class

			/**
             * @componentProperty {star|heart|lineStar|lineHeart|smiley|likeDislike|custom} ltPropAppearance
             * @version 3.1.0
			 * @condition ltPropType svg,toggle,multiple,switch
             */
			"ltPropAppearance" : prop("string"),				//star,heart,lineStar,lineHeart,smiley,likeDislike,custom

			/**
             * @componentProperty {boolean} ltPropReadOnly
             * @version 3.1.0
             * @default false
             * 
             */
			"ltPropReadOnly" : prop("boolean",{"default":false}),

			/**
             * @componentProperty {boolean} ltPropClearable
             * @version 3.1.0
             * @default false
             * 
             */
			"ltPropClearable" : prop("boolean",{"default":false}),

			/**
             * @componentProperty {boolean} ltPropHalfIncrement
             * @version 3.1.0
             * @default false
             * @condition ltPropType class,svg
			 * @condition ltPropAppearance star,heart,custom
             */
			"ltPropHalfIncrement" : prop("boolean",{"default":false}),

			/**
             * @componentProperty {string} ltPropEmptyIcon
             * @version 3.1.0
             * @default lyteRatingEmpty
			 * @condition ltPropType class,multiple,toggle,switch 
			 * @condition tPropAppearance custom,class ,class 
             */
			"ltPropEmptyIcon" : prop("string",{"default":"lyteRatingHeartEmptyIcon"}),

			/**
             * @componentProperty {string} ltPropFullIcon
             * @version 3.1.0
             * @default lyteRatingFilled
			 * @condition ltPropType class,multiple,toggle,switch 
			 * @condition tPropAppearance custom,class  
             */
			"ltPropFullIcon" : prop("string",{"default":"lyteRatingHeartRatedIcon"}),

			/**
             * @componentProperty {string} ltPropHalfIcon
             * @version 3.1.0
             * @default lyteRatingHalf
			 * @condition ltPropType class
             */
			"ltPropHalfIcon" : prop("string",{"default":"lyteRatingHeartRatedHalfIcon"}),

			/**
             * @componentProperty {string} ltPropHoverFullIcon
             * @version 3.1.0
             * @default lyteRatingHoverFilled
			 * @condition ltPropType class,multiple,toggle,switch
			 * @condition tPropAppearance custom,class 
             */
			"ltPropHoverFullIcon" : prop("string",{"default":"lyteRatingHeartHoverRatedIcon"}),

			/**
             * @componentProperty {string} ltPropHoverHalfIcon
             * @version 3.1.0
             * @default lyteRatingHoverHalf
			 * @condition ltPropType class
             */
			"ltPropHoverHalfIcon" : prop("string",{"default":"lyteRatingHeartHoverHalfIcon"}),

			/**
             * @componentProperty {boolean} ltPropYield
             * @version 3.1.0
             * @default false
             */
			"ltPropYield" : prop("boolean",{"default":false}),

			/**
             * @componentProperty {colorString} ltPropEmptyColor
             * @version 3.1.0
             */
			"ltPropEmptyColor" : prop("string"),

			/**
             * @componentProperty {colorString} ltPropFillColor
             * @version 3.1.0
             */
			"ltPropFillColor" : prop("string"),

			/**
             * @componentProperty {colorString} ltPropHoverColor
             * @version 3.1.0
             */
			"ltPropHoverColor" : prop("string"),

			/**
             * @componentProperty {string} ltPropWidth
             * @version 3.1.0
             * @default 22px
             * @suffix px,pt,cm,mm,vh,vm,em
			 * @condition ltPropType svg,class
			 * @condition ltPropAppearance star,heart,lineStar,lineHeart,custom,class
             */
			"ltPropWidth" : prop("string",{"default":"22px"}),

			/**
             * @componentProperty {string} ltPropHeight
             * @version 3.1.0
             * @default 22px
             * @suffix px,pt,cm,mm,vh,vm,em
			 * @condition ltPropType svg,class
			 * @condition ltPropAppearance star,heart,lineStar,lineHeart,custom,class
             */
			"ltPropHeight" : prop("string",{"default":"22px"}),

			/**
             * @componentProperty {number} ltPropPrecision
             * @version 3.1.0
             * @default 0.5
             */
			"ltPropPrecision" : prop("number",{"default":0.5}),
			/**
             * @componentProperty {string} ltPropStroke
             * @version 3.1.0
             */
			"ltPropStroke" : prop("string"),
			"countArray" : prop("array",{"default":[]}),
			"currentHover" : prop("number",{"default":0}),
			"emptyIconArray" : prop("array",{"default":[]}),
			"fullIconArray" : prop("array",{"default":[]}),
			"halfIconArray" : prop("array",{"default":[]}),
			"hoverFullIconArray" : prop("array",{"default":[]}),
			"hoverHalfIconArray" : prop("array",{"default":[]}),
			"starVB" : prop("string",{"default":"6.5 4.5 21 21"}),
			"lineStarVB" : prop("string",{"default":"5.5 2.5 23 23"}),
			"heartVB" : prop("string",{"default":"1.5 0.5 20 20"}),
			"lineHeartVB" : prop("string",{"default":"1 0 21 21"}),
			"ratingNum" : prop("number",{"default":0}),
			"ltPropPrevWrapper" : prop("string")
		}), arg1);		
	}

    keydown(event) {
		var focusedItem = document.activeElement;
		var isHalfIncr = this.getData('ltPropHalfIncrement');	
		var count = this.getData('ltPropCount');	
		if(event.keyCode === 39){
			var lrc = parseInt(focusedItem.dataset.lrc) + 1;
			if(isHalfIncr){
				
				var value = this.getData('ltPropValue');
				this.setData('ltPropValue',(value + 0.5) > count ? count : (value + 0.5));
				this.$node.setAttribute('aria-valuenow',this.getData('ltPropValue') );
				if(lrc !== (value + 1)){
					return;
				}
			}else{
				this.setData('ltPropValue',lrc > count ? count : lrc);
				this.$node.setAttribute('aria-valuenow',this.getData('ltPropValue'));
			}
			var nxt_elem = focusedItem.parentNode.querySelector("[data-lrc = '"+ lrc  +"']");
			if(nxt_elem){
				nxt_elem.focus();
			}
			
		}else if(event.keyCode === 37){
			var lrc = parseInt(focusedItem.dataset.lrc) - 1;
			if(isHalfIncr){
				var value = this.getData('ltPropValue');
				this.setData('ltPropValue',(value - 0.5) > count ? count : (value - 0.5) );
				this.$node.setAttribute('aria-valuenow',this.getData('ltPropValue') );
				if(lrc !== (value - 0.5)){
					return;
				}
			}else{
				this.setData('ltPropValue',lrc > count ? count : lrc );
				this.$node.setAttribute('aria-valuenow',this.getData('ltPropValue'));
			}
			var prev_elem = focusedItem.parentNode.querySelector("[data-lrc = '"+ lrc +"']");
			if(prev_elem ){
				prev_elem.focus();
			}
			
		}
	}

    setAriaAttributes() {
		this.$node.setAttribute('role','slider');
		this.$node.setAttribute('aria-valuenow',this.getData('ltPropValue'));
		this.$node.setAttribute('aria-valuemin',0);
		this.$node.setAttribute('aria-valuemax',this.getData('ltPropCount'));
		var rating = this.$node.querySelectorAll('lyte-rate-icon');
		rating[0].setAttribute('tabindex',0);
		var keydownFn = function (event) { this.keydown(event); }.bind(this);
		var focusin = function(event){
			_lyteUiUtils.addGlobalEventListener('keydown',keydownFn, true);
			var value = self.getData('ltPropValue');
			if(event.target.dataset.lrc == 1 && value == 0){
				value = value + 1;
				if(isHalfIncr){
					value -= 0.5;
				}
				self.setData('ltPropValue',value );
				self.$node.setAttribute('aria-valuenow',self.getData('ltPropValue'));
			}
		};
		var focusout = function(event){
			_lyteUiUtils.removeEventListenerGlobal('keydown',keydownFn, true);
		};
		var self = this;
		var isHalfIncr = this.getData('ltPropHalfIncrement');

		for(var index = 0 ; index < rating.length ; index++ ){
			if(index === 0){
				rating[index].setAttribute('tabindex', 0);
			}else{
				rating[index].setAttribute('tabindex', -1);
			}
			
			rating[index].addEventListener('focusin',focusin);
			rating[index].addEventListener('focusout',focusout)
		}
	}

    setCountArray() {
		var countArray = [];
		for(var i = 1; i <= this.getData('ltPropCount'); i++){
			var obj = {};
			obj.ind = i;
			if(this.getData('ltPropType') == "svg"){
				if(this.getData('ltPropAppearance') === "star"){
					obj.appearance = "#star";
					obj.viewbox = this.getData('starVB');
				}
				if(this.getData('ltPropAppearance') === "lineStar"){
					obj.appearance = "#star";
					obj.viewbox = this.getData('lineStarVB');
				}
				if(this.getData('ltPropAppearance') === "heart"){
					obj.appearance = "#heart";
					obj.viewbox = this.getData('heartVB');
				}
				if(this.getData('ltPropAppearance') === "lineHeart"){
					obj.appearance = "#heart";
					obj.viewbox = this.getData('lineHeartVB');
				}
			}
			countArray.push(obj);
		}
		this.setData('countArray',countArray);
	}

    getFillColor() {
		return "var(--lyte-rating-fill-color)";
	}

    getEmptyColor() {
		return "var(--lyte-rating-empty-color)";
	}

    getHoverColor() {
		return "var(--lyte-rating-hover-color)";
	}

    setRating(stopMethod) {

		this.clearHoverClasses();
		var currentRating = this.getData('ltPropValue'),
			ratedNodes = this.$node.querySelectorAll('.lyteRated'),
			nodes = this.$node.querySelectorAll('lyte-rate-icon');
		if(currentRating > nodes.length){
			currentRating = nodes.length;
		}
		if(this.getData('ltPropType') == "svg"){
			if(this.getData('ltPropHalfIncrement')){
				var stops;
				for(var i = 0; i < Math.floor(currentRating); i++){
					stops = nodes[i].querySelectorAll('linearGradient stop');
					stops[1].setAttribute('offset','100%');
					stops[2].setAttribute('offset','0%');
					stops[0].style.stopColor = this.getFillColor();
					stops[1].style.stopColor = this.getFillColor();
					nodes[i].classList.add(this.getData('ltPropFullIcon'),"lyteRated");
				}
				for(var i = Math.floor(currentRating); i < nodes.length; i++){
					stops = nodes[i].querySelectorAll('linearGradient stop');
					stops[1].setAttribute('offset','0%');
					stops[2].setAttribute('offset','0%');
					stops[0].style.stopColor = this.getEmptyColor();
					stops[1].style.stopColor = this.getEmptyColor();
					stops[2].style.stopColor = this.getEmptyColor();
					nodes[i].classList.remove(this.getData('ltPropFullIcon'),this.getData('ltPropHalfIcon'),"lyteRated");
				}
				if(currentRating > 0){
					stops = nodes[Math.ceil(currentRating)-1].querySelectorAll('linearGradient stop');
					stops[1].setAttribute('offset',(currentRating % 1 != 0 ? parseInt((currentRating - Math.floor(currentRating))*100) : 100 )+'%');
					stops[2].setAttribute('offset',parseInt((currentRating - Math.floor(currentRating) + 0.01)*100)+'%');
					stops[0].style.stopColor = this.getFillColor();
					stops[1].style.stopColor = this.getFillColor();
					stops[2].style.stopColor = this.getEmptyColor();
					nodes[Math.ceil(currentRating)-1].classList.add(this.getData('ltPropHalfIcon'),"lyteRated");
				}
			}
			else{
				if(ratedNodes.length > 0 && this.getData('ltPropHalfIcon') && ratedNodes[ratedNodes.length-1].classList.contains(this.getData('ltPropHalfIcon'))){
					ratedNodes[ratedNodes.length-1].classList.remove(this.getData('ltPropHalfIcon'));
					ratedNodes[ratedNodes.length-1].classList.add(this.getData('ltPropFullIcon'))
				}
				if(currentRating < ratedNodes.length){
					var obj;
					for(var i = currentRating; i < ratedNodes.length; i++){
						if(this.getData("ltPropAppearance") == "lineHeart"){
							// $L('use',ratedNodes[i]).attr("href","#lineHeart");
							obj = this.getData('countArray')[i];
							this.$addon.objectUtils(obj,"add","appearance","#lineHeart");
							this.$addon.objectUtils(obj,"add","viewbox",this.getData('lineHeartVB'));
						}
						if(this.getData("ltPropAppearance") == "lineStar"){
							// $L('use',ratedNodes[i]).attr("href","#lineStar");
							obj = this.getData('countArray')[i];
							this.$addon.objectUtils(obj,"add","viewbox",this.getData('lineStarVB'));
						}
						if(this.getData("ltPropAppearance") == "lineHeart" || this.getData("ltPropAppearance") == "lineStar"){
							$L('svg',ratedNodes[i]).css('stroke',this.getEmptyColor());
							$L('svg',ratedNodes[i]).css('fill',"transparent");
						}
						else{
							$L('svg',ratedNodes[i]).css('fill',this.getEmptyColor());
						}
						ratedNodes[i].classList.remove(this.getData('ltPropFullIcon'),"lyteRated");
						if(this.getData('ltPropHalfIcon')){
							ratedNodes[i].classList.remove(this.getData('ltPropHalfIcon'));
						}
					}
				}
				else if(currentRating > ratedNodes.length){
					var obj;
					for(var i = 0; i < currentRating && i < nodes.length; i++){
						if(this.getData("ltPropAppearance") == "lineHeart"){
							// $L('use',nodes[i]).attr("href","#heart");
							obj = this.getData('countArray')[i];
							this.$addon.objectUtils(obj,"add","appearance","#heart");
							this.$addon.objectUtils(obj,"add","viewbox",this.getData('heartVB'));
						}
						if(this.getData("ltPropAppearance") == "lineStar"){
							// $L('use',nodes[i]).attr("href","#star");
							obj = this.getData('countArray')[i];
							this.$addon.objectUtils(obj,"add","viewbox",this.getData('starVB'));
						}
						$L('svg',nodes[i]).css('fill',this.getFillColor());
						$L('svg',nodes[i]).css('stroke',"");
						nodes[i].classList.add(this.getData('ltPropFullIcon'),"lyteRated");
					}
				}
			}
		}
		else if(this.getData('ltPropType') == "switch"){
			if(ratedNodes.length){
				var prevValue = parseInt(ratedNodes[0].getAttribute('data-lrc'));
				ratedNodes[0].classList.remove(this.getClass(prevValue-1,'fi'),"lyteRated");
			}
			if(currentRating > 0){
				nodes[currentRating - 1].classList.add(this.getClass(currentRating-1,'fi'),"lyteRated");
			}
		}
		else if(this.getData('ltPropType') === "multiple" && this.getData('ltPropAppearance') != 'smiley'){
			if(currentRating % 1 !== 0){
				for(var i = 0; i < Math.floor(currentRating); i++){
					nodes[i].classList.add(this.getClass(i,'fi'),"lyteRated");
				}
				for(var i = Math.floor(currentRating); i < nodes.length; i++){
					nodes[i].classList.remove(this.getClass(i,'fi'),this.getClass(i,'hi'),"lyteRated");
				}
				nodes[Math.ceil(currentRating)-1].classList.add(this.getClass(Math.ceil(currentRating)-1,'hi'),"lyteRated");
			}
			else{
				if(ratedNodes.length > 0 && this.getData('halfIconArray').length && ratedNodes[ratedNodes.length-1].classList.contains(this.getClass(ratedNodes.length-1,'hi'))){
					ratedNodes[ratedNodes.length-1].classList.remove(this.getClass(ratedNodes.length-1,'hi'));
					ratedNodes[ratedNodes.length-1].classList.add(this.getClass(ratedNodes.length-1,'fi'))
				}
				if(currentRating < ratedNodes.length){
					for(var i = currentRating; i < ratedNodes.length; i++){
						ratedNodes[i].classList.remove(this.getClass(i,'fi'),"lyteRated");
						if(this.getData('halfIconArray').length){
							ratedNodes[i].classList.remove(this.getClass(i,'hi'));
						}
					}
				}
				else if(currentRating > ratedNodes.length){
					for(var i = ratedNodes.length; i < currentRating && i < nodes.length; i++){
						nodes[i].classList.add(this.getClass(i,'fi'),"lyteRated");
					}
				}
			}
		}
		else{
			if(currentRating % 1 !== 0){
				for(var i = 0; i < Math.floor(currentRating); i++){
					nodes[i].classList.add(this.getData('ltPropFullIcon'),"lyteRated");
				}
				for(var i = Math.floor(currentRating); i < nodes.length; i++){
					nodes[i].classList.remove(this.getData('ltPropFullIcon'),this.getData('ltPropHalfIcon'),"lyteRated");
				}
				nodes[Math.ceil(currentRating)-1].classList.add(this.getData('ltPropHalfIcon'),"lyteRated");
			}
			else{
				if(ratedNodes.length > 0 && this.getData('ltPropHalfIcon') && ratedNodes[ratedNodes.length-1].classList.contains(this.getData('ltPropHalfIcon'))){
					ratedNodes[ratedNodes.length-1].classList.remove(this.getData('ltPropHalfIcon'));
					ratedNodes[ratedNodes.length-1].classList.add(this.getData('ltPropFullIcon'))
				}
				if(currentRating < ratedNodes.length){
					for(var i = currentRating; i < ratedNodes.length; i++){
						ratedNodes[i].classList.remove(this.getData('ltPropFullIcon'),"lyteRated");
						if(this.getData('ltPropHalfIcon')){
							ratedNodes[i].classList.remove(this.getData('ltPropHalfIcon'));
						}
					}
				}
				else if(currentRating > ratedNodes.length){
					for(var i = ratedNodes.length; i < currentRating && i < nodes.length; i++){
						nodes[i].classList.add(this.getData('ltPropFullIcon'),"lyteRated");
					}
				}
			}
		}
		if(!stopMethod && this.getMethods("onClick")){
            this.executeMethod("onClick",this.getData('ltPropValue'),this);
        }
	}

    getClass(index, iconType) {
		if(iconType == "fi"){
			return this.$component.registeredHelpers.lyteUiGetArrayValueByIndex(this.getData('fullIconArray'),index);
		}
		if(iconType == "hi"){
			return this.$component.registeredHelpers.lyteUiGetArrayValueByIndex(this.getData('halfIconArray'),index);
		}
		if(iconType == "hfi"){
			return this.$component.registeredHelpers.lyteUiGetArrayValueByIndex(this.getData('hoverFullIconArray'),index);
		}
		if(iconType == "hhi"){
			return this.$component.registeredHelpers.lyteUiGetArrayValueByIndex(this.getData('hoverHalfIconArray'),index);
		}
	}

    clearHoverClasses() {
		var hoveredNodes = this.$node.querySelectorAll('.lyteRatingHover');
		for(var i = 0; i<hoveredNodes.length; i++){
			if(this.getData('ltPropType') == "switch"){
				hoveredNodes[i].classList.remove(this.getClass(parseInt(hoveredNodes[i].getAttribute('data-lrc'))-1, 'hfi'),'lyteRatingHover');
			}
			else if(this.getData('ltPropType') == "multiple" && this.getData('ltPropAppearance') != 'smiley'){
				hoveredNodes[i].classList.remove(this.getClass(parseInt(hoveredNodes[i].getAttribute('data-lrc'))-1, 'hfi'),'lyteRatingHover');
				if(this.getData('hoverHalfIconArray').length){
					hoveredNodes[i].classList.remove(this.getClass(parseInt(hoveredNodes[i].getAttribute('data-lrc'))-1, 'hhi'),'lyteRatingHover');
				}
			}
			else{
				hoveredNodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),this.getData('ltPropHoverHalfIcon'),'lyteRatingHover');
			}
		}
	}

    getCurrentValue(event, ele) {
		var value = parseInt(ele.getAttribute('data-lrc'));
		if(this.getData('ltPropHalfIncrement')){
			var elemOffset = ele.getBoundingClientRect();
			if(event.clientX <= elemOffset.left + (elemOffset.width/2)){
				value = (value-1)+0.5;
			}
		}
		return value;
	}

    onClick(event) {
		if(this.getData('ltPropClearable')){
			if(this.rated && this.getData('currentHover') == this.getData('ltPropValue')){
				this.setData('ltPropValue',0);
				this.rated = false;
				this.setData('currentHover',0)
			}
			else{

				var value = this.getCurrentValue(event,event.target.closest('lyte-rate-icon'));
				if(value != this.getData('currentHover')){
					this.setData('currentHover',value);
				}
				this.setData('ltPropValue',this.getData('currentHover'));
				this.rated = true;
			}
		}
		else{
			this.setData('ltPropValue',this.getData('currentHover'));
		}
	}

    onOver(event) {
		var target = event.target.closest('lyte-rate-icon'),
			value = parseInt(target.getAttribute('data-lrc')),
			hoveredNodes = this.getData('currentHover'),
			nodes = this.$node.querySelectorAll('lyte-rate-icon'),
			prevValue = hoveredNodes;
		if(this.getData('ltPropType') == "svg"){
			if(this.$component.registeredHelpers.lyteUiCheckHalfRatingSvg(this.getData('ltPropHalfIncrement'), this.getData('ltPropPrecision'))){
				var precision = this.getData('ltPropPrecision') * 100, stops,
				elemOffset = target.querySelector('use').getBoundingClientRect();
				hoveredNodes = Math.ceil(hoveredNodes);
				if((value - 1) < hoveredNodes){
					for(var i = value-1; i < hoveredNodes; i++){
						stops = nodes[i].querySelectorAll('linearGradient stop');
						stops[1].setAttribute('offset','0%');
						stops[2].setAttribute('offset','0%');
						stops[0].style.stopColor = this.getEmptyColor();
						stops[1].style.stopColor = this.getEmptyColor();
						stops[2].style.stopColor = this.getEmptyColor();
						nodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),this.getData('ltPropHoverHalfIcon'),"lyteRatingHover");
					}
				}
				else if((value - 1) > hoveredNodes){
					for(var i = hoveredNodes; i < value-1; i++){
						stops = nodes[i].querySelectorAll('linearGradient stop');
						stops[1].setAttribute('offset','100%');
						stops[2].setAttribute('offset','0%');
						stops[0].style.stopColor = this.getHoverColor();
						stops[1].style.stopColor = this.getHoverColor();
						stops[2].style.stopColor = this.getHoverColor();
						nodes[i].classList.add(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
					}
					for(var i = value-1; i < nodes.length; i++){
						stops = nodes[i].querySelectorAll('linearGradient stop');
						stops[1].setAttribute('offset','0%');
						stops[2].setAttribute('offset','0%');
						stops[0].style.stopColor = this.getEmptyColor();
						stops[1].style.stopColor = this.getEmptyColor();
						stops[2].style.stopColor = this.getEmptyColor();
						nodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),this.getData('ltPropHoverHalfIcon'),"lyteRatingHover");
					}
				}
				else{
					for(var i = value-1; i < nodes.length; i++){
						stops = nodes[i].querySelectorAll('linearGradient stop');
						stops[1].setAttribute('offset','0%');
						stops[2].setAttribute('offset','0%');
						stops[0].style.stopColor = this.getEmptyColor();
						stops[1].style.stopColor = this.getEmptyColor();
						stops[2].style.stopColor = this.getEmptyColor();
						nodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),this.getData('ltPropHoverHalfIcon'),"lyteRatingHover");
					}
				}
				target.classList.remove(this.getData('ltPropHoverFullIcon'),this.getData('ltPropHoverHalfIcon'),"lyteRatingHover");
				// if(event.clientX >= elemOffset.left && event.clientX <= elemOffset.right){
					var currWidth = Math.abs(event.clientX - Math.round(elemOffset.left)),
					totalWidth = Math.round(elemOffset.width),
					percentage = parseInt(Math.min(((currWidth/totalWidth)*100), 100));
					// console.log(currWidth,totalWidth,percentage)
					if(percentage != 0 && percentage % precision == 0){
						stops = target.querySelectorAll('linearGradient stop');
						stops[1].setAttribute('offset', percentage + '%');
						stops[2].setAttribute('offset', (percentage + 1) + '%');
						stops[0].style.stopColor = this.getHoverColor();
						stops[1].style.stopColor = this.getHoverColor();
						stops[2].style.stopColor = this.getEmptyColor();
						target.classList.add(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
						this.setData('currentHover',(value - 1)+(percentage/100));
					}
					else{
						var quo = parseInt(percentage / precision),
						nearest = Math.min((quo+1)*precision,100);
						stops = target.querySelectorAll('linearGradient stop');
						stops[1].setAttribute('offset', nearest + '%');
						stops[2].setAttribute('offset', (nearest + 1) + '%');
						stops[0].style.stopColor = this.getHoverColor();
						stops[1].style.stopColor = this.getHoverColor();
						stops[2].style.stopColor = this.getEmptyColor();
						target.classList.add(this.getData('ltPropHoverHalfIcon'),"lyteRatingHover");
						this.setData('currentHover',(value-1)+(nearest/100));
					}
				// }
			}
			else{
				var obj;
				if(value > hoveredNodes){
					for(var i = hoveredNodes; i < value; i++){
						nodes[i].classList.add(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
						if(this.getData("ltPropAppearance") == "lineHeart"){
							// $L('use',nodes[i]).attr("href","#heart");
							obj = this.getData('countArray')[i];
							this.$addon.objectUtils(obj,"add","appearance","#heart");
							this.$addon.objectUtils(obj,"add","viewbox",this.getData('heartVB'));
						}
						if(this.getData("ltPropAppearance") == "lineStar"){
							// $L('use',nodes[i]).attr("href","#star");
							obj = this.getData('countArray')[i];
							this.$addon.objectUtils(obj,"add","viewbox",this.getData('starVB'));
						}
						$L('svg',nodes[i]).css('fill',this.getHoverColor());
						$L('svg',nodes[i]).css('stroke',"");
					}
				}
				else if(value < hoveredNodes){
					for(var i = value; i < hoveredNodes; i++){
						nodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
						if(value < this.getData('ltPropValue')){
							if(this.getData("ltPropAppearance") == "lineHeart"){
								// $L('use',nodes[i]).attr("href","#heart");
								obj = this.getData('countArray')[i];
								this.$addon.objectUtils(obj,"add","appearance","#heart");
								this.$addon.objectUtils(obj,"add","viewbox",this.getData('heartVB'));
							}
							if(this.getData("ltPropAppearance") == "lineStar"){
								// $L('use',nodes[i]).attr("href","#star");
								obj = this.getData('countArray')[i];
								this.$addon.objectUtils(obj,"add","viewbox",this.getData('starVB'));
							}
							$L('svg',nodes[i]).css('fill',this.getFillColor());
							$L('svg',nodes[i]).css('stroke',"");
						}
						else{
							if(this.getData("ltPropAppearance") == "lineHeart"){
								// $L('use',nodes[i]).attr("href","#lineHeart");
								obj = this.getData('countArray')[i];
								this.$addon.objectUtils(obj,"add","appearance","#lineHeart");
								this.$addon.objectUtils(obj,"add","viewbox",this.getData('lineHeartVB'));
							}
							if(this.getData("ltPropAppearance") == "lineStar"){
								// $L('use',nodes[i]).attr("href","#lineStar");
								obj = this.getData('countArray')[i];
								this.$addon.objectUtils(obj,"add","viewbox",this.getData('lineStarVB'));
							}
							if(this.getData("ltPropAppearance") == "lineHeart" || this.getData("ltPropAppearance") == "lineStar"){
								$L('svg',nodes[i]).css('stroke',this.getEmptyColor());
								$L('svg',nodes[i]).css('fill',"transparent");
							}
							else{
								$L('svg',nodes[i]).css('fill',this.getEmptyColor());
							}
						}
					}
				}
				else{
					for(var i = 0; i < hoveredNodes; i++){
						if(!(nodes[i].classList.contains("lyteRatingHover"))){
							nodes[i].classList.add(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
							if(this.getData("ltPropAppearance") == "lineHeart"){
								// $L('use',nodes[i]).attr("href","#heart");
								obj = this.getData('countArray')[i];
								this.$addon.objectUtils(obj,"add","appearance","#heart");
								this.$addon.objectUtils(obj,"add","viewbox",this.getData('heartVB'));
							}
							if(this.getData("ltPropAppearance") == "lineStar"){
								// $L('use',nodes[i]).attr("href","#star");
								obj = this.getData('countArray')[i];
								this.$addon.objectUtils(obj,"add","viewbox",this.getData('starVB'));
							}
							$L('svg',nodes[i]).css('fill',this.getHoverColor());
							$L('svg',nodes[i]).css('stroke',"");
						}
					}
				}
				this.setData('currentHover',value);
			}
		}
		else if(this.getData('ltPropType') == "switch"){
			nodes[value-1].classList.add(this.$component.registeredHelpers.lyteUiGetArrayValueByIndex(this.getData('hoverFullIconArray'),value-1),'lyteRatingHover');
			this.setData('currentHover',value);
		}
		else if(this.getData('ltPropType') == "multiple" && this.getData('ltPropAppearance') != 'smiley'){
			if(this.getData('ltPropHalfIncrement')){
				hoveredNodes = Math.round(hoveredNodes);
				var elemOffset = target.getBoundingClientRect();
				if((value - 1) < hoveredNodes){
					for(var i = value-1; i < hoveredNodes; i++){
						nodes[i].classList.remove(this.getClass(i,'hfi'), this.getClass(i,'hhi'),"lyteRatingHover");
					}
				}
				else if((value - 1) > hoveredNodes){
					for(var i = hoveredNodes; i < value-1; i++){
						nodes[i].classList.add(this.getClass(i,'hfi'),"lyteRatingHover");
					}
				}
				target.classList.remove(this.getClass(value,'hfi'), this.getClass(value,'hhi'),"lyteRatingHover");
				if(event.clientX > elemOffset.left + (elemOffset.width/2)){
					target.classList.add(this.getClass(value,'hfi'),"lyteRatingHover");
					this.setData('currentHover',value);
				}
				else{
					target.classList.add(this.getClass(value,'hhi'),"lyteRatingHover");
					this.setData('currentHover',(value-1)+0.5);
				}
			}
			else{
				if(value > hoveredNodes){
					for(var i = hoveredNodes; i < value; i++){
						nodes[i].classList.add(this.getClass(i,'hfi'),"lyteRatingHover");
					}
				}
				else if(value < hoveredNodes){
					for(var i = value; i < hoveredNodes; i++){
						nodes[i].classList.remove(this.getClass(i,'hfi'),"lyteRatingHover");
					}
				}
				this.setData('currentHover',value);
			}
		}
		else{
			if(this.getData('ltPropHalfIncrement')){
				hoveredNodes = Math.round(hoveredNodes);
				var elemOffset = target.getBoundingClientRect();
				if((value - 1) < hoveredNodes){
					for(var i = value-1; i < hoveredNodes; i++){
						nodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),this.getData('ltPropHoverHalfIcon'),"lyteRatingHover");
					}
				}
				else if((value - 1) > hoveredNodes){
					for(var i = hoveredNodes; i < value-1; i++){
						nodes[i].classList.add(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
					}
				}
				target.classList.remove(this.getData('ltPropHoverFullIcon'),this.getData('ltPropHoverHalfIcon'),"lyteRatingHover");
				if(event.clientX > elemOffset.left + (elemOffset.width/2)){
					target.classList.add(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
					this.setData('currentHover',value);
				}
				else{
					target.classList.add(this.getData('ltPropHoverHalfIcon'),"lyteRatingHover");
					this.setData('currentHover',(value-1)+0.5);
				}
			}
			else{
				if(value > hoveredNodes){
					for(var i = hoveredNodes; i < value; i++){
						nodes[i].classList.add(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
					}
				}
				else if(value < hoveredNodes){
					for(var i = value; i < hoveredNodes; i++){
						nodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),"lyteRatingHover");
					}
				}
				this.setData('currentHover',value);
			}
		}
		if(prevValue != this.getData('currentHover') && this.getMethods("onHover")){
            this.executeMethod("onHover",this.getData('currentHover'),this);  
        }
	}

    onOut(event) {
		var target = event.target.closest('lyte-rate-icon');
		if(this.getData('ltPropType') == "switch"){
			target.classList.remove(this.$component.registeredHelpers.lyteUiGetArrayValueByIndex(this.getData('hoverFullIconArray'), parseInt(target.getAttribute('data-lrc')) - 1),'lyteRatingHover');
			this.setData('currentHover',0);
		}
		else{
			var children = this.$node.querySelectorAll('lyte-rate-icon');
			var left = children[0].getBoundingClientRect().left;
			var right = children[children.length-1].getBoundingClientRect().right;
			var offset = target.getBoundingClientRect();
			if(event.clientX <= left || event.clientX >= right || event.clientY <= offset.top || event.clientY >= offset.bottom){
				// console.log("mouseout",target);
				var nodes = this.$node.querySelectorAll('.lyteRatingHover');
				var obj;
				for(var i = 0; i < nodes.length; i++ ){
					if(this.getData('ltPropType') == "svg"){
						nodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),'lyteRatingHover');
						if(this.getData('ltPropHalfIncrement')){
							// console.log("WIP");
							this.setRating(true);
						}
						else{
							if(i < this.getData('ltPropValue')){
								$L('svg',nodes[i]).css('fill',this.getFillColor());
							}
							else{
								if(this.getData("ltPropAppearance") == "lineHeart"){
									// $L('use',nodes[i]).attr("href","#lineHeart");
									obj = this.getData('countArray')[i];
									this.$addon.objectUtils(obj,"add","appearance","#lineHeart");
									this.$addon.objectUtils(obj,"add","viewbox",this.getData('lineHeartVB'));
								}
								if(this.getData("ltPropAppearance") == "lineStar"){
									// $L('use',nodes[i]).attr("href","#lineStar");
									obj = this.getData('countArray')[i];
									this.$addon.objectUtils(obj,"add","viewbox",this.getData('lineStarVB'));
								}
								if(this.getData("ltPropAppearance") == "lineHeart" || this.getData("ltPropAppearance") == "lineStar"){
									$L('svg',nodes[i]).css('stroke',this.getEmptyColor());
									$L('svg',nodes[i]).css('fill',"transparent");
								}
								else{
									$L('svg',nodes[i]).css('fill',this.getEmptyColor());
									$L('svg',nodes[i]).css('stroke',"");
								}
							}
						}
					}
					else if(this.getData('ltPropType') == "multiple" && this.getData('ltPropAppearance') != 'smiley'){
						nodes[i].classList.remove(this.getClass(i,'hfi'),'lyteRatingHover');
						if(this.getData('halfIconArray').length){
							nodes[i].classList.remove(this.getClass(i,'hhi'));
						}
					}
					else{
						nodes[i].classList.remove(this.getData('ltPropHoverFullIcon'),'lyteRatingHover');
						if(this.getData('ltPropHoverHalfIcon')){
							nodes[i].classList.remove(this.getData('ltPropHoverHalfIcon'));
						}
					}
				}
				this.setData('currentHover',0);
			}
		}
		if(this.getMethods("onOut")){
            this.executeMethod("onOut",parseInt(target.getAttribute('data-lrc')),this);  
        }
	}

    static observers(arg1) {
        return Object.assign(super.observers({
			addReadOnlyClass : function(){
				var readonly = this.getData("ltPropReadOnly");
				if(readonly){
					this.$node.classList.add('lyteRatingReadOnly');
				}else{
					this.$node.classList.remove('lyteRatingReadOnly');
				}
			}.observes("ltPropReadOnly").on('didConnect'),
			addWrapperClass : function(){

				var wrapperClass = this.getData("ltPropWrapperClass");
				var prev_Wrapper = this.getData("ltPropPrevWrapper");
				if(prev_Wrapper){
					this.$node.classList.remove(prev_Wrapper);
				}
				if(wrapperClass){
					this.$node.classList.add(wrapperClass);
					this.setData("ltPropPrevWrapper",wrapperClass)
				}
			}.observes("ltPropWrapperClass").on('didConnect'),
            onColorchange : function(){
                if(this.getData('ltPropStroke')){
                    this.$node.style.setProperty('--lyte-rating-empty-color',this.getData('ltPropStroke'));
                }else if(this.getData('ltPropEmptyColor')){
                    this.$node.style.setProperty('--lyte-rating-empty-color',this.getData('ltPropEmptyColor'));
                }
                if(this.getData('ltPropFillColor')){
                    this.$node.style.setProperty('--lyte-rating-fill-color',this.getData('ltPropFillColor'));
                }
                if(this.getData('ltPropHoverColor')){
                    this.$node.style.setProperty('--lyte-rating-hover-color',this.getData('ltPropHoverColor'));
                }
            }.observes('ltPropFillColor','ltPropEmptyColor','ltPropHoverColor').on('didConnect'),

            initFn : function(){
                this.$node.classList.remove('heart','lineHeart','star','lineStar','custom','lyteRatingSmiley','three','five');
                if(this.getData('ltPropType') == "multiple"){
                    if(this.getData('ltPropAppearance') == "custom"){
                        var emptyIconArray = this.getData('ltPropEmptyIcon').split(",");
                        var fullIconArray = this.getData('ltPropFullIcon').split(",");
                        var hoverFullIconArray = this.getData('ltPropHoverFullIcon').split(",");
                        if(this.getData('ltPropHalfIncrement')){
                            var halfIconArray = this.getData('ltPropHalfIcon').split(",");
                            var hoverHalfIconArray = this.getData('ltPropHoverHalfIcon').split(",");
                            this.setData('halfIconArray',halfIconArray);
                            this.setData('hoverHalfIconArray',hoverHalfIconArray);
                        }
                        this.setData('emptyIconArray',emptyIconArray);
                        this.setData('fullIconArray',fullIconArray);
                        this.setData('hoverFullIconArray',hoverFullIconArray);
                    }
                    else{	//If not custom then by default it will be smiley
                        this.setData('ltPropAppearance','smiley');
                        this.$node.classList.add('lyteRatingSmiley');
                        if(this.getData('ltPropCount') == 3){
                            this.$node.classList.add('three');
                        }
                        else{
                            this.setData('ltPropCount',5);
                            this.$node.classList.add('five');
                        }
                        this.setData('ltPropEmptyIcon','lyteRatingEmpty');
                        this.setData('ltPropFullIcon','lyteRatingFilled');
                        this.setData('ltPropHoverFullIcon','lyteRatingHover');
                        this.setData('ltPropHalfIncrement', false);
                    }
                }
                else if(this.getData('ltPropType') == "switch"){
                    if(this.getData('ltPropAppearance') == "custom"){
                        var emptyIconArray = this.getData('ltPropEmptyIcon').split(",");
                        var fullIconArray = this.getData('ltPropFullIcon').split(",");
                        var hoverFullIconArray = this.getData('ltPropHoverFullIcon').split(",");
                        this.setData('emptyIconArray',emptyIconArray);
                        this.setData('fullIconArray',fullIconArray);
                        this.setData('hoverFullIconArray',hoverFullIconArray);
                        
                    }
                    else{	//If not custom then by default it will be likeDislike
                        this.setData('ltPropAppearance','likeDislike');
                        this.setData('ltPropEmptyIcon','lrcSwitchLikeEmpty,lrcSwitchDislikeEmpty');
                        this.setData('ltPropFullIcon','lrcSwitchLike,lrcSwitchDislike');
                        this.setData('ltPropHoverFullIcon','lrcSwitchHoverLike,lrcSwitchHoverDislike');
                        var emptyIconArray = this.getData('ltPropEmptyIcon').split(",");
                        var fullIconArray = this.getData('ltPropFullIcon').split(",");
                        var hoverFullIconArray = this.getData('ltPropHoverFullIcon').split(",");
                        this.setData('emptyIconArray',emptyIconArray);
                        this.setData('fullIconArray',fullIconArray);
                        this.setData('hoverFullIconArray',hoverFullIconArray);
                    }
                    this.setData('ltPropCount',2);
                    this.setData('ltPropHalfIncrement',false);
                    this.setData('ltPropClearable',true);
                }
                else if(this.getData('ltPropType') == "toggle"){
                    if(this.getData('ltPropAppearance') != "custom"){
                        this.setData('ltPropAppearance','heart');
                        this.setData('ltPropEmptyIcon','lrcToggleEmpty');
                        this.setData('ltPropFullIcon','lrcToggleFill');
                        this.setData('ltPropHoverFullIcon','lrcToggleHover2');
                    }
                    this.setData('ltPropCount',1);
                    this.setData('ltPropHalfIncrement',false);
                    this.setData('ltPropClearable',true);
                }
                else if(this.getData('ltPropType') == "svg"){
                    if(this.getData('ltPropAppearance') == "heart"){
                        this.$node.classList.add('heart');
                        this.setData('ltPropEmptyIcon','lyteRatingEmpty');
                        this.setData('ltPropFullIcon','lyteRatingFilled');
                        this.setData('ltPropHoverFullIcon','lyteRatingHoverFilled');
                        if(this.getData('ltPropHalfIncrement')){
                            this.setData('ltPropHalfIcon','lyteRatingHalf');
                            this.setData('ltPropHoverHalfIcon','lyteRatingHoverHalf');
                        }
                    }
                    else if(this.getData('ltPropAppearance') == "lineHeart"){
                        this.$node.classList.add('lineHeart');
                        this.setData('ltPropEmptyIcon','lyteRatingEmpty');
                        this.setData('ltPropFullIcon','lyteRatingFilled');
                        this.setData('ltPropHoverFullIcon','lyteRatingHoverFilled');
                        if(this.getData('ltPropHalfIncrement')){
                            this.setData('ltPropHalfIcon','lyteRatingHalf');
                            this.setData('ltPropHoverHalfIcon','lyteRatingHoverHalf');
                        }
                    }
                    else if(this.getData('ltPropAppearance') == "lineStar"){
                        this.$node.classList.add('lineStar');
                        this.setData('ltPropEmptyIcon','lyteRatingEmpty');
                        this.setData('ltPropFullIcon','lyteRatingFilled');
                        this.setData('ltPropHoverFullIcon','lyteRatingHoverFilled');
                        if(this.getData('ltPropHalfIncrement')){
                            this.setData('ltPropHalfIcon','lyteRatingHalf');
                            this.setData('ltPropHoverHalfIcon','lyteRatingHoverHalf');
                        }
                    }
                    else{
                        this.setData('ltPropAppearance','star');
                        this.$node.classList.add('star');
                        this.setData('ltPropEmptyIcon','lyteRatingEmpty');
                        this.setData('ltPropFullIcon','lyteRatingFilled');
                        this.setData('ltPropHoverFullIcon','lyteRatingHoverFilled');
                        if(this.getData('ltPropHalfIncrement')){
                            this.setData('ltPropHalfIcon','lyteRatingHalf');
                            this.setData('ltPropHoverHalfIcon','lyteRatingHoverHalf');
                        }
                    }
                    if(this.getData('ltPropType') == "svg" && this.getData('ltPropHalfIncrement')){
                        ratingNum += 1;
                        this.setData('ratingNum',ratingNum);
                    }
                }
                else if(this.getData('ltPropType') == "class"){
                    this.setData('ltPropAppearance','class');
                    if(this.getData('ltPropEmptyIcon') == "lyteDummyRC" || this.getData('ltPropFullIcon') == "lyteDummyRC" || this.getData('ltPropHoverFullIcon') == "lyteDummyRC"){
                        this.$node.classList.add('star');
                        this.setData('ltPropEmptyIcon','lyteRatingEmpty');
                        this.setData('ltPropFullIcon','lyteRatingFilled');
                        this.setData('ltPropHoverFullIcon','lyteRatingHoverFilled');
                        if(this.getData('ltPropHalfIncrement')){
                            this.setData('ltPropHalfIcon','lyteRatingHalf');
                            this.setData('ltPropHoverHalfIcon','lyteRatingHoverHalf');
                        }
                    }
                }
                if(this.getData('ltPropCount') != this.getData('countArray').length){
                    this.setCountArray();
                }
                this.setRating(true);
                //this.setAriaAttributes();
            }.observes("ltPropType","ltPropCount"/*,"ltPropHalfIncrement"*/).on("init"),

            appearanceChanged : function(){
                var value = this.getData('ltPropType');
                this.setData('ltPropType','');
                this.setData('ltPropType',value);
                this.setCountArray();
                this.setRating();
            }.observes("ltPropAppearance","ltPropHalfIncrement"),

            // setAria : function(){
            // 	this.setAriaAttributes();
            // }.on("didConnect"),
            didConnectFn : function(){
                var flag = false;
                if(!this.$node.rendered){
                    flag = true;
                    this.$node.rendered = true;
                }
                this.setRating(flag);
                
                if(!this.rendered){
                    this.setAriaAttributes();
                    if(this.getMethods("onRender")){
                        this.executeMethod("onRender",this);
                    }  
                    if(this.getData('ltPropValue') > 0){
                        this.rated = true;
                    }
                    this.rendered = true;
                }

            }.observes("ltPropValue").on("didConnect"),

            svgIconObserver : function(){
                this.setRating(true);
            }.observes('ltPropFillColor')
        }), arg1);
    }

    _() {
        _;
    }
}

LyteRatingComponent._template = "<template tag-name=\"lyte-rating\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropYield}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-yield yield-name=\"rateIcon\"></lyte-yield> </template><template case=\"{{expHandlers(ifNotEquals(ltPropAppearance,'smiley'),'&amp;&amp;',lyteUiIfEqualsAny(ltPropType,'switch','multiple'))}}\" is=\"case\" lc-id=\"lc_id_1\"> <template items=\"{{countArray}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-rate-icon class=\"{{lyteUiGetArrayValueByIndex(emptyIconArray,index)}}\" data-lrc=\"{{item.ind}}\" style=\"{{lyteUiConcat('width:',ltPropWidth,'; height:',ltPropHeight,';')}}\"></lyte-rate-icon> </template> </template><template case=\"{{ifEquals(ltPropType,'svg')}}\" is=\"case\" lc-id=\"lc_id_2\"> <template items=\"{{countArray}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-rate-icon class=\"{{lyteUiGetArrayValueByIndex(emptyIconArray,index)}}\" data-lrc=\"{{item.ind}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiCheckHalfRatingSvg(ltPropHalfIncrement,ltPropPrecision)}}\" is=\"case\" lc-id=\"lc_id_0\"> <svg viewBox=\"{{item.viewbox}}\" style=\"{{lyteUiConcat('width:',ltPropWidth,'; height:',ltPropHeight,lyteUiGetFillOrStroke(ltPropAppearance,'var(--lyte-rating-empty-color)',ltPropStroke))}}\"> <linearGradient id=\"{{lyteUiConcat('filledGradient',ratingNum,'__',item.ind)}}\" x1=\"0%\" y1=\"0%\" x2=\"100%\" y2=\"0%\"> <stop offset=\"0%\" style=\"stop-color: var(--lyte-rating-empty-color);\"></stop> <stop offset=\"0%\" style=\"stop-color: var(--lyte-rating-empty-color);\"></stop> <stop offset=\"0%\" style=\"stop-color: var(--lyte-rating-empty-color);\"></stop> </linearGradient> <use href=\"{{item.appearance}}\" fill=\"{{lyteUiConcat('url(#','filledGradient',ratingNum,'__',item.ind,')')}}\"></use> </svg> </template><template default=\"\"> <svg viewBox=\"{{item.viewbox}}\" style=\"{{lyteUiConcat('width:',ltPropWidth,'; height:',ltPropHeight,lyteUiGetFillOrStroke(ltPropAppearance,'var(--lyte-rating-empty-color)',ltPropStroke))}}\"> <use href=\"{{item.appearance}}\"></use> </svg> </template></template></lyte-rate-icon> </template> </template><template default=\"\"> <template items=\"{{countArray}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-rate-icon class=\"{{ltPropEmptyIcon}}\" data-lrc=\"{{item.ind}}\" style=\"{{lyteUiConcat('width:',ltPropWidth,'; height:',ltPropHeight,';')}}\"></lyte-rate-icon> </template> </template></template><svg version=\"1.1\" id=\"Layer_1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" x=\"0px\" y=\"0px\" viewBox=\"0 0 18 18\" style=\"display: none;\" xml:space=\"preserve\"> <g> <path id=\"heart\" d=\"M7.2,1C4.5,1,2.3,3.4,2.1,6.7c0,0.1-0.1,0.9,0.2,2.2c0.4,1.8,1.3,3.4,2.6,4.7l6.6,6.5l6.7-6.5 c1.3-1.3,2.2-2.9,2.6-4.7c0.3-1.2,0.2-2,0.2-2.2C20.7,3.4,18.6,1,15.9,1c-1.8,0-3.4,1-4.4,2.7C10.5,2,8.9,1,7.2,1L7.2,1z\"></path> </g> </svg> <svg version=\"1.1\" id=\"Layer_1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" x=\"0px\" y=\"0px\" viewBox=\"0 0 18 18\" style=\"display: none;\" xml:space=\"preserve\"> <style type=\"text/css\"> .st0 { fill: none; stroke-miterlimit: 10; } </style> <g> <path id=\"lineHeart\" class=\"st0\" d=\"M6.1,1C3.4,1,1.2,3.4,1,6.7c0,0.1-0.1,0.9,0.2,2.2c0.4,1.8,1.3,3.4,2.6,4.7l6.6,6.5l6.7-6.5 c1.3-1.3,2.2-2.9,2.6-4.7c0.3-1.2,0.2-2,0.2-2.2C19.7,3.4,17.6,1,14.9,1c-1.8,0-3.4,1-4.4,2.7C9.4,2,7.8,1,6.1,1L6.1,1z\"></path> </g> </svg> <svg version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" x=\"0px\" y=\"0px\" viewBox=\"5 5 89.6 80\" style=\"display: none;\" xml:space=\"preserve\"> <polygon id=\"star\" points=\"17,4.7 19.5,12 27,12 21,16.6 23.2,23.8 17,19.5 10.8,23.8 13,16.6 7,12 14.5,12 \"></polygon> </svg> </template>";;
LyteRatingComponent._dynamicNodes = [{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"i","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true},"lc_id_1":{"dN":[{"t":"a","p":[1],"cn":"lc_id_1"},{"t":"f","p":[1],"dN":[{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"lyteUiConcat","args":["'width:'","ltPropWidth","'; height:'","ltPropHeight","';'"]}}}},{"t":"cD","p":[1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_1"}],"cdp":{"t":"a","p":[1]},"dcn":true},"lc_id_2":{"dN":[{"t":"a","p":[1],"cn":"lc_id_2"},{"t":"f","p":[1],"dN":[{"t":"a","p":[1]},{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"lyteUiConcat","args":["'width:'","ltPropWidth","'; height:'","ltPropHeight",null]}}},"cn":"lc_id_0"},{"t":"a","p":[1,1],"cn":"lc_id_0"},{"t":"a","p":[1,3],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"lyteUiConcat","args":["'width:'","ltPropWidth","'; height:'","ltPropHeight",null]}}},"cn":"default"},{"t":"a","p":[1,1],"cn":"default"}]},"dc":{"lc_id_0":{},"default":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_2"}],"cdp":{"t":"a","p":[2]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"f","p":[1],"dN":[{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"lyteUiConcat","args":["'width:'","ltPropWidth","'; height:'","ltPropHeight","';'"]}}}},{"t":"cD","p":[1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"lc_id_1":{"dc":[0],"hc":true,"trans":true},"lc_id_2":{"dc":[0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0","lc_id_1","lc_id_2"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;

LyteRatingComponent._observedAttributes = [
    "ltPropValue",
    "ltPropCount",
    "ltPropWrapperClass",
    "ltPropType",
    "ltPropAppearance",
    "ltPropReadOnly",
    "ltPropClearable",
    "ltPropHalfIncrement",
    "ltPropEmptyIcon",
    "ltPropFullIcon",
    "ltPropHalfIcon",
    "ltPropHoverFullIcon",
    "ltPropHoverHalfIcon",
    "ltPropYield",
    "ltPropEmptyColor",
    "ltPropFillColor",
    "ltPropHoverColor",
    "ltPropWidth",
    "ltPropHeight",
    "ltPropPrecision",
    "ltPropStroke",
    "countArray",
    "currentHover",
    "emptyIconArray",
    "fullIconArray",
    "halfIconArray",
    "hoverFullIconArray",
    "hoverHalfIconArray",
    "starVB",
    "lineStarVB",
    "heartVB",
    "lineHeartVB",
    "ratingNum",
    "ltPropPrevWrapper"
];

if( !_lyteUiUtils.registeredCustomElements[ 'lyte-rate-icon' ] ){
    _lyteUiUtils.registeredCustomElements[ 'lyte-rate-icon' ] = true;

    class LyteRateIconCustomElements extends LyteUiComponentComponentRegistry.CustomElements {
        lookups() {
            return [{
                component: LyteUiComponentComponentRegistry
            }];
        }

        constructor() {
            super();
        }

        connectedCallback() {
            var lyteNode = this.$node;
            if(!lyteNode.prevent){
		        var comp = lyteNode.closest('lyte-rating').component;
		        lyteNode.addEventListener('click',comp.onClick.bind(comp));
		        lyteNode.addEventListener('mousemove',comp.onOver.bind(comp));
		        lyteNode.addEventListener('mouseout',comp.onOut.bind(comp));
		        if(comp.getData('ltPropYield')){
		        	if(comp.timeoutId){
		        		clearTimeout(comp.timeoutId);
		        		comp.timeoutId = false;
		        	}
		        	comp.timeoutId = setTimeout(function(){
		        		var nodes = lyteNode.closest('lyte-rating').querySelectorAll('lyte-rate-icon');
		        		for(var i = 0; i<nodes.length; i++){
		        			nodes[i].setAttribute('data-lrc',i+1);
		        		}
		        	}.bind(lyteNode),100);
		        }
		        lyteNode.prevent = true;
	        }
        }

        disconnectedCallback() {
	        
	    }

        _() {
            _;
        }
    }

    LyteRateIconCustomElements.options = {clone : {allCallbacks : false}};

    LyteRateIconCustomElements.register("lyte-rate-icon");
}

/**
 * @syntax nonYielded
 * <lyte-rating lt-prop-type = "svg" lt-prop-appearance = "star"> </lyte-rating>
 * </lyte-rating> 
 */
export { LyteRatingComponent };

LyteRatingComponent.register("lyte-rating", {
    hash: "LyteRatingComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
