// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/063ec30d7526c17f8e08696aeed9d6fc.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component, LyteUiComponentComponentRegistry } from "../component.js";

import {
	arrayUtils,
	appendChild,
	shouldIgnoreDisconnect,
	_setIgnoreDisconnect
} from "../../node_modules/@slyte/component/index.js";

import $L from "../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js"

/**
 * Renders a kanbanview
 * @component lyte-kanbanview
 * @version 3.1.0
 * @dependencies lyte-card,lyte-board
 * /components/lyte-card.js
 * /theme/compiledCSS/default/ltr/lyte-ui-card.css
 * /components/lyte-board.js
 * /theme/compiledCSS/default/ltr/lyte-ui-board.css
 * @methods onDragSelectForBoard, onDragSelectBoards, onBodyScroll, onDragSelectForCard, onRecordDropForBoard, onRecordDrop,onDragSelectCard
 * @utility getVisibleBoard
 */
class LyteKanbanviewComponent extends Component {
    constructor() {
        super();
        this._lyteUtilFunctions = ["getVisibleBoard"];
    }

    data(arg1) {
		return Object.assign(super.data({
			/**
			 * @componentProperty {array} ltPropBoardDetails
			 * @version 3.1.0
			 * @default []
			 */

			'ltPropBoardDetails': prop('array', {
				'default': []
			}),
			/**
			 * @typedef {object} sortable
			 * @property {boolean} board
			 * @property {boolean} card
			*/
			/**
			 * @componentProperty {object} ltPropSortable=
			 * @version 3.1.0
			 * @default {"board" : true, "card" : true }
			 */
			'ltPropSortable': prop('object', {
				'default': { "board": true, "card": true }

			}),
			/**
			 * @componentProperty {boolean} ltPropMoreStageRecord=false
			 * @version 3.1.0
			 * @default false
			 */
			'ltPropMoreStageRecord': prop('boolean', {
				'default': false
			}),
			/**
			 * @componentProperty {string} ltPropId
			 * @version 3.1.0
			 */
			'ltPropId': prop('string'),
			/** 
			 * @componentProperty {boolean} ltPropPreventBoardDetailObserver
			 * @version 3.1.0
			 * @default false
			 */
			'ltPropPreventBoardDetailObserver': prop('boolean', {
				'default': false
			}),
			/**
			 * @componentProperty {boolean} ltPropViewPort
			 * @version 3.1.0
			 * @default true
			 */
			'ltPropViewPort': prop('boolean', {
				'default': true
			}),
			/**
			 * @componentProperty {number} ltPropBoardScrollStopDuration
			 * @version 3.1.0
			 * @default 250
			 */
			'ltPropBoardScrollStopDuration': prop('number', {
				'default': 250
			}),
			/**
			 * @componentProperty {boolean} ltPropAria
			 * @version 3.1.0
			 * @default false
			 */
			'ltPropAria': prop('boolean', {
				'default': false
			}),
			/**
			 * @componentProperty {object} ltPropAriaAttributes
			 * @version 3.1.0
			 * @default {}
			 */
			'ltPropAriaAttributes': prop('object', {
				default: {}
			}),
			/**
			* @componentProperty string ltPropSortableCancel
			* @version 3.1.0
			*/
			'ltPropSortableCancel': prop('string'),
			'ltPropSortableRestrict': prop('string'),
			/**
			* @experimental dummyId
			*/
			'dummyId': prop('string', {
				'default': ''
			}),
			'ltPropKanbanItemClassKey': prop("string", { default: '' }),
			'ltPropIntersection': prop('object', {
				'default': { "board": false, "card": false }

			}),
			'ltPropBoardWidth': prop('string', { default: '300px' }),

			'visibleBoards': prop('array', {
				'default': []
			}),
			'lastTopIndex': prop('number', {
				'default': -1
			}),
			'lastBottomIndex': prop('number', {
				'default': -1
			}),
			'currentStartIndex': prop('number', {
				'default': 0
			}),
			'currentEndIndex': prop('number', {
				'default': 0
			}),
			'rowsPerLoad': prop('number', {
				'default': 10
			})
		}), arg1);
	}

    init() {


	}

    setVisibleArray() {
		var visibleBoards = this.getData('visibleBoards'),
			boardDetails = this.getData('ltPropBoardDetails'),
			rowsPerLoad = Math.max(this.getData('rowsPerLoad'), this.getVisibleRowCount() + 1)

		for (let i = 0; i < Math.min(rowsPerLoad, boardDetails.length); i++) {
			arrayUtils(visibleBoards, 'push', true)
		}
	}

    didDestroy() {
		this._observer && this._observer.disconnect()
		delete this._observer
		delete this._dir
		clearTimeout(this.timeout2);
		clearTimeout(this.debounceTimeout)
		clearTimeout(this.viewPortTimeOut)
		clearTimeout(this._sortableTimeout)
		clearTimeout(this._cardSortableTimeout)
	}

    didConnect() {
		var kanbanviewList = document.querySelectorAll('lyte-kanbanview')
		this._dir = window._lyteUiUtils.getRTL();
		if (this.getData('ltPropIntersection').board) {
			this.setVisibleArray()
		}
		if (this.getData('ltPropId') == undefined) {
			var pos = Object.values(kanbanviewList).indexOf(this.$node)
			this.setData('dummyId', 'dummyId' + pos)
		}
		else {
			this.setData('dummyId', this.getData('ltPropId'))
		}
		this.doBoardSortable()
		if (!this.getData('ltPropViewPort')) {
			this.doCardSortable()
		}
		this.$node.getVisibleBoard = function () {
			var scrollDiv = this.$node.querySelector('.lyteKanbanviewScrollDivSelector.' + this.getData('dummyId'))
			if (scrollDiv.scrollWidth < scrollDiv.clientWidth) {
				return this.getData('ltPropBoardDetails')
			}
			return this.getVisibleNode()
		}.bind(this)
		if (this.getData('ltPropIntersection').board) {
			setTimeout(function () {
				this.createObserver()
				this.addAllKanbanItemToObserver()
			}.bind(this), 200)

		}
	}

    addVisibleRows(kanbanItem) {

		const boardDetails = this.getData('ltPropBoardDetails')
		let visibleBoards = this.getData('visibleBoards')
		let indexToShow = kanbanItem.getAttribute('data-index')
		if (indexToShow > visibleBoards.length - 1) {
			arrayUtils(visibleBoards, 'push', true)
			// const sortableParent = this.$node.querySelectorAll( '.lyteKanbanNestedSortable.sortable-parent' )
			// const sortableElements = sortableParent[0].getConnectedWithArray()
			// this.setUpCard(sortableElements)

		} else if (visibleBoards[indexToShow] && kanbanItem._removedChild) {
			appendChild(kanbanItem, kanbanItem._removedChild)
			const board = kanbanItem.querySelector('lyte-board')
			board.toggleIgnore()
			delete kanbanItem._removedChild
		}

	}

    removeNotVisibleRows(row) {
		let visibleBoards = this.getData('visibleBoards')
		const index = row.getAttribute('data-index')
		if (visibleBoards[index] && !row._removedChild && !row.classList.contains('sortable-element-selected') && !row.classList.contains('lyteKanbanSortableParent')) {
			_setIgnoreDisconnect(true);

			if (row.children.length > 0 && row.children[0]) {
				const board = row.querySelector('lyte-board')
				board.toggleIgnore()
				row._removedChild = row.removeChild(row.children[0])
			}
			_setIgnoreDisconnect(false);
			return true
		}
		return false
	}

    getVisibleRowCount() {
		const containerHeight = this.$node.clientHeight;
		const rowHeight = parseInt(this.getData('ltPropBoardWidth')) || this.$node.querySelector('lyte-kanban-item').clientWidth; // Assuming all rows have the same height
		return Math.max(1, Math.floor(containerHeight / rowHeight));;
	}

    updateVisibility(newTopIndex) {
		const visibleCount = this.getVisibleRowCount();
		const bottomIndex = this.getData('currentEndIndex');
		let topIndex = this.getData('currentStartIndex');
		const lastTopIndex = this.getData('lastTopIndex')
		const lastBottomIndex = this.getData('lastBottomIndex')
		const bufferSize = 2
		const newBottomIndex = newTopIndex + visibleCount + bufferSize * 2 - 1;

		if (newTopIndex === lastTopIndex && newBottomIndex === lastBottomIndex) {
			return;
		} // No change, skip update


		// Hide rows out of new range
		// for (let i = lastTopIndex; i <= lastBottomIndex; i++) {
		// 	if (i < newTopIndex || i > newBottomIndex) {
		// 	// rows[i]?.style.display = "none";
		// 		this.removeNotVisibleRows(i)
		// 	}
		// }

		// // Show rows in new range
		// for (let i = newTopIndex; i <= newBottomIndex; i++) {
		// 	if (i < lastTopIndex || i > lastBottomIndex) {
		// 		this.addVisibleRows(i)
		// 	}
		// }

		const kanbanItems = this.$node.querySelectorAll('lyte-kanban-item:not(.lyteSortablePlaceholder)')
		kanbanItems.forEach((row, index) => {
			if (index >= newTopIndex && index <= newBottomIndex) {
				this.addVisibleRows(row)
			} else {
				this.removeNotVisibleRows(row)
			}
		});

		// topIndex = newTopIndex;
		// bottomIndex = newBottomIndex;
		this.setData('currentStartIndex', newTopIndex)
		this.setData('currentEndIndex', newBottomIndex)
		this.setData('lastTopIndex', newTopIndex)
		this.setData('lastBottomIndex', newBottomIndex)
	}

    intersectionObserverFunc(entries) {
		const self = this
		let topIndex = this.getData('currentStartIndex');
		let bottomIndex = this.getData('currentEndIndex');
		entries.forEach((entry) => {
			if (entry.isIntersecting) {
				// const rowIndex = parseInt(entry.target.getAttribute('data-index'))
				const row = entry.target
				// const visibleCount = this.getVisibleRowCount();
				// let newTopIndex = topIndex;

				// // **Scrolled up → Load previous rows**
				// if (rowIndex === topIndex && topIndex > 0) {
				// 	newTopIndex = Math.max(0, topIndex - 1);
				// }

				// // **Scrolled down → Load next rows**
				// if (rowIndex === bottomIndex && bottomIndex < this.getData('ltPropBoardDetails').length - 1) {
				// 	newTopIndex = Math.min(this.getData('ltPropBoardDetails').length - visibleCount, topIndex + 1);
				// }
				// this.setData('currentStartIndex',topIndex)
				// this.updateVisibility(newTopIndex);


				if (row.classList.contains('lyteSortablePlaceholder')) {
					return
				}
				this.addVisibleRows(row)
			} else {
				if (entry.target.classList.contains('lyteSortablePlaceholder')) {
					return
				}
				this.removeNotVisibleRows(entry.target)
			}
		})


	}

    createObserver() {
		const boardWidth = parseInt(this.getData('ltPropBoardWidth')) * 2
		const container = this.$node.querySelector('.lyteKanbanviewScrollDivSelector')
		const options = {
			root: container,
			threshold: 0, // Trigger when 10% of a row is visible
			rootMargin: "0px " + boardWidth + "px 0px " + boardWidth + "px"
		};
		this._observer = new IntersectionObserver(this.intersectionObserverFunc.bind(this), options);
	}

    addAllKanbanItemToObserver() {
		const kanbanItems = this.$node.querySelectorAll('.lyteKanbanViewItem')
		for (let idx = 0; idx < kanbanItems.length; idx++) {
			this.addRowObserve(kanbanItems[idx])
		}
	}

    removeAllKanbanItemToObserver() {
		const kanbanItems = this.$node.querySelectorAll('lyte-kanban-item')
		for (let idx = 0; idx < kanbanItems.length; idx++) {
			this.removeRowObserve(kanbanItems[idx])
		}
	}

    addRowObserve(row) {
		// console.log(row)
		!row._observed && this._observer.observe(row)
		row._observed = true
	}

    removeRowObserve(row) {
		this._observer.unobserve(row)
		row._observed = false
	}

    getVisibleNode() {

		// if(this._insideObserver){
		// 	return 
		// }

		// return;
		var bcr = this.$node.getBoundingClientRect(),
			originalRows = Array.from(this.$node.getElementsByTagName('lyte-board')),
			tValue = Math.max(bcr.left + (this._dir ? 0 : 10), -10),
			bValue = Math.min(window.innerWidth + 10, bcr.right - (this._dir ? 10 : 0)),
			visible = [], boardDetails = this.getData('ltPropBoardDetails');


		for (var i = 0; i < originalRows.length; i++) {
			var row = originalRows[i],
				_bcr = row.getBoundingClientRect(),
				index = this.getData('ltPropIntersection').board ? row.getData('ltPropIndex') : i;
			if (_bcr.right > tValue && _bcr.left < bValue) {
				if (originalRows[i].component.getMethods('onBoardVisible')) {
					originalRows[i].component.executeMethod('onBoardVisible', boardDetails, boardDetails[i])
				}

				visible.push(boardDetails[index]);

			}
		}

		return visible;

	}

    callOnBoardVisible() {

		var originalRows = Array.from(this.$node.getElementsByTagName('lyte-board'))
		for (var i = 0; i < originalRows.length; i++) {
			if (originalRows[i].component.getMethods('onBoardVisible')) {
				originalRows[i].component.executeMethod('onBoardVisible', this.getData('ltPropBoardDetails'), this.getData('ltPropBoardDetails')[i])
			}
		}
	}

    doCardSortable() {
		clearTimeout(this._cardSortableTimeout)
		this._cardSortableTimeout = setTimeout(function () {
			if (this.getData('ltPropBoardDetails').length > 0 && this.getData('ltPropSortable').card) {
				this.setupSortableForCard()
			}
		}.bind(this), 100)

	}

    doBoardSortable() {
		if (this.getData('ltPropSortable').board) {
			this.setupSortableForBoard()
		}
	}

    hasSameColumnReordered(source, destination) {
		var sourceIndex = source.getAttribute('index'),
			destinationIndex = destination.getAttribute('index')

		if (sourceIndex == destinationIndex) {
			return true
		}
		return false
	}

    onRecordDropForBoard(
		boardDetails,
		droppedElement,
		fromIndex,
		toIndex,
		source,
		destination,
		draggedElement
	) {

		if (this.getMethods('onRecordDropForBoard')) {
			this.executeMethod('onRecordDropForBoard', droppedElement, source, destination, fromIndex, toIndex, boardDetails, draggedElement[0])
		}
	}

    onRecordDrop(
		boardDetails,
		sourceArray,
		destArray,
		draggedElement,
		fromIndex,
		toIndex,
		source,
		destination,
		droppedElement
	) {
		var sourceIndex = source.getAttribute('index'),
			destinationIndex = destination.getAttribute('index'),
			sBoard, dBoard, card


		dBoard = destination.closest('lyte-board');
		card = droppedElement.querySelector('lyte-card')
		if (!this.hasSameColumnReordered(source, destination)) {
			sBoard = source.closest('lyte-board')
			boardDetails[sourceIndex].cards = sourceArray;
			sBoard.setData('ltPropContent', sourceArray);
		}
		sBoard = sBoard || dBoard
		boardDetails[destinationIndex].cards = destArray;
		dBoard.setData('ltPropContent', destArray);

		if (this.getMethods('onRecordDrop')) {
			this.executeMethod('onRecordDrop', card, sBoard, dBoard, fromIndex, toIndex, parseInt(sourceIndex), parseInt(destinationIndex), boardDetails, draggedElement[0])
		}
	}

    addSortableForCard() {
		//debounce
		clearTimeout(this.viewPortTimeOut)
		this.viewPortTimeOut = setTimeout(function () {
			// this.setupSortableForCard()

			this.setUpCard()
		}.bind(this), 250)
	}

    callSetUpCardFunc() {
		const sortableParent = this.$node.querySelectorAll('.lyteKanbanNestedSortable.sortable-parent')
		if (sortableParent.length != 0) {
			const sortableElements = sortableParent[0].getConnectedWithArray()
			this.setUpCard(sortableElements)
		}

	}

    setUpCard(boardWithSortable) {
		// console.log(boardWithSortable)
		var boardWithoutSortable = this.$node.querySelectorAll('.lyteKanbanNestedSortable:not(.sortable-parent)')
		boardWithSortable = boardWithSortable || this.$node.querySelectorAll('.lyteKanbanNestedSortable.sortable-parent')
		if (boardWithoutSortable.length > 0 && boardWithSortable.length > 0) {
			var self = this;
			var sortableObject = {
				scrollDivX: '.lyteKanbanviewScrollDivSelector.' + self.getData('dummyId'),
				isSameClass: true,
				dblTouchEvent: true,
				onReady: function (sortableElem) {
					var board = $L(sortableElem)[0].closest("lyte-board")
					if (self.getMethods('onReadyForCard')) {
						return self.executeMethod('onReadyForCard', sortableElem, board);
					}
				},
				onSelect: function (currentElem, fromIndex, source, event) {

					if ($L(currentElem).hasClass('lyteKanbanNoResultMsg')) {
						return false;
					}

					window._lyteUiUtils.closeAllPopups()
					var card = currentElem.querySelector('lyte-card')
					var res = true
					if (self.getMethods('onDragSelectForCard')) {
						res = self.executeMethod('onDragSelectForCard', card, fromIndex, source, event);

					}
					if (res && (self.getData('ltPropIntersection').card || self.getData('ltPropIntersection').board)) {
						var kanbanItem = $L(currentElem)[0].closest('lyte-kanban-item')
						kanbanItem && kanbanItem.classList.add('lyteKanbanSortableParent')
					}
					return res;
				},

				onPlaceholder: function (draggableElement, placeholderElement, source, destination) {
					var card = draggableElement.querySelector('lyte-card')
					if (self.getMethods('onPlaceholderForCard')) {
						return self.executeMethod('onPlaceholderForCard', card, placeholderElement, source, destination);
					}
					return true
				},
				onDrag: function (draggableElement, belowElem, event, placeholder) {
					self.prevent = true
					var card = draggableElement.querySelector('lyte-card')
					if (self.getMethods('onDragForCard')) {
						self.executeMethod('onDragForCard', card, belowElem, event, placeholder);
					}
					return true
				},
				onDrag: function (draggableElement, belowElem, event, placeholder) {
					self.prevent = true
					var card = draggableElement.querySelector('lyte-card')
					if (self.getMethods('onDragForCard')) {
						self.executeMethod('onDragForCard', card, belowElem, event, placeholder);
					}
				},
				onEnter: function (event, obj) {
					// console.log("onEnter")
					var element = arguments[1].sortable,
						boardDetails = self.getData('ltPropBoardDetails')

					if (element.getAttribute('cards-length') == 0) {
						var div = element.nextElementSibling
						if ($L(div).hasClass('lyteKanbanNoResultMsg')) {
							div.classList.add('lyteHide')
						}
					}
					if (self.getMethods('onEnterForCard')) {
						return self.executeMethod('onEnterForCard', boardDetails, event, obj);
					}
				},
				onLeave: function (event, obj) {
					var element = arguments[1].sortable,
						boardDetails = self.getData('ltPropBoardDetails')

					if (element.getAttribute('cards-length') == 0) {
						var div = element.nextElementSibling
						if ($L(div).hasClass('lyteKanbanNoResultMsg')) {
							div.classList.remove('lyteHide')
						}
					}
					if (self.getMethods('onLeaveForCard')) {
						return self.executeMethod('onLeaveForCard', boardDetails, event, obj);
					}
				},
				onBeforeDrop: function (droppableElement, belowElement, placeholderElement, fromIndex, toIndex, source, destination) {
					var sourceIndex = source.getAttribute('index'),
						destinationIndex = destination.getAttribute('index'),
						boardDetails = self.getData('ltPropBoardDetails'),
						sourceArray = boardDetails[sourceIndex].cards,
						destArray = boardDetails[destinationIndex].cards,
						flag = true;
					if (self.getMethods('onBeforeDropForCard')) {
						flag = self.executeMethod('onBeforeDropForCard', boardDetails, sourceArray, destArray, droppableElement, belowElement, placeholderElement, fromIndex, toIndex, source, destination);
					}
					if (!flag && (self.getData('ltPropIntersection').card || self.getData('ltPropIntersection').board)) {
						var kanbanItem = $L(droppableElement)[0].closest('lyte-kanban-item')
						kanbanItem && kanbanItem.classList.remove('lyteKanbanSortableParent')
					}
					return flag;
				},
				onDrop: function (droppedElement, destination, belowElement, fromIndex, toIndex, source) {
					var sourceIndex = source.getAttribute('index'),
						destinationIndex = destination.getAttribute('index'),
						boardDetails = self.getData('ltPropBoardDetails'),
						sourceArray = boardDetails[sourceIndex].cards,
						destArray = boardDetails[destinationIndex].cards,
						draggedElement
					if (destArray.length == 0) {
						toIndex = 0;
					}
					draggedElement = (self.hasSameColumnReordered(source, destination) ? arrayUtils(destArray, 'splice', fromIndex, 1) : arrayUtils(sourceArray, 'splice', fromIndex, 1))
					arrayUtils(destArray, 'splice', toIndex, 0, draggedElement[0])
					delete self.prevent

					self.onRecordDrop(boardDetails, sourceArray, destArray, draggedElement, fromIndex, toIndex, source, destination, droppedElement)
					if (self.getData('ltPropIntersection').card || self.getData('ltPropIntersection').board) {
						var kanbanItem = $L(source)[0].closest('lyte-kanban-item')
						kanbanItem && kanbanItem.classList.remove('lyteKanbanSortableParent')
					}
				}
			}
			if (this.getData('ltPropSortableCancel')) {
				sortableObject.cancel = this.getData('ltPropSortableCancel')

			}
			if (this.getData('ltPropSortableRestrict')) {
				sortableObject.restrict = this.getData('ltPropSortableRestrict')
			}
			boardWithSortable.forEach(function (element) {
				if (Array.from(boardWithoutSortable).indexOf(element) > -1) {
					return
				}
				element.addToConnectedWith(
					boardWithoutSortable, sortableObject);
			})
			// console.log(boardWithoutSortable)
			// console.log(boardWithSortable)
		}
		if (boardWithSortable.length == 0) {
			this.setupSortableForCard()
		}
	}

    setupSortableForBoard() {
		var self = this;
		var sortableObject = {
			scrollDivX: '.lyteKanbanviewScrollDivSelector.' + this.getData('dummyId'),
			dblTouchEvent: true,
			onReady: function (sortableElem) {
				if (self.getMethods('onReadyForBoard')) {
					return self.executeMethod('onReadyForBoard', sortableElem, this);
				}
			},
			onSelect: function (currentElem, fromIndex, source, event) {
				var board = currentElem.querySelector('lyte-board')

				if (self.getMethods('onDragSelectForBoard')) {
					return self.executeMethod('onDragSelectForBoard', board, fromIndex, source, event);
				}

				return true;
			},
			onDrag: function (draggableElement, belowElem, event, placeholder) {
				self.prevent = true
				var board = draggableElement.querySelector('lyte-board')

				if (self.getMethods('onDragForBoard')) {
					self.executeMethod('onDragForBoard', board, belowElem, event, placeholder);
				}
			},
			onEnter: function (event, obj) {
				var boardDetails = self.getData('ltPropBoardDetails')

				if (self.getMethods('onEnterForBoard')) {
					self.executeMethod('onEnterForBoard', boardDetails, event, obj);
				}
			},
			onLeave: function (event, obj) {
				var boardDetails = self.getData('ltPropBoardDetails')

				if (self.getMethods('onLeaveForBoard')) {
					self.executeMethod('onLeaveForBoard', boardDetails, event, obj);
				}
			},
			onBeforeDrop: function (droppableElement, belowElement, placeholderElement, fromIndex, toIndex, source, destination) {
				var boardDetails = self.getData('ltPropBoardDetails'),
					board = droppableElement.querySelector('lyte-board'),
					flag = true;
				if (self.getMethods('onBeforeDropForBoard')) {
					flag = self.executeMethod('onBeforeDropForBoard', boardDetails, droppableElement, belowElement, placeholderElement, fromIndex, toIndex, source, destination);
				}
				return flag;
			},
			onDrop: function (droppedElement, destination, belowElement, fromIndex, toIndex, source) {
				var boardDetails = self.getData('ltPropBoardDetails'),
					draggedElement, board = droppedElement.querySelector('lyte-board')



				draggedElement = arrayUtils(boardDetails, 'splice', fromIndex, 1)
				arrayUtils(boardDetails, 'splice', toIndex, 0, draggedElement[0])
				delete self.prevent

				self.setData('ltPropBoardDetails', boardDetails)
				self.onRecordDropForBoard(boardDetails, board, fromIndex, toIndex, source, destination, draggedElement)
				$L('.lyteKanbanNestedSortable ', this.$node).sortable("destroy")


			},
			onPlaceholder: function (draggableElement, placeholderElement, source, destination) {
				var boardDetails = self.getData('ltPropBoardDetails')

				var board = draggableElement.querySelector('lyte-board')
				if (self.getMethods('onPlaceholderForBoard')) {
					return self.executeMethod('onPlaceholderForBoard', board, placeholderElement, source, destination, boardDetails);
				}
				return true
			}
		}
		if (this.getData('ltPropSortableCancel')) {
			sortableObject.cancel = this.getData('ltPropSortableCancel')

		}
		if (this.getData('ltPropSortableRestrict')) {
			sortableObject.restrict = this.getData('ltPropSortableRestrict')
		}
		$L('.lyteKanbanviewScrollDivSelector ', this.$node).sortable(sortableObject);
	}

    setupSortableForCard() {
		var self = this;
		// if($L( '.lyteKanbanNestedSortable', this.$node )){
		var sortableObject = {
			scrollDivX: '.lyteKanbanviewScrollDivSelector.' + this.getData('dummyId'),
			connectedWith: '.lyteKanbanNestedSortable.' + this.getData('dummyId'),
			isSameClass: true,
			dblTouchEvent: true,
			onReady: function (sortableElem) {
				var board = $L(sortableElem)[0].closest("lyte-board")
				if (self.getMethods('onReadyForCard')) {
					return self.executeMethod('onReadyForCard', sortableElem, board);
				}
			},
			onSelect: function (currentElem, fromIndex, source, event) {
				if ($L(currentElem).hasClass('lyteKanbanNoResultMsg')) {
					return false;
				}

				window._lyteUiUtils.closeAllPopups()
				var card = currentElem.querySelector('lyte-card'), res = true

				if (self.getMethods('onDragSelectForCard')) {
					res = self.executeMethod('onDragSelectForCard', card, fromIndex, source, event);

				}
				if (res && (self.getData('ltPropIntersection').card || self.getData('ltPropIntersection').board)) {
					var kanbanItem = $L(currentElem)[0].closest('lyte-kanban-item')
					kanbanItem && kanbanItem.classList.add('lyteKanbanSortableParent')
				}
				return res;
			},
			onDrag: function (draggableElement, belowElem, event, placeholder) {
				self.prevent = true
				var card = draggableElement.querySelector('lyte-card')
				if (self.getMethods('onDragForCard')) {
					self.executeMethod('onDragForCard', card, belowElem, event, placeholder);
				}
			},
			onEnter: function (event, obj) {
				var element = arguments[1].sortable,
					boardDetails = self.getData('ltPropBoardDetails')

				if (element.getAttribute('cards-length') == 0) {
					var div = element.nextElementSibling
					if ($L(div).hasClass('lyteKanbanNoResultMsg')) {
						div.classList.add('lyteHide')
					}
				}
				if (self.getMethods('onEnterForCard')) {
					return self.executeMethod('onEnterForCard', boardDetails, event, obj);
				}
			},
			onLeave: function (event, obj) {
				var element = arguments[1].sortable,
					boardDetails = self.getData('ltPropBoardDetails')

				if (element.getAttribute('cards-length') == 0) {
					var div = element.nextElementSibling
					if ($L(div).hasClass('lyteKanbanNoResultMsg')) {
						div.classList.remove('lyteHide')
					}
				}
				if (self.getMethods('onLeaveForCard')) {
					return self.executeMethod('onLeaveForCard', boardDetails, event, obj);
				}
			},
			onBeforeDrop: function (droppableElement, belowElement, placeholderElement, fromIndex, toIndex, source, destination) {
				var sourceIndex = source.getAttribute('index'),
					destinationIndex = destination.getAttribute('index'),
					boardDetails = self.getData('ltPropBoardDetails'),
					sourceArray = boardDetails[sourceIndex].cards,
					destArray = boardDetails[destinationIndex].cards,
					flag = true;
				if (self.getMethods('onBeforeDropForCard')) {
					flag = self.executeMethod('onBeforeDropForCard', boardDetails, sourceArray, destArray, droppableElement, belowElement, placeholderElement, fromIndex, toIndex, source, destination);
				}
				if (!flag && (self.getData('ltPropIntersection').card || self.getData('ltPropIntersection').board)) {
					var kanbanItem = $L(droppableElement)[0].closest('lyte-kanban-item')
					kanbanItem && kanbanItem.classList.remove('lyteKanbanSortableParent')
				}
				return flag;
			},
			onDrop: function (droppedElement, destination, belowElement, fromIndex, toIndex, source) {
				var sourceIndex = source.getAttribute('index'),
					destinationIndex = destination.getAttribute('index'),
					boardDetails = self.getData('ltPropBoardDetails'),
					sourceArray = boardDetails[sourceIndex].cards,
					destArray = boardDetails[destinationIndex].cards,
					draggedElement
				if (destArray.length == 0) {
					toIndex = 0;
				}
				draggedElement = (self.hasSameColumnReordered(source, destination) ? arrayUtils(destArray, 'splice', fromIndex, 1) : arrayUtils(sourceArray, 'splice', fromIndex, 1))
				arrayUtils(destArray, 'splice', toIndex, 0, draggedElement[0])
				delete self.prevent

				self.onRecordDrop(boardDetails, sourceArray, destArray, draggedElement, fromIndex, toIndex, source, destination, droppedElement)
				if ((self.getData('ltPropIntersection').card || self.getData('ltPropIntersection').board)) {
					var kanbanItem = $L(source)[0].closest('lyte-kanban-item')
					kanbanItem && kanbanItem.classList.remove('lyteKanbanSortableParent')
				}
			},
			onPlaceholder: function (draggableElement, placeholderElement, source, destination) {
				var card = draggableElement.querySelector('lyte-card'),
					boardDetails = self.getData('ltPropBoardDetails')

				if (self.getMethods('onPlaceholderForCard')) {
					return self.executeMethod('onPlaceholderForCard', card, placeholderElement, source, destination, boardDetails);
				}
				return true
			}
		}
		if (this.getData('ltPropSortableCancel')) {
			sortableObject.cancel = this.getData('ltPropSortableCancel')

		}
		if (this.getData('ltPropSortableRestrict')) {
			sortableObject.restrict = this.getData('ltPropSortableRestrict')
		}
		$L('.lyteKanbanNestedSortable', this.$node).sortable(sortableObject);
		// }

	}

    isDragging() {
		if (this.prevent) {
			return true
		}
		return false
	}

    hasScrollEndReached(event) {
		var scrollDiv = this.$node.querySelector('.lyteKanbanviewScrollDivSelector.' + this.getData('dummyId')),
			boardDetails = this.getData('ltPropBoardDetails'),
			clientRect = this.$node.getBoundingClientRect(),
			lastClient = scrollDiv.lastElementChild.getBoundingClientRect()
		if ((!this._dir && lastClient.right - 3 < clientRect.right) || (this._dir && lastClient.left + 3 > Math.min(window.innerWidth, clientRect.left))) {

			if (this.getData('ltPropMoreStageRecord') && this.getMethods('onBodyScroll')) {
				this.executeMethod('onBodyScroll', this, boardDetails, event);

			}
		}

	}

    addSortableForNewBoards() {
		var sortableClass = this.$node.querySelectorAll('.lyteKanbanviewScrollDivSelector.' + this.getData('dummyId'))[0].getSortableClass(),
			boardWithoutSortable = this.$node.querySelectorAll('.lyteKanbanViewItem:not(.' + sortableClass + ')')

		$L(boardWithoutSortable).addClass("sortable-element " + sortableClass);
	}

    executeScrollStop(event) {

		var visible, boardDetails = this.getData('ltPropBoardDetails'),
			scrollDiv = this.$node.querySelector('.lyteKanbanviewScrollDivSelector.' + this.getData('dummyId'))
		if (scrollDiv.scrollWidth > scrollDiv.clientWidth) {
			visible = this.getVisibleNode()
		}
		else {
			visible = boardDetails
			this.callOnBoardVisible()

		}
	}

    static actions(arg1) {
		return Object.assign(super.actions({
			onBodyScroll: function (ev) {
				this.timeout2 = setTimeout(function () {
					this.hasScrollEndReached(ev)
				}.bind(this), 10);
				clearTimeout(this.debounceTimeout)
				this.debounceTimeout = setTimeout(function () {
					this.executeScrollStop(ev)
				}.bind(this), this.getData('ltPropBoardScrollStopDuration'))
			}
		}), arg1);
	}

    static observers(arg1) {
		return Object.assign(super.observers({
			ariaObs: function () {
				if (this.getData('ltPropAria')) {
					window._lyteUiUtils.setAttribute(this.$node.querySelector('.lyteKanbanviewScrollDivSelector'), this.getData('ltPropAriaAttributes') || {}, {});
				}
			}.observes('ltPropAriaAttributes').on('didConnect'),

			boardObs: function (changeObj) {
				if (this.getData('ltPropIntersection').board) {
					if (this.getData('visibleBoards').length < this.getData('rowsPerLoad')) {
						this.setVisibleArray()
					} else {
						if (changeObj.insertedItems && changeObj.insertedItems.length > 0 && this.getData('visibleBoards').length == changeObj.index) {
							arrayUtils(this.getData('visibleBoards'), 'push', true)
						}
						if (this.getData('visibleBoards').length > this.getData('ltPropBoardDetails').length) {
							arrayUtils(this.getData('visibleBoards'), 'removeAt', this.getData('visibleBoards').length - 1)
						}
					}

					if (this.getData('ltPropSortable').board) {
						this.addSortableForNewBoards()
					}
					const sortableParent = this.$node.querySelectorAll('.lyteKanbanNestedSortable.sortable-parent')

					if (!this.getData('ltPropPreventBoardDetailObserver') && this.getData('ltPropSortable').card && (!this.getData('ltPropIntersection').card || sortableParent.length == 0)) {
						clearTimeout(this._sortableTimeout)
						this._sortableTimeout = setTimeout(function () {
							this.setupSortableForCard()
							this.addAllKanbanItemToObserver()

						}.bind(this), 100)

					} else {
						this.addAllKanbanItemToObserver()

					}
					return
				}

				if (this.getData('ltPropSortable').board) {
					this.addSortableForNewBoards()
				}
				if (!this.getData('ltPropPreventBoardDetailObserver') && this.getData('ltPropSortable').card) {
					// $L( '.lyteKanbanviewScrollDivSelector ', this.$node ).sortable("destroy")
					clearTimeout(this._sortableTimeout)
					this._sortableTimeout = setTimeout(function () {
						this.setupSortableForCard()
					}.bind(this), 100)

				}
			}.observes('ltPropBoardDetails.[]')
		}), arg1);
	}

    _() {
        _;
    }
}

LyteKanbanviewComponent._template = "<template tag-name=\"lyte-kanbanview\"> <div class=\"lyteKanbanviewScrollDivSelector {{dummyId}} {{if(ltPropBoardIntersection,'','')}}\" onscroll=\"{{action('onBodyScroll',event)}}\"> <template items=\"{{ltPropBoardDetails}}\" item=\"item\" index=\"dataIndex\" is=\"for\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropIntersection.board}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-kanban-item class=\"lyteKanbanViewItem {{item[ltPropKanbanItemClassKey]}} {{if(item.collapse,'lyteKanbanBoardCollapse','')}}\" data-index=\"{{dataIndex}}\" style=\"width: {{ltPropBoardWidth}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{visibleBoards[dataIndex]}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-yield yield-name=\"kanbanYield\" lyte-board-item=\"{{item}}\" lyte-index=\"{{dataIndex}}\" lyte-kanban-id=\"{{dummyId}}\"></lyte-yield></template></template></lyte-kanban-item> </template><template default=\"\"> <div class=\"lyteKanbanViewItem {{item[ltPropKanbanItemClassKey]}} {{if(item.collapse,'lyteKanbanBoardCollapse','')}}\" data-index=\"{{dataIndex}}\"> <lyte-yield yield-name=\"kanbanYield\" lyte-board-item=\"{{item}}\" lyte-index=\"{{dataIndex}}\" lyte-kanban-id=\"{{dummyId}}\"></lyte-yield> </div> </template></template></template> </div> </template>";;
LyteKanbanviewComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"f","p":[1,1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"concat","args":["'width: '","ltPropBoardWidth"]}}},"cn":"lc_id_0"},{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"i","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"a","p":[1,1],"cn":"default"},{"t":"i","p":[1,1],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;

LyteKanbanviewComponent._observedAttributes = [
    "ltPropBoardDetails",
    "ltPropSortable",
    "ltPropMoreStageRecord",
    "ltPropId",
    "ltPropPreventBoardDetailObserver",
    "ltPropViewPort",
    "ltPropBoardScrollStopDuration",
    "ltPropAria",
    "ltPropAriaAttributes",
    "ltPropSortableCancel",
    "ltPropSortableRestrict",
    "dummyId",
    "ltPropKanbanItemClassKey",
    "ltPropIntersection",
    "ltPropBoardWidth",
    "visibleBoards",
    "lastTopIndex",
    "lastBottomIndex",
    "currentStartIndex",
    "currentEndIndex",
    "rowsPerLoad"
];

if (!window._lyteUiUtils.registeredCustomElements['lyte-kanban-item']) {
    window._lyteUiUtils.registeredCustomElements['lyte-kanban-item'] = true;

    class LyteKanbanItemCustomElements extends LyteUiComponentComponentRegistry.CustomElements {
        lookups() {
			return [{
				component: LyteUiComponentComponentRegistry
			}];
		}

        constructor() {
			super();
		}

        disconnectedCallback() {
			var lyteNode = this.$node;

			if (shouldIgnoreDisconnect()) {
				return
			}
			if (lyteNode._removedChild) {
				document.body.appendChild(lyteNode._removedChild);
				lyteNode._removedChild.remove();
			}
			var kanbanview = $L(lyteNode).closest('lyte-kanbanview').get(0);
			kanbanview && kanbanview.component.removeRowObserve(lyteNode);
		}

        _() {
            _;
        }
    }

    LyteKanbanItemCustomElements.options = { clone: { allCallbacks: false } };

    LyteKanbanItemCustomElements.register('lyte-kanban-item');
}

// if( !_lyteUiUtils.registeredCustomElements[ 'lyte-kanban-item ' ] ) {
// 	_lyteUiUtils.registeredCustomElements[ 'lyte-kanban-item ' ] = true; 
// 	Lyte.createCustomElement("lyte-kanban-item", {
//         static : {

//         },
//         connectedCallback: function() {
//             var query = "lyte-kanbanview",
//             kanbanview = this.closest ? this.closest( query ) : $L( this ).closest( query ).get( 0 );

//             if( kanbanview && kanbanview.getData( 'ltPropBoardIntersection' ) ){
//                 kanbanview.component.addRowObserve(this)
//             }
//         },
//         // createIntersection : function( arg ){
//         //         var table = $L( this ).closest( 'lyte-expresstable' ).get( 0 );
//         //         if( table.getData('ltPropStickyTable') ){
//         //             if( arg && !this._horizontalIntersectionDiv ){
//         //                 this.createIntersection( table );
//         //             } else if( !arg && this._horizontalIntersectionDiv ){
//         //                 var intersection = this._horizontalIntersectionDiv;
//         //                 this.removeIntersection( intersection, table );
//         //                 table.component.removeSticky( this );
//         //                 table.component.removeFixedClass( this );
//         //             }
//         //         }
//         // },
//         // removeIntersection : function( intersection, table ){
//         //         if( table ){
//         //             table.component._intersectionObs.unobserve( intersection );
//         //         }
//         //         intersection.remove();
//         //         delete intersection._cell;
//         //         delete this._horizontalIntersectionDiv;
//         //     },
//         disconnectedCallback : function(){
//             if(Lyte.Component.shouldIgnoreDisconnect()){
//                 return
//             }
//             var kanbanview = $L( this ).closest( 'lyte-kanbanview' ).get( 0 );
//         	kanbanview && kanbanview.component.removeRowObserve(this );
//         },


//     });
// }
/**
 *
 * @syntax yielded
 *	<lyte-kanbanview lt-prop-id="kanbanview">
 *   <template is="registerYield" yield-name="kanbanYield">
 *      <lyte-board lt-prop-kanban-id="kanbanview">
 *           <template is="registerYield" yield-name="headerItem">
 *           Board 1
 *           </template>
 *           <template is="registerYield" yield-name="contentItem">
 *               <lyte-card >
 *                   <template is="registerYield" yield-name="yield">
 *                       <lyte-card-body>
 *		    					Card 1
 *                       </lyte-card-body>
 *                   </template>
 *				</lyte-card>
 *				<lyte-card >
 *                   <template is="registerYield" yield-name="yield">
 *                       <lyte-card-body>
 *		    					Card 2
 *                       </lyte-card-body>
 *                   </template>
 *               </lyte-card>
 *          </template>
 *       </lyte-board>
 *       <lyte-board lt-prop-kanban-id="kanbanview">
 *           <template is="registerYield" yield-name="headerItem">
 *           Board 2
 *           </template>
 *           <template is="registerYield" yield-name="contentItem">
 *               <lyte-card >
 *                   <template is="registerYield" yield-name="yield">
 *                       <lyte-card-body>
 *		    					Card 1
 *                       </lyte-card-body>
 *                   </template>
 *				</lyte-card>
 *				<lyte-card >
 *                   <template is="registerYield" yield-name="yield">
 *                       <lyte-card-body>
 *		    					Card 2
 *                       </lyte-card-body>
 *                   </template>
 *               </lyte-card>
 *          </template>
 *       </lyte-board>
 *   </template>
 *   </lyte-kanbanview>
 */
export { LyteKanbanviewComponent };

LyteKanbanviewComponent.register("lyte-kanbanview", {
    hash: "LyteKanbanviewComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
