// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/18cbda377322d4cb30f6d33bad98aa67.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../component.js";

class LyteSvgComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
			ltPropViewBox: prop( 'string', { 'default': '' } ),
			ltPropClass: prop( 'string', { 'default': '' } ),
			ltPropPath: prop( 'string' ),
			ltPropId: prop( 'string', { 'default': '' } ),
			flag: prop( 'boolean', { default: false } )
		}), arg1);		
	}

    didConnect() {
		this.setData( 'flag', true );
	}

    _() {
        _;
    }
}

LyteSvgComponent._template = "<template tag-name=\"lyte-svg\"> <svg class=\"{{ltPropClass}}\" viewBox=\"{{ltPropViewBox}}\"> <use id=\"{{ltPropId}}\" xlink:href=\"{{lyteUiSetURL(ltPropPath,flag)}}\"></use> </svg> </template>";;
LyteSvgComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"a","p":[1,1]}];;
LyteSvgComponent._observedAttributes = ["ltPropViewBox", "ltPropClass", "ltPropPath", "ltPropId", "flag"];
export { LyteSvgComponent };
LyteSvgComponent.register("lyte-svg", {
    hash: "LyteSvgComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
