// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/26e5636e32bdcf3db6d3750921903a53.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../../../node_modules/@slyte/core/index.js";
import { objectUtils } from "../../../../../node_modules/@slyte/component/index.js";
import { Component } from "../../../../component.js";

class LyteCustomFilterComponent extends Component {
    constructor() {
        super();
    }

    init() {
		this.setData( 'ltPropCondition', { isValid : false, value : "", label : "", type : "custom" } );
	}

    data(arg1) {

		return Object.assign(super.data({
			ltPropCondition : prop( 'object' ),
			ltPropReset : prop( 'boolean', { default : false } ),
			ltPropData : prop( 'object', { default : {} } )
		}), arg1);		
	}

    static observers(arg1) {
        return Object.assign(super.observers({
            reset_obs : function( arg ){
                if( arg.newValue ){
                    var cb = "onCustomFilterReset",
                    __data = this.data,
                    Lc = objectUtils,
                    condition = __data.ltPropCondition;

                    Lc( condition, 'add', 'isValid', false );
                    Lc( condition, 'add', 'label', "" );
                    Lc( condition, 'add', 'value', "" );

                    this.getMethods( cb ) && this.executeMethod( cb, __data.ltPropData, condition, this.$node );
                    this.setData( arg.item, false );
                }
            }.observes( 'ltPropReset' )
        }), arg1);
    }

    _() {
        _;
    }
}

LyteCustomFilterComponent._template = "<template tag-name=\"lyte-custom-filter\"> <lyte-yield yield-name=\"lyte-custom-filter\" lt-prop-data=\"{{ltPropData}}\" lt-prop-condition=\"{{ltPropCondition}}\"></lyte-yield> </template>";;
LyteCustomFilterComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"i","p":[1],"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;
LyteCustomFilterComponent._observedAttributes = ["ltPropCondition", "ltPropReset", "ltPropData"];
export { LyteCustomFilterComponent };
LyteCustomFilterComponent.register("lyte-custom-filter", {
    hash: "LyteCustomFilterComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
