// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/2b422ea731a5666971918ec923c7e8d6.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../../../node_modules/@slyte/core/index.js";
import { objectUtils } from "../../../../../node_modules/@slyte/component/index.js";
import $L from "../../../../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";
import { Component } from "../../../../component.js";

class LyteTextFilterComponent extends Component {
    constructor() {
        super();
    }

    init() {
		var __data = this.data;
		__data.isSearch = Object.keys( __data.ltPropOptions ).length >= __data.ltPropSearchCount;

		if( window._lyteUiUtils.lyteUiIsEmpty( __data.ltPropCondition ) ){
			this.setData( 'ltPropCondition', { input : "", isNeg : false, isValid : false, value : "", label : "", type : "text" } );
		}else{
			this.updateInitialConditions();
		}

		this.setData( 'ltPropCondition', { input : "", isNeg : false, isValid : false, value : "", label : "", type : "text" } );
	}

    data(arg1) {
		var __string = "string",
		__array = "array",
		__object = "object",
		__boolean = "boolean",
		__number = 'number';

		return Object.assign(super.data({
			ltPropCondition : prop( __object ),
			ltPropPlaceholder : prop( __string, { default : "None" } ),
			ltPropOptions : prop( __object, { default : {
					set : window._lyteUiUtils.i18n( 'set', 'listview.filter', "Set" ),
					not_set : window._lyteUiUtils.i18n( 'not.set', 'listview.filter', "Not set" ),
					equal : window._lyteUiUtils.i18n( "equal", "listview.filter", "Equal" ),
					not_equal : window._lyteUiUtils.i18n( "not.equal", "listview.filter", "Not equal" ),
					begins_with : window._lyteUiUtils.i18n( 'bigins', 'listview.filter', "Starts with" ),
					contains : window._lyteUiUtils.i18n( "contains", 'listview.filter', "Contains" ),
					does_not_contains : window._lyteUiUtils.i18n( "not.contains", "listview.filter", "Does not contains" )
				} 
			} ),

			ltPropNoMatch : prop( __string, { default : window._lyteUiUtils.i18n( 'no.results.found', void 0, 'No Results Found' ) } ),

			ltPropSearchCount : prop( __number, { default : 8 } ),

			ltPropReset : prop( __boolean, { default : false } ),

			ltPropTempCondition : prop( __object ),

			selected : prop( __string, { default : "" } ),
			renderFirst : prop( __boolean ),
			noResult : prop( __boolean ),

			isSearch : prop( __boolean ),
			hideFirst : prop( __string, { default : "lyteSearchHidden" } )
		}), arg1);		
	}

    updateInitialConditions() {
		this.setData('ltPropTempCondition', this.data.ltPropCondition);
		this.setData( 'selected', this.data.ltPropCondition.value );

		let condition = this.data.ltPropCondition;
		let selected = this.data.selected;
		let isValid = condition.isValid;

		let renderFirst = false, second = '', match;

		switch( selected ){
			case 'equal':
			case 'not_equal':
			case 'starts_with':
			case 'contains':
			case 'does_not_contains':
				renderFirst = true;
				break;
		}

		this.setData( 'renderFirst', renderFirst );
	}

    static methods(arg1) {
        return Object.assign(super.methods({
            search : function( arg ){
                this.setData( 'noResult', arg.length == 0 );
            }
        }), arg1);
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            reset_obs : function( arg ){
                if( arg.newValue ){
                    this.setData( arg.item, false );
                    this.setData( 'selected', "" );
                }
            }.observes( 'ltPropReset' ),

            condn_obs : function( arg ){
                this.setData( 'selected', arg.newValue.value || "" );
            }.observes( 'ltPropCondition' ),

            selected_obs : function( arg ){
                var selected = arg ? arg.newValue : this.data.selected,
                renderFirst = true,
                isNeg = false,
                __data = this.data,
                hiddenclass = 'lyteSearchHidden',
                condition = __data.ltPropTempCondition,
                Lc = objectUtils,
                ns = "lyteTextFilter_";

                arg && $L( this.$node ).addClass( ns + selected ).removeClass( ns + arg.oldValue );

                switch( selected ){
                    case 'set' : {
                        renderFirst = false; 
                    }
                    break;
                    case 'not_set' : {
                        renderFirst = false;
                        isNeg = true;
                    }
                    break;
                    case 'does_not_contains' : 
                    case 'not_equal' : {
                        isNeg = true;
                    }
                    break;
                    case '' : {
                        renderFirst  = false;
                    }
                    break;
                }

                this.setData({
                    renderFirst : __data.renderFirst || renderFirst,
                    hideFirst : renderFirst ? '' : hiddenclass
                });

                Lc( condition, 'add', 'isNeg', isNeg );
                arg && Lc( condition, 'add', 'input', '' );
                Lc( condition, 'add', 'type', 'text' );

                if( !renderFirst ){
                    Lc( condition, 'add', 'isValid', !!selected );
                }

                Lc( condition, 'add', 'value', selected );
                Lc( condition, 'add', 'label', __data.ltPropOptions[ selected ] || "" );

            }.observes( 'selected' ).on('init'),

            obs : function(){
                var __data = this.data,
                condition = __data.ltPropTempCondition,
                value = condition.input,
                isValid = true;

                if( !value && !__data.hideFirst ){
                    isValid = false;
                } 

                objectUtils( condition, 'add', 'isValid', isValid );
                objectUtils( condition, 'add', 'class', ( condition.isValid && value  ) ? 'lyteListFilterSelected' : '' );


            }.observes( 'hideFirst', 'ltPropTempCondition.input' )
        }), arg1);
    }

    _() {
        _;
    }
}

LyteTextFilterComponent._template = "<template tag-name=\"lyte-text-filter\"> <lyte-dropdown lt-prop-selected=\"{{lbind(selected)}}\" lt-prop-placeholder=\"{{ltPropPlaceholder}}\"> <template is=\"registerYield\" yield-name=\"yield\"> <lyte-drop-box class=\"lyteTextFilterDropdown\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{isSearch}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-search lt-prop-query-selector=\"{ &quot;scope&quot; : &quot;.lyteTextFilterDropdown:not(.lyteDropdownHidden)&quot;, &quot;search&quot; : &quot;lyte-drop-item&quot; }\" on-search=\"{{method('search')}}\"></lyte-search> </template></template><lyte-drop-body> <template object=\"{{ltPropOptions}}\" value=\"value\" key=\"key\" is=\"forIn\" _new=\"true\"> <lyte-drop-item data-value=\"{{key}}\">{{value}}</lyte-drop-item> </template> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{isSearch}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"{{if(noResult,'lyteTextFilterNoResult','lyteTextFilterNoResult lyteSearchHidden')}}\">{{ltPropNoMatch}}</div> </template></template></lyte-drop-body> </lyte-drop-box> </template> </lyte-dropdown> <div class=\"lyteTextFilterWrapper\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{renderFirst}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-input lt-prop=\"{{stringify(ltPropFilterElementProperties.text)}}\" lt-prop-appearance=\"box\" class=\"lyteTextFilterFirst\" lt-prop-input-wrapper-class=\"{{hideFirst}}\" lt-prop-format=\"{{ltPropFormat}}\" lt-prop-value=\"{{lbind(ltPropTempCondition.input)}}\" lt-prop-bind-to-body=\"false\"></lyte-input> </template></template></div> </template>";;
LyteTextFilterComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"r","p":[1,1],"dN":[{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":4,"sibl":[3]},{"t":"a","p":[1,2,1]},{"t":"fI","p":[1,2,1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,0]},{"t":"cD","p":[1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":3,"sibl":[2]},{"t":"s","p":[1,2,3],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1]},{"t":"cD","p":[1,2],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[4,3,1,0],"hc":true,"trans":true,"in":2,"sibl":[1]},{"t":"cD","p":[1],"in":1,"sibl":[0]},{"t":"s","p":[3,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[2,1,0]}];;

LyteTextFilterComponent._observedAttributes = [
    "ltPropCondition",
    "ltPropPlaceholder",
    "ltPropOptions",
    "ltPropNoMatch",
    "ltPropSearchCount",
    "ltPropReset",
    "ltPropTempCondition",
    "selected",
    "renderFirst",
    "noResult",
    "isSearch",
    "hideFirst"
];

export { LyteTextFilterComponent };

LyteTextFilterComponent.register("lyte-text-filter", {
    hash: "LyteTextFilterComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
