// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/2b970b21a318a271465cc501cb2ab166.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import "../../node_modules/@slyte/component/src/directives/lyte-view.js";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component, LyteUiComponentComponentRegistry } from "../component.js";
import $L from "../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";

import {
	arrayUtils,
	appendChild,
	shouldIgnoreDisconnect,
	_setIgnoreDisconnect
} from "../../node_modules/@slyte/component/index.js";

/**
 * Renders an board
 * @component lyte-board
 * @version 3.1.0
 * @methods onBoardScroll
 * @utility getVisibleCard, collapse, expand
 */
class LyteBoardComponent extends Component {
    constructor() {
        super();
        this._lyteUtilFunctions = ["getVisibleCard", "collapse", "expand"];
    }

    data(arg1) {
		return Object.assign(super.data({
			/**
		  * @componentProperty {object} ltPropBoardDetail
		  * @version 3.1.0
		  */
			lyteViewPort: prop("boolean", { "default": true }),//No I18n
			'ltPropBoardDetail': prop('object'),
			/**
			* @componentProperty {number} ltPropIndex
			* @version 3.1.0
			*/
			'ltPropIndex': prop('number'),
			/**
			* @componentProperty {boolean} ltPropBoardSortable=true
			* @version 3.1.0
			*/
			'ltPropBoardSortable': prop('boolean', {
				'default': window._lyteUiUtils.resolveDefaultValue('lyte-board', 'boardSortable', true)
			}),
			/**
			* @componentProperty {string} ltPropClass
			* @version 3.1.0
			*/
			'ltPropClass': prop('string', {
				'default': window._lyteUiUtils.resolveDefaultValue('lyte-board', 'class', '')
			}),
			/**
			* @componentProperty {string} ltPropKanbanId=''
			* @version 3.1.0
			*/
			'ltPropKanbanId': prop('string', {
				'default': ''
			}),
			/**
			* @componentProperty {boolean} ltPropMoreStageRecord=false
			* @version 3.1.0
			*/
			'ltPropMoreStageRecord': prop('boolean', {
				'default': false
			}),
			/**
			* @componentProperty {string} ltPropNoResultMessage=''
			* @version 3.1.0
			*/
			'ltPropNoResultMessage': prop('string', { 'default': window._lyteUiUtils.resolveDefaultValue('lyte-dropdown', 'noResultMessage', window._lyteUiUtils.i18n('no.results.found')) }),
			/**
			* @componentProperty {boolean} ltPropLoadingYield=true
			* @version 3.1.0
			*/
			'ltPropLoadingYield': prop('boolean', {
				'default': true
			}),
			/**
			* @componentProperty {string} ltPropCardClassName
			* @version 3.1.0
			*/
			'ltPropCardClassName': prop('string', { 'default': undefined }),
			/**
			* @componentProperty {boolean} ltPropAria
			* @version 3.1.0
			*/
			'ltPropAria': prop('boolean', {
				'default': false
			}),
			/**
			 * @componentProperty {object} ltPropBoardAria
			 * @version 3.1.0
			 */
			'ltPropBoardAria': prop('object', {
				default: {}
			}),
			/**
			* @componentProperty {object} ltPropCardAria
			* @version 3.1.0
			*/
			'ltPropCardAria': prop('object', {
				default: {}
			}),
			'ltPropNoResultYield': prop('boolean', {
				default: false
			}),
			'ltPropCardIntersection': prop('boolean', {
				default: false
			}),
			'ltPropCardHeight': prop('string', { default: '150px' }),

			visibleCards: prop('array', { default: [] }),
			rowsPerLoad: prop('number', { default: 10 }),
			'lastTopIndex': prop('number', {
				'default': -1
			}),
			'lastBottomIndex': prop('number', {
				'default': -1
			}),
			'currentStartIndex': prop('number', {
				'default': 0
			}),
			'currentEndIndex': prop('number', {
				'default': 0
			}),
			/**
			* @experimental dummyId
			*/
			'dummyId': prop('string', {
				'default': ''
			}),
			/**
			* @experimental cardArray
			*/
			'cardArray': prop('array', {
				'default': []
			}),
			'ignoreObserve': prop('boolean', {
				default: false
			})
		}), arg1);
	}

    init() {

	}

    setVisibleArray() {
		let visibleCards = this.getData('visibleCards')
		var boardDetail = this.getData('ltPropBoardDetail'), len
		if (boardDetail && boardDetail.cards) {
			var rowsPerLoad = Math.max(this.getData('rowsPerLoad'), this.getVisibleRowCount() + 1)
			len = boardDetail.cards ? Math.min(boardDetail.cards.length, rowsPerLoad) : 0
		}
		for (let i = 0; i < len; i++) {
			arrayUtils(visibleCards, 'push', true)
		}
	}

    didConnect() {
		if (this.getData('ltPropCardIntersection')) {
			this.setData('lyteViewPort', false)
			this.setVisibleArray()
		}
		this.setCardArray()
		this.$node.getVisibleCard = function () {
			var scrollDiv = this.$node.getElementsByClassName('lyteKanbanNestedSortable')[0]
			if (scrollDiv.scrollHeight > scrollDiv.clientHeight) {
				return this.getVisibleNode()

			}
			return (this.getData('ltPropBoardDetail').cards ? this.getData('ltPropBoardDetail').cards : [])

		}.bind(this)
		this.$node.collapse = function () {
			var kanbanviewItem = $L(this.$node).closest('.lyteKanbanViewItem')[0]
			if (kanbanviewItem) {
				kanbanviewItem.classList.add('lyteKanbanBoardCollapse')
			}

		}.bind(this)
		this.$node.expand = function () {
			var kanbanviewItem = $L(this.$node).closest('.lyteKanbanViewItem')[0]
			if (kanbanviewItem) {
				kanbanviewItem.classList.remove('lyteKanbanBoardCollapse')
			}
		}.bind(this)
		this.$node.toggleIgnore = function () {
			this.setData('ignoreObserve', !this.getData('ignoreObserve'))
		}.bind(this)
		if (this.getData('ltPropCardIntersection')) {
			var kanbanview = $L(this.$node).closest('lyte-kanbanview')[0]
			if (kanbanview) {
				kanbanview.component.callSetUpCardFunc()
			}
			this.createObserver()
			this.addAllCardItemToObserver()

		}
	}

    addAria() {
		if (this.getData('ltPropAria')) {
			var self = this
			window._lyteUiUtils.setAttribute(this.$node.querySelector('.lyteBoardWrapper'), this.getData('ltPropBoardAria') || {}, {});
			this.$node.querySelectorAll('.lyteBoardItemContentData').forEach(function (item) {
				window._lyteUiUtils.setAttribute(item, self.getData('ltPropCardAria') || {}, {});
			})

		}
	}

    didDestroy() {
		this._observer && this._observer.disconnect()
		delete this._observer
		clearTimeout(this.timeout1);
		clearTimeout(this.debounceTimeout)
	}

    addAllCardItemToObserver() {
		const kanbanItems = this.$node.querySelectorAll('lyte-card-item')
		for (let idx = 0; idx < kanbanItems.length; idx++) {
			this.addRowObserve(kanbanItems[idx])
		}
	}

    removeAllCardItemToObserver() {
		const kanbanItems = this.$node.querySelectorAll('lyte-card-item')
		for (let idx = 0; idx < kanbanItems.length; idx++) {
			this.removeRowObserve(kanbanItems[idx])
		}
	}

    addRowObserve(row) {
		// console.log(row)
		this._observer.observe(row)
	}

    removeRowObserve(row) {
		this._observer.unobserve(row)
	}

    addVisibleRows(cardItem) {
		if (this.getData('ignoreObserve')) {
			return
		}
		const boardDetail = this.getData('ltPropBoardDetail')
		let visibleCards = this.getData('visibleCards')
		let indexToShow = cardItem.getAttribute('data-index')
		if (indexToShow > visibleCards.length - 1) {
			arrayUtils(visibleCards, 'push', true)
			this.addSortableForNewRecords()

		} else if (visibleCards[indexToShow] && cardItem._removedChild) {
			appendChild(cardItem, cardItem._removedChild)
			delete cardItem._removedChild
		}


	}

    removeNotVisibleRows(row) {
		if (this.getData('ignoreObserve')) {
			return
		}
		let visibleCards = this.getData('visibleCards')
		const index = row.getAttribute('data-index')
		const kanbanItem = row.closest('lyte-kanban-item')

		if (visibleCards[index] && !row._removedChild && !row.classList.contains('sortable-element-selected') && !row.classList.contains('lyteKanbanSortableParent')) {
			_setIgnoreDisconnect(true);

			if (row.children.length > 0 && row.children[0]) {
				row._removedChild = row.removeChild(row.children[0])
			}
			_setIgnoreDisconnect(false);
			return true
		}
		return false
	}

    getVisibleRowCount() {
		const containerHeight = this.$node.querySelector('.lyteBoardContainer').clientHeight;
		const rowHeight = parseInt(this.getData('ltPropCardHeight')) || this.$node.querySelector('lyte-card-item').clientHeight; // Assuming all rows have the same height
		return Math.max(1, Math.floor(containerHeight / rowHeight));;
	}

    updateVisibility(newTopIndex) {
		const visibleCount = this.getVisibleRowCount();
		const bottomIndex = this.getData('currentEndIndex');
		let topIndex = this.getData('currentStartIndex');
		const bufferSize = 2
		let cards = this.getData('ltPropBoardDetail') ? this.getData('ltPropBoardDetail').cards : []

		// const newBottomIndex = newTopIndex + visibleCount + bufferSize * 2 - 1;
		const newBottomIndex = Math.min(cards.length - 1, newTopIndex + visibleCount - 1);

		if (newTopIndex === topIndex) {
			return; // No change, skip update
		}

		// **Calculate Buffered Range**
		const start = Math.max(0, newTopIndex - bufferSize);
		const end = Math.min(cards.length - 1, newBottomIndex + bufferSize);

		const kanbanItems = this.$node.querySelectorAll('lyte-card-item')
		kanbanItems.forEach((row, index) => {
			if (index >= start && index <= end) {
				this.addVisibleRows(row)
			} else {
				this.removeNotVisibleRows(row)
			}
		});

		// topIndex = newTopIndex;
		// bottomIndex = newBottomIndex;
		this.setData('currentStartIndex', newTopIndex)
		this.setData('currentEndIndex', newBottomIndex)
		// this.setData('lastTopIndex', newTopIndex)
		// this.setData('lastBottomIndex', newBottomIndex)
	}

    intersectionObserverFunc(entries) {
		const self = this
		let topIndex = this.getData('currentStartIndex');
		let bottomIndex = this.getData('currentEndIndex');
		let boardDetail = this.getData('ltPropBoardDetail')
		let cards = boardDetail ? boardDetail.cards : []
		let bufferSize = 1
		if (!cards || cards.length == 0) {
			return
		}
		entries.forEach((entry) => {
			if (entry.isIntersecting) {
				// const rowIndex = Math.abs(entry.target.getAttribute('data-index'));

				// // const visibleCount = this.getVisibleRowCount();
				// let newTopIndex = topIndex;
				// if (rowIndex <= topIndex && topIndex > 0) {
				// 	newTopIndex = Math.max(0, topIndex - bufferSize);
				//   }

				//   // **Scroll Down: Load next rows**
				//   if (rowIndex >= bottomIndex  && bottomIndex < cards.length - 1) {
				// 	newTopIndex = Math.min(cards.length - bufferSize, topIndex + bufferSize);
				//   }

				// this.setData('currentStartIndex',topIndex)
				// this.updateVisibility(newTopIndex);
				if (entry.target.classList.contains('lyteSortablePlaceholder')) {
					return
				}
				this.addVisibleRows(entry.target)
			} else {
				if (entry.target.classList.contains('lyteSortablePlaceholder')) {
					return
				}
				this.removeNotVisibleRows(entry.target)
			}

		});
	}

    createObserver() {
		const cardHeight = parseInt(this.getData('ltPropCardHeight')) * 2
		const container = this.$node.querySelector('.lyteKanbanNestedSortable')
		const options = {
			root: container,
			threshold: 0.01, // Trigger when 10% of a row is visible
			rootMargin: cardHeight + "px 0px " + cardHeight + "px 0px"
		};
		this._observer = new window.IntersectionObserver(this.intersectionObserverFunc.bind(this), options);
	}

    getVisibleNode() {
		// return;
		var bcr = this.$node.querySelector('.lyteBoardContainer').getBoundingClientRect(),
			originalRows = Array.from(this.$node.getElementsByClassName('lyteBoardItemContentData')),
			tValue = Math.max(bcr.top + 10, -10),
			bValue = Math.min(window.innerWidth + 10, bcr.bottom),
			visible = [], boardDetails = this.getData('ltPropBoardDetail').cards;




		for (var i = 0; i < originalRows.length; i++) {
			var row = originalRows[i],
				_bcr = row.getBoundingClientRect(),
				index = this.getData('ltPropCardIntersection') ? parseInt(row.getAttribute('data-index')) : i;
			if (_bcr.bottom > tValue && _bcr.top < bValue) {
				visible.push(boardDetails[index]);

			}
		}


		return visible;

	}

    setCardArray() {
		var boardDetail = this.getData('ltPropBoardDetail')
		if (boardDetail && boardDetail.cards) {
			this.setData('cardArray', boardDetail.cards)
		}
	}

    addSortableForNewRecords() {
		var div = this.$node.querySelectorAll('.lyteKanbanNestedSortable.' + this.getData('ltPropKanbanId'))[0]
		if (div.classList.contains('sortable-parent')) {
			var sortableClass = div.getSortableClass(),
				cardWithoutSortable = this.$node.querySelectorAll('.lyteBoardItemContentData:not(.' + sortableClass + ')')
			$L(cardWithoutSortable).map(function (index, element) {
				element.parentNode.addToSortable(element);
			})
		}


	}

    addShadow() {
		this.$node.querySelector('.lyteBoardWrapper ').classList.add('lyteKanbanviewShadow');
		this.$node.querySelector('.lyteBoardHeader').classList.remove('lyteKanbanviewHeaderShadow'); // No I18n

	}

    hasScrollHeightReached(event) {
		if (event.target.scrollHeight - 10 <= (Math.ceil(event.target.offsetHeight) + Math.ceil(event.target.scrollTop))) {
			if (this.getData('ltPropMoreStageRecord') && this.getMethods('onBoardScroll')) {
				this.executeMethod('onBoardScroll', this.getData('ltPropBoardDetail'), this, event); //NO i18n
			}
		}
	}

    removeShadow() {
		this.$node.querySelector('.lyteBoardWrapper ').classList.add('lyteKanbanviewShadow');
		this.$node.querySelector('.lyteBoardHeader').classList.remove('lyteKanbanviewHeaderShadow'); // No I18n
	}

    executeScrollStop(event) {
		if (this.getMethods('onBoardScrollStop')) {
			var visible, boardDetail = this.getData('ltPropBoardDetail'),
				scrollDiv = this.$node.getElementsByClassName('lyteKanbanNestedSortable')[0]
			if (scrollDiv.scrollHeight > scrollDiv.clientHeight) {
				visible = this.getVisibleNode()
			} else {
				visible = boardDetail.cards
			}
			this.executeMethod('onBoardScrollStop', boardDetail, visible, this, scrollDiv.scrollTop, event);
		}
	}

    static actions(arg1) {
		return Object.assign(super.actions({
			boardScroll: function (event) {
				if (event.target.scrollTop != 0) {
					this.addShadow();
				}

				this.timeout1 = setTimeout(function () {

					this.hasScrollHeightReached(event)
				}.bind(this), 10);

				if (event.target.scrollTop == 0) {
					this.removeShadow()
				}
				//debounce
				clearTimeout(this.debounceTimeout)
				this.debounceTimeout = setTimeout(function () {
					this.executeScrollStop(event)
				}.bind(this), 100)
				// event.preventDefault()
				// event.stopPropagation();
			}
		}), arg1);
	}

    static observers(arg1) {
		return Object.assign(super.observers({
			ariaObs: function () {
				if (!this.getData('ltPropLoadingYield')) {
					this.addAria()
				}
			}.observes('ltPropBoardAria', 'ltPropCardAria', 'ltPropLoadingYield').on('didConnect'),

			viewPortObs: function () {
				if (!this.getData('lyteViewPort') && this.getData('ltPropBoardSortable')) {
					var kanbanview = $L(this.$node).closest('lyte-kanbanview')[0]
					if (kanbanview) {
						kanbanview.component.addSortableForCard()
					}
					this.addAria()
				}


			}.observes('lyteViewPort'),

			contentObs: function (changeObj) {
				this.setCardArray()
				if (this.getData('ltPropCardIntersection')) {
					if (this.getData('visibleCards').length < this.getData('rowsPerLoad')) {
						this.setVisibleArray()
					} else {
						if (changeObj.insertedItems && changeObj.insertedItems.length > 0 && changeObj.index >= this.getData('visibleCards').length) {
							arrayUtils(this.getData('visibleCards'), 'push', true)
						}
						if (changeObj.removedItems && changeObj.removedItems.length > 0 && changeObj.index <= this.getData('visibleCards').length - 1) {
							arrayUtils(this.getData('visibleCards'), 'removeAt', changeObj.index)
						}
					}
					if (this.getData('ltPropBoardSortable') && !this.getData('lyteViewPort')) {
						this.addSortableForNewRecords()
					}
					this.addAllCardItemToObserver()
					return
				}

				if (this.getData('ltPropBoardSortable') && !this.getData('lyteViewPort')) {
					this.addSortableForNewRecords()
				}

			}.observes('ltPropBoardDetail.cards.[]')
		}), arg1);
	}

    _() {
        _;
    }
}

LyteBoardComponent._template = "<template tag-name=\"lyte-board\"><template is=\"switch\" __vp=\"c-old\" l-c=\"true\" _new=\"true\"><template __vp=\"c-old\" case=\"{{lyteViewPort}}\" is=\"case\" lc-id=\"lc_id_0\"><dummy-port-element></dummy-port-element> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(ltPropLoadingYield,'!')}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteBoardWrapper\"> <div class=\"lyteKVLoadingHeaderPlaceholder\"></div> <div class=\"lyteBoardContainer\"> <div class=\"lyteKVLoadingPlaceholder\"></div> <div class=\"lyteKVLoadingPlaceholder\"></div> <div class=\"lyteKVLoadingPlaceholder\"></div> </div> </div> </template><template default=\"\"> <lyte-yield yield-name=\"loading\"></lyte-yield> </template></template><dummy-port-element></dummy-port-element></template><template default=\"\"> <div class=\"lyteBoardWrapper {{ltPropClass}}\"> <div class=\"lyteBoardHeader\"> <lyte-yield yield-name=\"headerItem\" board-detail=\"{{ltPropBoardDetail}}\" index=\"{{ltPropIndex}}\"></lyte-yield> </div> <div class=\"lyteBoardContainer\"> <div class=\"lyteKanbanNestedSortable {{ltPropKanbanId}}\" id=\"{{ltPropBoardDetail.id}}\" index=\"{{ltPropIndex}}\" data-loaded=\"{{dataLoaded}}\" cards-length=\"{{cardArray.length}}\" onscroll=\"{{action('boardScroll',event,this)}}\"> <template items=\"{{cardArray}}\" item=\"itemContent\" index=\"index1\" is=\"for\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropCardIntersection}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-card-item class=\"lyteBoardItemContentData {{lyteUiConcat(lyteUiAddSortableClass(ltPropBoardSortable,this),itemContent[ltPropCardClassName])}}\" data-index=\"{{index1}}\" style=\"width: {{ltPropCardWidth}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{visibleCards[index1]}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-yield yield-name=\"contentItem\" lyte-card-item=\"{{itemContent}}\" lyte-card-index=\"{{index1}}\"></lyte-yield></template></template></lyte-card-item> </template><template default=\"\"> <div class=\"lyteBoardItemContentData {{lyteUiConcat(lyteUiAddSortableClass(ltPropBoardSortable,this),itemContent[ltPropCardClassName])}}\"> <lyte-yield yield-name=\"contentItem\" lyte-card-item=\"{{itemContent}}\" lyte-card-index=\"{{index1}}\"></lyte-yield> </div> </template></template></template> </div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(cardArray.length,'==',0)}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(ltPropNoResultYield,'!')}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteKanbanNoResultMsg\">{{ltPropNoResultMessage}}</div> </template><template default=\"\"> <lyte-yield yield-name=\"noResultYield\" board-detail=\"{{ltPropBoardDetail}}\" index=\"{{ltPropIndex}}\"></lyte-yield> </template></template></template></template></div> <div class=\"lyteBoardFooter\"> <lyte-yield yield-name=\"footerItem\" board-detail=\"{{ltPropBoardDetail}}\" index=\"{{ltPropIndex}}\"></lyte-yield> </div> <div class=\"lyteBoardCollapse\"> <lyte-yield yield-name=\"collapseItem\" board-detail=\"{{ltPropBoardDetail}}\" index=\"{{ltPropIndex}}\"></lyte-yield> </div> </div> </template></template></template>";;
LyteBoardComponent._dynamicNodes = [{"t":"s","p":[0],"c":{"lc_id_0":{"dN":[{"t":"cD","p":[0],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"s","p":[2],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"i","p":[1],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[3],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"a","p":[1,1,1],"cn":"default"},{"t":"i","p":[1,1,1],"in":4,"sibl":[3],"cn":"default"},{"t":"a","p":[1,3,1],"cn":"default"},{"t":"a","p":[1,3,1,1],"cn":"default"},{"t":"f","p":[1,3,1,1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"concat","args":["'width: '","ltPropCardWidth"]}}},"cn":"lc_id_0"},{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"i","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"a","p":[1,1],"cn":"default"},{"t":"i","p":[1,1],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":3,"sibl":[2],"cn":"default"},{"t":"s","p":[1,3,3],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[1,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"i","p":[1],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":2,"sibl":[1],"cn":"default"},{"t":"a","p":[1,5,1],"cn":"default"},{"t":"i","p":[1,5,1],"in":1,"sibl":[0],"cn":"default"},{"t":"a","p":[1,7,1],"cn":"default"},{"t":"i","p":[1,7,1],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[2,1,0],"hc":true,"trans":true},"default":{"dc":[4,3,2,1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;

LyteBoardComponent._observedAttributes = [
    "lyteViewPort",
    "ltPropBoardDetail",
    "ltPropIndex",
    "ltPropBoardSortable",
    "ltPropClass",
    "ltPropKanbanId",
    "ltPropMoreStageRecord",
    "ltPropNoResultMessage",
    "ltPropLoadingYield",
    "ltPropCardClassName",
    "ltPropAria",
    "ltPropBoardAria",
    "ltPropCardAria",
    "ltPropNoResultYield",
    "ltPropCardIntersection",
    "ltPropCardHeight",
    "visibleCards",
    "rowsPerLoad",
    "lastTopIndex",
    "lastBottomIndex",
    "currentStartIndex",
    "currentEndIndex",
    "dummyId",
    "cardArray",
    "ignoreObserve"
];

LyteUiComponentComponentRegistry.registerHelper('lyteUiAddSortableClass', function (sortable, item) {
	if (sortable && item.parentNode && item.parentNode.getSortableClass) {
		return "sortable-element " + item.parentNode.getSortableClass() + (item.classList.contains("sortable-element-selected") ? " sortable-element-selected " : " ")
	}
	return ''
});

if (!window._lyteUiUtils.registeredCustomElements['lyte-card-item']) {
    window._lyteUiUtils.registeredCustomElements['lyte-card-item'] = true;

    class LyteCardItemCustomElements extends LyteUiComponentComponentRegistry.CustomElements {
        lookups() {
			return [{
				registry: LyteUiComponentComponentRegistry
			}];
		}

        constructor() {
			super();
		}

        disconnectedCallback() {
			debugger
			var lyteNode = this.$node;
			if (shouldIgnoreDisconnect()) {
				return
			}
			if (lyteNode._removedChild) {
				document.body.appendChild(lyteNode._removedChild);
				lyteNode._removedChild.remove();
			}
			var kanbanview = $L(lyteNode).closest('lyte-board').get(0);
			kanbanview && kanbanview.component.removeRowObserve(lyteNode);
		}

        _() {
            _;
        }
    }

    LyteCardItemCustomElements.options = { clone: { allCallbacks: false } };

    LyteCardItemCustomElements.register('lyte-card-item');
}

export { LyteBoardComponent };

LyteBoardComponent.register("lyte-board", {
    hash: "LyteBoardComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
