// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/31321560bfe44db139063254caad7c6d.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../../node_modules/@slyte/component/index.js";

class LyteCircularSliderComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
        return Object.assign(super.data({
            ltPropValue: prop('number', { default: 0 }),
            ltPropMin: prop('number', { default: 0 }),
            ltPropMax: prop('number', { default: 100 }),
            ltPropProgressWidth: prop('number', { default: 10 }),
            ltPropProgressColor: prop('string', { default: "#42a2eb" }),
            ltPropProgressBackgroundColor: prop('string', { default: "#DCE0E3" }),
            ltPropTextDisplay: prop('string', { default: "value" }),
            ltPropDigits: prop('number', { default: 0 }),
            ltPropStopAtEnd: prop('boolean', { default: false }),
            ltPropStartDirection: prop('string', { default: 'top' }),
            ltPropAtend: prop('string', { default: 'default' }),
            ltPropDisabled: prop('boolean', { default: false }),
            ltPropReadonly: prop('boolean', { default: false }),
            ltPropTooltip: prop('boolean', { default: true }),
            ltPropTooltipConfig: prop('object', { default: { "position": "top" } }),
            ltPropTooltipStyle: prop('string', { default: '' }),
            ltPropTooltipClass: prop('string', { default: '' }),
            ltPropStep: prop('number', { default: 1 }),
            ltPropDiscrete: prop('boolean', { default: false }),
            ltPropAriaLabel: prop('string', { default: '' }),
            ltPropAnimate: prop('boolean', { default: true }),

            dashOffset: prop('number', { default: 345.5749 }),
            textValue: prop('string'),
            radius: prop('number', { default: 55 }),
            dashArray: prop('number', { default: 345.5749 }),
            handlerStyle: prop('string', { default: '' }),
            handlerWrapperStyle: prop('string', { default: '' })
        }), arg1);
    }

    didConnect() {
        var $node = this.$node,
            rotateVal;
        switch (this.data.ltPropStartDirection) {
            case 'right': {
                rotateVal = 0;
                break;
            }
            case 'bottom': {
                rotateVal = 90;
                break;
            }
            case 'left': {
                rotateVal = 180;
                break;
            }
            default: {
                rotateVal = 270;
            }
        }
        $L('.lyteCirSliderWrapper', $node).get(0).setAttribute("transform", "rotate(" + rotateVal + ", 60, 60)");
        this.rotateVal = rotateVal;
        var halfWidth = this.data.ltPropProgressWidth / 2,
            radius = 60 - halfWidth;
        this.setData('radius', radius);
        this.setData('dashArray', 2 * Math.PI * radius);
        this.setData('handlerStyle', 'transform: translate(' + -(halfWidth + $L('.lyteCirSliderHandler', $node).get(0).offsetWidth / 2) + 'px, -50%)');
        this.changeValue(0, true, false, true);
    }

    didDestroy() {
        clearInterval(this.clickInterval);
        ['mainCircle', 'handlerWrapper', 'handler', 'transEnd', 'rotateVal', 'preventClick', 'up', 'move', 'prevAngle', 'actVal', 'avoid', 'prevHandlerPos', 'clickInterval', 'textValue'].forEach(function (item) {
            delete this[item];
        }, this);
    }

    static actions(arg1) {
        return Object.assign(super.actions({
            click: function (event) {
                if (this.preventClick || event.target.tagName != 'circle') {
                    delete this.preventClick;
                    return;
                }
                if (this.getMethods('onBeforeClick') && this.executeMethod('onBeforeClick', event, $node) == false) {
                    return;
                }
                this.angleFromPoint(event.x, event.y);
            },

            mousedown: function (event) {
                var $node = this.$node,
                    data = this.data;
                if (this.getMethods('onBeforeDragStart') && this.executeMethod('onBeforeDragStart', event, $node) == false) {
                    return;
                }
                this.up = this.mouseup.bind(this);
                this.move = this.mousemove.bind(this);
                this.mainCircle = $L('.lyteCirSliderOuterCircle', $node).get(0);
                // !data.ltPropStopAtEnd && this.angleFromPoint(event.x, event.y);
                if (this.prevAngle == undefined) {
                    var bcr = $L('.lyteCirSliderHandler', $node).get(0).getBoundingClientRect();
                    if (data.ltPropValue == data.ltPropMin) {
                        this.prevAngle = 0;
                    } else {
                        this.prevAngle = this.angleFromPoint(bcr.x + (bcr.width / 2), (bcr.y + (bcr.height / 2)), false, true).angle;
                    }
                }
                $L($node).addClass('lyteCircularSliderDrag');
                this.getMethods('onDragStart') && this.executeMethod('onDragStart', event, $node);
                document.addEventListener('mousemove', this.move);
                document.addEventListener('mouseup', this.up);
            },

            keydown: function (event) {
                var data = this.data,
                    value = this.textValue == undefined ? data.ltPropValue : this.textValue,
                    step = data.ltPropStep,
                    updVal;
                // getStepValue = function (value, inverse) {
                //     return (value % step == 0) ? value + (inverse ? -step : step) : inverse ? Math.floor(value / step) * step : Math.ceil(value / step) * step;
                // };
                if (this.getMethods('onBeforeKeydown') && this.executeMethod('onBeforeKeydown', event, this.$node) == false) {
                    return;
                }
                switch (event.key) {
                    case 'ArrowUp':
                    case 'ArrowRight': {
                        var newVal = value + step;
                        updVal = this.checkVal(newVal);
                        break;
                    }
                    case 'ArrowDown':
                    case 'ArrowLeft': {
                        var newVal = value - step;
                        updVal = this.checkVal(newVal);
                        break;
                    }
                    case 'Home': {//fn left
                        updVal = data.ltPropMin;
                        break;
                    }
                    case 'End': {
                        updVal = data.ltPropMax;
                        break;
                    }
                }
                if (updVal != undefined) {
                    this.changeValue(updVal, false, false, false, true);
                    event.preventDefault();
                }
            }
        }), arg1);
    }

    changeValue(custVal, fromObs, setOnlyText, override, avoidAnimation) {//this performs the value change
        var data = this.data,
            min = data.ltPropMin,
            max = data.ltPropMax,
            $node = this.$node,
            discrete = data.ltPropDiscrete,
            step = data.ltPropStep,
            steppedVal = Math.round(custVal / step) * step,
            setVal = this.checkVal(discrete ? steppedVal : custVal);

        if (!setOnlyText && !override && (setVal != custVal)) {
            this.setData('ltPropValue', setVal);
        } else {
            var textType = data.ltPropTextDisplay,
                digits = data.ltPropDigits,
                textVal = digits ? parseFloat(steppedVal).toFixed(digits) : discrete ? parseInt(steppedVal) : Math.floor(custVal),
                percent,
                valueChanged,
                textValue;
            percent = (setVal - min) / (max - min);
            this.textValue = textVal;
            if (textType == "value") {
                textValue = textVal;
            } else if (textType == 'percent') {
                textValue = (Math.round(((textVal - min) / (max - min)) * 100)) + '%';
            }
            if (setOnlyText) {
                this.setData('textValue', textValue);
            } else {
                valueChanged = this.data.ltPropValue != setVal;
                if (!fromObs && valueChanged) {
                    this.avoid = true;
                    this.setData('ltPropValue', setVal);
                }
                this.actVal = custVal;
                $L.fastdom.measure(function () {
                    var handlerPos = this.rotateVal + (percent * 360),
                        allowAnimation = !avoidAnimation && !this.up && this.prevHandlerPos && (Math.abs(this.prevHandlerPos - handlerPos) > 10),
                        animation;
                    $L.fastdom.mutate(function () {
                        if (allowAnimation) {
                            animation = this.animateSliderHandler();
                        }
                        if (!animation) {
                            this.setData('textValue', textValue);
                        }
                        this.prevHandlerPos = handlerPos;
                        this.setData('dashOffset', (1 - percent) * (data.dashArray));//dash offset
                        this.setData('handlerWrapperStyle', 'transform: rotate(' + handlerPos + 'deg)');//handler rotation
                        valueChanged && this.getMethods('onValueChange') && this.executeMethod('onValueChange', textType ? textValue : textVal, setVal, $node);
                    }.bind(this));
                }.bind(this));
            }
        }
    }

    angleFromPoint(x, y, drag, readonly) {
        var pos = (this.mainCircle && this.mainCircle.getBoundingClientRect()) || $L('.lyteCirSliderOuterCircle', this.$node).get(0).getBoundingClientRect(),
            angle = Math.atan2((y - (pos.y + (pos.height / 2))), (x - (pos.x + (pos.width / 2)))) * (180 / Math.PI),
            data = this.data,
            anglediff,
            value;
        angle < 0 && (angle += 360);
        angle += 360 - this.rotateVal;
        angle > 360 && (angle -= 360);
        anglediff = angle - (this.prevAngle == undefined ? 0 : this.prevAngle);
        data.ltPropStopAtEnd && drag && Math.abs(anglediff) > 30 && (angle = anglediff < 0 ? 360 : 0);
        if (!readonly && this.prevAngle == angle && this.transEnd) {
            this.transitionend();
        }
        value = data.ltPropMin + (Math.abs(data.ltPropMax - (data.ltPropMin)) * (angle / 360));
        if (readonly) {
            return { angle: angle, value: value };
        }
        this.prevAngle = angle;
        this.changeValue(value);
    }

    mousemove(event) {
        this.angleFromPoint(event.x, event.y, true);
    }

    mouseup(event) {
        document.removeEventListener('mousemove', this.move);
        document.removeEventListener('mouseup', this.up);
        delete this.move;
        delete this.up;
        delete this.mainCircle;
        this.preventClick = true;
        $L(this.$node).removeClass('lyteCircularSliderDrag');
        this.getMethods('onDragEnd') && this.executeMethod('onDragEnd', event, this.$node);
    }

    checkVal(value) {
        var data = this.data,
            atend = data.ltPropAtend,
            max = data.ltPropMax,
            min = data.ltPropMin,
            oldValue = data.ltPropValue;
        if (value == undefined || isNaN(value)) {
            return min;
        }
        if (value < max && value > min) {
            return value;
        }
        if (data.ltPropStopAtEnd) {
            if (value > max) {
                value = max;
            } else if (value < min) {
                value = min;
            }
        } else {
            if (value == max || value == min) {
                if (atend == 'end') {
                    value = max;
                } else if (atend == 'start') {
                    value = min;
                } else {
                    if (oldValue != max && oldValue != min) {
                        value = Math.abs(oldValue - max) < Math.abs(oldValue - min) ? max : min;
                    } else {
                        value = oldValue;
                    }
                }
            }
            if (value > max) {
                value = min + Math.abs(value - max);
            } else if (value < min) {
                value = max - Math.abs(min - value);
            }
        }
        return value;
    }

    transitionend(event) {
        setTimeout(() => {
            requestAnimationFrame(() => {
                $L(this.$node).removeClass('lyteCircularSliderAnimation');
                this.handlerWrapper.removeEventListener('transitionend', this.transEnd);
                delete this.transEnd;
                delete this.handlerWrapper;
                delete this.handler;
                clearInterval(this.clickInterval);
                $L('.lyteCirSliderHandler', this.$node).get(0).focus();
                this.getMethods('onAnimationEnd') && this.executeMethod('onAnimationEnd', event, this.$node);
                this.changeValue(this.data.ltPropValue, false, true);
            })
        })
    }

    animateSliderHandler() {
        if (this.data.ltPropAnimate) {
            var firstTime = true;
            if (this.transEnd) {
                firstTime = false;
                this.handlerWrapper.removeEventListener('transitionend', this.transEnd);
            }
            $L(this.$node).addClass('lyteCircularSliderAnimation');
            this.handlerWrapper = $L('.lyteCirSliderHandlerWrapper', this.$node).get(0);
            if (!parseFloat(getComputedStyle(this.handlerWrapper).transitionDuration)) {
                delete this.handlerWrapper;
                $L(this.$node).removeClass('lyteCircularSliderAnimation');
                return;
            }
            this.transEnd = this.transitionend.bind(this);
            this.handlerWrapper.addEventListener('transitionend', this.transEnd);
            this.handler = $L('.lyteCirSliderHandler', this.$node).get(0);
            firstTime && this.getMethods('onAnimationStart') && this.executeMethod('onAnimationStart', $node);
            clearInterval(this.clickInterval);
            this.clickInterval = setInterval(function () {
                var handlerPos = this.handler.getBoundingClientRect();
                this.changeValue(this.angleFromPoint(handlerPos.x + handlerPos.width / 2, handlerPos.y + handlerPos.height / 2, false, true).value, false, true);
            }.bind(this));
            return true;
        }
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            valObs: function (args) {
                if (this.avoid) {
                    delete this.avoid;
                    return;
                }
                if (args) {
                    this.changeValue(this.data.ltPropValue, true);
                } else {
                    setTimeout(() => {
                        requestAnimationFrame(() => {
                            this.changeValue(this.data.ltPropValue, true);
                        })
                    }, 16)
                }
            }.observes('ltPropValue').on('didConnect')
        }), arg1);
    }

    _() {
        _;
    }
}

LyteCircularSliderComponent._template = "<template tag-name=\"lyte-circular-slider\" onclick=\"{{action('click',event)}}\" class=\"{{if(ltPropDisabled,'lyteCircularSliderDisabled ','')}}{{if(ltPropReadonly,'lyteCircularSliderReadOnly','')}}\"> <svg class=\"lyteCirSliderSvg\" viewBox=\"0 0 120 120\"> <g class=\"lyteCirSliderWrapper\"> <circle class=\"lyteCirSliderOuterCircle\" cx=\"60\" cy=\"60\" r=\"{{radius}}\" fill=\"none\" stroke=\"{{ltPropProgressBackgroundColor}}\" stroke-width=\"{{ltPropProgressWidth}}\"></circle> <circle class=\"lyteCirSliderStrokeCircle\" cx=\"60\" cy=\"60\" r=\"{{radius}}\" fill=\"none\" stroke=\"{{ltPropProgressColor}}\" stroke-width=\"{{ltPropProgressWidth}}\" stroke-dasharray=\"{{dashArray}}\" stroke-dashoffset=\"{{dashOffset}}\"></circle> </g> <template is=\"switch\" lyte-switch=\"true\" l-c=\"true\" _new=\"true\"></template></svg> <div class=\"lyteCirSliderHandlerWrapper\" style=\"{{handlerWrapperStyle}}\"> <span class=\"lyteCirSliderHandler\" style=\"{{handlerStyle}}\" lt-prop-title=\"{{if(ltPropTooltip,if(textValue,textValue,ltPropValue),'')}}\" lt-prop-tooltip-config=\"{{if(ltPropTooltip,ltPropTooltipConfig)}}\" lt-prop-tooltip-style=\"{{if(ltPropTooltip,ltPropTooltipStyle)}}\" lt-prop-tooltip-class=\"{{if(ltPropTooltip,ltPropTooltipClass)}}\" aria-valuemin=\"{{ltPropMin}}\" aria-valuemax=\"{{ltPropMax}}\" aria-valuenow=\"{{if(textValue,textValue,ltPropValue)}}\" aria-valuetext=\"{{expHandlers(ltPropAriaLabel,'+',if(textValue,textValue,ltPropValue))}}\" aria-label=\"{{if(textValue,textValue,ltPropValue)}}\" role=\"slider\" onmousedown=\"{{action('mousedown',event)}}\" onkeydown=\"{{action('keydown',event)}}\" tabindex=\"{{if(ltPropDisabled,expHandlers(1,'-'),0)}}\"> </span> </div> </template>";;
LyteCircularSliderComponent._dynamicNodes = [{"t":"a","p":[1,1,1]},{"t":"a","p":[1,1,3]},{"t":"s","p":[1,3],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"svg":true,"actualTemplate":"<template is=\"switch\"><template case=\"{{expHandlers(textValue,'!=',undefined)}}\" is=\"case\" lc-id=\"lc_id_0\" depth=\"1\"><svg><text x=\"60\" y=\"60\" class=\"lyteCircularSliderValue\">{{textValue}}</text></svg></template></template>"},{"t":"a","p":[3],"a":{"style":{"name":"style","dynamicValue":"handlerWrapperStyle"}}},{"t":"a","p":[3,1],"a":{"style":{"name":"style","dynamicValue":"handlerStyle"}}}];;

LyteCircularSliderComponent._observedAttributes = [
    "ltPropValue",
    "ltPropMin",
    "ltPropMax",
    "ltPropProgressWidth",
    "ltPropProgressColor",
    "ltPropProgressBackgroundColor",
    "ltPropTextDisplay",
    "ltPropDigits",
    "ltPropStopAtEnd",
    "ltPropStartDirection",
    "ltPropAtend",
    "ltPropDisabled",
    "ltPropReadonly",
    "ltPropTooltip",
    "ltPropTooltipConfig",
    "ltPropTooltipStyle",
    "ltPropTooltipClass",
    "ltPropStep",
    "ltPropDiscrete",
    "ltPropAriaLabel",
    "ltPropAnimate",
    "dashOffset",
    "textValue",
    "radius",
    "dashArray",
    "handlerStyle",
    "handlerWrapperStyle"
];

export { LyteCircularSliderComponent };
LyteCircularSliderComponent.register("lyte-circular-slider", {
    hash: "LyteCircularSliderComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
