// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/34010b278b34f816d39e54ad9023822c.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../component.js";
import $L from "../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";
import { objectUtils,arrayUtils } from "../../node_modules/@slyte/component/index.js";

/**
 * Renders an DatemultiSelect
 * @component lyte-datemultiselect
 * @methods  onDateBeforeSelect,onDateAfterSelect,onDateBeforeRemove,onDateAfterRemove,onClearAll,onBeforeShow,onShow,onBeforeClose,onClose;
 */
class LyteDatemultiselectComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {

		return Object.assign(super.data({

			'ltPropBindToBody': prop('boolean', { 'default': false }),

			'ltPropType': prop('string', { 'default': 'popup' }),

			'ltPropSelectedDates': prop('array', { 'default': [] }),

			'ltPropCurrentMonth': prop('array', { 'default': [] }),

			'ltPropCurrentDates': prop('array', { 'default': [] }),

			'ltPropBoundary': prop('object', { 'default': {} }),

			'ltPropCalendar': prop('object', { 'default': {} }),

			'ltPropIsOpen': prop('boolean', { 'default': false }),

			'ltPropFormat': prop('string', { 'default': 'MM/DD/YYYY' }),

			'countDays': prop('string', { 'default': "" }),

			'month': prop('string', { 'default': "" }),

			'date': prop('string', { 'default': "" }),

			'year': prop('string', { 'default': "" }),

			'itemsArray': prop('array', { 'default': [] }),

			'defaultFormatDates': prop('array', { 'default': "[]" }),

			'formattedDates': prop('string', { 'default': "" }),

			'selectDays': prop('string', { 'default': "Selected Days" }),

			'ltPropPlaceholder': prop('string', { 'default': 'Select Days' }),

			'isDateSelected': prop('boolean', { 'default': false }),

			'lyteSelectionContainer': prop('boolean', { 'default': false }),

			'ltPropAllowMultiple': prop('boolean', { 'default': false })

		}), arg1);
	}

    didConnect(event) {
		var that = this;
		var lyteDMSType = this.getData('ltPropType');
		this.$node.close = function (event) {
			that.close(event);
		}
		this.$node.open = function (event) {
			that.open(event);
		}
		let getSelectedDays = this.getData('ltPropSelectedDates');
		if (getSelectedDays.length > 0) {
			getSelectedDays.forEach((element) => {
				const ltPropFormat = this.getData('ltPropFormat');
				const sliceMonth = element.slice(0, 2);
				const convertMonth = Number(sliceMonth);
				const sliceDate = element.slice(3, 5);
				const sliceYear = element.slice(6, 10);
				const convertDate = Number(sliceDate);
				const convertYear = Number(sliceYear);

				let formattedDate;

				var format_1 = "DD/Month/YYYY",
				format_2 = "Month/DD/YYYY";

				switch (ltPropFormat) {
					case format_1:
						var dates = new Date(convertYear, convertMonth - 1, convertDate);
						var options = {
							day: '2-digit',
							month: 'short',
							year: 'numeric'
						};
						formattedDate = dates.toLocaleDateString('en-US', options);
						var spliceDate = formattedDate.slice(4, 6);
						var spliceMonth = formattedDate.slice(0, 3);
						var spliceYear = formattedDate.slice(8, 12);
						formattedDate = spliceDate + " " + spliceMonth + "," + spliceYear;
						break;

					case "DD/MM/YYYY":
					case "MM/DD/YYYY":
					case "DD/YYYY/MM":
					case "YYYY/MM/DD":
					case "YYYY/DD/MM":
						formattedDate = $L.moment(element, 'MM/DD/YYYY').format(ltPropFormat);
						break;

					case format_2:
						var dates2 = new Date(convertYear, convertMonth - 1, convertDate);
						var options2 = {
							day: '2-digit',
							month: 'short',
							year: 'numeric'
						};
						formattedDate = dates2.toLocaleDateString('en-US', options2);
						break;
				}

				if (formattedDate) {
					arrayUtils(this.getData('itemsArray'), 'push', formattedDate);

				}
			});
			if (lyteDMSType == "inline") {
				this.bindWormhole(event);
				this.openDMSWrap(event);

			}

			this.setData('countDays', this.getData('ltPropSelectedDates').length);
			this.setData('isDateSelected', true);
		}
	}

    openDMSWrap(event) {
		let placePosition = $L(this.$node).find('.lyteDmsWrap')[0];
		var getSelectedDays = this.getData('ltPropSelectedDates');
		var selectedDays = this.getData('itemsArray');
		var button = $L(this.$node).find('.lyteDMSButton')[0];
		var calendar = $L(this.$node).find('.lyteDmsCalander')[0];
		var lyteSelectionContainer = $L(this.$node).find('.lyteDmsSelectionContainer')
		if (getSelectedDays.length > 0) {
			this.setData("lyteSelectionContainer", true);
			if (selectedDays.length == 1 || selectedDays.length == 0) {
				$L(placePosition).find('.lyteDMSRemoveAll')[0].style.display = "none";
			}
			else if (selectedDays.length > 1) {
				$L(placePosition).find('.lyteDMSRemoveAll')[0].style.display = "block";
			}
		}
		else if (getSelectedDays == "") {
			this.setData("lyteSelectionContainer", false);
		}
	}

    openWormhole(event) {
		window._lyteDMS.activeDMS = this.$node;
		let wormhole = this.childComp;
		let addClass = $L(this.$node)
		let ltPropType = this.getData('ltPropType')
		wormhole.style.display = "block";
		addClass.addClass("lyteDMSWormholeOpen");
		$L('.lyteDMSWormholeOpen')[0].style.display = "block";
		this.wormholePosition(event);
		$L('.lyteDMSButton')[0].style.width =  $L('.lyteDmsCalander')[0].clientWidth-(10 + 10)+"px";
		if (ltPropType == "popup") {
			var getSelectedDays = this.getData('ltPropSelectedDates');
			let placePosition = $L(wormhole).find('.lyteDmsWrap')[0];
			var selectedDays = this.getData('itemsArray');
		if (getSelectedDays.length > 0) {
			this.setData("lyteSelectionContainer", true);
			if (selectedDays.length == 1 || selectedDays.length == 0) {
				$L(placePosition).find('.lyteDMSRemoveAll')[0].style.display = "none";
			}
			else if (selectedDays.length > 1) {
				$L(placePosition).find('.lyteDMSRemoveAll')[0].style.display = "block";
			}
		}
		else if (getSelectedDays == "") {
			this.setData("lyteSelectionContainer", false);
			}
		}
		else if (ltPropType == "inline") {
			let lyteDmsCalender = $L(this.$node).find('.lyteDMSRemoveAll')[0];
			var selectedDays = this.getData('itemsArray');
			var getSelectedDays = this.getData('ltPropSelectedDates');
			this.setData("lyteSelectionContainer", true);
			if (selectedDays.length == 1 || selectedDays.length == 0) {
				lyteDmsCalender.style.display = "none";
			}
			else if (selectedDays.length > 1) {
				lyteDmsCalender.style.display = "block";
			}

			else if (getSelectedDays == "") {
				this.setData("lyteSelectionContainer", false);
			}

		}
	}

    wormholePosition() {
		let wormhole = this.childComp;
		let placePosition = $L(wormhole).find('.lyteDmsWrap')[0];
		var button = $L(this.$node).find('.lyteDMSButton')[0];
		var buttonBoundingClients = button.getBoundingClientRect();
		$L('body').addClass('lyteBodyWrapper');
		placePosition.style.position = "absolute";
		placePosition.style.top = document.documentElement.scrollTop + button.getBoundingClientRect().bottom + "px";
		placePosition.style.left = buttonBoundingClients.left + "px";
	}

    open(event) {
		var ret = this.fireBeforeCallback('onBeforeOpen', event);
		var _this = this;
		if (ret) {
			return;
		}
		this.setData('ltPropIsOpen', true);
		var lyteDMSType = this.getData('ltPropType');
		var removeClass = $L(this.$node);
		let allowMultiple = this.getData('ltPropAllowMultiple');
		let button = Array.from($L('lyte-datemultiselect'));
		button.forEach(function (button) {
			if (button.classList.contains('lyteDMSWormholeOpen') && button != _this.$node && allowMultiple == false) {
			button.component.childComp.style.display = "none";
		}
			});
		if (lyteDMSType == "popup") {
			this.bindWormhole(event);
			let appendWormhole = this.childComp;
			if (appendWormhole.style.display == "none") {
				this.openWormhole(event);
				
			}
		
			else if (appendWormhole.style.display == "block") {
				this.close(event);
				appendWormhole.style.display = "none";
				removeClass[0].classList.remove("lyteDMSWormholeOpen");
				if (this.getMethods('onClose')) {
					this.executeMethod('onClose', event);
				}
			}

		}

		if (this.getMethods('onShow')) {
			this.executeMethod('onShow', event);
		}
		if (this.getMethods('onBeforeOpen')) {
			this.executeMethod('onBeforeOpen', event);
		}
		if (this.getMethods('onBeforeShow')) {
			this.executeMethod('onBeforeShow', event);
		}

	}

    getCalendar() {
		var child = this.childComp,
			calendar = child.querySelector('lyte-calendar');

		return calendar;

	}

    close(event) {
		var ret = this.fireBeforeCallback('onBeforeHide', event);
		let wormhole = this.childComp;
		if (ret) {
			return;
		}
		this.setData('ltPropIsOpen', false);
		this.bindWormhole(event);
		wormhole.style.display = "none";
		if (this.getMethods('onHide')) {
			this.executeMethod('onHide', event, wormhole);
		}
		if (this.getMethods('onBeforeClose')) {
			this.executeMethod('onBeforeClose', event, wormhole);
		}
		window._lyteDMS.activeDMS = null;
	}

    fireBeforeCallback() {
		var callbackName = arguments[0], ret;
		if (this.getMethods(callbackName)) {
			ret = this.executeMethod.apply(this, arguments) === false;
		}
		return ret;

	}

    getIndex() {
		let calender = this.getData('ltPropSelectedDates');
		let index = calender.findIndex((element) => element == this.getData('formattedDates'));
		arrayUtils(calender, 'removeAt', index, 1);
		arrayUtils(this.getData('itemsArray'), 'removeAt', index, 1);
	}

    removeAtEnd(date) {
		let calender = this.getData('ltPropSelectedDates');
		let index = calender.findIndex((element) => element == date);
		arrayUtils(calender, 'removeAt', index, 1);
		arrayUtils(this.getData('itemsArray'), 'removeAt', index, 1);
	}

    showCalendar() {
		this.childComp.classList.remove('lyteDMSHide');
		window._lyteDMS.activeDMS = this.$node;
		this.getCalendar().revertToToday();
	}

    position() {
		$L(this.childComp).placement({
			append: false,
			originElement: this.$node
		});
	}

    remove(item, event) {

		var dataValue = item.getAttribute('data-value'),
			selectedDates = this.getData('itemsArray') || [],
			indexToRemove = selectedDates.indexOf(dataValue),
			ret = this.fireBeforeCallback('onBeforeRemove', event, dataValue);

		if (ret || indexToRemove === -1) {
			return;
		}

		arrayUtils(selectedDates, 'removeAt', indexToRemove);
		if (this.getMethods('onRemove', event)) {
			this.executeMethod('onRemove', event, dataValue);
		}
		if (this.getMethods('onDateBeforeRemove', event)) {
			this.executeMethod('onDateBeforeRemove', event, dataValue);
		}
		if (this.getMethods('onDateAfterRemove', event)) {
			this.executeMethod('onDateAfterRemove', event, dataValue);
		}
		if (this.getMethods('onClearAll')) {
			this.executeMethod('onClearAll', event, dataValue);
		}
	}

    exceedsBoundary() {
		var boundary = this.getBoundary(),
			button = this.$node,
			buttonBoundingClients = button.getBoundingClientRect();

		if (buttonBoundingClients.left < boundary.left
			|| buttonBoundingClients.right > boundary.right
			|| buttonBoundingClients.top < boundary.top
			|| buttonBoundingClients.bottom > boundary.bottom) {
			return true;
		}
	}

    getBoundary() {
		var bounds = this.getData('ltPropBoundary');
		return {
			left: bounds.left || 0,
			right: bounds.right || window.innerWidth,
			top: bounds.top || 0,
			bottom: bounds.bottom || window.innerHeight
		};
	}

    bindWormhole() {
		if (!this.getData('ltPropBindToBody')) {
			this.setData('ltPropBindToBody', true);
		}
	}

    static actions(arg1) {
        return Object.assign(super.actions({
            openCalendar: function (event) {
                this.open(event);
            },
            removeSelectedDays: function (event, item) {
                if (this.getData().ltPropType == "popup") {
                let lyteDmsCalender = $L(this.childComp);
                let calender = this.getData('ltPropSelectedDates');
                let shortHands = lyteDmsCalender.find('lyte-calendar')[0].component.data.shortHands;
                const ltPropFormat = this.getData('ltPropFormat');
                let selectedDays = this.getData('itemsArray');
                let defaultFormatDate = item;
                    let sliceMonth, sliceDate, sliceYear, findMonth, monthIndex, month, day, year, sliceDateMonth, sliceMonthMonth, sliceYearMonth, findMonthMonth, monthIndexMonth, monthMonth, dayMonth, yearMonth;
                switch (ltPropFormat) {
                    case "DD/Month/YYYY":
                         sliceMonth = item.slice(3, 6);
                         sliceDate = item.slice(0, 2);
                         sliceYear = item.slice(7, 12);
                         findMonth = shortHands.findIndex((element) => element == sliceMonth);
                         monthIndex = findMonth + 1;
                         month = monthIndex.toString().padStart(2, '0');
                         day = sliceDate.toString().padStart(2, '0');
                         year = sliceYear.toString();
                        defaultFormatDate = `${month}/${day}/${year}`;
                        break;

                    case "DD/YYYY/MM":
                    case "MM/YYYY/DD":
                    case "YYYY/DD/MM":
                    case "DD/MM/YYYY":
                        defaultFormatDate = $L.moment(item, ltPropFormat).format("MM/DD/YYYY");
                            break;

                    case "Month/DD/YYYY":
                         sliceDateMonth = item.slice(4, 6);
                         sliceMonthMonth = item.slice(0, 3);
                         sliceYearMonth = item.slice(8, 12);
                         findMonthMonth = shortHands.findIndex((element) => element == sliceMonthMonth);
                         monthIndexMonth = findMonthMonth + 1;
                         monthMonth = monthIndexMonth.toString().padStart(2, '0');
                         dayMonth = sliceDateMonth.toString().padStart(2, '0');
                         yearMonth = sliceYearMonth.toString();
                        defaultFormatDate = `${monthMonth}/${dayMonth}/${yearMonth}`;
                        break;
                }

                if (calender.find((element) => element == defaultFormatDate)) {
                    this.setData('formattedDates', defaultFormatDate);
                    this.getIndex();
                }
                var ret = this.fireBeforeCallback('onBeforeRemove', event, item);
                let countDays = this.getData('ltPropSelectedDates').length;
                this.setData('countDays', countDays);
                this.setData('countDays', this.getData('ltPropSelectedDates').length);
                if (selectedDays.length == 1 || selectedDays.length == 0) {
                    lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "none";
                }
                else if (selectedDays.length > 1) {
                    lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "block";
                }
                }
                if (this.getData().ltPropType == "inline") {
                    let lyteDmsCalender = $L(this.$node);
                    let calender = this.getData('ltPropSelectedDates');
                    let shortHands = lyteDmsCalender.find('lyte-calendar')[0].component.data.shortHands;
                    const ltPropFormat = this.getData('ltPropFormat');
                    let selectedDays = this.getData('itemsArray');
                    let defaultFormatDate = item;
                    let sliceMonth, sliceDate, sliceYear, findMonth, monthIndex, month, day, year, sliceDateMonth, sliceMonthMonth, sliceYearMonth, findMonthMonth, monthIndexMonth, monthMonth, dayMonth, yearMonth;
                    switch (ltPropFormat) {
                        case "DD/Month/YYYY":
                            sliceMonth = item.slice(3, 6);
                            sliceDate = item.slice(0, 2);
                            sliceYear = item.slice(7, 12);
                            findMonth = shortHands.findIndex((element) => element == sliceMonth);
                            monthIndex = findMonth + 1;
                            month = monthIndex.toString().padStart(2, '0');
                            day = sliceDate.toString().padStart(2, '0');
                            year = sliceYear.toString();
                            defaultFormatDate = `${month}/${day}/${year}`;
                            break;

                        case "DD/YYYY/MM":
                        case "MM/YYYY/DD":
                        case "YYYY/DD/MM":
                        case "DD/MM/YYYY":
                            defaultFormatDate = $L.moment(item, ltPropFormat).format("MM/DD/YYYY");
                            break;

                        case "Month/DD/YYYY":
                            sliceDateMonth = item.slice(4, 6);
                            sliceMonthMonth = item.slice(0, 3);
                            sliceYearMonth = item.slice(8, 12);
                            findMonthMonth = shortHands.findIndex((element) => element == sliceMonthMonth);
                            monthIndexMonth = findMonthMonth + 1;
                            monthMonth = monthIndexMonth.toString().padStart(2, '0');
                            dayMonth = sliceDateMonth.toString().padStart(2, '0');
                            yearMonth = sliceYearMonth.toString();
                            defaultFormatDate = `${monthMonth}/${dayMonth}/${yearMonth}`;
                            break;
                    }

                    if (calender.find((element) => element == defaultFormatDate)) {
                        this.setData('formattedDates', defaultFormatDate);
                        this.getIndex();
                    }
                    var ret = this.fireBeforeCallback('onBeforeRemove', event, item);
                    let countDays = this.getData('ltPropSelectedDates').length;
                    this.setData('countDays', countDays);
                    this.setData('countDays', this.getData('ltPropSelectedDates').length);
                    if (selectedDays.length == 1 || selectedDays.length == 0) {
                        lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "none";
                    }
                    else if (selectedDays.length > 1) {
                        lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "block";
                    }
                }


                if (this.getMethods('onRemove')) {
                    this.executeMethod('onRemove', event, item);
                }
                if (this.getMethods('onDateBeforeRemove')) {
                    this.executeMethod('onDateBeforeRemove', event, item);
                }
                if (this.getMethods('onDateAfterRemove')) {
                    this.executeMethod('onDateAfterRemove', event, item);
                }
            },
            clearSelectedDays: function (event, item) {
                let ltPropType = this.getData('ltPropType');
                if (ltPropType == "popup") {
                let lyteDmsCalender = $L(this.childComp);
                this.setData('ltPropSelectedDates', []);
                this.setData('itemsArray', []);
                lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "none";
                var ret = this.fireBeforeCallback('onBeforeRemove', event, item);
                let countDays = this.getData('itemsArray').length;
                this.setData('countDays', countDays);
                }
                else if (ltPropType == "inline") {
                    let lyteDmsCalender = $L(this.$node).find('.lyteDMSRemoveAll')[0];
                    let ltPropType = this.getData('ltPropType');
                    this.setData('ltPropSelectedDates', []);
                    this.setData('itemsArray', []);
                    lyteDmsCalender.style.display = "none";
                    var ret = this.fireBeforeCallback('onBeforeRemove', event, item);
                    let countDays = this.getData('itemsArray').length;
                    this.setData('countDays', countDays);
                }

                if (this.getMethods('onRemove')) {
                    this.executeMethod('onRemove', event, item);
                }

                if (this.getMethods('onClearAll')) {
                    this.executeMethod('onClearAll', event, item);
                }
            }
        }), arg1);
    }

    static methods(arg1) {
        return Object.assign(super.methods({
            beforeRemove: function (event, date) {
                var ret = this.fireBeforeCallback('onBeforeRemove', event, date);

                if (ret) {
                    return false;
                }
                let calender = this.getData('ltPropSelectedDates');
                if (calender.find((element) => element == date)) {
                    this.removeAtEnd(date);

                }
                if (this.getMethods('onDateBeforeRemove')) {
                    this.executeMethod('onDateBeforeRemove', event, date);
                }
            },
            beforeAdd: function (event, date,) {
                var ret = this.fireBeforeCallback('onBeforeAdd', event, date);

                if (ret) {
                    return false;
                }
                if (this.getMethods('onDateBeforeSelect')) {
                    this.executeMethod('onDateBeforeSelect', event, date);
                }
            },
            setFlags: function (wormhole) {
                this.childComp = wormhole;
            },
            handleAddition: function (event, date) {
                this.getData('ltPropFormat');
                var sliceMonth = date.slice(0, 2);
                var convertMonth = Number(sliceMonth);
                var sliceDate = date.slice(3, 5);
                var sliceYear = date.slice(6, 10);
                var convertDate = Number(sliceDate);
                var convertYear = Number(sliceYear);
                var ltPropFormat = this.getData('ltPropFormat');
                let selectedDays = this.getData('itemsArray');
                let formattedDate;
                let dates, options, spliceDate, spliceMonth, spliceYear, dates2, options2;
                let ltPropType = this.getData('ltPropType')

                if (ltPropType == 'popup') {
                    this.openWormhole();
                    this.wormholePosition();
                }
                switch (ltPropFormat) {
                    case "DD/Month/YYYY":
                         dates = new Date(convertYear, convertMonth - 1, convertDate);
                         options = {
                            day: '2-digit',
                            month: 'short',
                            year: 'numeric'
                        };
                        formattedDate = dates.toLocaleDateString('en-US', options);
                         spliceDate = formattedDate.slice(4, 6);
                         spliceMonth = formattedDate.slice(0, 3);
                         spliceYear = formattedDate.slice(8, 12);
                        formattedDate = spliceDate + " " + spliceMonth + "," + spliceYear;
                        break;

                    case "DD/MM/YYYY":
                    case "MM/DD/YYYY":
                    case "DD/YYYY/MM":
                    case "YYYY/MM/DD":
                    case "YYYY/DD/MM":
                        formattedDate = $L.moment(date, 'MM/DD/YYYY').format(ltPropFormat);
                        break;

                    case "Month/DD/YYYY":
                         dates2 = new Date(convertYear, convertMonth - 1, convertDate);
                         options2 = {
                            day: '2-digit',
                            month: 'short',
                            year: 'numeric'
                        };
                        formattedDate = dates2.toLocaleDateString('en-US', options2);
                        break;
                }

                if (formattedDate) {
                    arrayUtils(this.getData('itemsArray'), 'push', formattedDate);
                }
                this.setData('isDateSelected', true);
                this.setData("lyteSelectionContainer", true);
                let countDays = this.getData('itemsArray').length;
                this.setData('countDays', countDays);
                if (ltPropType == "popup") {
                    let lyteDmsCalender = $L(this.childComp).find('.lyteDMSRemoveAll')[0];
                if (selectedDays.length == 1 || selectedDays.length == 0) {
                        lyteDmsCalender.style.display = "none";
                }
                else if (selectedDays.length > 1) {
                        lyteDmsCalender.style.display = "block";
                }
                }
                else if (ltPropType == "inline") {
                    let lyteDmsCalender = $L(this.$node).find('.lyteDMSRemoveAll')[0];
                    if (selectedDays.length == 1 || selectedDays.length == 0) {
                        lyteDmsCalender.style.display = "none";
                    }
                    else if (selectedDays.length > 1) {
                        lyteDmsCalender.style.display = "block";
                    }
                }

                if (this.getMethods('onDateAfterSelect')) {
                    this.executeMethod('onDateAfterSelect', event, date);
                }
                if (this.getMethods('onAdd')) {
                    this.executeMethod('onAdd', event, date);
                }
            },
            handleRemoval: function (event, date) {
                let calender = this.getData('ltPropSelectedDates');
                let selectedDays = this.getData('itemsArray');
                let ltPropType = this.getData('ltPropType');
                if (calender.find((element) => element == date)) {
                    this.removeAtEnd(date);
                }
                var ret = this.fireBeforeCallback('onBeforeRemove', event, date);
                let countDays = this.getData('ltPropSelectedDates').length;
                this.setData('countDays', countDays);
                if (ltPropType == "popup") {
                    let lyteDmsCalender = $L(this.childComp);
                    if (selectedDays.length == 1 || selectedDays.length == 0) {
                        lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "none";
                    }
                    else if (selectedDays.length > 1) {
                        lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "block";
                    }

                }
                else if (ltPropType == "inline") {
                    let lyteDmsCalender = $L(this.$node).find('.lyteDMSRemoveAll')[0];
                if (selectedDays.length == 1 || selectedDays.length == 0) {
                    lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "none";
                }
                else if (selectedDays.length > 1) {
                    lyteDmsCalender.find('.lyteDMSRemoveAll')[0].style.display = "block";
                }
                }
                if (this.getMethods('onRemove')) {
                    this.executeMethod('onRemove', event, date);
                }
                if (this.getMethods('onDateAfterRemove')) {
                    this.executeMethod('onDateAfterRemove', event, date);
                }
            }


        }), arg1);
    }

    didDestroy() {
		let wormhole = this.childComp;
		if (wormhole) {
			wormhole.remove();
		}
		if (this.$node === window._lyteDMS.activeDMS) {
			window._lyteDMS.activeDMS = null;
		}

		if (this.$node === window._lyteDMS.closedDMS) {
			window._lyteDMS.closedDMS = null;
		}
	}

    _() {
        _;
    }
}

LyteDatemultiselectComponent._template = "<template tag-name=\"lyte-datemultiselect\"> <div class=\"lyteDMSButton\" onclick=\"{{action(&quot;openCalendar&quot;,event)}}\" tabindex=\"0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{isDateSelected}}\" is=\"case\" lc-id=\"lc_id_0\"><span>{{countDays}} {{selectDays}}</span></template><template default=\"\"><span class=\"lyteDMSButtonPlaceholder\">{{ltPropPlaceholder}}</span></template></template></div> <div class=\"lyteDmsRowSection\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(ltPropType,'==',&quot;inline&quot;)}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteDmsWrap\"> <lyte-calendar lt-prop-header-type=\"dropdown\" class=\"lyteDmsCalander\" lt-prop=\"{{stringify(ltPropCalendar)}}\" lt-prop-dropdown=\"{&quot;boxClass&quot;:&quot;lyteDmsDropBoxClass&quot;}\" lt-prop-current-dates=\"{{lbind(ltPropSelectedDates)}}\" lt-prop-multiple=\"true\" on-before-add=\"{{method('beforeAdd')}}\" on-before-remove=\"{{method('beforeRemove')}}\" on-date-selected=\"{{method('handleAddition')}}\" on-date-removed=\"{{method('handleRemoval')}}\"> </lyte-calendar> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteSelectionContainer}}\" is=\"case\" lc-id=\"lc_id_0\"><div class=\"lyteDmsSelectionContainer\"> <div class=\"lyteDmsCalcWrap\"> <p class=\"lyteDmsDayCalc\">Selected Days</p> <p class=\"lyteDmsDateCalc\">{{countDays}}</p> </div> <div class=\"lyteDmsRightPanelContentWrap\"> <template items=\"{{itemsArray}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-dms-item class=\"lyteDmsSelectedDays\"> {{item}} <lyte-dms-remove></lyte-dms-remove> <div onclick=\"{{action('removeSelectedDays',event,item)}}\" class=\"lyteDMSRemoveIcon\">X</div> </lyte-dms-item> </template> </div> <p class=\"lyteDMSRemoveAll\" onclick=\"{{action('clearSelectedDays',event,item)}}\">Clear All</p> </div></template></template></div> </template><template case=\"{{expHandlers(ltPropType,'==',&quot;popup&quot;)}}\" is=\"case\" lc-id=\"lc_id_1\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropBindToBody}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-wormhole class=\"lyteDMSContainer lyteDMSWormhole\" on-before-append=\"{{method('setFlags')}}\" style=\"display: none;\"> <template is=\"registerYield\" yield-name=\"lyte-content\"> <div class=\"lyteDmsWrap\"> <lyte-calendar lt-prop-header-type=\"dropdown\" class=\"lyteDmsCalander\" lt-prop=\"{{stringify(ltPropCalendar)}}\" lt-prop-dropdown=\"{&quot;boxClass&quot;:&quot;lyteDmsDropBoxClass&quot;}\" lt-prop-current-dates=\"{{lbind(ltPropSelectedDates)}}\" lt-prop-multiple=\"true\" on-before-add=\"{{method('beforeAdd')}}\" on-before-remove=\"{{method('beforeRemove')}}\" on-date-selected=\"{{method('handleAddition')}}\" on-date-removed=\"{{method('handleRemoval')}}\"> </lyte-calendar> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteSelectionContainer}}\" is=\"case\" lc-id=\"lc_id_0\"><div class=\"lyteDmsSelectionContainer\"> <div class=\"lyteDmsCalcWrap\"> <p class=\"lyteDmsDayCalc\">Selected Days</p> <p class=\"lyteDmsDateCalc\">{{countDays}}</p> </div> <div class=\"lyteDmsRightPanelContentWrap\"> <template items=\"{{itemsArray}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-dms-item class=\"lyteDmsSelectedDays\"> {{item}} <lyte-dms-remove></lyte-dms-remove> <div onclick=\"{{action('removeSelectedDays',event,item)}}\" class=\"lyteDMSRemoveIcon\">X</div> </lyte-dms-item> </template> </div> <p class=\"lyteDMSRemoveAll\" onclick=\"{{action('clearSelectedDays',event,item)}}\">Clear All</p> </div></template></template></div> <lyte-datemultiselect-freeze></lyte-datemultiselect-freeze> </template> </lyte-wormhole> </template></template></template></template></div> </template>";;
LyteDatemultiselectComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,0],"cn":"lc_id_0"},{"t":"tX","p":[0,2],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"tX","p":[0,0],"cn":"default"}]},"dc":{"lc_id_0":{},"default":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0]},{"t":"s","p":[3,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1,1],"cn":"lc_id_0"},{"t":"cD","p":[1,1],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"s","p":[1,3],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,1,3,0],"cn":"lc_id_0"},{"t":"a","p":[0,3,1],"cn":"lc_id_0"},{"t":"f","p":[0,3,1],"dN":[{"t":"tX","p":[1,1]},{"t":"cD","p":[1,3],"in":1,"sibl":[0]},{"t":"a","p":[1,5]},{"t":"cD","p":[1],"in":0}],"dc":[1,0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"},{"t":"a","p":[0,5],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true},"lc_id_1":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"r","p":[1,1],"dN":[{"t":"a","p":[1,1]},{"t":"cD","p":[1,1],"in":2,"sibl":[1]},{"t":"s","p":[1,3],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,1,3,0],"cn":"lc_id_0"},{"t":"a","p":[0,3,1],"cn":"lc_id_0"},{"t":"f","p":[0,3,1],"dN":[{"t":"tX","p":[1,1]},{"t":"cD","p":[1,3],"in":1,"sibl":[0]},{"t":"a","p":[1,5]},{"t":"cD","p":[1],"in":0}],"dc":[1,0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"},{"t":"a","p":[0,5],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"cD","p":[3],"in":0}],"dc":[2,1,0],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_1"}],"cdp":{"t":"a","p":[1]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true},"lc_id_1":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0","lc_id_1"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;

LyteDatemultiselectComponent._observedAttributes = [
    "ltPropBindToBody",
    "ltPropType",
    "ltPropSelectedDates",
    "ltPropCurrentMonth",
    "ltPropCurrentDates",
    "ltPropBoundary",
    "ltPropCalendar",
    "ltPropIsOpen",
    "ltPropFormat",
    "countDays",
    "month",
    "date",
    "year",
    "itemsArray",
    "defaultFormatDates",
    "formattedDates",
    "selectDays",
    "ltPropPlaceholder",
    "isDateSelected",
    "lyteSelectionContainer",
    "ltPropAllowMultiple"
];

window._lyteDMS = {
	closeDMS: function () {
		if (!_lyteDMS.activeDMS) {
			return;
		}

		_lyteDMS.activeDMS.close();
	},

	isFocusedOpen: function () {
		var focusedDMS = _lyteDMS.getFocusedDMS();

		if (focusedDMS) {
			return focusedDMS.ltProp('isOpen');
		}
	},

	getFocusedDMS: function () {
		var activeElement = document.activeElement;

		if (activeElement.classList.contains('lyteDMSButton')) {
			activeElement = $L(activeElement).closest('datemultiselect-comp').get(0);

			return activeElement;
		}
	},

	openFocused: function (event) {
		var focusedDMS = _lyteDMS.getFocusedDMS();

		if (focusedDMS) {
			focusedDMS.open(event);
		}
	}
};

// if (!window._lyteUiUtils.registeredCustomElements['lyte-dms-remove']) {
//     window._lyteUiUtils.registeredCustomElements['lyte-dms-remove'] = true;

//     /**
// 	   * @customElement lyte-dms-remove
// 	   */

//    class LyteDmsRemoveCustomElements extends HTMLElement {
//        static options = {clone : {allCallbacks : false}};

//        lookups() {
//            return [{
//                registry: LyteUiComponentComponentRegistry
//            }];
//        }

//        constructor() {
//            super();
//        }

//        static get observedAttributes() {
//            return [];
//        }

//        connectedCallback() {
//            var lyteNode = this.$node;
//            lyteNode.addEventListener('click', function (event) {
//                var dms = $L(lyteNode).closest('ldatemultiselect').get(0);
//                dms.component.remove(lyteNode.closest('lyte-dms-item'), event);
//            });
//        }
//    }

//     LyteDmsRemoveCustomElements.register("lyte-dms-remove");
// }


window._lyteUiUtils.addGlobalEventListener('click', function (event) {
	var target = event.target,
		parent = $L(target).closest('.lyteDMSContainer, .lyteDMSButton, .lyteDmsCalander, .lyteDmsSelectionContainer, .lyteDmsDropBoxClass');
	if (!parent.get(0)) {
		_lyteDMS.closeDMS();
		
	}
}, true);

window.addEventListener('scroll', function (event) {
	let button = Array.from($L(".lyteDMSButton"));
	
	button.forEach(function (button) {
		let dms = button.parentElement;
	var type = dms.component.data.ltPropType;
		var allowMultiple = dms.component.data.ltPropAllowMultiple;
		let comp = Array.from($L('lyte-datemultiselect'));
		comp.forEach(function (button) {
			if (button.classList.length > 1 && allowMultiple == false) {
				button.component.childComp.style.display = "none";
			}
		});
		if (type == 'popup') {
		dms.component.bindWormhole(event);
		let wormhole = dms.component.childComp;
		let button = $L('.lyteDMSButton')[0];
		let buttonBoundingClients = button.getBoundingClientRect();
		let calendar = $L('.lyteDmsCalander')[0];
		let calendarBoundingClients = calendar.getBoundingClientRect();
		for (const object in calendarBoundingClients) {
			if (buttonBoundingClients.top < 0 && wormhole !== undefined) {
				dms.component.wormholePosition(event);
				dms.component.close(event);
				wormhole.style.display = "none";
			}
				else if (buttonBoundingClients.top > 0 ) {
				dms.component.wormholePosition(event);
			}
		}
		}
		else {
				dms.component.openDMSWrap(event);
			
	}
	});

	
	var activeDMS = _lyteDMS.activeDMS, comp,
		closedDMS = _lyteDMS.closedDMS;

	if (activeDMS) {
		comp = activeDMS.component;

		// comp.position();

		if (comp.exceedsBoundary()) {
			activeDMS.close(event);
			_lyteDMS.closedDMS = activeDMS;
		}

		return;
	}

	if (closedDMS) {
		comp = closedDMS.component;

		if (!comp.exceedsBoundary()) {
			closedDMS.open(event);
			_lyteDMS.closedDMS = null;
		}
	}
}, true);

window._lyteUiUtils.addGlobalEventListener('keydown', function (event) {
	var key = event.code,
		spaceKey = 'Space',
		escapeKey = 'Escape',
		tabKey = 'Tab';

	if (key === spaceKey && _lyteDMS.getFocusedDMS() && !_lyteDMS.isFocusedOpen()) {
		_lyteDMS.openFocused(event);
		event.preventDefault();
	}
	else if ((key === escapeKey || key === tabKey) && _lyteDMS.activeDMS) {
		_lyteDMS.activeDMS.close();
	}
});

window.addEventListener('resize', function () {
	window.clearTimeout(_lyteDMS.debounceId);

	_lyteDMS.debounceId = setTimeout(function () {
		var activeDMS = _lyteDMS.activeDMS, comp;

		if (activeDMS) {
			comp = activeDMS.component;
			comp.position();
		}
	}, 100);
});

export { LyteDatemultiselectComponent };

LyteDatemultiselectComponent.register("lyte-datemultiselect", {
    hash: "LyteDatemultiselectComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
