// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/440b22d3a8c9d3775b29a21b25c3c580.js
//ignorei18n_start
import { Service } from "../../../core/src/service.js";
import { ComponentRegistry, _LC, LyteYield } from "../../index.js";
import { deepCopyObject } from "../../../core/src/lyte-utils.js";

class Hide extends Service {
    static register(){
        ComponentRegistry.addFakeDirective(this,"hide");
    }
    constructor(){
        super()
        this.yield = {
            self : this,
            connect : function(comp){
                if(comp._callee && _LC.directive.getTransitionArg(comp._callee, "hide-tag")){
                    this.self.template(comp, _LC.hiddenYieldsDiv);
                }
            },
            disConnect : function(comp){
                if(Array.from(_LC.hiddenYieldsDiv.content.childNodes).indexOf(comp)!=-1){
                    _LC.hiddenYieldsDiv.content.removeChild(comp)
                }
                if(comp._hiddenTemplate && comp._hiddenTemplate.length){
                    this.self.callDisconnectedCallback(comp);
                }
            }
        };
    }
    appendTemp(){
        document.body.appendChild(_LC.hiddenComponentsDiv);
        document.body.appendChild(_LC.hiddenYieldsDiv);
    }
    static triggerGlobalEvent(target, eventName, ev, hasHandled, ignore){
        if(target && target._hiddenBoundEvents && target._hiddenBoundEvents[eventName] && !ev.cancelBubble && (ignore || ev.currentTarget == document.body)){
            let actObj = target._hiddenBoundEvents[eventName];
            let cloneActObj = deepCopyObject(actObj);
            _LC.skipArgProcessing(cloneActObj, eventName, target);
            _LC.throwAction.call(target.parentElement._callee, target.parentElement._callee, eventName, cloneActObj, undefined, undefined, target, ev, hasHandled);
        }
    }
    connect(comp, content){
        let tempParent = comp._callee;
        if(tempParent){
            this.template(comp, _LC.hiddenComponentsDiv ,content)
        }else{
            console.error(' Error:Parent node of current template is not of type TEMPLATE . Cannot hide this template');
        }
    }
    rmHelpersSpecificIndex(node, index){
        if(typeof index=='number' && node._hiddenTemplate[index] && node._hiddenTemplate.length){
            for(let i in node._hiddenTemplate[index]){
                node._hiddenTemplate[index][i].disconnectedCallback();
            }
            if(node.getAttribute("is") == "for"){
                node._hiddenTemplate.splice(index, 1); 
            }else{
                delete node._hiddenTemplate[index];
            }
        }
    }
    forRmHelpers(node){
        if(node._hiddenTemplate && node._hiddenTemplate.length) {
            for(let i=0;i<node._hiddenTemplate.length;i++) {
                for(let j=0;j<node._hiddenTemplate[i].length;j++) {
                    node._hiddenTemplate[i][j].disconnectedCallback();
                }
                node._hiddenTemplate[i] = [];
            }
            node._hiddenTemplate = [];
        }
    }
    ifRmHelpers(node){
        if(node._hiddenTemplate && node._hiddenTemplate.length){
            this.callDisconnectedCallback(node);
        }
    }
    setTxtNode(comp, yld){
        if(_LC.directive.getTransitionArg(comp, "hide-tag")){
            var emptyTextNode = document.createTextNode("");
            _LC.replaceWithPf(yld, emptyTextNode);
        } 
    }
    static shouldStop(node, event){
        let stopBubble = false;
        if(node._hiddenBoundEvents && node._hiddenBoundEvents[event.type] && node._hiddenBoundFlag){
            stopBubble = true;
            node._hiddenBoundFlag = false;
        }
        return stopBubble;
    }
    addHiddenBoundEvents(compNode, content){
        if(!compNode._boundEvents){return;}
        for(let i=0; i<content.childNodes.length ;i++){
            if(compNode._boundEvents){
                content.childNodes[i]._hiddenBoundEvents = compNode._boundEvents;
                if(content.childNodes[i]._boundEvents){
                    content.childNodes[i]._hiddenBoundFlag = true;
                }
            }
        }
    }
    callDisconnectedCallback(node){
        for(let i=0; i<node._hiddenTemplate.length; i++){
            node._hiddenTemplate[i].disconnectedCallback();
        }
        node._hiddenTemplate=[];
    }
    template(compNode, hiddenDomDiv, content){
        _LC.setIgnoreDisconnect(true);
        compNode._ignoreDisconnect=true;
        compNode.__lyteIgnore=true;
        let parentElement=compNode._callee;
        let parentEle=compNode.parentElement;
        let domContent=document.createDocumentFragment();
        let len=compNode.childNodes.length;
        let firstChild=document.createTextNode("");
        let lastChild=document.createTextNode("");
        compNode._firstChild=firstChild;
        compNode._lastChild=lastChild;
    
        for(let i=0;i<len;i++){
            compNode.childNodes[0]._hiddenBoundEvents=compNode._boundEvents;
            if(compNode.childNodes[0]._boundEvents){
                compNode.childNodes[0]._hiddenBoundFlag=true;
            }
            domContent.append(compNode.childNodes[0]);
        }
        let elePos=Array.from(parentEle.childNodes).indexOf(compNode);
        parentEle.insertBefore(firstChild,parentEle.childNodes[elePos]);
    
    
        elePos=Array.from(parentEle.childNodes).indexOf(compNode);
        if(domContent.childNodes.length){
            parentEle.insertBefore(domContent,parentEle.childNodes[elePos]);
            elePos=Array.from(parentEle.childNodes).indexOf(compNode);
        }
    
        if(content){
            this.addHiddenBoundEvents(compNode, content);
            parentEle.insertBefore(content,parentEle.childNodes[elePos]);
        }
    
        elePos=Array.from(parentEle.childNodes).indexOf(compNode);
        parentEle.insertBefore(lastChild,parentEle.childNodes[elePos]);
        let parentHelper = compNode._parentHelper;
        if(parentHelper){
            if(parentHelper.tagName == "LYTE-YIELD"){
                parentHelper._hiddenTemplate.push(compNode);
            }
            else if(parentHelper._forContent){
                if(compNode._cx){
                    let ind = compNode._cx.itemIndex;
                    // let eleArr = parentHelper._forContent[ind];
                    // eleArr.push(content.childNodes[i]);
                    if(!parentHelper._hiddenTemplate[ind]){
                        parentHelper._hiddenTemplate[ind] = [];
                    }   
                    parentHelper._hiddenTemplate[ind].push(compNode);
                }
            } else if(parentHelper._caseContent) {
                // compNode._parentHelper._caseContent.push(content.childNodes[i]);
                parentHelper._hiddenTemplate.push(compNode);
            } 
            // else if(parentHelper._yieldContent) {
            //     compNode._parentHelper._yieldContent.push(content.childNodes[i]);
            // } else if(parentHelper._renderedComponent) {
    
            // }
        } 
        else if(compNode._callee){
            compNode._callee._hiddenTemplate.push(compNode);
        }
        hiddenDomDiv.content.appendChild(compNode);
        compNode._ignoreDisconnect=false;
        _LC.setIgnoreDisconnect(false);
        compNode.__lyteIgnore=false;
    }
    didConnect(regIns){
        ComponentRegistry.hide.define.remove(LyteYield);
        let commonCls = regIns._registeredCommonClass;
        for(let cls in commonCls){
            ComponentRegistry.hide.define.remove(commonCls[cls]);
            ComponentRegistry.hide.define.getContent(commonCls[cls]);
        }
    }
}

Hide.__lMod = "Save";

Hide.define = {
    remove : function(classDef){
        const $remove = Element.prototype.remove;
        Object.defineProperty(classDef.prototype,"remove",{
            configurable : true,
            writable : true,
            value : function(){
                if(this._firstChild == undefined || this._firstChild.parentElement == null){
                    $remove.call(this);
                    this._firstChild = undefined;
                    this._lastChild = undefined;
                    return;
                }
                firstNode = this._firstChild;
                parent = firstNode.parentElement;
                lastNode = this._lastChild;
                while(firstNode.nextSibling != lastNode){
                    parent.removeChild(firstNode.nextSibling);
                }
                this._firstChild = undefined;
                this._lastChild = undefined;
                parent.removeChild(firstNode);
                parent.removeChild(lastNode);
                return;
            }
        })
    },
    getContent : function(customCrmComponent){
        Object.defineProperty(customCrmComponent.prototype,"getContent",{
            configurable : true,
            writable : true,
            value : function(){
                arr = [];
                if(this._firstChild == undefined){
                    console.error("Cannot access template through getContent");
                }
                firstNode = this._firstChild;
                lastNode = this._lastChild;
                firstNode = firstNode.nextSibling;
                while(firstNode!=lastNode){
                    arr.push(firstNode);
                    firstNode=firstNode.nextSibling;
                }
                return arr;
            }
        });
    }
};
Hide.register();
export {Hide};
//ignorei18n_end
