// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/4966f55aadfc52767346e5c9b0dbe2d6.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../node_modules/@slyte/core/index.js";
import { objectUtils, arrayUtils } from "../../../node_modules/@slyte/component/index.js";
import { Component, LyteUiComponentComponentRegistry } from "../../component.js";
window._lyteSchedulerStartId = 0;
window._lyteSchedulerEndId = 0;

class LyteSchedulerComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {array} ltPropEvent
             */
			'ltPropEvent' : prop('array' ,{ default : [] } ),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {string} ltPropFormat
             */
			'ltPropFormat' : prop('string',{ default : 'DD-MM-YYYY hh:mm A'}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {boolean} ltPropDraggable
			 * @default false
             */
			'ltPropDraggable' : prop('boolean',{default:false}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {number} ltPropMaxAllDayEvent
			 * @default 5
             */
			'ltPropMaxAllDayEvent' : prop('number',{default: 5}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {number} ltPropMaxTimelineEvent
			 * @default 2
             */
			'ltPropMaxTimelineEvent' : prop('number',{default: 2}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {boolean} ltPropMultiUserView
			 * @default false
             */
			'ltPropMultiUserView' : prop('boolean',{default: false}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {array} ltPropSchedulerUser
             */
			'ltPropSchedulerUser' : prop('array',{default:[]}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {array} ltPropBusinessDays
             */
			'ltPropBusinessDays' : prop('array',{default:[1,2,3,4,5]}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {array} ltPropBusinessHour
             */
			'ltPropBusinessHour' : prop('array',{default:['09:00 AM','06:00 PM']}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {string} ltPropSchedulerView
			 * @default month
             */
			'ltPropSchedulerView' : prop('string',{default:'month'}),
			/**
             * This specifies the type of the scheduler.
             * @componentProperty {boolean} ltPropSchedulerHeaderYield
			 * @default false
             */
			'ltPropSchedulerHeaderYield' : prop('boolean',{ default : false }),
			'ltPropHiddenEventPopoverFreeze' : prop('boolean',{default : false}),
			'ltPropCalendarPopoverFreeze' : prop('boolean',{default : false}),
			'ltPropCalendarHeaderType' : prop('boolean',{default : false}),
			'ltPropHideCalendarNav' : prop('boolean',{default : false}),
			'ltPropWeekEventOverlap' : prop('boolean',{default : false}),
			'ltPropEventGap' : prop('number',{default : 5}),
			'ltPropTimelineNmore' : prop('boolean',{default: true}),
			"ltPropHiddenPopoverWrapper" : prop('string',{default : ''}),
			'ltPropCalendarPopoverWrapper' : prop('string',{default : ''}),
			'ltPropDragThershold' : prop('number',{default : 5}),
			'ltPropContinentalTimeFormat' : prop('boolean',{default : false}),
			'ltPropyieldNav' : prop('boolean',{default : false}),
			'ltPropTitleYield' : prop('boolean',{ default : false }),
			'ltPropShowWeekNumber' : prop('boolean',{dafault: false}),
			'ltPropPopoverAllowMultiple' : prop('boolean',{default: false}),
			'ltPropWorkingDayOnly' : prop('boolean',{default: false}),
			'ltPropSchedulerLabelFormat' : prop('string'),	
			'ltPropShowCurrentTime' : prop('boolean',{default : true}),
			'ltPropManipulatedEvent' : prop('array',{ default : [] }),
			'ltPropLabel' :  prop('array',{default : ['SUN','MON','TUE','WED','THU','FRI','SAT']}),
			'ltPropCurrMonth' : prop('number'),
			'ltdisplayCurrDate' : prop('string'),
			'ltPropCurDay' :prop('string'),
			'ltPropDisplayYear' : prop('number'),
			'ltPropWeekStart' : prop('number'),
			'openSchedulerPopover' : prop('boolean',{default: false}),
			'ltPropTimeLine' : prop('number',{default : 30}),
			'timeLine' : prop('array',{ default:['12am','1am','2am','3am','4am','5am','6am','7am','8am','9am','10am','11am','12pm','1pm','2pm','3pm','4pm','5pm','6pm','7pm','8pm','9pm','10pm','11pm']}),
			'timelineArray' : prop('array',{default:[]}),
			'currDay' : prop('string'),
			'ltCurrStartDate' : prop('string'),
			'ltCurrEndDate' : prop('string'),
			'ltPropCurrentDate' : prop('object',{default : new Date()}),
			'labelFF' :  prop('array',{default : ['Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday']}),
			'label' : prop('array',{default : ['Sun','Mon','Tue','Wed','Thu','Fri','Sat']}),
			'month' :  prop( 'array', { default : [ 'January', 'February', 'March', 'April', 'May','June' ,'July','August','September','October','November','December' ] } ),
			'montharray' : prop('array', { default : [] } ),
			'DateArray' : prop('array', { default : [] } ),
			"days" : prop('array',{default:['sunday','monday','tuesday','wednesday','thursday','friday','saturday']}),
			"startDateOfView" : prop('string',{default : ''}),
			"hiddenObj" : prop('object',{default:{}}),
			"eventArr" : prop('object',{default:{}}),
			"todayButton": prop('string',{default:window._lyteUiUtils.i18n('today')}),
			'ltPropSchedulerSearch': prop('boolean',{default : false}),
			'ltPropSchedulerSearchYield' : prop('boolean',{default : false}),
			"selectionType" : prop('string',{default:''}),
			'ltPropSortUser' : prop('boolean',{default: false}),
			'checkboxArr' : prop('array',{default:[]}),
			"selectedUserArr" : prop('array',{default:[]}),
			 //lyte-menu data
			"ltPropSchedulerMenu" : prop('boolean',{default: false}),
			"ltPropSchedulerMenuYield" : prop('boolean',{default: false}),
			"ltPropSchedulerMenuContent" : prop('array',{default:[]}),
			"ltPropSchedulerMenuUserValue" : prop('string'),
			"ltPropSchedulerMenuSystemValue" : prop('string'),
			"ltPropSchedulerMenuDescription" : prop('string'),
			"ltPropSchedulerMenuId" : prop('string'),
			"ltPropSchedulerMenuClass" : prop('string'),
			"ltPropSchedulerMenuPosition" : prop('string'),
			"ltPropSchedulerMenuWidth" : prop('string',{default:'auto'}),
			"ltPropSchedulerMenuHeight" : prop('string',{default:'auto'}),
			"ltPropSchedulerMenuCallout" : prop('boolean', {default : false}),
			"ltPropSchedulerMenuFreeze" : prop('boolean',{default : false}),
			"ltPropSchedulerMenuQueryClass" : prop('string',{default: 'lyteMenuSelected'}),
			"ltPropSchedulerMenuBindToBody" : prop('boolean',{default : true}),
			"ltPropSchedulerMenuAnimate" : prop('boolean',{default : false}),
			"ltPropSchedulerMenuWrapperClass" : prop('string'),
			"ltPropSchedulerMenuAria" : prop('boolean', { default : false}),
			"ltPropSchedulerMenuAriaAttribute" : prop('object' , { default : { role : 'menu'} }) ,
			"isEventAdded" : prop('boolean', { default : false}),
			"isResizing" : prop('boolean', { default : false}),
			"dragDir" : prop('string', { default : ''}),
			"currEventTag": prop('string', { default : ''}),
			"breakpoint": prop('number', { default : 0}),
			"ybreakpoint": prop('number', { default : 0}),
			"curPgX": prop('number', { default : 0}),
			"curPgY": prop('number', { default : 0}),
			"quadrant": prop('number', { default : 0}),
			"ltPropEventResize": prop('boolean', { default : false}),
			"tempResizeClientY": prop('number'),
			"hiddenUser" : prop('object',{default: {}})
		}), arg1);		
	}

    init() {
		this.generateTimeLine();
		var date = this.getData('ltPropCurrentDate');
		if(!date){
			date = new Date();
			this.data.ltPropCurrentDate = new Date();
		}
		this.setData('ltdisplayCurrDate',('0' + date.getDate()).slice(-2)+ ' ' + window._lyteUiUtils.i18n(this.data.month[date.getMonth()]) + ' ' + date.getFullYear());
		this.setData('ltPropCurDay', window._lyteUiUtils.i18n( this.data.days[date.getDay()] ) );
		this.setData('ltPropCurrMonth',date.getMonth());
		this.setData('ltPropDisplayYear',date.getFullYear());	
	}

    didConnect() {
		var _this = this;
		this.$node.addEvent = function(event){
			if(event.start){
				var events  = _this.getData('ltPropEvent');
				
				if(!$u.findWhere(events,{'id':event.id})){
					arrayUtils( events , 'push' , event );
				}
			}
		}
		this.$node.replaceEvent = function(event,replaceEvent){
			var events = _this.getData('ltPropEvent');
			var delIdx = events.findIndex(function( elem ){
				return elem.id === event.id ? true : false;
			});
			arrayUtils( events, 'replaceAt', delIdx, replaceEvent );
		}
		this.$node.UpdateEvent = function(eventid, key , value){
			var events  = _this.getData('ltPropEvent');
			var eventData = $u.findWhere(events,{'id':eventid});
			objectUtils( eventData , 'add' , key, value );
			_this.callCurrentView();
		}
		this.$node.deleteEvent = function(event_id){
			if(event_id){
				var events  = _this.getData('ltPropEvent');
				var event_data = $u.findWhere(events,{'id':event_id});

				if(event_data){
					arrayUtils( events , 'removeObjects' , event_data );
				}

			}
		}
        this.$node.today = function(){
			_this._navigated = false;
			_this.ResetTimeLine();
			$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',true);
			var date = _this.getData('ltPropCurrentDate') || new Date();
			_this.setData('ltPropCurrMonth',date.getMonth());
			_this.setData('ltPropDisplayYear',date.getFullYear());
			_this.callCurrentView(date,false,_this.getData('ltPropWeekStart'));
			
			if(_this.getMethods('onTodayClick')){
				_this.executeMethod('onTodayClick',$L.moment(date),_this.$node);
			}
		}
		this.$node.moveTo = function(date){
			_this._navigated =  true;
			var day = $L.moment(date,'DD-MM-YYYY');
			if(_this.getData('ltPropSchedulerView') == 'month'){
				day.startOf('month');
			}
			_this.callCurrentView(day.getDObj(),false,_this.getData('ltPropWeekStart'));
			_this.findActiveKeys(day);
		}
		this.$node.moveToDay = function(date){		
			date = $L.moment(date,'DD-MM-YYYY').getDObj();
			var view = _this.getData('ltPropSchedulerView');
			_this.ResetTimeLine();
			if($L.moment(date,'DD-MM-YYYY').format('DD-MM-YYYY') ===  $L.moment(new Date()).format('DD-MM-YYYY')){
				_this._navigated = false;
				$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',true)
			}else{
				_this._navigated = true;
				$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false)
			}
			if(view !== 'day'){
				_this.moveToDay = date;
				_this.setData('ltPropSchedulerView','day');
			}else{
				_this.callCurrentView(date,false,_this.getData('ltPropWeekStart'));
			}
			_this.findActiveKeys(date);
		}
		this.$node.selectUser = function(userid){
			if(typeof userid == 'string'){
				_this.adduser(userid);
			}else if(Array.isArray(userid)){
				userid.forEach(function(user){
					_this.adduser(user);
				})
			}
		}
		this.$node.unselectUser = function(userid){
			if(typeof userid == 'string'){
				_this.removeUser(userid);
			}else if(Array.isArray(userid)){
				userid.forEach(function(user){
					_this.removeUser(user);
				})
			}
		}
		this.$node.clearSelectedUser = function(){
			_this.clearUser();
		}
		this.$node.hideHiddenPopover = function(){
			var popover = $L('#lyteSchedulerHiddenEvent',_this.$node)[0];
			if(popover){
				popover.setData('ltPropShow',false);
			}
		}
		this.$node.openHiddenUserPopover = function( hiddenEvent, user, target, popoverProp ){
			var userDetails =  $u.clone($u.findWhere(_this.getData('ltPropSchedulerUser'),{'id':user}));
			userDetails.eventCount = hiddenEvent.length;
			_this.setData('hiddenUser',userDetails);
			_this.setData('hiddenEvent',$u.clone(hiddenEvent));
			_this.setData('ltPropUserHiddenEvent',true);
			$L('#schedulerPopover',_this.$node)[0] && $L('#schedulerPopover',_this.$node)[0].setAttribute('id',0);
			target.setAttribute('id','schedulerPopover');
			var popover = _this.$node.querySelector('#lyteSchedulerHiddenEvent');
			popover.ltProp(popoverProp);
			popover.setData('ltPropShow',true);			
			var poptitle = $L('.lyteSchedulerMoreEventsPopTitleMonthVal',popover.component.actualModalDiv)[0];
			var popDay = $L('.lyteSchedulerMoreEventsPopTitleDayVal',popover.component.actualModalDiv)[0];
			poptitle.innerHTML = ' ';
			popDay.innerHTML = ' ';
			var year = _this.data.ltPropDisplayYear;
			var month = _this.data.ltPropCurrMonth;
			popDay.appendChild( document.createTextNode(window._lyteUiUtils.i18n(_this.data.month[month]) + ' ' + year ));
			if(_this.getMethods('openHiddenUserEvent')){
				_this.executeMethod('openHiddenUserEvent',event.target);
			}
		}
		this.addEventForNav();
		this.callCurrentView(undefined,false,this.getData('ltPropWeekStart'));
		this.addScroll();
		this.MultiDayViewNav();
		this.findActiveKeys( $L.moment(this.getData('startDateOfView'),'DD-MM-YYYY') );
		this.$node.rendered = true;
		this._previousView = this.getData('ltPropSchedulerView');
		this._currentView = this.getData('ltPropSchedulerView');
	}

    didDestroy() {
		if( $L && $L.schedulerEventResize ){
			$L.schedulerEventResize( "destroy", this );
		}
	}

    adduser(userid) {
		var selectedList = $u.clone(this.getData('selectedUserArr'));
		var users  = this.getData('ltPropSchedulerUser');
		// var userData = $u.findWhere(users,{'id':userid});
		// var index = users.indexOf(userData);
		if(!selectedList.includes(userid)){
			selectedList.push(userid)
			this.setData('selectedUserArr',selectedList);
		}
	}

    removeUser(userid) {
		var selectedList = $u.clone(this.getData('selectedUserArr'));
		var users  = this.getData('ltPropSchedulerUser');
		// var userData = $u.findWhere(users,{'id':userid});
		// var index = users.indexOf(userData);
		var SelectedUserIndex = selectedList.indexOf(userid);
		if( SelectedUserIndex !== -1 ){
			selectedList.splice( SelectedUserIndex, 1 );
			this.setData('selectedUserArr',selectedList);
		}
	}

    addEventForNav() {
		var leftNav = $L('.lyteSchedulerNavLeft',this.$node)[0];
		var rightNav = $L('.lyteSchedulerNavRight',this.$node)[0];
		if(leftNav){
			leftNav.addEventListener('click', this.schedulerNav.bind(this));
		}
		if(rightNav){
			rightNav.addEventListener('click', this.schedulerNav.bind(this));
		}
	}

    addScroll() {
		var _this = this;
		var isMultiView = this.getData('ltPropMultiUserView');
		var view = this.getData('ltPropSchedulerView');
		if(isMultiView && (view == 'week' || view == 'day')){
			var prevScroll;
			$L('.lyteSchedulerMultiUserView',this.$node)[0].addEventListener('scroll',function(){
				var scrollTop = this.scrollTop;
				if((this.scrollHeight - scrollTop) <= this.offsetHeight && prevScroll !== scrollTop){
					if(_this.getMethods('onScrollEnd')){
						_this.executeMethod('onScrollEnd',_this.$node);
					}
				}
				prevScroll = scrollTop;
			});
		}
	}

    toggleEventResize() {
		if( this.getData('ltPropEventResize') ){
			if(this.getData('ltPropSchedulerView') === 'month' && this.getData('ltPropMultiUserView') === true) {
				return;
			}
			if( $L && $L.schedulerEventResize ){
				$L.schedulerEventResize( "connect", this );
			}
		}
		else{
			this.destoryEventResize();
		}
	}

    destoryEventResize() {
		if( $L && $L.schedulerEventResize ){
			$L.schedulerEventResize( "destroy", this );
		}
	}

    findDayViewPosition(previousView, currentView) {
		if( (previousView == 'day' && (['week','month'].includes(currentView))) ){
			this._DaypreviousDate = this.getData('startDateOfView');
		}
		if((previousView == 'week' && currentView == 'month')){
			this._WeekpreviousDate = this.getData('startDateOfView');
		}
		if( ((currentView == 'day' && this._DaypreviousDate) || (currentView == 'week' && this._WeekpreviousDate)) ){
			if(currentView == 'day'){
				var prevDate = $L.moment(this._DaypreviousDate,'DD-MM-YYYY').get('time');
			}else{
				var prevDate = $L.moment(this._WeekpreviousDate,'DD-MM-YYYY').get('time');
			}
			
			if(previousView == 'month'){
				var startDate = new Date(this.data.ltPropDisplayYear,this.data.ltPropCurrMonth,1).getTime();
				var endDate = new Date(this.data.ltPropDisplayYear,this.data.ltPropCurrMonth + 1,0).getTime();
			}else if(previousView == 'week'){
				var dateArray =  this.data.ltPropDateArray;
				var startDate = dateArray[0].val.get('time');
				var endDate = dateArray[dateArray.length - 1].val.get('time');
			}
			if(startDate <=  prevDate && endDate >= prevDate){
				this._MovetoPreviousDate = true;
			}else{
				this._MovetoPreviousDate = false;
			}
		}
	}

    findWeekoftheYear(now, weekStartDay) {
		var onejan = new Date(now.getFullYear(), 0, 1);

		var dayOffset = (onejan.getDay() - weekStartDay + 7) % 7;
		var firstWeekStart = new Date(onejan);
		firstWeekStart.setDate(onejan.getDate() - dayOffset);
		
		var diffDays = Math.floor((now - firstWeekStart) / 86400000);
		var week = Math.ceil((diffDays + 1) / 7);
	
		if (week === 53) {
			var lastdate = new Date(now.getFullYear(), 11, 31);
			var lastWeekOffset = (lastdate.getDay() - weekStartDay + 7) % 7;
			if (lastWeekOffset !== 6) {
				week = 1;
			}			
		}
		return week;
	}

    findActiveKeys(startDate) {
		// var startDate = $L.moment(this.getData('startDateOfView'),'DD-MM-YYYY');
		var today_btn = $L('.schedulerToday',this.$node)[0];
		var today = $L.moment(new Date());
		if(today_btn.getData('ltPropDisabled')){
			today_btn.classList.add('lyteSchedulerActiveBtn');
			$L('.lyteSchedulerNavRight')[0].classList.remove('lyteSchedulerActiveBtn');
			$L('.lyteSchedulerNavLeft')[0].classList.remove('lyteSchedulerActiveBtn');
		}else if(today.fromNow(startDate).past){
			today_btn.classList.remove('lyteSchedulerActiveBtn');
			$L('.lyteSchedulerNavRight')[0].classList.remove('lyteSchedulerActiveBtn');
			$L('.lyteSchedulerNavLeft')[0].classList.add('lyteSchedulerActiveBtn');
		}else{
			today_btn.classList.remove('lyteSchedulerActiveBtn');
			$L('.lyteSchedulerNavLeft')[0].classList.remove('lyteSchedulerActiveBtn');
			$L('.lyteSchedulerNavRight')[0].classList.add('lyteSchedulerActiveBtn');
		}
	}

    schedulerNav(event) {

		var schedulerView =  this.getData('ltPropSchedulerView');	
		this._navigated = true;	
		var weekStart = this.getData('ltPropWeekStart');
		weekStart = weekStart === undefined ? 0 : weekStart;
		var _this = this;
		var isMultiView = this.getData('ltPropMultiUserView');
		switch(schedulerView){
			case 'month':
				if(event.target.classList.contains('lyteSchedulerNavLeft')){
					var displayMonth = this.data.ltPropCurrMonth - 1;
				}else{
					var displayMonth = this.data.ltPropCurrMonth + 1;
				}
				var date = new Date(this.data.ltPropDisplayYear,displayMonth);
				this.setData('startDateOfView',$L.moment(date).format('DD-MM-YYYY'));
				this.monthView($L.moment(date).getDObj());
				if(displayMonth ===  $L.moment(this.getData('ltPropCurrentDate')).getDObj().getMonth() && this.data.ltPropDisplayYear ===  $L.moment(this.getData('ltPropCurrentDate')).getDObj().getFullYear()){
					$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',true);
				}else{
					$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false);
				}
				this._WeekpreviousDate = false;
				this._DaypreviousDate = false;
				break;
			case 'week':
				var start_date = this.getData('ltPropDateArray')[0].date;
				var date =  new Date(this.data.ltPropDisplayYear,this.data.ltPropCurrMonth,start_date);
				var dayCount = this.data.ltPropWorkingDayOnly ? 7 : 7;
				if(event.target.classList.contains('lyteSchedulerNavLeft')){
					date.setDate(date.getDate()-dayCount);
				}else{
					date.setDate(date.getDate()+dayCount);
				}
				if(this.findWeekoftheYear(date,weekStart) == this.findWeekoftheYear($L.moment(this.getData('ltPropCurrentDate')).getDObj(),weekStart)){
					$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',true);	
				}else{
					$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false);
				}
				this.weekView($L.moment(date).getDObj(),this.getData('ltPropWeekStart'));
				this._DaypreviousDate = this.getData('startDateOfView');
				break;
			case 'day':
				if(isMultiView){
					var events = $L('lyte-scheduler-event-tag');
					for(var index =0 ;index < events.length;index++){
						events[index].remove();
					}
				}
				var curr_date = $L.moment(this.getData('startDateOfView'),'DD-MM-YYYY').getDObj();
				var date =  new Date(curr_date.getFullYear(),curr_date.getMonth(),curr_date.getDate());
				if(event.target.classList.contains('lyteSchedulerNavLeft')){
					date.setDate(date.getDate()-1);
				}else{
					date.setDate(date.getDate()+1);
				}
				if($L.moment(date).format('DD-MM-YYYY') ===  $L.moment(this.getData('ltPropCurrentDate')).format('DD-MM-YYYY')){
					this._navigated = false;
					$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',true);
				}else{
					this._navigated = true;
					$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false);
				}
				this.ResetTimeLine();
				var start_date = $L.moment(date);
				var _this = this;
				this.setData('startDateOfView',$L.moment(date).format('DD-MM-YYYY'));
				_this.setCurrentTimeLine(start_date)				
				start_date = start_date.getDObj();
				this.dayView(start_date);
				break;
		}
		this.findActiveKeys($L.moment(date));
		this.setData( 'ltdisplayCurrDate', ('0' + date.getDate()).slice(-2) + ' ' + window._lyteUiUtils.i18n(this.data.month[ date.getMonth() ])  );
		this.setData('ltPropCurDay',window._lyteUiUtils.i18n(this.data.days[date.getDay()]));
		$L.fastdom.mutate(function(){
			_this.addDragAndDrop();
			_this.destoryEventResize();
			_this.toggleEventResize();
		})
		if(this.getMethods('onNavClick')){
			this.executeMethod('onNavClick',$L.moment(date));
		}
	}

    callCurrentView(date, init, weekStart) {
		this._DayNumberHeight = 0;
		var schedulerView = this.getData('ltPropSchedulerView');
		var _this = this;
		weekStart = weekStart === undefined ? 0 : weekStart;
		var isMultiView = this.getData('ltPropMultiUserView');
		var AllDayMaxEvent = this.getData('ltPropMaxAllDayEvent');
		var TimelineMaxEvent;
		if(this.getData('ltPropTimelineNmore')){
			TimelineMaxEvent = this.getData('ltPropMaxTimelineEvent')
		}else{
			TimelineMaxEvent = Infinity;
		}
		
		var format = this.getData('ltPropFormat');
		var eventManipulator = new window._lyteUiUtils.eventManipulator( format );
		var businessDays = this.getData('ltPropBusinessDays');
		var isWorkingOnly = this.getData('ltPropWorkingDayOnly');
		var popover = $L('#calendarPopover',this.$node)[0];
		switch(schedulerView){
			case 'month':
				if( this._navigated && this.data.startDateOfView){
					var start_date = $L.moment(new Date(this.data.ltPropDisplayYear,this.data.ltPropCurrMonth,1));
					date = date ? date :  new Date(start_date.format('YYYY'),start_date.format('MM') - 1,start_date.format('DD'));
				}else if( !date ){
					date = this.getData('ltPropCurrentDate');
				}
				if(popover.component.actualModalDiv){
					this.setData('selectionType','month');
					var calendar = $L('lyte-calendar',popover.component.actualModalDiv)[0];
					if(calendar){
						calendar.setData('ltPropSelectionType','month');
						if(calendar.getAttribute('view-type') == 'dateView'){
							$L('.lyteDrillCalHeaderButton',popover.component.actualModalDiv)[0].click();
						}
					}
				}else{
					this.setData('selectionType','month');
				}
				
				this.setData('todayButton','thisMonth');
				var eventManipulator = new window._lyteUiUtils.eventManipulator( format );
				if(isMultiView){
					var event = eventManipulator.generateEventObjArrNmore( this.getData('ltPropManipulatedEvent'), format, schedulerView,  weekStart , isMultiView , AllDayMaxEvent, null, null, null, this.getData('selectedUserArr'));
				}else{	
					var event = eventManipulator.generateEventObjArr( this.getData('ltPropManipulatedEvent'), format, schedulerView,  weekStart , isMultiView , AllDayMaxEvent);
				}
				
				this.setData('ltPropProcessedData',event.eventObj);
				this.setData('hiddenObj',event.hiddenObj);
				this.monthView($L.moment( date,'DD-MM-YYYY').getDObj() );
				break;
			case 'week':
				if(isMultiView){
					var event = eventManipulator.generateEventObjArr( this.getData('ltPropManipulatedEvent'), format, schedulerView,  weekStart , isMultiView ,AllDayMaxEvent,businessDays,isWorkingOnly);
				}else{	
					var event = eventManipulator.generateEventObjArrNmore( this.getData('ltPropManipulatedEvent'), format, schedulerView,  weekStart , isMultiView , AllDayMaxEvent ,TimelineMaxEvent, businessDays, isWorkingOnly);
				}
				if(this._MovetoPreviousDate){
					var start_date =  $L.moment( this._WeekpreviousDate,'DD-MM-YYYY');
					this._WeekpreviousDate = null;
					this._MovetoPreviousDate = false;
					date = date ? date :  new Date(start_date.format('YYYY'),start_date.format('MM') - 1,start_date.format('DD'));
				}else if( this._navigated && this.data.startDateOfView){
					var start_date = $L.moment( this.data.startDateOfView,'DD-MM-YYYY');
					date = date ? date :  new Date(start_date.format('YYYY'),start_date.format('MM') - 1,start_date.format('DD'));
				}else if( !date ){
					date = this.getData('ltPropCurrentDate');
				}
				this.setData('ltPropProcessedData',event.eventObj);
				this.setData('hiddenObj',event.hiddenObj);
				this.weekView($L.moment( date,'DD-MM-YYYY').getDObj(),weekStart);
				if(this.getData('ltPropWeekStart') !== undefined){
					this.setData('selectionType','week');
				}else{
					this.setData('selectionType','day');
				}
				this.setData('todayButton','thisWeek');
				break;
			case 'day':	
				if(this._MovetoPreviousDate){
					var start_date =  $L.moment( this._DaypreviousDate,'DD-MM-YYYY');
					this._DaypreviousDate = null;
					this._MovetoPreviousDate = false;
					date = date ? date :  new Date(start_date.format('YYYY'),start_date.format('MM') - 1,start_date.format('DD'));
				}else if( this._navigated && this.data.startDateOfView){
					var start_date = $L.moment( this.data.startDateOfView,'DD-MM-YYYY');
					date = date ? date :  new Date(start_date.format('YYYY'),start_date.format('MM') - 1,start_date.format('DD'));
				}else if( !date ){
					date = this.getData('ltPropCurrentDate');
				}
				if(!isMultiView){
					TimelineMaxEvent = Infinity;
				}
				var event = eventManipulator.generateEventObjArrNmore( this.getData('ltPropManipulatedEvent'), format, schedulerView,  weekStart , isMultiView ,AllDayMaxEvent,TimelineMaxEvent);
				var currDate = $L.moment( date );
				this.MultiDayViewNav();
				this.setData('startDateOfView',currDate.format('DD-MM-YYYY'));
				this.setData('ltPropProcessedData',event.eventObj);
				this.dayView(currDate.getDObj());
				this.setData('hiddenObj',event.hiddenObj);
				this.setData('selectionType','day');
				this.setData('todayButton',window._lyteUiUtils.i18n('today'));
				this.setCurrentTimeLine($L.moment( date ));
				break;
		}
		if(!init){
			this.isToday(new Date());
		}
		var isToday = $L('.schedulerToday',this.$node)[0].getData('ltPropDisabled');
		date = (init ||isToday ) ? new Date() : $L.moment(this.data.startDateOfView,'DD-MM-YYYY').getDObj();
		if(schedulerView ==  'month' && !(init ||isToday ) ){
			date.setMonth(this.data.ltPropCurrMonth);
			date.setDate(1);
			date.setYear(this.data.ltPropDisplayYear)
		}
		this.findActiveKeys(date);
		this.setData('ltdisplayCurrDate', ('0' + date.getDate()).slice(-2) + ' ' + window._lyteUiUtils.i18n(this.data.month[date.getMonth()]) + ' ' + date.getFullYear());
		this.setData('ltPropCurDay',window._lyteUiUtils.i18n(this.data.days[date.getDay()]));
		$L.fastdom.mutate(function(){
			_this.addDragAndDrop();
		})
	}

    isToday(date) {
		var view = this.getData('ltPropSchedulerView');
		var dateArray = this.getData('ltPropDateArray');
		var startDate;
		var end_date;
		if(view == 'month'){
			startDate = $L.moment(dateArray[0][0].val.get('time'));
			if(startDate.get('date') !== 1){
				startDate.startOf('month');
				startDate.add(1,'month');
			}
			end_date = dateArray[dateArray.length - 1][6].val;
		}else if(view == 'week'){
			startDate = dateArray[0].val;
			end_date = dateArray[dateArray.length-1].val;
		}else{
			startDate = $L.moment(this.getData('startDateOfView'),'DD-MM-YYYY');
			end_date = $L.moment(this.getData('startDateOfView'),'DD-MM-YYYY');
		}
		if(this.isDateEqual(startDate.getDObj()) || (!startDate.fromNow($L.moment(date)).past && end_date.fromNow($L.moment(date)).past) || this.isDateEqual(end_date.getDObj())){
			$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',true);
			this._navigated = false;
		 }else{
			$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false);
			this._navigated = true;
		 }
	}

    monthView(date) {
		var date = date || new Date();
		var displayHead = date.getMonth();
		this.setData('ltPropCurrMonth', displayHead);
		this.setData('ltdisplayHead',this.data.month[displayHead] +  ' ' + date.getFullYear());
		this.setData('ltPropDisplayYear', date.getFullYear());
		this.generateMonth(new Date(date.getFullYear(), date.getMonth(), 1));
	}

    generateMonth(startOfMonth) {
		var result=[];
		var current_month = startOfMonth.getMonth();
		var eventArr  = this.getData('ltPropProcessedData');
		var businessDay = this.getData('ltPropBusinessDays');
		var date = new Date(startOfMonth.getFullYear(), startOfMonth.getMonth(), 1); 
		var weekstart = this.getData('ltPropWeekStart') === undefined ? 0 : this.getData('ltPropWeekStart');
		var cur_Month = date.getMonth();
		var start_diff = date.getDay()  - weekstart ;
		if( date.getDay() > weekstart ){
			date.setDate(date.getDate() - (start_diff));
		}else{
			date.setDate(date.getDate() - ((7 + start_diff) % 7));
		}
		for(var row = 0 ; row < 6; row++ ){
			var week = new Array(7);
			for(var index = 0 ; index < 7 ; index++){
				var year = date.getFullYear();
				var month = ( date.getMonth() + 1 );
				var curr_date = ('0' + date.getDate()).slice(-2) + '-' + ('0' + month).slice(-2) + '-' + year;

				var is_business = businessDay.includes(date.getDay()) ? 'lyteSchedulerBusinessDay' : 'lyteSchedulerNonBusinessDay' ;
				week[ index ] = {'val' : $L.moment(curr_date,'DD-MM-YYYY') ,'date' : date.getDate(), 'buisness' : is_business,'events' : eventArr[curr_date] , 'currentDate' : this.isDateEqual( curr_date ),'month' :month-1,'year': year};
				if(current_month == date.getMonth()){
					week[index].current_month = true;
				}
				date.setDate( date.getDate() + 1 );
			}
			result.push( week );
			if(cur_Month != date.getMonth()){
				break;
			}
		}
		var startDate = result[0][0].val;
		this.setData('startDateOfView',startDate.format('DD-MM-YYYY'));
		startDate = startDate.getDObj(); 
		var _this = this;
		$L.fastdom.mutate( function(){
			_this.setDisplayDate($L("scheduler-start-date",this.$node)[0],startOfMonth);
		});
		this.setData('ltPropDateArray', result);
	}

    weekView(date, weekStart) {
		weekStart = weekStart ? weekStart : 0;
		var date = date || new Date();
		this.generateWeek(date, weekStart);
		var right = 'right';
		if(window._lyteUiUtils.getRTL()){
			right = 'left';
		}
		if(!this.getData('ltPropMultiUserView')){
			var _this = this;
			setTimeout(function(){
				var allDayline = $L('.lyteSchedulerAllDayHighlight',_this.$node)[0];
				var alldayth = $L('.lyteSchedulerSingleUserAllDay',_this.$node)[0];
				if(allDayline && alldayth){
					allDayline.style.top =  alldayth.getBoundingClientRect().bottom - allDayline.offsetParent.getBoundingClientRect().top - 1 + 'px';
					allDayline.style[right] = window._lyteUiUtils.getScrollBarWidth() + 'px';
				}
			},10)
		}
	}

    generateWeek(date, weekStart) {
		var startofWeek = new Date(date.getFullYear(),date.getMonth(), date.getDate() );
		var start_diff = date.getDay()  - weekStart  ;
		var isWorkingOnly = this.getData('ltPropWorkingDayOnly');
		if( date.getDay() > weekStart ){
			startofWeek.setDate(startofWeek.getDate() - (start_diff));
		}else{
			startofWeek.setDate(startofWeek.getDate() - ((7 + start_diff) % 7));
		}
		var result = [];
		var businessDay = this.getData('ltPropBusinessDays');

		var eventArr  = this.getData('ltPropProcessedData');
		var nxt_date = new Date(startofWeek);
		
		var count = 7;
		for ( var index = 0; index < count; index++ ) {
			var year = nxt_date.getFullYear();
			var month = ( nxt_date.getMonth() + 1 );
			var curr_date = ('0' + nxt_date.getDate()).slice(-2) + '-' + ('0' + month).slice(-2) + '-' + year;
			if(isWorkingOnly && !businessDay.includes(nxt_date.getDay())){
				nxt_date.setDate(nxt_date.getDate()+1);
				count--;
				index--;
				continue;
			}
			var is_business = businessDay.includes(nxt_date.getDay()) ? 'lyteSchedulerBusinessDay' : 'lyteSchedulerNonBusinessDay' ;
			result[index] = {'val' : $L.moment(curr_date,'DD-MM-YYYY') ,'date' : nxt_date.getDate(),'month' : nxt_date.getMonth(),'day' : nxt_date.getDay(),'buisness': is_business,'events' : eventArr[curr_date], 'currentDate' : this.isDateEqual(curr_date)};
			if(eventArr.allDay && eventArr.allDay[curr_date]){
				result[index].allDay = eventArr.allDay[curr_date];
			}
			nxt_date.setDate(nxt_date.getDate()+1);
		}
		startofWeek = new Date(result[0].val.getDObj().toString());
		var lastofweek = nxt_date.setDate(nxt_date.getDate()-1);
		this.setData('ltPropCurrMonth',startofWeek.getMonth());
		this.setData('ltPropDisplayYear', startofWeek.getFullYear());
		this.setData('startDateOfView',result[0].val.format('DD-MM-YYYY'));
		var _this = this;
		$L.fastdom.mutate( function(){
			var startofWeek = $L.moment(_this.getData('startDateOfView'),'DD-MM-YYYY').getDObj();
			_this.setDisplayDate($L("scheduler-start-date",this.$node)[0],startofWeek);
			_this.setDisplayDate($L("scheduler-end-date",this.$node)[0],result[result.length-1].val.getDObj());
		});	
		var start_Date = $L.moment(startofWeek.getTime());
		startofWeek = new Date( startofWeek.setDate(startofWeek.getDate() + 6) );
		$L('.lyteSchedulerWeekViewWrapper')[0].style.setProperty('--lyte-scheduler-table-col-count',result.length);
		this.setData('ltPropDateArray',result);
		this.setCurrentTimeLine(start_Date,$L.moment(lastofweek));
	}

    generateTimeLine() {
		var timeLine = this.$node.getData('ltPropTimeLine');
		var timelineArray = [];
		var index = 0;
		var mins = 60 - timeLine;
		timelineArray[index++] = 0; 
		while(mins){
			timelineArray[index] =  timelineArray[ index - 1 ] + timeLine;
			mins -= timeLine;		
			index++;
			
		}
		
		this.$node.setData('timelineArray',timelineArray);
	}

    MultiDayViewNav() {
		var scrollLeftNav = this.$node.querySelector('.lyteSchedulerLeftNav');
		var fixedCol = this.$node.querySelector('.lyteSchedulerFixedCol');
		if(scrollLeftNav){
			scrollLeftNav.style.left = fixedCol.offsetWidth + 'px';
			scrollLeftNav.classList.add('lyteSchedulerhideScrollNav');
			var scrollRightNav = this.$node.querySelector('.lyteSchedulerRightNav');
			var scrollDiv = this.$node.querySelector('.lyteSchedulerViewWrapper');
			var _this = this;
			scrollDiv.addEventListener('scroll',function(event){
				
				var MaxScrollWidth = this.scrollWidth - this.offsetWidth;
				if(this.scrollLeft == 0){
					scrollLeftNav.classList.add('lyteSchedulerhideScrollNav');
				}
				if(this.scrollLeft !== 0){
					scrollLeftNav.classList.remove('lyteSchedulerhideScrollNav');
				}
				if(this.scrollLeft == MaxScrollWidth){
					scrollRightNav.classList.add('lyteSchedulerhideScrollNav');
				}
				if(this.scrollLeft !== MaxScrollWidth){
					scrollRightNav.classList.remove('lyteSchedulerhideScrollNav');
				}
			});
		}
	}

    ResetTimeLine() {
		var allDay = $L('.lyteSchedulerAllDayEvent .scheduler-event-div');
		for(var index = 0; index < allDay.length;index++){
			var div = allDay[index];
			if(div){
				div.style.height = '0px';
			}
		}
		var timeline = $L('.lyteSchedulerMultiUserViewEventElem',this.$node);
		for(var index = 0 ; index < timeline.length ; index++){
			timeline[index].style.height = '0px';
		}
	}

    dayView(date) {
		var date =  date || new Date();
		this.generateDay(date);
		var _this = this;
		var right = 'right';
		if(window._lyteUiUtils.getRTL()){
			right = 'left';
		}
		if(!this.getData('ltPropMultiUserView')){
			setTimeout(function(){
				var allDayline = $L('.lyteSchedulerAllDayHighlight',_this.$node)[0];
				var alldayth = $L('.lyteSchedulerSingleUserAllDay',_this.$node)[0];
				if(allDayline && alldayth){
					allDayline.style.top =  alldayth.getBoundingClientRect().bottom - allDayline.offsetParent.getBoundingClientRect().top + 'px';
					allDayline.style[right] = window._lyteUiUtils.getScrollBarWidth() + 'px';
				}
			},10)
		}else{	
			var allDayline = $L('.lyteSchedulerAllDayHighlight',this.$node)[0];
			var fixedTh = $L('.lyteSchedulerAllDayTh',this.$node)[0];
			var fixedHeaderCs = window.getComputedStyle(fixedTh)
			if(allDayline && fixedTh && $L('.lyteSchedulerAllDayEvent',this.$node)[0]){
				allDayline.style.left = fixedTh.getBoundingClientRect().left - allDayline.offsetParent.getBoundingClientRect().left + 'px';
				allDayline.style.width = $L('.lyteSchedulerAllDayEvent',this.$node)[0].offsetWidth - parseInt(fixedHeaderCs.borderRight) + 'px';
				allDayline.style.bottom = window._lyteUiUtils.getScrollBarWidth() + 'px';
			}			
		}
	}

    generateDay(date) {
		this.setData('ltPropDisplayYear',date.getFullYear());
		var Day = this.data.days[date.getDay()];
		this.setData('currDay',Day);
		var events = this.getData('ltPropProcessedData');
		var month = date.getMonth()+1;
		var curr_date = ('0' + date.getDate()).slice(-2) + '-' + ('0' + month).slice(-2) + '-' + date.getFullYear();
		var _this = this;
		var array = {};
		array =  events[curr_date] || {};
		if(events.allDay && events.allDay[curr_date]){
			array.allDay = events.allDay[curr_date];
		}
		this.setData('ltPropDateObj',array);
		$L.fastdom.mutate( function(){
			_this.setDisplayDate($L("scheduler-start-date",$L(_this.$node).closest('lyte-scheduler')[0])[0],date);
		});

		$L('.lyteSchedulerDate').addClass(this.isDateEqual(curr_date) ? 'lyteSchedulerCurrDate' : '');
	}

    showCurrentTime(startdate, endDate) {
		var view = this.getData('ltPropSchedulerView');
		var currentDate = $L.moment(this.getData('ltPropCurrentDate'));
		if( view === 'day' ){
			return (startdate.format('DD-MM-YYYY') ===  currentDate.format('DD-MM-YYYY') );
		}else if( view === 'week' ){
			return !startdate.fromNow(currentDate).past && endDate.fromNow(currentDate).past;
		}
	}

    setCurrentTimeLine(date, endDate) {
		if(this.getData('ltPropShowCurrentTime')){
			if( this.showCurrentTime(date,endDate) && !this.getData('ltPropMultiUserView') ){
				this._navigated = false;
				if($L('.lyteschedulerCurrentTimeLine',this.$node)[0]){
					$L('.lyteschedulerCurrentTimeLine',this.$node)[0].style.display = 'flex';
					var currentLine = $L('.lyteschedulerCurrentTimeLine',this.$node)[0];
					if(this.currentTimeout){
						clearTimeout(this.currentTimeout);
					}
					if(currentLine){
						var current_Time = $L.moment(new Date());
						var min = current_Time.get('minutes');
						var Div_min = this.getData('ltPropTimeLine') * parseInt((min/this.getData('ltPropTimeLine')));
						var timeformat  = '#T' + current_Time.format('hh') + ('0' + Div_min).slice(-2) +  current_Time.format('A') ;
						var div = $L(timeformat,this.$node)[0];
						var divRect = div.getBoundingClientRect();
						currentLine.style.top =  divRect.top - currentLine.offsetParent.getBoundingClientRect().top + (divRect.height/this.getData('ltPropTimeLine')) * (min - Div_min)  + 'px';
						if($L('.lyteSchedulerCurrentTime',this.$node)[0]){
							$L('.lyteSchedulerCurrentTime',this.$node)[0].remove();
						} 
						if(this.getData('ltPropContinentalTimeFormat')){
							currentLine.insertBefore(this.createTextNode(current_Time.format('HH:mm'),'lyteSchedulerCurrentTime'),currentLine.firstChild)
						}else{
							currentLine.insertBefore(this.createTextNode(current_Time.format('hh:mm') + ' ' + window._lyteUiUtils.i18n(current_Time.format('A')),'lyteSchedulerCurrentTime'),currentLine.firstChild)
						}
						var _this = this;
						this.currentTimeout = setTimeout(function(){
							_this.removeCurrentTimeline();
							if(_this.getData('ltPropSchedulerView') == 'week'){
								var result = _this.data.ltPropDateArray;
								_this.setCurrentTimeLine($L.moment(_this.getData('startDateOfView'),"DD-MM-YYYY"),result[result.length - 1].val);
							}else{
								_this.setCurrentTimeLine($L.moment(_this.getData('startDateOfView'),"DD-MM-YYYY"));
							}
							
						},(60 - parseInt(current_Time.get('seconds')))*1000)
					}
				}
			}else{
				if(this.currentTimeout){
					clearTimeout(this.currentTimeout);
				}
				if($L('.schedulerToday',this.$node).length){
					$L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false)
				}
				this.removeCurrentTimeline();
			}
			
		}else{
			if(this.currentTimeout){
				clearTimeout(this.currentTimeout);
			}
			this.removeCurrentTimeline();
		}	
		
	}

    removeCurrentTimeline() {
		if($L('.lyteschedulerCurrentTimeLine',this.$node)[0]){
			$L('.lyteschedulerCurrentTimeLine',this.$node)[0].style.display = 'none';
			if($L('.lyteSchedulerCurrentTime',this.$node)[0]){
				$L('.lyteSchedulerCurrentTime',this.$node)[0].remove();
			} 
		}
	}

    isDateEqual(date) {
		var curr_date = this.getData( 'ltPropCurrentDate' );
		date = $L.moment( date , 'DD-MM-YYYY' ).getDObj();
		return curr_date.getDate() === date.getDate() && curr_date.getMonth() === date.getMonth() && curr_date.getFullYear() === date.getFullYear();
	}

    getTimediff(start, end) {
		var timeDiff =  $L.moment(end,this.getData('ltPropFormat')).fromNow($L.moment(start,this.getData('ltPropFormat')));
		if(timeDiff.hours){
			timeDiff = timeDiff.hours.value * 60;
		}else{
			timeDiff = timeDiff.value;
		}
		return timeDiff;
	}

    findEndDate(newStartDate, eventData) {
		var diff = this.getTimediff(eventData.start,eventData.end);
		return $L.moment( newStartDate.getDObj().toString() ).add(diff,'minutes');
	}

    callonDateChange(eventData, StartDate) {
		if(this.getMethods('onDateChange')){
			this.executeMethod('onDateChange',eventData,StartDate,this.findEndDate(StartDate,eventData),this.$node)
		}
	}

    addDragAndDrop() {
		if(!this.getData('ltPropDraggable')){
			return;
		}
		var _this = this;
		var  allDayEvent = false;
		var diff = 0,first = true;
		var total_length;
		var returnval = true;
		var isMultiView  = this.getData('ltPropMultiUserView');
		var view = this.getData('ltPropSchedulerView');
		var userdata_width = $L('.lyteSchedulerFixedCol').length && $L('.lyteSchedulerFixedCol')[0].offsetWidth;
		var _eventData;
		var width;
		var _animationFrame;
		var _eventList = this.getData('ltPropEvent');
		var timelineInterval = this.getData('ltPropTimeLine');
		var ScrollDivRect;
		var prevScrollTop;
		var currentDrop;
		var header_height;
		var Draginit = true;
		
		if((view == 'week' || view == 'day') && !isMultiView){
			var height = this.$node.querySelector("#T1200AM").getBoundingClientRect().height / 4;
			var scrollDiv =  this.$node.querySelector(".lyteSchedulerViewWrapper");
			if($L('.lyteSchedulerDate',this.$node)[0]){
				width = $L('.lyteSchedulerDate',this.$node)[0].offsetWidth;
			}
			var MaxScrollHeight =  scrollDiv.scrollHeight - (scrollDiv.offsetHeight );
 			var first_tdWidth =  scrollDiv.querySelector('td').getBoundingClientRect().width;
 		}else if(view == 'day'){
			var timelineEvent = $L('.lyteSchedulerTimeLineEvent',this.$node)[0];
			var height = 0;
			if(timelineEvent){
				var height = timelineEvent.offsetHeight;
				width = timelineEvent.offsetWidth / 6;
			}
 			var scrollDiv =  this.$node.querySelector(".lyteSchedulerViewWrapper");
 			var MaxScrollWidth =  scrollDiv.scrollWidth - ( scrollDiv.offsetWidth );
 		}	
		var verticalScrollAnimation = function(scrollDiv,element,MaxScrollWidth,event,width){
			var scrollLeft = scrollDiv.scrollLeft;
			var elementOffset = element.getBoundingClientRect();
			var scrollClientRect = scrollDiv.getBoundingClientRect();
			var current_pos = elementOffset.left;
			if(event.clientX >= (scrollClientRect.right - 50) && scrollDiv.scrollLeft <= MaxScrollWidth){
				scrollDiv.scrollLeft += 10;
			}
			if((scrollClientRect.left + userdata_width + 180 - 20 ) >= (event.clientX ) &&  (scrollLeft > 0)){
				scrollDiv.scrollLeft -= 10;
			}
			if( (event.clientX - current_pos) >= width &&  scrollLeft <= MaxScrollWidth ){
				if((element.getBoundingClientRect().left + element.offsetWidth + width) < (ScrollDivRect.left + ScrollDivRect.width )){
					var count = parseInt((event.clientX - current_pos)/width);
					offsetLeft = offsetLeft + count * width;
					_this.StartDate.add(count * 10,'minutes')
				}
				prevScrollLeft = scrollDiv.scrollLeft;
			} 
			if( (current_pos - event.clientX) > width && scrollLeft > 0 && ( elementOffset.left  ) >= ( scrollClientRect.left + userdata_width + 180 )){
				var count = parseInt((current_pos - event.clientX)/width);
				offsetLeft = offsetLeft - count * width;
				_this.StartDate.subtract(count * 10,'minutes')
				prevScrollLeft = scrollDiv.scrollLeft;
			}
			element.style.top = offsetTop + 'px';
			element.style.left = offsetLeft + 'px';
			if(_animationFrame){
				cancelAnimationFrame(_animationFrame);
			}
			_animationFrame = requestAnimationFrame(verticalScrollAnimation.bind(_this,scrollDiv,element,MaxScrollWidth,event,width))
		}
		var scrollAnimation = function(scrollDiv,element,MaxScrollHeight,event,height){
			var scrollTop = scrollDiv.scrollTop;
			var timelineBOttom = $L('.lyteSchedulerViewMainTable',this.$node)[0].getBoundingClientRect().bottom;
			var elementOffset = element.getBoundingClientRect();
			var scrollClientRect = scrollDiv.getBoundingClientRect();
			var current_pos = elementOffset.top;
			if(event.clientY >= (scrollClientRect.bottom - 50) && scrollDiv.scrollTop <= MaxScrollHeight){
				scrollDiv.scrollTop += 10;
			}
			if((scrollClientRect.top + header_height - 20 ) >= (event.clientY ) &&  (scrollTop > 0)){
				scrollDiv.scrollTop -= 10;
			}
			if( (event.clientY - current_pos) >= height &&  scrollTop <= MaxScrollHeight ){
				if((elementOffset.bottom + height) < timelineBOttom){
					var count = parseInt((event.clientY - current_pos)/height);
					offsetTop = offsetTop + count * height;
					_this.StartDate.add(count * (timelineInterval/4),'minutes')
				}
				prevScrollTop = scrollDiv.scrollTop;
			} 
			if( (current_pos - event.clientY) >= height && scrollTop > 0 && ( elementOffset.top - height ) >= ( scrollClientRect.top + header_height )){
				var count = parseInt((current_pos - event.clientY)/height);
				offsetTop = offsetTop - count * height;
				_this.StartDate.subtract(count * (timelineInterval/4),'minutes')
				prevScrollTop = scrollDiv.scrollTop;
			}
			element.style.left = offsetLeft + 'px';
        	element.style.top = offsetTop  + 'px';
			if(_animationFrame){
				cancelAnimationFrame(_animationFrame);
			}
			_animationFrame = requestAnimationFrame(scrollAnimation.bind(this,scrollDiv,element,MaxScrollHeight,event,height))
		}
		var event_tag = this.$node.querySelectorAll('lyte-scheduler-event-tag');
		event_tag.length && $L(event_tag).draggable({
			restrict : '.lyteSchedulerNoDrag',
			threshold : _this.getData('ltPropDragThershold'),
			cancel : ".lyteSchedulerDragCancel, .lyteSchedulerGrabResize",
			onStart : function(element, event){
				if(!isMultiView){
					header_height = _this.$node.querySelector('.lyteSchedulerViewMainTable thead') ;
					header_height = header_height ? header_height.offsetHeight : _this.$node.querySelector('.lyteSchedulerDayViewAllDayRow').offsetHeight;
				}
				_eventData = $u.findWhere(_eventList,{id: element.dataset.id});
				startTop = event.clientY;
            	startLeft = event.clientX;
            	offsetTop = element.offsetTop - parseInt(window.getComputedStyle(element).marginTop) - 0.5;
            	offsetLeft = element.offsetLeft;
            	element.style.zIndex = 10000;
				allDayEvent = element.classList.contains('all-day-event') ? true  : false;
				if(scrollDiv){
					ScrollDivRect = scrollDiv.getBoundingClientRect();
				}
			},
			onDragStart : function ( element, event) {
				var id =  element.dataset.id;
				var scheduler = $L(element).closest('lyte-scheduler')[0];
				var disable_elem = scheduler.querySelectorAll('#u'+ id );
            	$L(disable_elem).not(element).addClass('hideSchedulerEvent');
            	if(!isMultiView){
					if(!element.parent){
						element.parent = $L(element).closest('td')[0];
						element.style.left = '5px';
						offsetLeft = 5;
					}else{
						element.style.left =  element.offsetParent.getBoundingClientRect().left - element.getBoundingClientRect().left   +"px";
						offsetLeft = element.offsetLeft;
					}
				}else{
					if(!element.parent){
						element.parent = $L(element).closest('tr')[0];
					}
				}
            	if(!isMultiView && !_eventData.allDayEvent ){
            		element.style.right =  '5px';
            	}   
				if(!element.classList.contains('lyteSchedulerAllDayEventTag')){
					element.style.width = 'calc(100% - 10px)';
				}
				_this.StartDate = $L.moment(_eventData.start,_this.getData('ltPropFormat'));  
				_this.end_date = $L.moment(_eventData.end,_this.getData('ltPropFormat'));
			},
			onDrag : function( element ,dragElem,event ){
				if(!element.classList.contains('lyteSchedulerAllDayEventTag') && Draginit){
					element.style.width = 'calc(100% - 10px)';
					Draginit = false;
				}
				var x = event.clientX,
            		y = event.clientY,
            		elem_offset = element.getBoundingClientRect();
				var _elemOrginDiv = element.parent;
				_elemOrginDiv = _elemOrginDiv ? _elemOrginDiv : element.offsetParent;
				var _elemOffset = _elemOrginDiv.getBoundingClientRect();
            	if((view == 'week' ||  view == 'day') && !element.classList.contains('lyteSchedulerAllDayEventTag') && !isMultiView){
            		if(  view == 'week' && (_elemOffset.left + _elemOffset.width ) <= x  &&(_elemOffset.left + _elemOffset.width ) <= (ScrollDivRect.left + ScrollDivRect.width) && element.parent.nextElementSibling){
            			offsetLeft = offsetLeft + width ;
            			if(!allDayEvent){
							_this.StartDate.add(1,'date')
            				element.parent = element.parent.nextElementSibling ? element.parent.nextElementSibling : element.parent;
							_this.callonDateChange(_eventData,_this.StartDate)
            			}
            		}
            		if( view == 'week' && x < (_elemOffset.left - 5) && (_elemOffset.left - 5 ) > (ScrollDivRect.left + first_tdWidth) && element.parent.previousElementSibling  ){
            			offsetLeft = offsetLeft - width ;
            			if(!allDayEvent){
							_this.StartDate.subtract(1,'date')
							_this.callonDateChange(_eventData,_this.StartDate)
            				element.parent = element.parent.previousElementSibling ? element.parent.previousElementSibling : element.parent;
            			}
            		}
           			if(!allDayEvent){
	            		if( (y - startTop) > height && (element.getBoundingClientRect().bottom + height) < (ScrollDivRect.top + ScrollDivRect.height )){
							var count = parseInt((y - startTop)/height);
							_this.StartDate.add((count * (timelineInterval/4)),'minutes')
	            			offsetTop = offsetTop + count * height;
	            			startTop = y - (y - startTop)%height;
							_this.callonDateChange(_eventData,_this.StartDate)
	            		}
	            		if( (startTop - y) > height && (element.getBoundingClientRect().top - height ) > ( ScrollDivRect.top + header_height )){
							var count =  parseInt((startTop - y)/height);
							_this.StartDate.subtract((count * (timelineInterval/4)),'minutes')
	            			offsetTop = offsetTop - count * height;
	            			startTop = y - (y - startTop)%height;
							_this.callonDateChange(_eventData,_this.StartDate)
	            		}
	            	}
					if(_animationFrame){
						cancelAnimationFrame(_animationFrame);
					}
					mouseposition = event.clientY - element.getBoundingClientRect().top;
					prevScrollTop = scrollDiv.scrollTop;
					element.style.left = offsetLeft + 'px';
        			element.style.top = offsetTop  + 'px';
					_animationFrame = window.requestAnimationFrame(scrollAnimation.bind(this,scrollDiv,element,MaxScrollHeight,event,height))
					if(_this.getMethods('onDragEvent')){
						_this.executeMethod('onDragEvent',_eventData,_this.StartDate,_this.findEndDate(_this.StartDate,_eventData),_this.$node);
					}
					return false;
				}else if(view === 'day' && !element.classList.contains('lyteSchedulerAllDayEventTag')){
					_this.ispreventDrag = true;
					var scrollLeft = scrollDiv.scrollLeft;
					if( (element.parent.getBoundingClientRect().top + element.parent.getBoundingClientRect().height) <= y  && ( element.getBoundingClientRect().top + element.getBoundingClientRect().height + height) < (ScrollDivRect.top + ScrollDivRect.height)){
            			offsetTop = offsetTop + height ;
            			if(element.parent.nextElementSibling){
            				element.parent = element.parent.nextElementSibling ? element.parent.nextElementSibling : element.parent;
            			}
            		}
            		if( y < (elem_offset.top) && (elem_offset.top - height) > (ScrollDivRect.top + element.offsetHeight  ) ){
            			offsetTop = offsetTop - height;
            			if(element.parent.previousElementSibling){
            				element.parent = element.parent.previousElementSibling ? element.parent.previousElementSibling : element.parent;
            			}
            		}
					if( (x - startLeft) > width && (element.getBoundingClientRect().left + element.offsetWidth  ) < (ScrollDivRect.left + ScrollDivRect.width ) ){
						var count =  parseInt((x - startLeft)/width);
	            		offsetLeft = offsetLeft + (width * count);
						_this.StartDate.add(10 * count,'minutes')
	            		startLeft = x - (x - startLeft) % width;
						_this.callonDateChange(_eventData,_this.StartDate)
	            	}
	            	if( (startLeft - x) > width   && ( element.getBoundingClientRect().left ) > (ScrollDivRect.left + userdata_width + 180) ){
						var count =  parseInt((startLeft - x)/width);
						offsetLeft = offsetLeft - (count * width);
						_this.StartDate.subtract(10 * count,'minutes')
	            		startLeft = x + ( startLeft - x ) % width;
						_this.callonDateChange(_eventData,_this.StartDate)
	            	}
					if(_animationFrame){
						cancelAnimationFrame(_animationFrame);
					}
					if(_this.getMethods('onDragEvent')){
						_this.executeMethod('onDragEvent',_eventData,_this.StartDate,_this.findEndDate(_this.StartDate,_eventData),_this.$node);
					}
					_animationFrame = window.requestAnimationFrame(verticalScrollAnimation.bind(_this,scrollDiv,element,MaxScrollWidth,event,width))
					return false;
        		}
			},
			onStop : function(){
				delete _this.ispreventDrag;
				delete _this.startDate;
				cancelAnimationFrame(_animationFrame);
				Draginit = false;
				return returnval;
			}
		});
	 	var allDayEvent = this.$node.querySelectorAll('.lyteSchedulerAllDayEvent');

		allDayEvent.length && $L(allDayEvent).droppable({
			onLeave : function () {
				var hoverDiv = _this.$node.querySelectorAll('.lyteSchedulerhoverDate');
				$L(hoverDiv).removeClass('lyteSchedulerhoverDate');
			},
			onDrag : function( draggableElem, droppableElem ){
				var startDate =  draggableElem.parentElement.dataset.date;
				var id =  draggableElem.dataset.id;
				var format = 'DD-MM-YYYY';
				var view = _this.getData('ltPropSchedulerView');
				if(_this.ispreventDrag ){
					return;
				}
				if(view == 'week' ){
					var curr_row = droppableElem.parentElement;
					var date_div = curr_row.querySelectorAll('.lyteSchedulerAllDayEvent');
				}else{
					var date_div = _this.$node.querySelectorAll('.lyteSchedulerAllDayEvent');
				}	
				var events = _this.data.ltPropManipulatedEvent;
				var event =  $u.findWhere(events,{'id': id});
				var org_start = $L.moment(event.start,_this.getData('ltPropFormat')).format('DD-MM-YYYY');
				total_length = _this.getEventInterval(event.start,event.end,_this.getData('ltPropFormat'));	
				if(currentDrop !== droppableElem){
					currentDrop = droppableElem;
					var start_date = $L.moment(droppableElem.children[0].dataset.date,format).subtract(Math,min((diff - 1),0),'date',true);
					start_date.set('time',_this.StartDate.get('time'));
					var end_date = $L.moment(start_date.format('DD-MM-YYYY'),format).add(Math.min((total_length - 1),0),'milliseconds');
					end_date.set('time',_this.end_date.get('time'));
					var event_data = $u.findWhere(events,{id: draggableElem.dataset.id});
					if(_this.getMethods('onDateChange')){
						_this.executeMethod('onDateChange',event_data,start_date,end_date,_this.$node)
					}
					_this.StartDate = start_date;
					_this.end_date = end_date;
				} 
				if(_this.getMethods('onDragEvent')){
					_this.executeMethod('onDragEvent',event,_this.StartDate,_this.end_date,_this.$node)
				}
				if(event ){
					var drop_pos = _this.getEventInterval(date_div[0].children[0].dataset.date,droppableElem.children[0].dataset.date,'DD-MM-YYYY');
					if((org_start !== startDate) && first){
						diff =  _this.getEventInterval(event.start , $L.moment(startDate,'DD-MM-YYYY') ,_this.getData('ltPropFormat'));
						first = false;
					}
					total_length = _this.getEventInterval(event.start,event.end,_this.getData('ltPropFormat'));	
					for(var index = diff ; index >= 0; index--){
						if(date_div[ drop_pos - index ]){
							date_div[ drop_pos - index ].classList.add('lyteSchedulerhoverDate');
						}
					}

					for(var index = 0; index < ( ( total_length ) - diff)  ; index++){
						if(date_div[ drop_pos + index ]){
							date_div[ drop_pos + index ].classList.add('lyteSchedulerhoverDate');
						}
					}
				}
				return true;
			},
			onDrop : function( draggedElem, droppableElem ){
				
				var date_div = _this.$node.querySelectorAll('.lyteSchedulerDropPlaceholder');
				$L(date_div).removeClass('lyteSchedulerDropPlaceholder');
				var user = droppableElem.parentElement.dataset.userid;
				var format = 'DD-MM-YYYY';
				var start_date = $L.moment(droppableElem.children[0].dataset.date,format).subtract((( diff ) >= 0 ? diff : 0),'date',true);

				var end_date = $L.moment(start_date.format('DD-MM-YYYY'),format).modify(( total_length - 1 ),'date',true);
				var events  = _this.getData('ltPropManipulatedEvent');
				var event_data = $u.findWhere(events,{id: draggedElem.dataset.id});
				var orginalStart =  $L.moment(event_data.start,_this.data.ltPropFormat);
				var originEnd =  $L.moment(event_data.end,_this.data.ltPropFormat);
				start_date.set('hours',orginalStart.get('hours'));
				start_date.set('minutes',orginalStart.get('minutes'));
				end_date.set('hours',originEnd.get('hours'));
				end_date.set('minutes',originEnd.get('minutes'));
				if(_this.getMethods('onDropEvent')){
					window.returnVal = _this.executeMethod('onDropEvent',event_data,start_date,end_date,user);
				}
				return window.returnVal;

			},
			tolerance :"touch"
		});
		
		var timeLineEvent = this.$node.querySelectorAll('.lyteSchedulerTimeLineEvent');
		if( timeLineEvent.length ){

 			$L(timeLineEvent).droppable({
				onDrop : function(draggedElem,droppableElem){
					if(draggedElem.classList.contains('all-day-event')){
						return;	
					}
					var start_time ;
					var end_time;
					if(!isMultiView){
						var events  = _this.getData('ltPropManipulatedEvent');
						var event_data = $u.findWhere(events,{id: draggedElem.dataset.id});
						var start_date = draggedElem.parent.children[0].dataset.date;	
						start_time = droppableElem.dataset.time;
						if(start_time){
							start_time = _this.StartDate;
							var timeDiff = _this.getTimediff(event_data.start,event_data.end);
							if(( timeDiff ) ){
								var date = $L.moment(start_time.format(_this.getData('ltPropFormat')),_this.getData('ltPropFormat')).add( timeDiff,'minutes',true);
								end_time = date;
							}
						}
					}else{
						var user = draggedElem.parent.dataset.userid;
						var start_date = _this.getData('startDateOfView');
						start_time = droppableElem.children[0].dataset.time;
						var time_division = 60 / _this.getData('ltPropTimeLine');
						if(start_time){

							var interval = Math.round( draggedElem.offsetWidth / ( droppableElem.offsetWidth / time_division ) );
							var date = start_date + ' ' + start_time;
							date = $L.moment(date,_this.getData('ltPropFormat'));
							start_time = $L.moment(start_date + ' ' + start_time,_this.getData('ltPropFormat'));
							
							var start_time_min = parseInt((draggedElem.getBoundingClientRect().left - droppableElem.getBoundingClientRect().left) / (droppableElem.offsetWidth / time_division));
							start_time_min *=  _this.getData('ltPropTimeLine');
							if(start_time_min){
								date.add(start_time_min,'minutes',true);
								start_time = $L.moment(date.format(_this.getData('ltPropFormat')),_this.getData('ltPropFormat'));
							}
							if(interval){

								end_time = date.add( ( parseInt(interval)  ) * _this.getData('ltPropTimeLine') ,'minutes',true);
							}
							var events  = _this.getData('ltPropManipulatedEvent');
							var event_data = window.$u.findWhere(events,{id: draggedElem.dataset.id});
						}


					}
					_this.StartDate = null;
					if(_this.getMethods('onDropEvent')){

						_this.executeMethod('onDropEvent',event_data,start_time,end_time,user);
					}
					return false;
				},
				tolerance :"touch",
				restrict : ".lyteSchedulerEventTimelinetag"
			});
 		}
	}

    getEventInterval(start, end, format, returntime) {
		var start_date = start;
		if(typeof start_date == 'string'){
			start_date = $L.moment(start_date, format)._dateObj;
		}
		var end_date = end;
		if(typeof end_date == 'string'){
			end_date = $L.moment(end ,format)._dateObj;
		}
		var difference = end_date.getTime() - start_date.getTime();
		if(returntime){
			return difference;
		}
    	var TotalDays = Math.ceil(difference / (1000 * 3600 * 24));
		return TotalDays;
	}

    hide(position_elem, event_div) {
		if((position_elem.offsetHeight + event_div.getBoundingClientRect().height) >= position_elem.parentElement.offsetHeight){
			event_div.classList.add('schedulerhide');
			event_div.style.top = 0;
			event_div.style.left = 0; 
			return true;
		}
	}

    hideUsers(hideuserarray) {
		var userList =  this.getData('ltPropSchedulerUser');
		var table_tr = this.$node('.UserCol');
		hideuserarray.forEach(function(index){
			table_tr[index].classList.add('LyteSchdeulerHideUser');
		});
	}

    setDisplayDate(dateDiv, date, showdate, popover) {
		if(!dateDiv){
			return;
		}
		if( dateDiv.children.length){
		
			while(dateDiv.children.length){
				dateDiv.children[0].remove();
			}		
		}
		var tempMoment = $L.moment(date.toString());
		if(this.getData('ltPropSchedulerView') !== 'month' || showdate){
			dateDiv.appendChild(this.createTextNode(tempMoment.format('DD'), 'lyteSchNavigatorDateLabel'));
		}
		if(this.getData('ltPropSchedulerView') == 'month'){
			dateDiv.appendChild(this.createTextNode(window._lyteUiUtils.i18n(tempMoment.format('MMMM')), 'lyteSchNavigatorMonthLabel'));
		}else{
			dateDiv.appendChild(this.createTextNode(window._lyteUiUtils.i18n(tempMoment.format('MMM')), 'lyteSchNavigatorMonthLabel'));
		}
		dateDiv.appendChild(this.createTextNode(date.getFullYear(), 'lyteSchNavigatorYearLabel'));
		if(this.getData('ltPropSchedulerView') == 'day' && !popover){
			dateDiv.appendChild(this.createTextNode(window._lyteUiUtils.i18n(tempMoment.format('dddd').toLowerCase()), 'lyteSchNavigatorDayLabel'));
		}
	}

    createTextNode(text, className) {
		var span = document.createElement('span');
		if(className) {
			span.classList.add(className);
		}
		var textnode =  document.createTextNode(text);
		span.appendChild(textnode);
		return span; 
	}

    clearUser() {
		this.setData('selectedUserArr',[]);
		var search_popover = $L('#usersreachPopover',this.$node)[0];
		var userlist = $L('.lyteSchedulerUserSearch', search_popover.actualModalDiv);
		for(index = 0; index < userlist.length; index++){
			var checkbox = $L('lyte-checkbox',userlist[index])[0];
			checkbox.setData('ltPropChecked',false);
		}
		var  schedulerUser = $L('[data-userid]');
		schedulerUser.removeClass('lyteSchedulerUserHide');
		schedulerUser.removeClass('lyteSchedulerUserSelected');
	}

    findDateonMenuClick(originElem, view) {
		var isMultiUser = this.getData('ltPropMultiUserView')
		var date;
		if(view == 'day'){
			date = this.getData('startDateOfView');
		}else if(view == 'week' && !isMultiUser){
			date = parseInt(originElem.dataset.index);
			date = this.data.ltPropDateArray[date].val.format('DD-MM-YYYY');
		}else{
			date = originElem.querySelector('.scheduler-event-div').dataset.date;
		}
		return date;
	}

    findTimeonMenuClick(event, isMultiView, view) {
		var time;
		if(!isMultiView && view !== 'month'){
			if(!event.target.classList.contains('lyteSchedulerAllDayEvent')){
				time = event.target.closest('.lyteSchedulerTimeLineEvent').dataset.time;
			}
		}else if(isMultiView && view == 'day'){
			if(event.target.classList.contains('lyteSchedulerTimeLineEvent')){
				time = event.target.querySelector('.lyteSchedulerMultiUserViewEventElem').dataset.time;
			}
		}
		return time;
	}

    static actions(arg1) {
        return Object.assign(super.actions({
            onSchedulerEventHover : function(event,eventTag,eventObj,user){
                if(!eventObj){
                    eventObj = eventTag.getData('ltPropEvent');
                }
                var UserData;
                if(user){
                    UserData = $u.findWhere(this.getData('ltPropSchedulerUser'),{id: user});
                }
                if(this.getMethods('onEventHover')){
                    this.executeMethod('onEventHover',event,eventObj,eventTag,UserData);
                }
            },
            onSchedulerEventClick : function(event,eventTag,user){
                if(user){
                    var eventObj = user;
                }else{
                    var eventObj = eventTag.getData('ltPropEvent');
                }
                if(this.getMethods('onEventClick')){
                    this.executeMethod('onEventClick',event,eventObj,eventTag);
                }
            },
            onSchedulerEventMouseDown : function(event,eventTag){
                if( !this.getData('isEventAdded')|| !event.target.closest('.lyteSchedulerGrabResize') || eventTag.classList.contains('lyteSchedulerParticipantEvent')  || eventTag.classList.contains('lyteSchedulerNoDrag') || ( this.getData('ltPropSchedulerView') === 'month' && this.getData('ltPropMultiUserView') === true ) ){
                    return;
                }
                eventTag.classList.add('lyteSchedulerOnResize');
                eventTag.style.zIndex = 10000;
                var view = this.getData('ltPropSchedulerView');
                var isMultiUser = this.getData('ltPropMultiUserView');
                this.setData('dragDir','');
                var eventObj = eventTag.getData('ltPropEvent');
                if( $L && $L.schedulerEventResize ){
                    if( (view === 'day' && !isMultiUser) ||
                     (view === 'week' && !isMultiUser && !event.allDayEvent)  ) {
                        if( event.target.tagName === 'SPAN' && event.target.classList.contains('lyteSchedulerGrabResize')){
                            event.target.classList.add('lyteSchedulerGrabResizeShow');
                            $L.schedulerEventResize( event, this, eventObj, eventTag );
                        }
                    }
                    if( (view === 'day' && isMultiUser )  ||
                     ( view === 'week' && !isMultiUser && event.allDayEvent ) || 
                     ( view === 'week' && isMultiUser ) || 
                     ( view === 'month' && !isMultiUser ) ) {

                        if( event.target.tagName === 'SPAN' && event.target.classList.contains('lyteSchedulerHorizontalResizeIcon')){
                            event.target.classList.add('lyteSchedulerLeftOrRightResizeIcon');
                            $L.schedulerEventResize( event, this, eventObj, eventTag );
                        }
                    }
                    else{
                        $L.schedulerEventResize( event, this, eventObj, eventTag );
                    }
                }
            },
            clearSelection : function(){
                this.clearUser();
            },
            cancelUserSelect : function(){
                var popover = $L('#usersreachPopover',this.$node)[0];
                popover.ltProp('show',false);
                this.setData('checkboxArr',[]);
            },
            SelectUserSelect : function(){
                var popover = $L('#usersreachPopover',this.$node);
                popover[0].ltProp('show',false);
                var selectedUser = $L('[data-userid]:not(.lyteSchedulerUserSelected)');
                selectedUser.addClass('lyteSchedulerUserHide');	
                var cloneArr = $u.clone(this.getData('checkboxArr'));
                this.setData('selectedUserArr',cloneArr);
            },
            hideandShowCalender : function( event  ){
                var originElem;
                originElem = event.target.closest('lyte-scheduler-icon').getAttribute('id');
                var startDateOfView = this.getData('startDateOfView');
                var view = this.getData('ltPropSchedulerView');
                var isMonth = view == 'month';
                if(isMonth){
                    startDateOfView = $L('.lyteSchedulerCurrentMonthDate .lyteSchedulerMonthEventsWrap',this.$node)[0].dataset.date;
                }
                if(!this.getData('ltPropHideCalendarNav')){
                    $L('#'+originElem).addClass('lyteSchNavigatorCalOpened');
                    var popover = $L('#calendarPopover',this.$node)[0];
                    this.setData('ltPropShowCalendar',true);
                    popover.setData('ltPropOriginElem','#'+originElem);
                    if(window._lyteUiUtils.getRTL()){
                        popover.setData('ltPropPlacement', 'bottomRight');
                    }
                    var calendar = $L('lyte-calendar',popover.component.actualModalDiv)[0];
                    calendar.setData('ltPropCurrentDate',startDateOfView);
                    calendar.setData('viewType','dateView');
                    if(view == 'week'){
                        var dateArray = this.data.ltPropDateArray;
                        calendar.setData('ltPropCurrentWeek',[startDateOfView,dateArray[dateArray.length - 1].val.format('DD-MM-YYYY')])
                    }
                    if(isMonth){
                        var drilldownSpan = $L('.lyteCalsCalMon .lyteDrillCalHeaderButton',popover.component.actualModalDiv)[0];
                        if(calendar && calendar.getAttribute('view-type') == 'dateView' && drilldownSpan){
                            drilldownSpan.click();
                        }
                    }
                    $L('#calendarPopover')[0].setData('ltPropShow',true);
                }
                if(this.getMethods('onCalendarIconClick')){
                    this.executeMethod('onCalendarIconClick',originElem,event.target,startDateOfView)
                }
            },
            scrollnav : function(event){
                var scroll_width = this.$node.querySelector('.lyteSchedulerTimeLineEvent').offsetWidth;
                var scrollDiv = this.$node.querySelector('.scrollContainer');
                if(event.target.classList.contains('lyteSchedulerLeftNav')){
                    scrollDiv.scrollLeft -= scroll_width; 
                }else{
                    scrollDiv.scrollLeft += scroll_width;
                }
            },
            today : function(event){
                this.$node.today();
                this.findActiveKeys($L.moment(new Date()));
            },
            hiddenUserEvent : function(event,UserEvent,user){
                this.$node.openHiddenUserPopover(UserEvent, user, event.target, {'placement': 'right' , 'type' : 'box'})
            },
            hiddenEvent : function(event,hiddenObj,date,user,time,view,isAllDay){
                var group = event.target.dataset.group;
                var displayEvent = this.getData('ltPropProcessedData');
                var isMultiUser = this.getData('ltPropMultiUserView');
                this.setData('ltPropUserHiddenEvent',false);
                this.setData('hiddenUser',undefined);
                if(isAllDay){
                    displayEvent = displayEvent.allDay;
                }
                if(time){
                    date = typeof date == 'string' ? $L.moment(date,'DD-MM-YYYY') : date; 
                    if(view == 'day' && isMultiUser && displayEvent[date.format('DD-MM-YYYY')] && displayEvent[date.format('DD-MM-YYYY')][user.id]){
                        var hiddenEvent = hiddenObj[time];
                        var eventData = displayEvent[date.format('DD-MM-YYYY')][user.id][time];
                    }else{
                        var hiddenEvent = hiddenObj[time];
                        var eventData = displayEvent[time];
                    }
                }else{
                    date = typeof date == 'string' ? $L.moment(date,'DD-MM-YYYY') : date; 
                    var hiddenEvent = hiddenObj[date.format('DD-MM-YYYY')];
                    var eventData = displayEvent[date.format('DD-MM-YYYY')];
                    if(user){
                        hiddenEvent = hiddenEvent[user];
                        eventData = eventData[user];
                    }
                }
                if(group){
                    hiddenEvent = hiddenEvent[group];
                    eventData = eventData[group];
                }
                if(view == 'day' ){
                    var tempArr = [];
                    if(eventData){
                        if(isMultiUser && !isAllDay){
                            eventData.forEach(function(eventArray){
                                eventArray.forEach(function(event){
                                    tempArr.push(event)
                                })
                            })
                        }else{
                            eventData.forEach(function(event){
                                tempArr.push(event)
                            })
                        }
                        
                        this.setData('hiddenEvent',tempArr.concat(hiddenEvent));
                    }
                }else{
                    if(Array.isArray(eventData)){
                        var tempArray = [];
                        eventData.forEach(function(event,index){
                            if(event.dummy_id){
                                event =  $u.clone($u.findWhere(this.getData('ltPropEvent'),{id: event.dummy_id}));
                                tempArray.push(event);
                                eventData[index] = event;
                            }else if(event){
                                tempArray.push(event);
                            }
                        }.bind(this))
                        this.setData('hiddenEvent',tempArray.concat(hiddenEvent));
                    }else{
                        var userData = this.getData('ltPropSchedulerUser');
                        var objectToArray = {...eventData,...hiddenEvent};
                        var ArrayFormat = [];
                        for(let user in objectToArray){
                            var obj = $u.clone($u.findWhere(userData,{id: user}));
                            if(!obj){
                                continue;
                            }
                            if(obj.events){
                                obj.events.push(objectToArray[user])
                            }else{
                                obj.events = [];
                                obj.events.push(objectToArray[user])
                            }
                            ArrayFormat.push(obj);
                        }
                        this.setData('hiddenEvent',ArrayFormat);
                    }
                    
                }
                console.log(this.getData('hiddenEvent'))
                $L('#schedulerPopover',this.$node)[0] && $L('#schedulerPopover',this.$node)[0].setAttribute('id',0);
                event.target.setAttribute('id','schedulerPopover');
                var popover = this.$node.querySelector('#lyteSchedulerHiddenEvent');
                popover.setData('ltPropPlacement','bottom');	
                popover.setData('ltPropType','callout');
                popover.setData('ltPropShow',true);
                var poptitle = $L('.lyteSchedulerMoreEventsPopTitleMonthVal',popover.component.actualModalDiv);
                date =  date.getDObj();
                var popDay = $L('.lyteSchedulerMoreEventsPopTitleDayVal',popover.component.actualModalDiv);
                this.setDisplayDate(poptitle[0],date,true,true);
                popDay[0].innerHTML = '';
                popDay[0].appendChild(document.createTextNode(window._lyteUiUtils.i18n(this.data.days[date.getDay()])));
                if(this.getMethods('openHiddenEvent')){
                    this.executeMethod('openHiddenEvent',event.target);
                }
            },
            onDateClick : function(event,item){
                if(this.getMethods('onDateClick')){
                    this.executeMethod('onDateClick',item);
                }
            },
        }), arg1);
    }

    static methods(arg1) {
        return Object.assign(super.methods({
            onSchedulerNavToday : function(event,startDate,endDate,component,istoday){
                if(istoday){
                    this.$node.today();
                    $L('#calendarPopover',this.$node)[0].setData('ltPropShow',false);
                }
            },
            eventRender : function(event_div){
                if(this.getMethods('afterEventRender')){
                    this.executeMethod('afterEventRender',event_div,this.$node);
                }
            },
            onUserSearchClose : function(event){
                var search = $L('.lyteSchedulerFirstColHeader',this.$node)[0];
                if(event.target === search || search.contains(event.target)){
                    return false;
                }else{
                    return true;
                }

            },
            boxUnchecked : function(input, component, event, useraction){
                var arr = this.getData('checkboxArr');
                var val = component.getData('ltPropVal');
                var index =  arr.indexOf(val);
                if(index > -1){
                    arr.splice( index, 1 );
                }
                this.setData('checkboxArr', arr);
            },
            boxChecked : function(input, component, event, useraction){
                var arr = this.getData('checkboxArr');
                var val = component.getData('ltPropVal');
                arr.push(val);
                this.setData('checkboxArr', arr);
            },
            weekSelected : function(event,week){
                if(this.getData('ltPropSchedulerView') == 'week'){
                    $L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false)
                    var popover = $L('#calendarPopover')[0];
                    this._navigated =  true;
                    var date = $L.moment(week,'DD-MM-YYYY');
                    this.callCurrentView(date.getDObj(),false,this.getData('ltPropWeekStart'));
                    this.setData('selectionType','');
                    this.setData('selectionType','week');
                    popover.setData('ltPropShow',false);
                    if(this.getMethods('onCalendarNav')){
                        this.executeMethod('onCalendarNav',date);
                    }
                }
            },
            monthSelected : function(event,month,cal){
                if(this.getData('ltPropSchedulerView') == 'month'){
                    $L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false)
                    var popover = $L('#calendarPopover')[0];
                    popover.setData('ltPropShow',false);
                    this._navigated =  true;
                    var date = $L.moment('01-'+month+'-'+cal.getData('viewDate').getFullYear(),'DD-MM-YYYY');
                    this.callCurrentView(date.getDObj(),false,this.getData('ltPropWeekStart'));
                    this.setData('selectionType','');
                    this.setData('selectionType','month');
                    // popover.setData('ltPropShow',false);
                    if(this.getMethods('onCalendarNav')){
                        this.executeMethod('onCalendarNav',date);
                    }
                }
            },
            dateselect : function(event,date,component){
                
                $L('.schedulerToday',this.$node)[0].setData('ltPropDisabled',false);
                var originElm = $L('.lyteSchNavigatorCalOpened',this.$node);
                var date_obj =  $L.moment(date,'DD-MM-YYYY').getDObj();
                var popover = $L('#calendarPopover')[0];
                popover.setData('ltPropShow',false);
                this._navigated =  true;
                var start_day;
                
                if(this.getData('ltPropSchedulerView') == 'week'){
                    if(originElm.hasClass('lyteSchedulerStartDate')){
                        start_day = date_obj.getDay();
                    }else{
                        start_day = date_obj.getDay() + 1;
                    }
                    var dayLabel = this.getData('ltPropLabel');
                    var split = dayLabel.slice(start_day);
                    dayLabel = split.concat( dayLabel.slice(0,start_day));
                    this.setData('label',dayLabel);
                    this.callCurrentView($L.moment(date,'DD-MM-YYYY').getDObj(),false,start_day);
                }else{
                    this.ResetTimeLine();
                    this.callCurrentView($L.moment(date,'DD-MM-YYYY').getDObj());
                }
                if(this.getMethods('onCalendarNav')){
                    this.executeMethod('onCalendarNav',$L.moment(date,'DD-MM-YYYY'));
                }
            },
            menuonclick : function( value , event , element , menuOriginElem , clickedItemAndSubmenu_detail){
                var view = this.getData('ltPropSchedulerView');
                var isMultiView = this.getData('ltPropMultiUserView');
                var time;
                if(!isMultiView && view !== 'month' && !menuOriginElem.classList.contains('lyteSchedulerAllDayEvent')){
                    var table = $L('.lyteSchedulerViewMainTable tbody',this.$node)[0];
                    var tr = $L('.lyteSchedulerViewMainTable tbody .lyteSchedulerTimeLineEvent',this.$node);
                    var AllDay_tr = 0;
                    if(view == 'day'){
                        AllDay_tr = $L('.lyteSchedulerAllDayEvent',this.$node)[0].offsetHeight;
                    }
                    
                    var current_tr = (event.target.closest('lyte-menu-body').getBoundingClientRect().top - (table.getBoundingClientRect().top + AllDay_tr)) / tr[0].offsetHeight;
                    if(current_tr >= 0){
                        time = tr[parseInt(current_tr)].dataset.time;
                    }else{
                        time = undefined;
                    }
                }else if(isMultiView && view == 'day'){
                    if(menuOriginElem.classList.contains('lyteSchedulerTimeLineEvent')){
                        time = menuOriginElem.querySelector('.lyteSchedulerMultiUserViewEventElem').dataset.time;
                    }
                }
                var date = this.findDateonMenuClick(menuOriginElem,view);
                if(this.getMethods('onMenuClick')){
                    this.executeMethod('onMenuClick',date,value , event , element , menuOriginElem , clickedItemAndSubmenu_detail,time);
                }
            },
            menuonbeforeopen : function( menu , event , originElem ){
                var event_tag = event.target;
                var flag = true;
                var view = this.getData('ltPropSchedulerView');
                while(event_tag && event_tag.tagName !== 'TR'){
                    if(event_tag.tagName === "LYTE-SCHEDULER-EVENT-TAG"){
                        flag = false;
                        break;
                    }
                    event_tag = event_tag.parentElement;
                    if(!event_tag){
                        return false;
                    }
                }
                if(event.which == 1 && !event.target.classList.contains("lyteSchedulerEventMoreBtn") && !event.target.classList.contains("lyteSchedulerDayNumber") && flag){
                    var date = this.findDateonMenuClick(originElem,view);
                    if(this.getData('ltPropMultiUserView')  && view !== 'month' ){
                        var userid = event.target.closest('tr').dataset.userid;
                        var userData = $u.findWhere(this.getData('ltPropSchedulerUser'),{'id':userid});
                        userData = $u.clone(userData);
                    }
                    if(this.getMethods('onBeforeMenuOpen')){
                        this.executeMethod('onBeforeMenuOpen',date,event,this.findTimeonMenuClick(event,this.getData('ltPropMultiUserView'),view),userData);
                    }
                }else{
                    return false;
                }
            },
            menuonopen : function( menu , event , originElem ){
            
                var menu_width =  menu.component.childComp.getBoundingClientRect().width;
                var menu_height =  menu.component.childComp.getBoundingClientRect().height;
                var Leftboundaries =  window.innerWidth;
                var Topboundaries = window.innerHeight;
                if(Topboundaries < event.clientY + menu_height){
                    menu.component.childComp.style.top = Topboundaries + document.documentElement.scrollTop - menu_height  + 'px';
                }else{
                    menu.component.childComp.style.top = event.clientY + document.documentElement.scrollTop + 'px';
                }			
                if(window._lyteUiUtils.getRTL()){
                    menu.component.childComp.style.right = 'unset' ;	
                }
                if(Leftboundaries < event.clientX + menu_width ){
                    menu.component.childComp.style.left = Leftboundaries - menu_width + 'px';
                }else{
                    menu.component.childComp.style.left = event.clientX + 'px';
                }
                var view = this.getData('ltPropSchedulerView');
                var date = this.findDateonMenuClick(originElem,view);
                if(this.getData('ltPropMultiUserView')  && view !== 'month' ){
                    var userid = originElem.closest('tr').dataset.userid;
                    var userData = $u.findWhere(this.getData('ltPropSchedulerUser'),{'id':userid});
                    userData = $u.clone(userData);
                }
                if(this.getMethods('onMenuOpen')){
                    this.executeMethod('onMenuOpen',date,event,userData);
                }
            },
            menuonbeforeclose : function ( menu , event  ) {
                if(this.getMethods('onBeforeMenuClose')){
                    this.executeMethod('onBeforeMenuClose', menu , event );
                }
            },
            menuonclose: function( menu , event  ){
                if(this.getMethods('onMenuClose')){
                    this.executeMethod('onMenuClose',menu , event);
                }
            },		
            ChangeSchedulerView: function(value){
                var previousView = this.getData('ltPropSchedulerView');
                this._fromButtonSwitch = true;
                this.findDayViewPosition(previousView,value)			
                this.setData('ltPropDateArray',[]);
                this.setData('ltPropDateObj',{});
                this.setData('ltPropSchedulerView',value);
            },
            closeschedulerpopover : function(event,element){
                var targetElem = this.$node.querySelector('#schedulerPopover');
                targetElem.setAttribute('id','');
                this.getMethods('onHiddenEventPopoverHide') && this.executeMethod('onHiddenEventPopoverHide',element);
            },
            closeCalender : function(event,popover){
                $L(popover.getData('ltPropOriginElem'),this.$node).removeClass('lyteSchNavigatorCalOpened');
            },
            openUserSearch : function(event,search){
                var popover = $L('#usersreachPopover',this.$node);
                search.closest('.lyteSchedulerFirstColHeader').classList.add('lyteSchedulerUserSearchActive');
                if(!popover[0].getData('ltPropShow')){
                    popover[0].ltProp('show',true);
                }
            },
            closeUserPopover : function(){
                $L('.lyteSchedulerUserSearchActive',this.$node)[0].classList.remove('lyteSchedulerUserSearchActive');
            },
            onPopoverShow : function(element){
                this.getMethods('onHiddenEventPopoverShow') && this.executeMethod('onHiddenEventPopoverShow',element);
            }
        }), arg1);
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            eventResize: function(){
                this.toggleEventResize();
            }.observes('ltPropEventResize').on('didConnect'),

            selectedUserArr : function(){
                if(this.getData('ltPropSchedulerView') == 'month' && this.getData('ltPropMultiUserView') === true){
                    this.callCurrentView(null, null, this.getData('ltPropWeekStart'));
                }	
            }.observes('selectedUserArr.[]'),

            userSortable : function(){
                var view = this.getData('ltPropSchedulerView');
                var sortableCol = $L('.lyteSchedulerViewMainTable tbody',this.$node);
                if( (view == 'week'|| view == 'day') && this.getData('ltPropMultiUserView') ){
                    if(this.getData('ltPropSortUser')){
                        sortableCol.sortable({
                            onDragStart:function(){
                            },helper: function(elem){
                                var table_td = elem.querySelector('.UserCol lyte-scheduler-user');
                                var clone_tr = document.createElement('lyte-tr');
                                var clone_td  = document.createElement('lyte-td');
                                clone_td.appendChild(table_td.cloneNode(true));
                                clone_tr.appendChild(clone_td);
                                clone_tr.appendChild(document.createElement('lyte-td'));
                                return clone_tr;
                            },onBeforeDrop: function( dragElem , bellowElem , placeholder, from, to, soruce ){
                                var table_tr = soruce.querySelectorAll('tr');
                                dragElem.remove();
                                bellowElem.parentNode.insertBefore(table_tr[from],bellowElem.nextSibling);
                            },onDrop : function(dragElem){
                                dragElem.remove();
                            },placeholder : "lyteSchdeulerPlaceholder",
                            threshold: '10'
                        });
                    }else {
                        if(sortableCol[0]){
                            sortableCol.sortable("destroy");
                        }
                    }
                }
            }.observes('ltPropSortUser').on('didConnect'),

            ChangeWeekOrder : function(){
                var start_day = this.getData('ltPropWeekStart');
                if(start_day){
                    var dayLabel = this.getData('ltPropLabel');
                    var split = dayLabel.slice(start_day);
                    dayLabel = split.concat( dayLabel.slice(0,start_day));
                    this.setData('label',dayLabel);
                    var dayfullform = this.getData('days');
                    split = dayfullform.slice(start_day);
                    dayfullform = split.concat( dayfullform.slice(0,start_day));
                    this.setData('labelFF',dayfullform);
                }
                
            }.observes('ltPropWeekStart').on('init'),

            changeEvent : function(){
                var eventManipulator = new window._lyteUiUtils.eventManipulator( this.getData('ltPropFormat') );
                eventManipulator.addEvents( this.data.ltPropEvent, this.getData('ltPropSchedulerView'), this.getData('ltPropMultiUserView') );
                this.ResetTimeLine();
                this.setData('ltPropManipulatedEvent',eventManipulator.events);
            }.observes('ltPropEvent.[]').on('init'),

            schedulerView : function(){
                var date = null;
                var view = this.getData('ltPropSchedulerView');
                this._previousView = this._currentView;
                this._currentView = this.getData('ltPropSchedulerView');
                if(!this._fromButtonSwitch){
                    this.findDayViewPosition(this._previousView,this._currentView);
                }else{
                    this._fromButtonSwitch = false;
                }
                if(this.moveToDay && view == 'day'){
                    date = this.moveToDay;
                    this.callCurrentView(date , false ,this.getData('ltPropWeekStart'));
                }else{
                    this.callCurrentView(null , !this.$node.rendered ,this.getData('ltPropWeekStart'));
                }
                var startView;
                if(view == 'month'){
                    var current_date = this.getData('ltPropCurrentDate');
                    var date;
                    if(current_date.getMonth() == this.data.ltPropCurrMonth && this.data.ltPropDisplayYear == current_date.getFullYear()){
                        date = new Date( current_date.toString() );
                    }else{
                        date =  new Date(this.data.ltPropDisplayYear,this.data.ltPropCurrMonth,1);
                    }
                    startView = $L.moment(date);
                }else{
                    startView = $L.moment(this.getData('startDateOfView'),'DD-MM-YYYY')
                }
                if(this.getMethods('onViewChange')){
                    this.executeMethod('onViewChange',startView,this.getData('ltPropSchedulerView'));
                }
                this.$node.rendered = true;
                this.addScroll();
                this.destoryEventResize();
                this.toggleEventResize();
            }.observes('ltPropSchedulerView'),

            toggleMultiUserView : function(){
                this.callCurrentView(undefined,undefined,this.getData('ltPropWeekStart'));
                this.addScroll();
            }.observes('ltPropMultiUserView'),

            changeCurrentDate : function(){
                this.callCurrentView(null , !this.$node.rendered ,this.getData('ltPropWeekStart'));
            }.observes('ltPropCurrentDate'),

            callDragDrop : function(){
                this.addDragAndDrop();
            }.observes('ltPropDraggable').on('didConnect'),

            addSchedulerEvent: function(){
                var view = this.getData('ltPropSchedulerView');
                var date = undefined;
                if(view == 'month'){
                    date  =  new Date(this.getData('ltPropDisplayYear'),this.getData('ltPropCurrMonth'),1);
                }
                this.callCurrentView(date,undefined,this.getData('ltPropWeekStart'));
                this.addDragAndDrop();
            }.observes('ltPropManipulatedEvent'),

            onWorkingViewChange : function(){
                this.ResetTimeLine();
                this.callCurrentView(null , !this.$node.rendered ,this.getData('ltPropWeekStart'));
            }.observes('ltPropWorkingDayOnly'),

            showCurrentTimeObserver: function(){
                var result;
                if(this.getData('ltPropSchedulerView') == 'week'){
                    result = this.data.ltPropDateArray;
                    result = result[result.length-1].val;
                }
                this.setCurrentTimeLine($L.moment(this.getData('startDateOfView'),'DD-MM-YYYY'),result);
            }.observes('ltPropShowCurrentTime')
        }), arg1);
    }

    _() {
        _;
    }
}

LyteSchedulerComponent._template = "<template tag-name=\"lyte-scheduler\"> <lyte-scheduler-header class=\"{{if(ltPropShowWeekNumber,'lyteSchedulerWeekNumberShown','')}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSchedulerHeaderYield}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-yield yield-name=\"scheduler-header\"></lyte-yield> </template><template default=\"\"> <div class=\"lyteSchedulerDefaultHeader\"> <lyte-scheduler-view> <lyte-button-group lt-prop-selected=\"{{ltPropSchedulerView}}\" lt-prop-type=\"radiobutton\" lt-prop-appearance=\"fill\" on-select=\"{{method(&quot;ChangeSchedulerView&quot;)}}\"> <template is=\"registerYield\" yield-name=\"yield\"> <lyte-button lt-prop-value=\"day\" class=\"lyteSchedulerDayViewButton\"> <template is=\"registerYield\" yield-name=\"text\"> {{lyteUiI18n(\"day\")}} </template> </lyte-button> <lyte-button lt-prop-value=\"week\" class=\"lyteSchedulerWeekViewButton\"> <template is=\"registerYield\" yield-name=\"text\"> {{lyteUiI18n(\"week\")}} </template> </lyte-button> <lyte-button lt-prop-value=\"month\" class=\"lyteSchedulerMonthViewButton\"> <template is=\"registerYield\" yield-name=\"text\"> {{lyteUiI18n(\"month\")}} </template> </lyte-button> </template> </lyte-button-group> </lyte-scheduler-view> <div class=\"lyteSchedulerDateNavigator\"> <lyte-scheduler-icon class=\"lyteSchedulerStartIcon\" onclick=\"{{action(&quot;hideandShowCalender&quot;,event)}}\" tabindex=\"0\" role=\"button\" lt-prop-aria-keydown=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropyieldNav}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-yield yield-name=\"scheduler-nav-icon\"></lyte-yield> </template></template></lyte-scheduler-icon> <scheduler-start-date class=\"lyteSchedulerStartDate\" tabindex=\"-1\" aria-hidden=\"true\"> {{ltCurrStartDate}} </scheduler-start-date> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(ltPropSchedulerView,'==','week')}}\" is=\"case\" lc-id=\"lc_id_0\"> <span class=\"lyteSchedulerTitleSeparator\" tabindex=\"-1\" aria-hidden=\"true\">To</span> <scheduler-end-date class=\"lyteSchedulerEndDate\" tabindex=\"-1\" aria-hidden=\"true\"> {{ltCurrEndDate}} </scheduler-end-date> </template></template></div> <lyte-scheduler-title tabindex=\"0\"> <span class=\"lyteSchedulerTitleDate\">{{ltdisplayCurrDate}}</span> <span class=\"lyteSchedulerTitleDay\">{{lyteUiI18n(ltPropCurDay)}}</span> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(ltPropSchedulerView,'==','day')}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropShowWeekNumber}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerWeekNumber lyteSchedulerDayView\"> {{lyteSchedulerUiFindWeek(startDateOfView,ltPropSchedulerView)}} </span></template></template></template></template></lyte-scheduler-title> <lyte-scheduler-nav tabindex=\"-1\"> <span role=\"button\" class=\"lyteSchedulerNavLeft\" tabindex=\"0\" lt-prop-aria-keydown=\"true\"> <span class=\"lyteVisuallyHidden\"> Previous {{ltPropSchedulerView}} </span> </span> <lyte-button onclick=\"{{action('today',event)}}\" class=\"schedulerToday\"> <template is=\"registerYield\" yield-name=\"text\"> {{lyteUiI18n(todayButton)}} </template> </lyte-button> <span role=\"button\" class=\"lyteSchedulerNavRight\" tabindex=\"0\" lt-prop-aria-keydown=\"true\"> <span class=\"lyteVisuallyHidden\"> Next {{ltPropSchedulerView}} </span> </span> </lyte-scheduler-nav> <lyte-yield yield-name=\"scheduler-selected-User\" class=\"lyteSchedulerSelectedUser\"></lyte-yield> </div> </template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(ltPropMultiUserView,'&amp;&amp;',expHandlers(ltPropSchedulerSearchYield,'!'))}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(selectedUserArr.length,'>',0)}}\" is=\"case\" lc-id=\"lc_id_0\"> <div> <div> <template items=\"{{selectedUserArr}}\" item=\"userid\" index=\"index\" is=\"for\" _new=\"true\"> <div> <lyte-yield yield-name=\"scheduler-user-icon\" scheduler-user=\"{{lyteUiSchedulerfindUser(ltPropSchedulerUser,userid)}}\"> </lyte-yield> </div> </template> </div> <div onclick=\"{{action('clearSelection')}}\"> clear </div> </div> </template></template></template></template></lyte-scheduler-header> <lyte-scheduler-body> <div class=\"lyteSchedulerWrapperClass\"><template _new=\"true\" value=\"{{ltPropSchedulerView}}\" is=\"switch\"><template case=\"month\" is=\"case\"><lyte-scheduler-month> <div class=\"{{if(ltPropMultiUserView,'lyteSchedulerMultiUserView','')}} lyteSchedulerMonthViewWrapper lyteSchedulerViewWrapper\"> <table class=\"lyteSchedulerViewMainTable lyteSchedulerMonthView\"> <tbody is=\"switch\" lyte-switch=\"true\" l-c=\"true\" _new=\"true\"></tbody><tbody class=\"lyteSchedulerMonthViewBody\"> <tr is=\"for\" lyte-for=\"true\" items=\"{{ltPropDateArray}}\" item=\"week\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody> </table> </div> </lyte-scheduler-month></template><template case=\"week\" is=\"case\"><lyte-scheduler-week> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropMultiUserView}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteSchedulerMultiUserView lyteSchedulerWeekViewWrapper lyteSchedulerViewWrapper {{if(ltPropWorkingDayOnly,'lyteSchedulerBusinessDayOnly','')}}\"> <table class=\"lyteSchedulerViewMainTable lyteSchedulerWeekView \"> <thead> <tr class=\"lyteSchedulerWeekViewHeader\"> <th class=\"lyteSchedulerFirstColHeader\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSchedulerSearch}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-search lt-prop-query-selector=\"{&quot;scope&quot; : &quot;lyte-popover-content&quot;, &quot;search&quot; : &quot;.lyteSchedulerUserSearch&quot;}\" on-focus=\"{{method('openUserSearch')}}\"> </lyte-search> </template><template case=\"{{ltPropSchedulerSearchYield}}\" is=\"case\" lc-id=\"lc_id_1\"> <lyte-yield yield-name=\"lyteSchedulerSearchYield\"> </lyte-yield> </template></template></th> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropDateArray}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </thead> <tbody> <tr is=\"for\" lyte-for=\"true\" items=\"{{ltPropSchedulerUser}}\" item=\"user\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody> </table> </div> </template><template default=\"\"> <div class=\"lyteSchedulerWeekViewWrapper lyteSchedulerViewWrapper {{if(ltPropWorkingDayOnly,'lyteSchedulerBusinessDayOnly','')}}\"> <div class=\"lyteSchedulerAllDayHighlight\"></div> <div class=\"lyteSchedularWeekViewInnerWrapper\"> <table class=\"lyteSchedulerViewMainTable lyteSchedulerWeekView \"> <thead> <tr class=\"lyteSchedulerWeekViewHeader\"> <th class=\"lyteSchedulerWeekTh\"></th> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropDateArray}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> <tr class=\"lyteSchedulerWeekViewAllDayRow\"> <td> <div class=\"lyteSchedulerSingleUserAllDay\"> {{lyteUiI18n('all-day')}} ({{lyteUiSchedulerAllDayCount(ltPropProcessedData,hiddenObj,startDateOfView,ltPropSchedulerView,ltPropMultiUserView)}}) </div> </td> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropDateArray}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </thead> <tbody> <tr is=\"for\" lyte-for=\"true\" items=\"{{timeLine}}\" item=\"time\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody> </table> <div class=\"lyteSchedulerOverlay\"> <table class=\"lyteSchedulerWeekOverlayTable\" cellpadding=\"0\"> <tbody> <tr> <td style=\"width:100px\"></td> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropDateArray}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </tbody> </table> </div> <div class=\"lyteschedulerCurrentTimeLine\"> <div class=\"lyteschedulerLine\"></div> </div> </div> </div> </template></template></lyte-scheduler-week></template><template case=\"day\" is=\"case\"><lyte-scheduler-day> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropMultiUserView}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"scrollContainer lyteSchedulerMultiUserView lyteSchedulerDayViewWrapper lyteSchedulerViewWrapper\"> <table> <thead> <tr class=\"lyteSchedulerMultiUserRow\"> <th class=\"lyteSchedulerFirstColHeader lyteSchedulerFixedCol\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSchedulerSearch}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-search lt-prop-query-selector=\"{&quot;scope&quot; : &quot;lyte-popover-content&quot;, &quot;search&quot; : &quot;.lyteSchedulerUserSearch&quot;}\" on-focus=\"{{method('openUserSearch')}}\"> </lyte-search> </template><template case=\"{{ltPropSchedulerSearchYield}}\" is=\"case\" lc-id=\"lc_id_1\"> <lyte-yield yield-name=\"lyteSchedulerSearchYield\"> </lyte-yield> </template></template></th> <th class=\"lyteSchedulerAllDayTh\"> <div class=\"lyteSchedulerMultiUserAllDay\">{{lyteUiI18n('all-day')}} ({{lyteUiSchedulerAllDayCount(ltPropProcessedData,hiddenObj,startDateOfView,ltPropSchedulerView,ltPropMultiUserView)}}) </div> </th> <td is=\"for\" lyte-for=\"true\" items=\"{{timeLine}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </thead> <tbody> <tr is=\"for\" lyte-for=\"true\" items=\"{{ltPropSchedulerUser}}\" item=\"user\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody> </table> </div> <div class=\"lyteSchedulerAllDayHighlight\"></div> <scheduler-scroll-nav class=\"lyteSchedulerLeftNav\" onclick=\"{{action('scrollnav',event)}}\"></scheduler-scroll-nav> <scheduler-scroll-nav class=\"lyteSchedulerRightNav\" onclick=\"{{action('scrollnav',event)}}\"></scheduler-scroll-nav> </template><template default=\"\"> <div class=\"lyteSchedulerDayViewWrapper lyteSchedulerViewWrapper\"> <div class=\"lyteSchedularDayViewInnerWrapper\"> <table class=\"lyteSchedulerViewMainTable lyteSchedulerDayView\"> <tbody><tr class=\"lyteSchedulerDayViewAllDayRow\"> <td is=\"switch\" lyte-switch=\"true\" l-c=\"true\" _new=\"true\"></td><td class=\"lyteSchedulerDate lyteSchedulerMenuQuerySelector lyteSchedulerAllDayEvent\"> <div class=\"scheduler-event-div\" data-date=\"{{startDateOfView}}\"> <template items=\"{{ltPropDateObj.allDay}}\" item=\"event\" index=\"index\" is=\"for\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{event.allDayEvent}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerEvent(event)}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-scheduler-event-tag event-index=\"{{index}}\" class=\"{{event.class}} {{if(expHandlers(event.editable,'!'),'lyteSchedulerNoDrag','')}}\" lt-prop-event=\"{{event}}\" data-id=\"{{event.id}}\" id=\"u{{event.id}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',event,this)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-event=\"{{event}}\"> </lyte-yield> </lyte-scheduler-event-tag></template></template></template></template></template> <div class=\"lyteSchedulerEventMoreBtn lyteSchedulerEventMoreBtnHide \" onclick=\"{{action(&quot;hiddenEvent&quot;,event,hiddenObj.allDay,startDateOfView,undefined,undefined,ltPropSchedulerView,true)}}\"> {{lyteUiSchedulerhiddenEvent(hiddenObj.allDay,startDateOfView,this,user.id)}} </div> </div> </td> </tr> <tr is=\"for\" lyte-for=\"true\" items=\"{{timeLine}}\" item=\"time\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody></table> <div class=\"lyteSchedulerOverlay\"> <table class=\"lyteSchedulerDayOverlayTable\"> <thead> <tr class=\"lyteSchedulerDayViewAllDayRow\"> <td> <div class=\"lyteSchedulerSingleUserAllDay\">{{lyteUiI18n('all-day')}}</div> </td> <td class=\"lyteSchedulerDate lyteSchedulerAllDayEvent\"> <div class=\"scheduler-event-div\"> <div></div> </div> </td> </tr> </thead> <tbody> <tr> <td></td> <td class=\"lyteSchedulerDate\"> <div class=\"scheduler-event-div\" data-date=\"{{startDateOfView}}\"> <template object=\"{{ltPropDateObj}}\" value=\"column\" key=\"key\" is=\"forIn\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(key,'!==','allDay')}}\" is=\"case\" lc-id=\"lc_id_0\"> <template items=\"{{column}}\" item=\"colArray\" index=\"index\" is=\"for\" _new=\"true\"><template items=\"{{colArray}}\" item=\"event\" index=\"rowindex\" is=\"for\" _new=\"true\"> <lyte-scheduler-event-tag event-index=\"{{index}}\" class=\"{{event.event.class}} {{if(expHandlers(event.event.editable,'!'),'lyteSchedulerNoDrag','')}}\" lt-prop-event=\"{{event.event}}\" data-id=\"{{event.event.id}}\" id=\"u{{event.event.id}}\" lt-prop-col=\"{{event.col}}\" lt-prop-row=\"{{rowindex}}\" lt-prop-intersect=\"{{event.intersect}}\" lt-prop-prevelem=\"{{event.prevElem}}\" lt-prop-count=\"{{column.length}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',event,this)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-event=\"{{event.event}}\"> </lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{event.event.editable}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerGrabResize\"></span></template></template></lyte-scheduler-event-tag> </template> </template> </template></template></template> </div> </td> </tr> </tbody> </table> </div> <div class=\"lyteschedulerCurrentTimeLine\"> <div class=\"lyteschedulerLine\"></div> </div> </div> <div class=\"lyteSchedulerAllDayHighlight\"></div> </div> </template></template></lyte-scheduler-day></template></template></div> </lyte-scheduler-body> <lyte-beta-popover id=\"lyteSchedulerHiddenEvent\" lt-prop-allow-multiple=\"{{ltPropPopoverAllowMultiple}}\" lt-prop-origin-elem=\"#schedulerPopover\" lt-prop-header-padding=\"\" lt-prop-content-padding=\"\" lt-prop-freeze=\"{{ltPropHiddenEventPopoverFreeze}}\" lt-prop-wrapper-class=\"lyteSchedulerMoreEventsPopupWrapper {{if(ltPropUserHiddenEvent,'lyteSchedulerUsersMoreEventsPopupWrapper','')}} {{ltPropHiddenPopoverWrapper}}\" on-close=\"{{method(&quot;closeschedulerpopover&quot;)}}\" lt-prop-allow=\"\" lt-prop-scrollable=\"true\" on-show=\"{{method('onPopoverShow')}}\"> <template is=\"registerYield\" yield-name=\"popover\"> <lyte-popover-header> <div class=\"lyteSchedulerMoreEventsPopTitleMonthVal\"></div> <div class=\"lyteSchedulerMoreEventsPopTitleDayVal\"></div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{hiddenUser}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-yield yield-name=\"scheduler-header-hidden-event\" scheduler-hidden-user=\"{{hiddenUser}}\"> </lyte-yield></template></template></lyte-popover-header> <lyte-popover-content> <lyte-yield yield-name=\"scheduler-hidden-event\" scheduler-hidden-event=\"{{hiddenEvent}}\"> </lyte-yield> </lyte-popover-content> </template> </lyte-beta-popover> <lyte-beta-popover lt-prop-content-padding=\"0px\" lt-prop-allow-multiple=\"{{ltPropPopoverAllowMultiple}}\" lt-prop-freeze=\"{{ltPropCalendarPopoverFreeze}}\" lt-prop-type=\"box\" lt-prop-show-close-button=\"false\" id=\"calendarPopover\" lt-prop-bind-to-body=\"true\" lt-prop-placement=\"bottomLeft\" on-close=\"{{method(&quot;closeCalender&quot;)}}\" lt-prop-wrapper-class=\"{{ltPropCalendarPopoverWrapper}}\"> <template is=\"registerYield\" yield-name=\"popover\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropShowCalendar}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-popover-content style=\"padding: 0px;\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(expHandlers(ltPropSchedulerView,'==','month'),'||',expHandlers(ltPropCalendarHeaderType,'!'))}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-calendar lt-prop-header-type=\"drilldown\" on-month-selected=\"{{method(&quot;monthSelected&quot;)}}\" on-week-selected=\"{{method(&quot;weekSelected&quot;)}}\" lt-prop-selection-type=\"{{selectionType}}\" on-date-selected=\"{{method(&quot;dateselect&quot;)}}\" lt-prop-format=\"DD-MM-YYYY\" lt-prop-start-week-day=\"{{if(ifEquals(ltPropWeekStart,undefined),0,ltPropWeekStart)}}\" class=\"lyteSchedulerCalendar\" on-navigate=\"{{method('onSchedulerNavToday')}}\"></lyte-calendar> </template><template default=\"\"> <lyte-calendar lt-prop-header-type=\"dropdown\" on-month-selected=\"{{method(&quot;monthSelected&quot;)}}\" on-week-selected=\"{{method(&quot;weekSelected&quot;)}}\" lt-prop-selection-type=\"{{selectionType}}\" on-date-selected=\"{{method(&quot;dateselect&quot;)}}\" lt-prop-format=\"DD-MM-YYYY\" lt-prop-start-week-day=\"{{if(ifEquals(ltPropWeekStart,undefined),0,ltPropWeekStart)}}\" class=\"lyteSchedulerCalendar\" on-navigate=\"{{method('onSchedulerNavToday')}}\"></lyte-calendar> </template></template></lyte-popover-content></template></template></template> </lyte-beta-popover> <lyte-beta-popover id=\"usersreachPopover\" lt-prop-origin-elem=\".lyteSchedulerUserSearchActive\" lt-prop-allow-multiple=\"{{ltPropPopoverAllowMultiple}}\" lt-prop-close-on-body-click=\"false\" on-before-close=\"{{method('onUserSearchClose')}}\" lt-prop-freeze=\"false\" lt-prop-type=\"box\" lt-prop-show-close-button=\"false\" lt-prop-placement=\"bottom\" on-close=\"{{method('closeUserPopover')}}\"> <template is=\"registerYield\" yield-name=\"popover\"> <lyte-popover-content> <template items=\"{{ltPropSchedulerUser}}\" item=\"user\" index=\"index\" is=\"for\" _new=\"true\"> <div class=\"lyteSchedulerUserSearch\"> <lyte-checkbox lt-prop-name=\"checkbox\" lt-prop-val=\"{{user.id}}\" on-checked=\"{{method('boxChecked')}}\" on-unchecked=\"{{method('boxUnchecked')}}\"></lyte-checkbox> <lyte-yield yield-name=\"scheduler-user-search\" scheduler-user=\"{{user}}\"></lyte-yield> </div> </template> </lyte-popover-content> <lyte-popover-footer style=\"padding: 10px 25px 25px;\"> <lyte-button onclick=\"{{action('cancelUserSelect')}}\"> <template is=\"registerYield\" yield-name=\"text\"> Cancel </template> </lyte-button> <lyte-button lt-prop-appearance=\"primary\" onclick=\"{{action('SelectUserSelect')}}\"> <template is=\"registerYield\" yield-name=\"text\"> Done </template> </lyte-button> </lyte-popover-footer> </template> </lyte-beta-popover> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSchedulerMenu}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSchedulerMenuYeild}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-menu lt-prop-yield=\"{{ltPropSchedulerMenuYield}}\" lt-prop-event=\"click\" lt-prop-query=\".lyteSchedulerMenuQuerySelector\" lt-prop-user-value=\"{{ltPropSchedulerMenuUserValue}}\" lt-prop-content=\"{{ltPropSchedulerMenuContent}}\" lt-prop-system-value=\"{{ltPropSchedulerMenuSystemValue}}\" lt-prop-description=\"{{ltPropScheluerMenuDescription}}\" lt-prop-id=\"{{ltPropSchedulerMenuId}}\" lt-prop-class=\"{{ltPropSchedulerMenuClass}}\" lt-prop-position=\"{{ltPropSchedulerMenuPosition}}\" lt-prop-width=\"{{ltPropSchedulerMenuWidth}}\" lt-prop-height=\"{{ltPropSchedulerMenuHeight}}\" lt-prop-callout=\"{{ltPropSChedulerMenuCallout}}\" lt-prop-freeze=\"{{ltPropSchedulerMenuFreeze}}\" lt-prop-query-class=\"{{ltPropSchedulerMenuQueryClass}}\" lt-prop-animate=\"{{ltPropSchedulerMenuAnimate}}\" lt-prop-wrapper-class=\"{{ltPropSchedulerMenuWrapperClass}}\" lt-prop-bind-to-body=\"{{ltPropSchedulerMenuBindToBody}}\" on-before-open=\"{{method('menuonbeforeopen')}}\" on-open=\"{{method('menuonopen')}}\" on-before-close=\"{{method('menuonbeforeclose')}}\" on-close=\"{{method('menuonclose')}}\" on-menu-click=\"{{method('menuonclick')}}\"> <template is=\"registerYield\" yield-name=\"yield\"> <lyte-yield yield-name=\"scheduler-menu\"></lyte-yield> </template> </lyte-menu> </template><template default=\"\"> <lyte-menu lt-prop-yield=\"{{ltPropSchedulerMenuYield}}\" lt-prop-set-css=\"false\" lt-prop-event=\"click\" lt-prop-query=\".lyteSchedulerMenuQuerySelector\" lt-prop-user-value=\"{{ltPropSchedulerMenuUserValue}}\" lt-prop-content=\"{{ltPropSchedulerMenuContent}}\" lt-prop-system-value=\"{{ltPropSchedulerMenuSystemValue}}\" lt-prop-description=\"{{ltPropScheluerMenuDescription}}\" lt-prop-id=\"{{ltPropSchedulerMenuId}}\" lt-prop-class=\"{{ltPropSchedulerMenuClass}}\" lt-prop-position=\"{{ltPropSchedulerMenuPosition}}\" lt-prop-width=\"{{ltPropSchedulerMenuWidth}}\" lt-prop-height=\"{{ltPropSchedulerMenuHeight}}\" lt-prop-callout=\"{{ltPropSChedulerMenuCallout}}\" lt-prop-freeze=\"{{ltPropSchedulerMenuFreeze}}\" lt-prop-query-class=\"{{ltPropSchedulerMenuQueryClass}}\" lt-prop-animate=\"{{ltPropSchedulerMenuAnimate}}\" lt-prop-wrapper-class=\"{{ltPropSchedulerMenuWrapperClass}}\" lt-prop-bind-to-body=\"{{ltPropSchedulerMenuBindToBody}}\" on-before-open=\"{{method('menuonbeforeopen')}}\" on-open=\"{{method('menuonopen')}}\" on-before-close=\"{{method('menuonbeforeclose')}}\" on-close=\"{{method('menuonclose')}}\" on-menu-click=\"{{method('menuonclick')}}\"> <template is=\"registerYield\" yield-name=\"yield\"> <lyte-yield yield-name=\"scheduler-menu\"></lyte-yield> </template> </lyte-menu> </template></template></template></template></template>";;
LyteSchedulerComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"i","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1,1,1],"cn":"default"},{"t":"r","p":[1,1,1,1],"dN":[{"t":"r","p":[1,1],"dN":[{"t":"tX","p":[1]}],"in":5,"sibl":[4]},{"t":"cD","p":[1],"in":4,"sibl":[3]},{"t":"r","p":[3,1],"dN":[{"t":"tX","p":[1]}],"in":3,"sibl":[2]},{"t":"cD","p":[3],"in":2,"sibl":[1]},{"t":"r","p":[5,1],"dN":[{"t":"tX","p":[1]}],"in":1,"sibl":[0]},{"t":"cD","p":[5],"in":0}],"dc":[4,2,0],"hc":true,"trans":true,"in":12,"sibl":[11],"cn":"default"},{"t":"cD","p":[1,1,1],"in":11,"sibl":[10],"cn":"default"},{"t":"cD","p":[1,1],"in":10,"sibl":[9],"cn":"default"},{"t":"a","p":[1,3,1],"cn":"default"},{"t":"s","p":[1,3,1,1],"c":{"lc_id_0":{"dN":[{"t":"i","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":9,"sibl":[8],"cn":"default"},{"t":"cD","p":[1,3,1],"in":8,"sibl":[7],"cn":"default"},{"t":"tX","p":[1,3,3,1],"cn":"default"},{"t":"cD","p":[1,3,3],"in":7,"sibl":[6],"cn":"default"},{"t":"s","p":[1,3,5],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[3,1],"cn":"lc_id_0"},{"t":"cD","p":[3],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":6,"sibl":[5],"cn":"default"},{"t":"tX","p":[1,5,1,0],"cn":"default"},{"t":"tX","p":[1,5,3,0],"cn":"default"},{"t":"s","p":[1,5,5],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":5,"sibl":[4],"cn":"default"},{"t":"cD","p":[1,5],"in":4,"sibl":[3],"cn":"default"},{"t":"tX","p":[1,7,1,1,1],"cn":"default"},{"t":"a","p":[1,7,3],"cn":"default"},{"t":"r","p":[1,7,3,1],"dN":[{"t":"tX","p":[1]}],"in":3,"sibl":[2],"cn":"default"},{"t":"cD","p":[1,7,3],"in":2,"sibl":[1],"cn":"default"},{"t":"tX","p":[1,7,5,1,1],"cn":"default"},{"t":"cD","p":[1,7],"in":1,"sibl":[0],"cn":"default"},{"t":"i","p":[1,9],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{"dc":[12,11,10,9,8,7,6,4,2,1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":11,"sibl":[10]},{"t":"s","p":[1,2],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1,1,1],"cn":"lc_id_0"},{"t":"f","p":[1,1,1],"dN":[{"t":"a","p":[1,1]},{"t":"i","p":[1,1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"},{"t":"a","p":[1,3],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":10,"sibl":[9]},{"t":"cD","p":[1],"in":9,"sibl":[8]},{"t":"a","p":[3,1,0]},{"t":"s","p":[3,1,0],"c":{"month":{"dN":[{"t":"a","p":[0,1],"cn":"month"},{"t":"s","p":[0,1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1,1,1],"cn":"lc_id_0"},{"t":"f","p":[1,1,1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,1]}],"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <th class=\"lyteSchedulerMonthTh {{item.buisness}}\"> {{lyteUiI18n(lyteUiSchedulerLabelFormat('ddd',item.month,item.date,ltPropDisplayYear))}} </th> </tr></tbody></table></template>","tagName":"TR","cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"actualTemplate":"<template is=\"switch\"><template case=\"{{expHandlers(ltPropMultiUserView,'!')}}\" is=\"case\" lc-id=\"lc_id_0\" depth=\"1\"><table> <thead> <tr class=\"lyteSchedulerMonthViewHead\"> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropDateArray[0]}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </thead> </table></template></template>","in":2,"sibl":[1],"cn":"month"},{"t":"a","p":[0,1,1,2,1],"cn":"month"},{"t":"f","p":[0,1,1,2,1],"dN":[{"t":"a","p":[1,1]},{"t":"f","p":[1,1],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"s","p":[1,1,1],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"cn":"lc_id_0"},{"t":"a","p":[2],"cn":"lc_id_0"},{"t":"tX","p":[2,1,0],"cn":"lc_id_0"},{"t":"tX","p":[2,3,0],"cn":"lc_id_0"},{"t":"a","p":[4],"cn":"lc_id_0"},{"t":"tX","p":[4,1,0],"cn":"lc_id_0"},{"t":"tX","p":[4,3,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"tX","p":[1,0],"cn":"default"}]},"dc":{"lc_id_0":{},"default":{}},"hd":true,"co":["lc_id_0"],"in":3,"sibl":[2]},{"t":"s","p":[1,1,2],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1]},{"t":"a","p":[1,1,3]},{"t":"i","p":[1,1,3],"in":1,"sibl":[0]},{"t":"s","p":[1,1,5],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"fI","p":[1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"a","p":[0,1],"cn":"lc_id_0"},{"t":"i","p":[0,1],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"s","p":[0,3],"c":{"lc_id_0":{"dN":[{"t":"r","p":[1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,1]}],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[2,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"},{"t":"a","p":[3],"cn":"lc_id_0"},{"t":"tX","p":[3,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"f","p":[1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"a","p":[0,1],"cn":"lc_id_0"},{"t":"i","p":[0,1],"in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"s","p":[0,3],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"s","p":[0,4],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[3,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"default"},{"t":"a","p":[3],"cn":"default"},{"t":"tX","p":[3,1],"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[1,0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <td class=\"lyteSchedulerDate lyteSchedulerMenuQuerySelector {{if(ifEquals(item.current_month,true),&quot;lyteSchedulerCurrentMonthDate&quot;,&quot;lyteSchedulerOtherMonthDate&quot;)}} lyteSchedulerAllDayEvent {{item.buisness}} {{if(ifEquals(item.currentDate,true),&quot;lyteSchedulerCurrDate&quot;,&quot;&quot;)}}\"> <div class=\"lyteSchedulerMonthEventsWrap scheduler-event-div\" data-date=\"{{lyteUiSchedulerGetDate(item.val)}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropMultiUserView}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.currentDate}}\" is=\"case\" lc-id=\"lc_id_0\"><div class=\"lyteSchedulerMonthTodayLabel\">Today</div></template></template><div class=\"lyteSchedulerMonthYearLabel\" onclick=\"{{action('onDateClick',event,item.val)}}\"> <span>{{lyteUiSchedulerMonthshortForm(item.month)}}</span> <span>{{item.year}}</span> </div> <div class=\"lyteSchedulerMultiUserMonthViewDateDayLabel\" onclick=\"{{action('onDateClick',event,item.val)}}\"> <span class=\"lyteSchedulerMultiUserMonthViewDateLabel\">{{item.date}}</span> <span class=\"lyteSchedulerMultiUserMonthViewDayLabel\">{{lyteUiSchedulerLabelFormat('ddd',item.month,item.date,ltPropDisplayYear)}}</span> </div> </template><template default=\"\"> <span class=\"lyteSchedulerDayNumber\" onclick=\"{{action('onDateClick',event,item.val)}}\">{{item.date}}</span> </template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(index,'!')}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropShowWeekNumber}}\" is=\"case\" lc-id=\"lc_id_0\"><div class=\"lyteSchedulerWeekNumber\"> {{lyteSchedulerUiFindWeek(item.val)}} </div></template></template></template></template><lyte-yield yield-name=\"scheduler-date\" scheduler-date=\"{{item.val}}\"></lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.events}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropMultiUserView}}\" is=\"case\" lc-id=\"lc_id_0\"> <template object=\"{{item.events}}\" value=\"value\" key=\"key\" is=\"forIn\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerisSearch(selectedUserArr,key,Index)}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-scheduler-event-tag class=\"{{value.class}} {{if(expHandlers(item.editable,'!'),'lyteSchedulerNoDrag','')}}\" lt-prop-event=\"{{value}}\" data-id=\"{{value.id}}\" id=\"u{{key}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',value,this,key)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this,value,key)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-user=\"{{lyteUiSchedulergetUserData(key,ltPropSchedulerUser)}}\" scheduler-event=\"{{value}}\"> </lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(value.length,'>',1)}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"yield\" yield-name=\"scheduler-event-badge\"> <div onclick=\"{{action(&quot;hiddenUserEvent&quot;,event,value,key)}}\" class=\"lyteSchedulerEventMoreBtn lyteSchedulerUserMoreBtn\"> {{value.length}} </div> </template> </template></template></lyte-scheduler-event-tag></template></template></template> <div class=\"lyteSchedulerEventMoreBtn lyteSchedulerEventMoreBtnHide\" onclick=\"{{action(&quot;hiddenEvent&quot;,event,hiddenObj,item.val)}}\"> {{lyteUiSchedulerhiddenEvent(hiddenObj,lyteUiSchedulerGetDate(item.val),this)}} </div> </template><template default=\"\"> <template items=\"{{item.events}}\" item=\"event\" index=\"index\" is=\"for\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerEvent(event)}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-scheduler-event-tag class=\"{{event.class}} {{if(expHandlers(event.editable,'!'),'lyteSchedulerNoDrag','')}}\" lt-prop-event=\"{{event}}\" data-id=\"{{event.id}}\" id=\"u{{event.id}}\" event-index=\"{{index}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',event,this)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-event=\"{{event}}\"> </lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerIsResize(week,event,ltPropFormat,'start')}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerHorizontalResizeIcon lyteSchedulerHorizontalLeftResize\"></span></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerIsResize(week,event,ltPropFormat,'end')}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerHorizontalResizeIcon lyteSchedulerHorizontalRightResize\"></span></template></template></lyte-scheduler-event-tag></template></template></template> <div class=\"lyteSchedulerEventMoreBtn lyteSchedulerEventMoreBtnHide\" onclick=\"{{action(&quot;hiddenEvent&quot;,event,hiddenObj,item.val)}}\"> {{lyteUiSchedulerhiddenEvent(hiddenObj,lyteUiSchedulerGetDate(item.val),this)}} </div> </template></template></template></template></div> </td> </tr></tbody></table></template>","tagName":"TR","in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody> <tr> <td is=\"for\" lyte-for=\"true\" items=\"{{week}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </tbody></table></template>","tagName":"TBODY","in":1,"sibl":[0],"cn":"month"},{"t":"cD","p":[0],"in":0,"cn":"month"}]},"week":{"dN":[{"t":"s","p":[0,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"s","p":[1,1,1,1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true},"lc_id_1":{"dN":[{"t":"i","p":[1],"in":0,"cn":"lc_id_1"}],"cdp":{"t":"a","p":[1]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"lc_id_1":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0","lc_id_1"],"hc":true,"trans":true,"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"a","p":[1,1,1,1,3],"cn":"lc_id_0"},{"t":"f","p":[1,1,1,1,3],"dN":[{"t":"a","p":[1]},{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"sibl":[0]},{"t":"s","p":[1,2],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[1,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"tX","p":[1,0],"cn":"default"},{"t":"tX","p":[3,1,0],"cn":"default"},{"t":"tX","p":[3,3,0],"cn":"default"}]},"dc":{"lc_id_0":{},"default":{}},"hd":true,"co":["lc_id_0"]}],"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <th class=\"lyteSchedulerWeekTh {{item.buisness}} {{if(ifEquals(item.currentDate,true),&quot;lyteSchedulerCurrDate&quot;,&quot;&quot;)}}\" onclick=\"{{action('onDateClick',event,item.val)}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(index,'!')}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropShowWeekNumber}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerWeekNumber\"> {{lyteSchedulerUiFindWeek(item.val)}} </span></template></template></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSchedulerLabelFormat}}\" is=\"case\" lc-id=\"lc_id_0\"> <div> {{lyteUiSchedulerLabelFormat(ltPropSchedulerLabelFormat,item.month,item.date,ltPropDisplayYear)}} </div> </template><template default=\"\"> <div class=\"lyteSchedulerMonthLabel\">{{lyteUiI18n(month[item.month])}} </div> <div> <span class=\"lyteSchedulerDateLabel\">{{lyteUiI18n(item.date)}}</span> <span class=\"lyteSchedulerDayLabel\">{{lyteUiI18n(label[index])}}</span> </div> </template></template></th> </tr></tbody></table></template>","tagName":"TR","in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"a","p":[1,1,3,1],"cn":"lc_id_0"},{"t":"f","p":[1,1,3,1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"a","p":[1,1,1],"cn":"lc_id_0"},{"t":"a","p":[1,1,1,1],"cn":"lc_id_0"},{"t":"i","p":[1,1,1,1],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"cD","p":[1,1,1],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"a","p":[1,3],"cn":"lc_id_0"},{"t":"f","p":[1,3],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"s","p":[1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"f","p":[1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"a","p":[0,1],"cn":"lc_id_0"},{"t":"i","p":[0,1],"in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"s","p":[0,3],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"s","p":[0,4],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[3,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"t":"a","p":[1,3]},{"t":"tX","p":[1,3,1]}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <td class=\"{{item.buisness}} {{if(ifEquals(item.currentDate,true),&quot;lyteSchedulerCurrDate&quot;,&quot;&quot;)}} lyteSchedulerDate lyteSchedulerMenuQuerySelector lyteSchedulerAllDayEvent\"> <div class=\"scheduler-event-div\" data-date=\"{{lyteUiSchedulerGetDate(item.val)}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.events}}\" is=\"case\" lc-id=\"lc_id_0\"> <template items=\"{{item.events[user.id]}}\" item=\"event\" index=\"index\" is=\"for\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerEvent(event)}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-scheduler-event-tag event-index=\"{{index}}\" class=\"{{event.class}} {{if(expHandlers(event.editable,'!'),'lyteSchedulerNoDrag','')}}\" lt-prop-event=\"{{event}}\" data-id=\"{{event.id}}\" id=\"u{{event.id}}_{{user.id}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',event,this)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this,null,user.id)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-event=\"{{event}}\"> </lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerIsResize(ltPropDateArray,event,ltPropFormat,'start')}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerHorizontalResizeIcon lyteSchedulerHorizontalLeftResize\"></span></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerIsResize(ltPropDateArray,event,ltPropFormat,'end')}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerHorizontalResizeIcon lyteSchedulerHorizontalRightResize\"></span></template></template></lyte-scheduler-event-tag></template></template></template> </template></template></div> <div class=\"lyteSchedulerEventMoreBtn lyteSchedulerEventMoreBtnHide\" onclick=\"{{action(&quot;hiddenEvent&quot;,event,hiddenObj,item.val,user.id)}}\"> {{lyteUiSchedulerhiddenEvent(hiddenObj,lyteUiSchedulerGetDate(item.val),this,user.id)}} </div> </td> </tr></tbody></table></template>","tagName":"TR","in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[2,1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"actualTemplate":"<template is=\"switch\"><template case=\"{{lyteUiSchedulerisSearch(selectedUserArr,user.id)}}\" is=\"case\" lc-id=\"lc_id_0\" depth=\"2\"><table><tbody> <tr data-userid=\"{{user.id}}\" class=\"lyteSchedulerMultiUserRow\"> <td class=\"lyteSchedulerUserCol\"> <lyte-scheduler-user lt-prop-userid=\"{{user.id}}\"> <lyte-yield yield-name=\"scheduler-user\" scheduler-user=\"{{user}}\" scheduler-event-count=\"{{lyteUiSchedulerEventCount(ltPropDateArray,user,hiddenObj)}}\"> </lyte-yield> </lyte-scheduler-user> </td> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropDateArray}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </tbody></table></template></template>","in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody> <tr is=\"switch\" lyte-switch=\"true\" l-c=\"true\" _new=\"true\"></tr></tbody></table></template>","tagName":"TBODY","in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"a","p":[1,3,1,1,1,3],"cn":"default"},{"t":"f","p":[1,3,1,1,1,3],"dN":[{"t":"a","p":[1]},{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[1,1],"cn":"lc_id_0"},{"t":"s","p":[3],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"tX","p":[1,1],"cn":"default"},{"t":"tX","p":[3,1,0],"cn":"default"},{"t":"tX","p":[3,3,0],"cn":"default"},{"t":"s","p":[5],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"cn":"default"}]},"dc":{"lc_id_0":{},"default":{}},"hd":true,"co":["lc_id_0"]}],"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <th class=\"lyteSchedulerWeekTh {{if(ifEquals(item.currentDate,true),&quot;lyteSchedulerCurrDate&quot;,&quot;&quot;)}} {{item.buisness}}\" onclick=\"{{action('onDateClick',event,item.val)}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSchedulerLabelFormat}}\" is=\"case\" lc-id=\"lc_id_0\"> <div> {{lyteUiSchedulerLabelFormat(ltPropSchedulerLabelFormat,item.month,item.date,ltPropDisplayYear,item)}} </div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(index,'!')}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropShowWeekNumber}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerWeekNumber\"> {{lyteSchedulerUiFindWeek(item.val)}} </span></template></template></template></template></template><template default=\"\"> <div class=\"lyteSchedulerMonthLabel\"> {{lyteUiI18n(month[item.month])}} </div> <div> <span class=\"lyteSchedulerDateLabel\">{{item.date}}</span> <span class=\"lyteSchedulerDayLabel\">{{lyteUiI18n(label[index])}}</span> </div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(index,'!')}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropShowWeekNumber}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerWeekNumber\"> {{lyteSchedulerUiFindWeek(item.val)}} </span></template></template></template></template></template></template></th> </tr></tbody></table></template>","tagName":"TR","in":3,"sibl":[2],"cn":"default"},{"t":"tX","p":[1,3,1,1,3,1,1,1],"cn":"default"},{"t":"tX","p":[1,3,1,1,3,1,1,3],"cn":"default"},{"t":"a","p":[1,3,1,1,3,3],"cn":"default"},{"t":"f","p":[1,3,1,1,3,3],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"s","p":[1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"f","p":[1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"a","p":[0,1],"cn":"lc_id_0"},{"t":"i","p":[0,1],"in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"s","p":[0,3],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"s","p":[0,4],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[3,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"},{"t":"a","p":[3],"cn":"lc_id_0"},{"t":"tX","p":[3,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <td class=\"lyteSchedulerDate lyteSchedulerMenuQuerySelector lyteSchedulerAllDayEvent {{item.buisness}}\" data-index=\"{{index}}\"> <div class=\"scheduler-event-div\" data-date=\"{{lyteUiSchedulerGetDate(item.val)}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.allDay}}\" is=\"case\" lc-id=\"lc_id_0\"> <template items=\"{{item.allDay}}\" item=\"event\" index=\"index\" is=\"for\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{event}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerEvent(event)}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-scheduler-event-tag event-index=\"{{index}}\" class=\"{{event.class}} {{if(expHandlers(event.editable,'!'),'lyteSchedulerNoDrag','')}}\" data-group=\"{{key}}\" lt-prop-event=\"{{event}}\" data-id=\"{{event.id}}\" id=\"u{{event.id}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',event,this)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-event=\"{{event}}\"> </lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerIsResize(ltPropDateArray,event,ltPropFormat,'start')}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerHorizontalResizeIcon lyteSchedulerHorizontalLeftResize\"></span></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiSchedulerIsResize(ltPropDateArray,event,ltPropFormat,'end')}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerHorizontalResizeIcon lyteSchedulerHorizontalRightResize\"></span></template></template></lyte-scheduler-event-tag></template></template></template></template></template> <div class=\"lyteSchedulerEventMoreBtn lyteSchedulerEventMoreBtnHide\" onclick=\"{{action(&quot;hiddenEvent&quot;,event,hiddenObj.allDay,item.val,undefined,undefined,ltPropSchedulerView,true)}}\"> {{lyteUiSchedulerhiddenEvent(hiddenObj.allDay,lyteUiSchedulerGetDate(item.val),this)}} </div> </template></template></div> </td> </tr></tbody></table></template>","tagName":"TR","in":2,"sibl":[1],"cn":"default"},{"t":"a","p":[1,3,1,3,1],"cn":"default"},{"t":"f","p":[1,3,1,3,1],"dN":[{"t":"a","p":[1]},{"t":"f","p":[1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,1,1,1]},{"t":"a","p":[1,3]},{"t":"f","p":[1,3],"dN":[{"t":"a","p":[1]}],"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <td data-index=\"{{index}}\" class=\"lyteSchedulerMenuQuerySelector\" data-date=\"{{lyteUiSchedulerGetDate(item.val)}}\"></td> </tr></tbody></table></template>","tagName":"TR"}],"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody> <tr id=\"T{{lyteUiSchedulerTimeFormat(time,min,null,false,true)}}\" data-time=\"{{lyteUiSchedulerTimeFormat(time,min,null,false)}}\" class=\"lyteSchedulerDate lyteSchedulerTimeLineEvent {{lyteUiSchedulerBusinessHour(ltPropBusinessHour,this)}}\"> <td style=\"width: 100px;\"> <div class=\"lyteSchedulerHourLabel\"> {{lyteUiSchedulerTimeFormat(time,min,ltPropContinentalTimeFormat,true)}} </div> <div class=\"lyteSchedulerHourQuarterLine\"></div> <div class=\"lyteSchedulerHourHalfLine\"></div> <div class=\"lyteSchedulerHourQuarterLine\"></div> </td> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropDateArray}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </tbody></table></template>","tagName":"TBODY"}],"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody> <tr is=\"for\" lyte-for=\"true\" items=\"{{timelineArray}}\" item=\"min\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody></table></template>","tagName":"TBODY","in":1,"sibl":[0],"cn":"default"},{"t":"a","p":[1,3,3,1,1,1,3],"cn":"default"},{"t":"f","p":[1,3,3,1,1,1,3],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"s","p":[1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"fI","p":[1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"f","p":[0],"dN":[{"t":"a","p":[0]},{"t":"f","p":[0],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"i","p":[1,1],"in":2,"sibl":[1]},{"t":"s","p":[1,3],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[2,0],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"s","p":[2],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"tX","p":[0,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[1],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <td class=\"lyteSchedulerDate {{if(ifEquals(item.currentDate,true),&quot;lyteSchedulerCurrDate&quot;,&quot;&quot;)}} {{item.buisness}}\"> <div class=\"scheduler-event-div\" data-date=\"{{lyteUiSchedulerGetDate(item.val)}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.events}}\" is=\"case\" lc-id=\"lc_id_0\"> <template object=\"{{item.events}}\" value=\"events\" key=\"key\" is=\"forIn\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(key,'!==','allDayEvent')}}\" is=\"case\" lc-id=\"lc_id_0\"><template items=\"{{events}}\" item=\"colArray\" index=\"index\" is=\"for\" _new=\"true\"><template items=\"{{colArray}}\" item=\"event\" index=\"rowIndex\" is=\"for\" _new=\"true\"> <lyte-scheduler-event-tag class=\"{{event.event.class}} {{if(expHandlers(event.event.editable,'!'),'lyteSchedulerNoDrag','')}}\" data-group=\"{{key}}\" lt-prop-event=\"{{event.event}}\" data-id=\"{{event.event.id}}\" id=\"u{{event.event.id}}\" lt-prop-col=\"{{event.col}}\" lt-prop-row=\"{{rowIndex}}\" lt-prop-intersect=\"{{event.intersect}}\" lt-prop-prevelem=\"{{event.prevElem}}\" lt-prop-count=\"{{events.length}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',event,this)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-event=\"{{event.event}}\"> </lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{event.event.editable}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerGrabResize\"></span></template></template></lyte-scheduler-event-tag> </template> </template> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{hiddenObj[key]}}\" is=\"case\" lc-id=\"lc_id_0\"><div class=\"lyteSchedulerEventMoreBtn lyteSchedulerEventMoreBtnHide\" data-group=\"{{key}}\" onclick=\"{{action(&quot;hiddenEvent&quot;,event,hiddenObj,item.val)}}\"> {{lyteUiSchedulerhiddenEvent(hiddenObj,lyteUiSchedulerGetDate(item.val),this,'',key)}} </div></template></template></template></template></template> </template></template></div> </td> </tr></tbody></table></template>","tagName":"TR","in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[2,0],"hc":true,"trans":true},"default":{"dc":[2,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"week"},{"t":"cD","p":[0],"in":0,"cn":"week"}]},"day":{"dN":[{"t":"s","p":[0,1],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1,1,1,1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true},"lc_id_1":{"dN":[{"t":"i","p":[1],"in":0,"cn":"lc_id_1"}],"cdp":{"t":"a","p":[1]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"lc_id_1":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0","lc_id_1"],"hc":true,"trans":true,"in":4,"sibl":[3],"cn":"lc_id_0"},{"t":"tX","p":[1,1,1,1,3,1,0],"cn":"lc_id_0"},{"t":"tX","p":[1,1,1,1,3,1,2],"cn":"lc_id_0"},{"t":"a","p":[1,1,1,1,5],"cn":"lc_id_0"},{"t":"f","p":[1,1,1,1,5],"dN":[{"t":"tX","p":[1,1]}],"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <th> {{lyteUiSchedulerTimeFormat(item,0,ltPropContinentalTimeFormat,true)}} <div class=\"lyteSchedulerHourSplitLineWrap\"> <span class=\"lyteSchedulerHourSplitLine\"></span> <span class=\"lyteSchedulerHourSplitLine\"></span> <span class=\"lyteSchedulerHourSplitLine\"></span> <span class=\"lyteSchedulerHourSplitLine\"></span> <span class=\"lyteSchedulerHourSplitLine\"></span> </div> <div class=\"lyteSchedulerHourSeparatorLine\"></div> </th> </tr></tbody></table></template>","tagName":"TR","in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"a","p":[1,1,3,1],"cn":"lc_id_0"},{"t":"f","p":[1,1,3,1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"a","p":[1,1,1],"cn":"lc_id_0"},{"t":"a","p":[1,1,1,1],"cn":"lc_id_0"},{"t":"i","p":[1,1,1,1],"in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"cD","p":[1,1,1],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"a","p":[1,3,1],"cn":"lc_id_0"},{"t":"a","p":[1,3,1,1],"cn":"lc_id_0"},{"t":"f","p":[1,3,1,1],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"i","p":[1,1],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[1,0],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"a","p":[1,3,1,3],"cn":"lc_id_0"},{"t":"tX","p":[1,3,1,3,1],"cn":"lc_id_0"},{"t":"a","p":[1,5],"cn":"lc_id_0"},{"t":"f","p":[1,5],"dN":[{"t":"a","p":[1,1]},{"t":"s","p":[1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"f","p":[1],"dN":[{"t":"a","p":[0]},{"t":"f","p":[0],"dN":[{"t":"s","p":[0],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"a","p":[1,1],"cn":"lc_id_0"},{"t":"i","p":[1,1],"in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"s","p":[1,3],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"s","p":[1,4],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[3,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"t":"a","p":[1,3]},{"t":"tX","p":[1,3,1]}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr> <td class=\"lyteSchedulerTimeLineEvent lyteSchedulerDate lyteSchedulerMenuQuerySelector\"> <div id=\"T{{lyteUiSchedulerTimeFormat(time,null,null,false,true)}}\" data-time=\"{{lyteUiSchedulerTimeFormat(time,min,null,false)}}\" class=\"lyteSchedulerMultiUserViewEventElem scheduler-event-div\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropDateObj[user.id][time]}}\" is=\"case\" lc-id=\"lc_id_0\"> <template items=\"{{ltPropDateObj[user.id][time]}}\" item=\"eventRow\" index=\"rowIndex\" is=\"for\" _new=\"true\"><template items=\"{{eventRow}}\" item=\"event\" index=\"index\" is=\"for\" _new=\"true\"><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(expHandlers(lyteUiIsEmptyObject(event),'!'),'&amp;&amp;',expHandlers(event.hiddenid,'!'))}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-scheduler-event-tag class=\"{{event.class}} {{if(expHandlers(event.editable,'!'),'lyteSchedulerNoDrag','')}}\" lt-prop-event=\"{{event}}\" event-index=\"{{rowIndex}}\" data-group=\"{{schedulerGroup}}\" data-id=\"{{event.id}}\" id=\"u{{event.id}}_{{user.id}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',event,this)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this,null,user.id)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-event=\"{{event}}\"> </lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{event.editable}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerHorizontalResizeIcon lyteSchedulerHorizontalLeftResize\"></span></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{event.editable}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteSchedulerHorizontalResizeIcon lyteSchedulerHorizontalRightResize\"></span></template></template></lyte-scheduler-event-tag> </template></template> </template> </template> </template></template></div> <div class=\"lyteSchedulerEventMoreBtn lyteSchedulerEventMoreBtnHide\" onclick=\"{{action(&quot;hiddenEvent&quot;,event,hiddenObj[startDateOfView][user.id],startDateOfView,user,time,ltPropSchedulerView)}}\"> {{lyteUiSchedulerhiddenEvent(hiddenObj,startDateOfView,this,user.id,null,time,ltPropSchedulerView)}} </div> </td> </tr></tbody></table></template>","tagName":"TR","in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[3,2,1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"actualTemplate":"<template is=\"switch\"><template case=\"{{lyteUiSchedulerisSearch(selectedUserArr,user.id)}}\" is=\"case\" lc-id=\"lc_id_0\" depth=\"2\"><table><tbody> <tr class=\"lyteSchedulerMultiUserRow\" data-userid=\"{{user.id}}\"> <td class=\"lyteSchedulerUserCol lyteSchedulerFixedCol\"> <lyte-scheduler-user lt-prop-user=\"{{user.id}}\"> <lyte-yield yield-name=\"scheduler-user\" scheduler-user=\"{{user}}\" scheduler-event-count=\"{{lyteUiSchedulerEventCount(ltPropDateObj,user,hiddenObj,startDateOfView)}}\"> </lyte-yield> </lyte-scheduler-user> </td> <td class=\"lyteSchedulerDate lyteSchedulerMenuQuerySelector lyteSchedulerAllDayEvent\"> <div class=\"scheduler-event-div \" data-date=\"{{startDateOfView}}\"> <template items=\"{{ltPropDateObj.allDay[user.id]}}\" item=\"event\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-scheduler-event-tag event-index=\"{{index}}\" class=\"{{event.class}} {{if(expHandlers(event.editable,'!'),'lyteSchedulerNoDrag','')}}\" lt-prop-event=\"{{event}}\" data-id=\"{{event.id}}\" id=\"u{{event.id}}_{{user.id}}\" event-render=\"{{method('eventRender')}}\" onclick=\"{{action('onSchedulerEventClick',event,this)}}\" onmouseover=\"{{action('onSchedulerEventHover',event,this,null,user.id)}}\" onmousedown=\"{{action('onSchedulerEventMouseDown',event,this)}}\"> <lyte-yield yield-name=\"scheduler-event\" scheduler-event=\"{{event}}\"> </lyte-yield> </lyte-scheduler-event-tag> </template> <div class=\"lyteSchedulerEventMoreBtn lyteSchedulerEventMoreBtnHide\" onclick=\"{{action(&quot;hiddenEvent&quot;,event,hiddenObj.allDay,startDateOfView,user.id,null,ltPropSchedulerView,true)}}\"> {{lyteUiSchedulerhiddenEvent(hiddenObj.allDay,startDateOfView,this,user.id)}} </div> </div> </td> <td is=\"for\" lyte-for=\"true\" items=\"{{timeLine}}\" item=\"time\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </tbody></table></template></template>","in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody> <tr is=\"switch\" lyte-switch=\"true\" l-c=\"true\" _new=\"true\"></tr></tbody></table></template>","tagName":"TBODY","in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"a","p":[5],"cn":"lc_id_0"},{"t":"cD","p":[5],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"a","p":[7],"cn":"lc_id_0"},{"t":"cD","p":[7],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"s","p":[1,1,1,1,0,1],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"tX","p":[0,1,1],"cn":"default"},{"t":"tX","p":[0,1,3],"cn":"default"}]},"dc":{"lc_id_0":{},"default":{}},"hd":true,"co":["lc_id_0"],"actualTemplate":"<template is=\"switch\"><template case=\"{{ltPropMultiUserView}}\" is=\"case\" lc-id=\"lc_id_0\" depth=\"3\"><table><tbody><tr><td>{{lyteUiI18n('all-day')}}</td></tr></tbody></table></template><template default=\"\" depth=\"3\"><table><tbody><tr><td> <div class=\"lyteSchedulerSingleUserAllDay\"> {{lyteUiI18n('all-day')}} ({{lyteUiSchedulerAllDayCount(ltPropProcessedData,hiddenObj,startDateOfView,ltPropSchedulerView,ltPropMultiUserView)}}) </div> </td></tr></tbody></table></template></template>","in":3,"sibl":[2],"cn":"default"},{"t":"a","p":[1,1,1,1,0,2,1],"cn":"default"},{"t":"a","p":[1,1,1,1,0,2,1,1],"cn":"default"},{"t":"f","p":[1,1,1,1,0,2,1,1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"a","p":[0,1],"cn":"lc_id_0"},{"t":"i","p":[0,1],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":2,"sibl":[1],"cn":"default"},{"t":"a","p":[1,1,1,1,0,2,1,3],"cn":"default"},{"t":"tX","p":[1,1,1,1,0,2,1,3,1],"cn":"default"},{"t":"a","p":[1,1,1,1,2],"cn":"default"},{"t":"f","p":[1,1,1,1,2],"dN":[{"t":"a","p":[1]},{"t":"f","p":[1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,1,1,1]}],"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody> <tr id=\"T{{lyteUiSchedulerTimeFormat(time,min,null,false,true)}}\" data-time=\"{{lyteUiSchedulerTimeFormat(time,min,null,false)}}\" class=\"lyteSchedulerDate lyteSchedulerMenuQuerySelector lyteSchedulerTimeLineEvent {{lyteUiSchedulerBusinessHour(ltPropBusinessHour,this)}}\"> <td> <div class=\"lyteSchedulerHourLabel\"> {{lyteUiSchedulerTimeFormat(time,min,ltPropContinentalTimeFormat,true)}} </div> <div class=\"lyteSchedulerHourQuarterLine\"></div> <div class=\"lyteSchedulerHourHalfLine\"></div> <div class=\"lyteSchedulerHourQuarterLine\"></div> </td> <td>&nbsp;</td> </tr> </tbody></table></template>","tagName":"TBODY"}],"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody> <tr is=\"for\" lyte-for=\"true\" items=\"{{timelineArray}}\" item=\"min\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody></table></template>","tagName":"TBODY","in":1,"sibl":[0],"cn":"default"},{"t":"tX","p":[1,1,3,1,1,1,1,1,0],"cn":"default"},{"t":"a","p":[1,1,3,1,3,1,3,1],"cn":"default"},{"t":"a","p":[1,1,3,1,3,1,3,1,1],"cn":"default"},{"t":"fI","p":[1,1,3,1,3,1,3,1,1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"f","p":[1],"dN":[{"t":"a","p":[0]},{"t":"f","p":[0],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"i","p":[1,1],"in":2,"sibl":[1]},{"t":"s","p":[1,3],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[2,0],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[4,2,1,0],"hc":true,"trans":true},"default":{"dc":[2,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"day"},{"t":"cD","p":[0],"in":0,"cn":"day"}]}},"d":{},"dc":{"month":{"dc":[1,0],"hc":true,"trans":true},"week":{"dc":[1,0],"hc":true,"trans":true},"day":{"dc":[1,0],"hc":true,"trans":true}},"co":["month","week","day"],"hc":true,"trans":true,"in":8,"sibl":[7]},{"t":"cD","p":[3],"in":7,"sibl":[6]},{"t":"a","p":[5]},{"t":"r","p":[5,1],"dN":[{"t":"s","p":[1,5],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"i","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":3,"sibl":[2]},{"t":"cD","p":[1],"in":2,"sibl":[1]},{"t":"a","p":[3,1]},{"t":"i","p":[3,1],"in":1,"sibl":[0]},{"t":"cD","p":[3],"in":0}],"dc":[3,2,1,0],"hc":true,"trans":true,"in":6,"sibl":[5]},{"t":"cD","p":[5],"in":5,"sibl":[4]},{"t":"a","p":[7]},{"t":"r","p":[7,1],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"s","p":[0,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"cD","p":[1],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":4,"sibl":[3]},{"t":"cD","p":[7],"in":3,"sibl":[2]},{"t":"a","p":[9]},{"t":"r","p":[9,1],"dN":[{"t":"a","p":[1,1]},{"t":"f","p":[1,1],"dN":[{"t":"a","p":[1,1]},{"t":"cD","p":[1,1],"in":1,"sibl":[0]},{"t":"a","p":[1,3]},{"t":"i","p":[1,3],"in":0}],"dc":[1,0],"hc":true,"trans":true,"in":6,"sibl":[5]},{"t":"cD","p":[1],"in":5,"sibl":[4]},{"t":"a","p":[3,1]},{"t":"r","p":[3,1,1],"dN":[],"in":4,"sibl":[3]},{"t":"cD","p":[3,1],"in":3,"sibl":[2]},{"t":"a","p":[3,3]},{"t":"r","p":[3,3,1],"dN":[],"in":2,"sibl":[1]},{"t":"cD","p":[3,3],"in":1,"sibl":[0]},{"t":"cD","p":[3],"in":0}],"dc":[6,5,3,1,0],"hc":true,"trans":true,"in":2,"sibl":[1]},{"t":"cD","p":[9],"in":1,"sibl":[0]},{"t":"s","p":[11],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"r","p":[1,1],"dN":[{"t":"i","p":[1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"r","p":[1,1],"dN":[{"t":"i","p":[1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"default"},{"t":"cD","p":[1],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true},"default":{"dc":[1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[11,10,9,8,7,6,5,4,3,2,1,0]}];;

LyteSchedulerComponent._observedAttributes = [
    "ltPropEvent",
    "ltPropFormat",
    "ltPropDraggable",
    "ltPropMaxAllDayEvent",
    "ltPropMaxTimelineEvent",
    "ltPropMultiUserView",
    "ltPropSchedulerUser",
    "ltPropBusinessDays",
    "ltPropBusinessHour",
    "ltPropSchedulerView",
    "ltPropSchedulerHeaderYield",
    "ltPropHiddenEventPopoverFreeze",
    "ltPropCalendarPopoverFreeze",
    "ltPropCalendarHeaderType",
    "ltPropHideCalendarNav",
    "ltPropWeekEventOverlap",
    "ltPropEventGap",
    "ltPropTimelineNmore",
    "ltPropHiddenPopoverWrapper",
    "ltPropCalendarPopoverWrapper",
    "ltPropDragThershold",
    "ltPropContinentalTimeFormat",
    "ltPropyieldNav",
    "ltPropTitleYield",
    "ltPropShowWeekNumber",
    "ltPropPopoverAllowMultiple",
    "ltPropWorkingDayOnly",
    "ltPropSchedulerLabelFormat",
    "ltPropShowCurrentTime",
    "ltPropManipulatedEvent",
    "ltPropLabel",
    "ltPropCurrMonth",
    "ltdisplayCurrDate",
    "ltPropCurDay",
    "ltPropDisplayYear",
    "ltPropWeekStart",
    "openSchedulerPopover",
    "ltPropTimeLine",
    "timeLine",
    "timelineArray",
    "currDay",
    "ltCurrStartDate",
    "ltCurrEndDate",
    "ltPropCurrentDate",
    "labelFF",
    "label",
    "month",
    "montharray",
    "DateArray",
    "days",
    "startDateOfView",
    "hiddenObj",
    "eventArr",
    "todayButton",
    "ltPropSchedulerSearch",
    "ltPropSchedulerSearchYield",
    "selectionType",
    "ltPropSortUser",
    "checkboxArr",
    "selectedUserArr",
    "ltPropSchedulerMenu",
    "ltPropSchedulerMenuYield",
    "ltPropSchedulerMenuContent",
    "ltPropSchedulerMenuUserValue",
    "ltPropSchedulerMenuSystemValue",
    "ltPropSchedulerMenuDescription",
    "ltPropSchedulerMenuId",
    "ltPropSchedulerMenuClass",
    "ltPropSchedulerMenuPosition",
    "ltPropSchedulerMenuWidth",
    "ltPropSchedulerMenuHeight",
    "ltPropSchedulerMenuCallout",
    "ltPropSchedulerMenuFreeze",
    "ltPropSchedulerMenuQueryClass",
    "ltPropSchedulerMenuBindToBody",
    "ltPropSchedulerMenuAnimate",
    "ltPropSchedulerMenuWrapperClass",
    "ltPropSchedulerMenuAria",
    "ltPropSchedulerMenuAriaAttribute",
    "isEventAdded",
    "isResizing",
    "dragDir",
    "currEventTag",
    "breakpoint",
    "ybreakpoint",
    "curPgX",
    "curPgY",
    "quadrant",
    "ltPropEventResize",
    "tempResizeClientY",
    "hiddenUser"
];

if(!window._lyteUiUtils.registeredCustomElements['lyte-scheduler-icon']){
    window._lyteUiUtils.registeredCustomElements['lyte-scheduler-icon'] = true;

    class LyteSchedulerIconCustomElements extends LyteUiComponentComponentRegistry.CustomElements {
        lookups() {
            return [{
                component: LyteUiComponentComponentRegistry
            }];
        }

        constructor() {
            super();
        }

        connectedCallback() {
            var lyteNode = this.$node;
            lyteNode.setAttribute('id','SchedulerNavIcon_'+_lyteSchedulerNavId++);
        }

        _() {
            _;
        }
    }

    LyteSchedulerIconCustomElements.options = {clone : {allCallbacks : false}};

    LyteSchedulerIconCustomElements.register("lyte-scheduler-icon");
}

LyteUiComponentComponentRegistry.registerHelper("lyteUiSchedulerhiddenEvent",function( hiddenObj , date , elem , user, group,time,view){
	if(view == 'day'){
		if(hiddenObj[date]){
			hiddenObj = hiddenObj[date];
			if(hiddenObj[user]){
				hiddenObj = hiddenObj[user];
			}
		}
	}
	var hiddenBtn =  $L(elem).closest('.lyteSchedulerEventMoreBtn')[0];
	if( hiddenObj && hiddenObj[time] ){
		if(hiddenBtn){
			hiddenBtn.classList.remove('lyteSchedulerEventMoreBtnHide');
		}
		return hiddenObj[time].length +' '+ _lyteUiUtils.i18n('more','note');
	}else if( hiddenObj && hiddenObj[date] ){
		var events = hiddenObj[date];
		if(user){
			events = events[user];
		}
		if(events && group){
			if(!events[group]){
				return ;
			}
			var hiddenBtn =  $L(elem).closest('.lyteSchedulerEventMoreBtnHide')[0];
			if(hiddenBtn){
				hiddenBtn.classList.remove('lyteSchedulerEventMoreBtnHide');
			}
			return events[group].length +' '+ _lyteUiUtils.i18n('more','note');
		}else if(events  && (events.length || Object.keys(events).length)){
			var length = events.length || Object.keys(events).length;
			var hiddenBtn =  $L(elem).closest('.lyteSchedulerEventMoreBtnHide')[0];
			if(hiddenBtn){
				hiddenBtn.classList.remove('lyteSchedulerEventMoreBtnHide');
			}
			return length +' '+_lyteUiUtils.i18n('more','note');
		}	
	}else{
		if(hiddenBtn){ 
			hiddenBtn.classList.add('lyteSchedulerEventMoreBtnHide');
		}
	}	
});

LyteUiComponentComponentRegistry.registerHelper("lyteUiSchedulerEvent",function( event ){
	
	if(event && Object.keys(event).length && !event.dummy_id){
		return true;
	}
	return false;
});

LyteUiComponentComponentRegistry.registerHelper('lyteUiSchedulerLabelFormat',function(format,month,date,year){
	
	var D = new Date( year , month, date);
	return $L.moment(D).i18N(format); 
});

LyteUiComponentComponentRegistry.registerHelper('lyteUiSchedulerisSearch',function(array,userid){
	if( !array.length || array.includes(userid)){
		return true;
	}else{
		return false;
	}

})

LyteUiComponentComponentRegistry.registerHelper('lyteUiSchedulerEventCount',function(event,user,hiddenObj,startDateOfView){
	var count = 0;	
	if(Array.isArray(event)){
		event.forEach(function(date){
			if(!date.val){
				return;
			}
			if(date.events && date.events[user.id] ){
				count += date.events[user.id].count;
				
			}
			var formatedDate = date.val.format('DD-MM-YYYY');
			if(hiddenObj && hiddenObj[formatedDate] && hiddenObj[formatedDate][user.id]){
				count += hiddenObj[formatedDate][user.id].count;
			}
		});
	}else{
		if(!event || !Object.keys(event).length){
			return 0;
		}
		if(event[user.id]){
			count += event[user.id].count ;
		}
		if(event.allDay && event.allDay[user.id]){
			count += event.allDay[user.id].count ;
		}
		if(hiddenObj.allDay && hiddenObj.allDay[startDateOfView] && hiddenObj.allDay[startDateOfView][user.id] ){
			count += hiddenObj.allDay[startDateOfView][user.id].count ;
		}
	}	
	return count;
})

LyteUiComponentComponentRegistry.registerHelper('lyteUiSchedulerAllDayCount',function(eventData,hiddenObj,startDateOfView,view,isMultiView){
	var count = 0;
	if(eventData ){
		if(view == 'day' ){
			if(isMultiView && eventData.allDay && eventData.allDay[startDateOfView]){
				var events = eventData.allDay[startDateOfView];
				var keys = Object.keys(events);
				var eventArray = [];
				keys.forEach(function(user){
					eventArray = eventArray.concat(events[user]);
					if(hiddenObj && hiddenObj.allDay && hiddenObj.allDay[startDateOfView] && hiddenObj.allDay[startDateOfView][user]){
						eventArray = eventArray.concat( hiddenObj.allDay[startDateOfView][user] )
					}
				})
			}else if(eventData.allDay && eventData.allDay[startDateOfView]){
				var events = eventData.allDay[startDateOfView];
				if(hiddenObj.allDay && hiddenObj.allDay[startDateOfView]){
					events = events.concat(hiddenObj.allDay[startDateOfView]);
				}	
				var eventArray = events;			
			}
			for(var Iindex = 0;events && Iindex < eventArray.length;Iindex++){
				if(eventArray[Iindex] && Object.keys(eventArray[Iindex]).length && !eventArray[Iindex].dummy_id && !eventArray[Iindex].overdue){
					count++;
				}
			}
		}else if(!isMultiView && view == 'week' && eventData.allDay){			
			for(var index = 0; index < 7; index++){
				var events = eventData.allDay[startDateOfView];
				for(var Iindex = 0;events && Iindex < events.length;Iindex++){
					if(events[Iindex] && Object.keys(events[Iindex]).length && !events[Iindex].dummy_id && !events[Iindex].overdue){
						count++;
					}
				}
				if(hiddenObj && hiddenObj.allDay && hiddenObj.allDay[startDateOfView]){
					count += hiddenObj.allDay[startDateOfView].length;
				}
				startDateOfView = $L.moment(startDateOfView,'DD-MM-YYYY').add(1,'date').format('DD-MM-YYYY');
			}
		}
	}
	return count;
});

LyteUiComponentComponentRegistry.registerHelper('lyteUiSchedulergetUserData',function(userid,userData){
	var User = $u.findWhere(userData,{'id':userid});
	return User; 
});

LyteUiComponentComponentRegistry.registerHelper('lyteSchedulerUiFindWeek',function(date,view){
	var now;
	if(!date){
		return;
	}
	if(typeof date == 'string'){
		now  = $L.moment(date,'DD-MM-YYYY');
	}else{
		now = date;
	}
	now = now.getDObj();
	const onejan = new Date(now.getFullYear(), 0, 1);
	var week = Math.ceil((((now.getTime() - onejan.getTime()) / 86400000) + onejan.getDay() -6  ) / 7);
	if(week == 53){
		var lastdate = new Date( now.getFullYear() , 11 , 31 );
		if(lastdate.getDay() != 6){
			week = 1;
		}			
	}
	if(view && view == 'day'){
		return _lyteUiUtils.i18n('week')+ ' ' +week;
	}
	return _lyteUiUtils.i18n('short.week') + ' ' +week;
});

LyteUiComponentComponentRegistry.registerHelper('lyteUiSchedulerfindUser',function(userArray,userid){
	var user = $u.findWhere(userArray,{id: userid});
	return user;
})

LyteUiComponentComponentRegistry.registerHelper('lyteUiSchedulerIsResize',function(week,event,format,position){
	var lastdate;
	var firstDate;
	if(!event.editable){
		return false;
	}
	lastdate = week[week.length - 1].val;
	firstDate = week[0].val;
	if(position == 'end'){
		var currDate = $L.moment(event.end,format);
	}else{
		var currDate = $L.moment(event.start,format);
	}
	if(lastdate.fromNow(currDate).past && !firstDate.fromNow(currDate).past){
		return true;
	}
	return false;
})

export { LyteSchedulerComponent };

LyteSchedulerComponent.register("lyte-scheduler", {
    hash: "LyteSchedulerComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
