// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/55eff92b4f1bca84367d57b894b048a9.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../../node_modules/@slyte/component/index.js";

/**
 * Displays hovercard with shortcut for originElement
 * @component lyte-shortcut-wrapper
 * @version 3.104.0
 * @methods (from @3.106.0) :  beforeRender, afterRender, onHovercardShow, onHovercardHide, onHovercardBeforeHide, onBeforeHovercardShow
 * @dependency lyte-hovercard
 * components/javascript/lyte-hovercard.js
 * components/styles/lyte-hovercard.css
 * @dependency lyte-hovercard
 * components/javascript/lyte-popover.js
 * components/styles/lyte-popover.css
 * @dependency lyte-wormhole
 * components/javascript/lyte-wormhole.js
 * components/styles/lyte-wormhole.css
 */

class LyteShortcutWrapperComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
        return Object.assign(super.data({
            /**
             * @componentProperty {boolean} ltPropYield=false
            * @version 3.106.0
             */
            'ltPropYield': prop('boolean', {
                'default': false
            }),
            /** 
             * @componentProperty {number} ltPropClass=''
             * @version 3.107.0
             */
            'ltPropClass': prop('string', { 'default': _lyteUiUtils.resolveDefaultValue('lyte-hovercard', 'class', '') }),
            /** 
             * @componentProperty {number} ltPropShowDelay=0
             * @version 3.107.0
             */
            'ltPropShowDelay': prop('number', { 'default': 0 }),
            /** 
             * @componentProperty {number} ltPropHideDelay=0
             * @version 3.107.0
             */
            'ltPropHideDelay': prop('number', { 'default': 0 }),
            /** 
             * @componentProperty {number} ltPropMaxDisplayTime=5000
             * @version 3.107.0
             */
            'ltPropMaxDisplayTime': prop('number', { 'default': 5000 }),
            /**
             * @componentProperty {string} ltPropOriginElem
             * OriginElement Query for hovercard
             * @version 3.104.0
            */
            'ltPropOriginElem': prop('string', { 'default': '' }),
            /** 
             * @componentProperty {string} ltPropPlacement=''
             * @version 3.106.0
            */
            'ltPropPlacement': prop('string', { 'default': '' }),
            /**
             * @componentProperty {string} ltPropPopoverWrapperClass
             * @version 3.106.0
             */
            'ltPropPopoverWrapperClass': prop('string', { 'default': _lyteUiUtils.resolveDefaultValue('lyte-hovercard', 'popoverWrapperClass', '') }),

            /**
             * The keys for which the shortcut has been registered
             * @version 3.104.0
             */
            'shortcutKeys': prop('string', {
                'default': ''
            }),
            /**
             * The formatted keys to display in hovercard
             * @version 3.104.0
             */
            'shortcutKeysToDisplay': prop('string', {
                'default': ''
            }),
            /**
             * The title for the hovercard
             * @version 3.104.0
             */
            'titleToDisplay': prop('string', {
                'default': ''
            }),
            /** 
             * The user-given hovercard entry
             * @version 3.106.0
            */
            'properties': prop('object', { 'default': {} })
        }), arg1);
    }

    static actions(arg1) {
        return Object.assign(super.actions({
            // Functions for event handling
        }), arg1);
    }

    static methods(arg1) {
        return Object.assign(super.methods({
            // Functions which can be used as callback in the component.
            onBeforeRender: function (hovercard) {
                var res = this.callMethod('beforeRender', hovercard);
                if (!res) {
                    return false
                }
                return true
            },
            onAfterRender: function (hovercard) {
                var res = this.callMethod('afterRender', hovercard);
                if (!res) {
                    return false
                }
                return true
            },
            hovercardShow: function (hovercard) {
                var res = this.callMethod('onHovercardShow', hovercard);
                if (!res) {
                    return false
                }
                return true
            },
            hovercardHide: function (hovercard) {
                var res = this.callMethod('onHovercardHide', hovercard);
                if (!res) {
                    return false
                }
                return true
            },
            beforeHovercardShow: function (hovercard) {
                var res = this.callMethod('onBeforeHovercardShow', hovercard);
                if (!res) {
                    return false
                }
                return true
            },
            hovercardBeforeHide: function (hovercard) {
                var res = this.callMethod('onHovercardBeforeHide', hovercard);
                if (!res) {
                    return false
                }
                return true
            }
        }), arg1);
    }

    callMethod(callbackName, hovercard) {
        if (this.getMethods(callbackName)) {
            var res = this.executeMethod(callbackName, this.$node, hovercard);
            if (!res) {
                return false
            }
        }
        return true
    }

    _() {
        _;
    }
}

LyteShortcutWrapperComponent._template = "<template tag-name=\"lyte-shortcut-wrapper\"> <lyte-hovercard lt-prop-auto-show=\"true\" lt-prop-origin-elem=\"{{ltPropOriginElem}}\" lt-prop-popover-wrapper-class=\"{{ltPropPopoverWrapperClass}}\" lt-prop-placement=\"{{ltPropPlacement}}\" lt-prop-class=\"{{ltPropClass}}\" lt-prop-show-delay=\"{{ltPropShowDelay}}\" lt-prop-hide-delay=\"{{ltPropHideDelay}}\" lt-prop-max-display-time=\"{{ltPropMaxDisplayTime}}\" before-render=\"{{method('onBeforeRender')}}\" after-render=\"{{method('onAfterRender')}}\" on-hovercard-show=\"{{method('hovercardShow')}}\" on-hovercard-hide=\"{{method('hovercardHide')}}\" on-before-hovercard-show=\"{{method('beforeHovercardShow')}}\" on-hovercard-before-hide=\"{{method('hovercardBeforeHide')}}\"> <template is=\"registerYield\" yield-name=\"hoverCardYield\"> <lyte-hovercard-content> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropYield}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-yield yield-name=\"hovercard\" key=\"{{shortcutKeys}}\" title=\"{{titleToDisplay}}\" properties=\"{{properties}}\"></lyte-yield> </template><template default=\"\"> <div class=\"\"> {{titleToDisplay}} </div> <div class=\"lyteWrapperContent\"> {{shortcutKeysToDisplay}} </div> </template></template></lyte-hovercard-content> </template> </lyte-hovercard> </template>";;
LyteShortcutWrapperComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"r","p":[1,1],"dN":[{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"i","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"tX","p":[1,1],"cn":"default"},{"t":"tX","p":[3,1],"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[1,0],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0},{"type":"dc","trans":true,"hc":true,"p":[1,0]}];;

LyteShortcutWrapperComponent._observedAttributes = [
    "ltPropYield",
    "ltPropClass",
    "ltPropShowDelay",
    "ltPropHideDelay",
    "ltPropMaxDisplayTime",
    "ltPropOriginElem",
    "ltPropPlacement",
    "ltPropPopoverWrapperClass",
    "shortcutKeys",
    "shortcutKeysToDisplay",
    "titleToDisplay",
    "properties"
];

export { LyteShortcutWrapperComponent };

LyteShortcutWrapperComponent.register("lyte-shortcut-wrapper", {
    hash: "LyteShortcutWrapperComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
