// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/5c78002b0f5dab51dfeced70f287a2f6.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../../../node_modules/@slyte/core/index.js";
import { objectUtils, arrayUtils } from "../../../../../node_modules/@slyte/component/index.js";
import $L from "../../../../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";
import { Component } from "../../../../component.js";

class LyteMultiselectFilterComponent extends Component {
    constructor() {
        super();
    }

    init() {
		if( this.data.ltPropCondition ){
			this.setData( 'ltPropCondition', { type : "multiselect", value : [], isValid : false, selected : [] } );
		}else{
			this.updateInitialConditions();
		}
	}

    updateInitialConditions() {
		this.setData( 'ltPropTempCondition', { type : "multiselect", value : [], isValid : false, selected : [] } );
	}

    data(arg1) {
		var __string = "string",
		__array = "array",
		__object = "object",
		__boolean = "boolean",
		__number = 'number';

		return Object.assign(super.data({
			ltPropCondition : prop( __object ),
			ltPropTempCondition : prop( __object ),
			ltPropSearchCount : prop( __number, { default : 8 } ),
			ltPropReset : prop( __boolean, { default : false } ),
			ltPropNoMatch : prop( __string, { default : window._lyteUiUtils.i18n( 'no.results.found', void 0, 'No Results Found' ) } ),

			ltPropData : prop( __object, { default : {} } ),

			ltPropPicklistOptions : prop( __array, { default : [] } ),
			ltPropRenderContent : prop( __array , { default : [] } ),
			ltPropDynamic : prop( __boolean, { default : true } ),

			picklistOptions : prop( __array ),
			noResult : prop( __boolean ),
			isSearch : prop( __boolean ),
			first : prop( __string, { default : "" } )

		}), arg1);		
	}

    call_valueobs() {
		var __data = this.data,
		value = __data.first.toLowerCase(),
		picklistOptions = __data.ltPropPicklistOptions,
		fn = function( __arr ){
			var table = this.$node.getElementsByTagName( "lyte-table" )[ 0 ];

			table && table.scrollTable( 0, 0 );

			this.setData({
				isSearch : __data.isSearch || __arr.length >= __data.ltPropSearchCount,
				noResult : __arr.length == 0,
				picklistOptions : __arr
			});
		}.bind( this );

		if( __data.ltPropDynamic ){
			var cb = "onPicklistConstruct";

			if( this.getMethods( cb ) ){
				var ret = this.executeMethod( cb, __data.first, __data.ltPropData, __data.picklistOptions );

				if( ret == false ){
					return;
				} else if( ret ){
					if( ret.then ){
					 	return ret.then( fn );
					}
					return fn( ret );
				}
			}
		}
		
		fn( value ? picklistOptions.filter( function( item ){
			var __name = ( item.name || '' ).toLowerCase(),
			__email = ( item.email || '' ).toLowerCase();

			return __name.indexOf( value ) != -1 || __email.indexOf( value ) != -1;

		}) : picklistOptions );
	}

    uncheck(value, to_value) {
		var __data = this.data,
		picklistOptions = __data.picklistOptions,
		index = picklistOptions.findIndex( function( item ){
			return item.value == value;
		});

		if( index != -1 ){
			var cur = picklistOptions[ index ];

			if( cur.checked != to_value ){
				objectUtils( cur, 'add', 'checked', to_value );
			}
		}
	}

    static methods(arg1) {
        return Object.assign(super.methods({
            change : function( item, input ){
                var condition = this.data.ltPropTempCondition,
                arr = condition.value,
                selected = condition.selected
                /*,
                args = [ arr ],
                args1 = [ selected ]*/;

                if( input.checked ){
                    // args.push( 'push', item.value );
                    // args1.push( 'push', item );
                    arrayUtils( arr, 'push', item.value );
                    arrayUtils( selected, 'push', item );
                } else {
                    var __index = arr.indexOf( item.value );
                    // args.push( 'removeAt', __index );
                    // args1.push( 'removeAt', __index );
                    arrayUtils( arr, 'removeAt', __index );
                    arrayUtils( selected, 'removeAt', __index );

                }

                // Lyte.arrayUtils.apply( __Lyte, args );
                // Lyte.arrayUtils.apply( __Lyte, args1 );
                
                objectUtils( condition, 'add', 'isValid', !!arr.length );
                objectUtils( condition, 'add', 'class', condition.isValid ? 'lyteListFilterSelected' : '' );

            },

            scrollEnd : function(){
                var cb = 'fetchMoreData',
                __data = this.data;

                return this.getMethods( cb ) && this.executeMethod( cb, __data.first, __data.ltPropData, __data.picklistOptions );
            }
        }), arg1);
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            init_obs : function(){
                var __data = this.data,
                picklistOptions = __data.ltPropPicklistOptions || [];

                this.setData({
                    isSearch : picklistOptions.length >= __data.ltPropSearchCount,
                    picklistOptions : picklistOptions,
                    noResult : picklistOptions.length == 0
                });

                if( window._lyteUiUtils.lyteUiIsEmpty(__data.ltPropCondition) ){
                    this.setData('ltPropCondition', { type : "multiselect", value : [], isValid : false, selected : [] } );
                }else{
                    this.updateInitialConditions();
                }

                if( __data.ltPropDynamic ){
                    this.call_valueobs();
                }

            }.observes( 'ltPropPicklistOptions' ).on( 'init' ),

            value_obs : function(){
                this.call_valueobs();
            }.observes( 'first' ),

            cond_obs : function( arg ){
                var new_arr = ( ( arg.newValue || {} ).value || [] ).slice(),
                old_arr = ( ( arg.oldValue || {} ).value || [] ),
                __this = this;

                ( ( arg.oldValue || {} ).value || [] ).forEach( function( item ){
                    if( new_arr.indexOf( item ) == -1 ){
                        __this.uncheck( item, false );
                    }
                });

                new_arr.forEach( function( item ){
                    if( old_arr.indexOf( item ) == -1 ){
                        __this.uncheck( item, true );
                    }
                });
            }.observes( 'ltPropCondition' ),

            content_obs : function(){
                this.call_valueobs();
            }.observes('ltPropRenderContent'),

            reset_obs : function( arg ){
                if( arg.newValue ){

                    var __data = this.data,
                    Lc = objectUtils,
                    ns = 'checked',
                    value = __data.ltPropTempCondition.value,
                    __length = value.length - 1;

                    this.setData( arg.item, false );
                    this.setData( 'first', "" );
                    
                    for( var i = __length; i >= 0; i-- ){
                        var __cur = value[ i ];
                        this.uncheck( __cur, false );
                    }

                    this.call_valueobs();
                }
            }.observes( 'ltPropReset' )
        }), arg1);
    }

    _() {
        _;
    }
}

LyteMultiselectFilterComponent._template = "<template tag-name=\"lyte-multiselect-filter\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{isSearch}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-input class=\"lyteFiltersMultiselectSearch\" lt-prop-type=\"search\" lt-prop-appearance=\"box\" lt-prop-placeholder=\"Search\" lt-prop-value=\"{{lbind(first)}}\"></lyte-input> </template></template><div class=\"lyteMultiselectFilterOptions\"> <lyte-table class=\"lyteMultiselectFilterTable\" lt-prop-yield=\"true\" lt-prop-infinite-scroll=\"true\" lt-prop-content-length=\"25\" lt-prop-content=\"{{picklistOptions}}\" scroll-end=\"{{method('scrollEnd')}}\"> <template is=\"registerYield\" yield-name=\"yield\"> <lyte-table-structure> <lyte-tbody> <template items=\"{{ltPropData}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-tr> <lyte-td> <lyte-checkbox lt-prop-label=\"{{item.body.label}}\" lt-prop-value=\"{{item.body.value}}\" lt-prop-name=\"multifilter\" on-changed=\"{{method('change',item.body)}}\" lt-prop-checked=\"{{lbind(item.body.checked)}}\"></lyte-checkbox> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.body.email}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteMultipleEmail\">{{item.body.email}}</div> </template></template></lyte-td> </lyte-tr> </template> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{isSearch}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"{{if(noResult,'lyteMultipleFilterNoResult','lyteMultipleFilterNoResult lyteSearchHidden')}}\">{{ltPropNoMatch}}</div> </template></template></lyte-tbody> </lyte-table-structure> </template> </lyte-table> </div> </template>";;
LyteMultiselectFilterComponent._dynamicNodes = [{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":2,"sibl":[1]},{"t":"a","p":[2,1]},{"t":"r","p":[2,1,1],"dN":[{"t":"a","p":[1,1,1]},{"t":"f","p":[1,1,1],"dN":[{"t":"a","p":[1,1,1]},{"t":"cD","p":[1,1,1],"in":3,"sibl":[2]},{"t":"s","p":[1,1,3],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[1,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1]},{"t":"cD","p":[1,1],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[3,1,0],"hc":true,"trans":true,"in":3,"sibl":[2]},{"t":"s","p":[1,1,3],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1]},{"t":"cD","p":[1,1],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[3,1,0],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"cD","p":[2,1],"in":0},{"type":"dc","trans":true,"hc":true,"p":[2,1,0]}];;

LyteMultiselectFilterComponent._observedAttributes = [
    "ltPropCondition",
    "ltPropTempCondition",
    "ltPropSearchCount",
    "ltPropReset",
    "ltPropNoMatch",
    "ltPropData",
    "ltPropPicklistOptions",
    "ltPropRenderContent",
    "ltPropDynamic",
    "picklistOptions",
    "noResult",
    "isSearch",
    "first"
];

export { LyteMultiselectFilterComponent };

LyteMultiselectFilterComponent.register("lyte-multiselect-filter", {
    hash: "LyteMultiselectFilterComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
