// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/61733e285a2a17acc884c6c6d734d770.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../../../node_modules/@slyte/core/index.js";
import $L from "../../../../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";
import { objectUtils } from "../../../../../node_modules/@slyte/component/index.js";
import { Component } from "../../../../component.js";

class LyteBooleanFilterComponent extends Component {
    constructor() {
        super();
    }

    init() {
		var __data = this.data;
		__data.isSearch = Object.keys( __data.ltPropOptions ).length >= __data.ltPropSearchCount;
		this.setData( 'ltPropCondition', { isValid : false, isNeg : false, value : "", label : "", type : "boolean" } );
	}

    data(arg1) {
		var __string = "string",
		__array = "array",
		__object = "object",
		__boolean = "boolean",
		__number = 'number';

		return Object.assign(super.data({
			ltPropCondition : prop( __object ),
			ltPropPlaceholder : prop( __string, { default : window._lyteUiUtils.i18n( 'none', "listview.filter", "None" ) } ),
			ltPropOptions : prop( __object, { default : {
					is : window._lyteUiUtils.i18n( "selected", void 0, "Selected" ),
					is_not : window._lyteUiUtils.i18n( "not.selected", "listview.filter", "Not selected" )
				} 
			} ),

			ltPropNoMatch : prop( __string, { default : window._lyteUiUtils.i18n( 'no.results.found', void 0, 'No Results Found' ) } ),

			ltPropSearchCount : prop( __number, { default : 8 } ),

			ltPropReset : prop( __boolean, { default : false } ),

			selected : prop( __string, { default : "" } ),
			noResult : prop( __boolean ),

			isSearch : prop( __boolean )
		}), arg1);		
	}

    static methods(arg1) {
        return Object.assign(super.methods({
            search : function( arg ){
                this.setData( 'noResult', arg.length == 0 );
            }
        }), arg1);
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            condn_obs : function( arg ){
                this.setData( 'selected', arg.newValue.value || "" );
            }.observes( 'ltPropCondition' ),

            reset_obs : function( arg ){
                if( arg.newValue ){
                    this.setData( arg.item, false );
                    this.setData( 'selected', "" );
                }
            }.observes( 'ltPropReset' ),

            selected_obs : function( arg ){
                var selected = arg.newValue,
                __data = this.data,
                condition = __data.ltPropTempCondition,
                Lc = objectUtils,
                ns = "lyteTextFilter_";

                $L( this.$node ).addClass( ns + selected ).removeClass( ns + arg.oldValue );

                Lc( condition, 'add', 'isNeg', selected == "is_not" );
                Lc( condition, 'add', 'isValid', !!selected );
                Lc( condition, 'add', 'value', selected );
                Lc( condition, 'add', 'label', __data.ltPropOptions[ selected ] || "" );
                Lc( condition, 'add', 'class', condition.isValid ? 'lyteListFilterSelected' : '' );

            }.observes( 'selected' )
        }), arg1);
    }

    _() {
        _;
    }
}

LyteBooleanFilterComponent._template = "<template tag-name=\"lyte-boolean-filter\"> <lyte-dropdown lt-prop-selected=\"{{lbind(selected)}}\" lt-prop-placeholder=\"{{ltPropPlaceholder}}\"> <template is=\"registerYield\" yield-name=\"yield\"> <lyte-drop-box class=\"lyteBooleanFilterDropdown\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{isSearch}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-search lt-prop-query-selector=\"{ &quot;scope&quot; : &quot;.lyteBooleanFilterDropdown:not(.lyteDropdownHidden)&quot;, &quot;search&quot; : &quot;lyte-drop-item&quot; }\" on-search=\"{{method('search')}}\"></lyte-search> </template></template><lyte-drop-body> <template object=\"{{ltPropOptions}}\" value=\"value\" key=\"key\" is=\"forIn\" _new=\"true\"> <lyte-drop-item data-value=\"{{key}}\">{{value}}</lyte-drop-item> </template> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{isSearch}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"{{if(noResult,'lyteBooleanFilterNoResult','lyteBooleanFilterNoResult lyteSearchHidden')}}\">{{ltPropNoMatch}}</div> </template></template></lyte-drop-body> </lyte-drop-box> </template> </lyte-dropdown> </template>";;
LyteBooleanFilterComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"r","p":[1,1],"dN":[{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":4,"sibl":[3]},{"t":"a","p":[1,2,1]},{"t":"fI","p":[1,2,1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,0]},{"t":"cD","p":[1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":3,"sibl":[2]},{"t":"s","p":[1,2,3],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1]},{"t":"cD","p":[1,2],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[4,3,1,0],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0},{"type":"dc","trans":true,"hc":true,"p":[1,0]}];;

LyteBooleanFilterComponent._observedAttributes = [
    "ltPropCondition",
    "ltPropPlaceholder",
    "ltPropOptions",
    "ltPropNoMatch",
    "ltPropSearchCount",
    "ltPropReset",
    "selected",
    "noResult",
    "isSearch"
];

export { LyteBooleanFilterComponent };

LyteBooleanFilterComponent.register("lyte-boolean-filter", {
    hash: "LyteBooleanFilterComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
