// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/8decb42e69f6b979bb756bb82c8fbfd6.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import "../../../node_modules/@slyte/component/src/directives/lyte-hide.js";
import { prop } from "../../../node_modules/@slyte/core/index.js";
import { Component, LyteUiComponentComponentRegistry } from "../../component.js";

class LyteTreeTableChildComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
			ltPropData : prop('array' , {
				default : []
			}),
			hideTemplate : prop('boolean', {
				default : true
			}),
			indexVar: prop('string', {
				default: ''
			})
		}), arg1);		
	}

    static actions(arg1) {
        return Object.assign(super.actions({
            // Functions for event handling
        }), arg1);
    }

    static methods(arg1) {
        return Object.assign(super.methods({
            // Functions which can be used as callback in the component.
        }), arg1);
    }

    _() {
        _;
    }
}

LyteTreeTableChildComponent._template = "<template tag-name=\"lyte-tree-table-child\"> <template items=\"{{ltPropData}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-yield yield-name=\"treeTableRow\" index-var=\"{{lyteTableTreeIndexHelp(indexVar,index)}}\" level=\"{{lyteTableTreeLevel(indexVar,index)}}\" list-value=\"{{item}}\" index=\"{{index}}\"></lyte-yield> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(item.children,'&amp;&amp;',expHandlers(item.collapsed,'!==',true))}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-tree-table-child @hide-tag=\"true\" lt-prop-data=\"{{item.children}}\" index-var=\"{{lyteTableTreeIndexHelp(indexVar,index)}}\" level=\"{{lyteTableTreeLevel(indexVar,index)}}\"> <template is=\"registerYield\" yield-name=\"treeTableRow\" from-parent=\"\" list-value=\"{{item}}\" index=\"{{index}}\"></template> </lyte-tree-table-child> </template></template></template> </template>";;
LyteTreeTableChildComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"f","p":[1],"dN":[{"t":"a","p":[1]},{"t":"i","p":[1],"in":1,"sibl":[0]},{"t":"s","p":[3],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"trans":true,"in":1,"chld":[0],"cn":"lc_id_0"},{"t":"a","p":[1,1],"cn":"lc_id_0"},{"t":"r","p":[1,1],"dN":[],"in":0,"cn":"lc_id_0"},{"t":"cD","p":[1],"_igTs":true,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[1],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[1,0],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;
LyteTreeTableChildComponent._observedAttributes = ["ltPropData", "hideTemplate", "indexVar"];

LyteUiComponentComponentRegistry.registerHelper('lyteTableTreeLevel' , function(parentIndex , index){
	return ((parentIndex +" "+ index).trim()).split(' ').length
})

LyteUiComponentComponentRegistry.registerHelper('lyteTableTreeIndexHelp' , function(parentIndex , index){
	return ((parentIndex +" "+ index).trim())
})

export { LyteTreeTableChildComponent };
LyteTreeTableChildComponent.register("lyte-tree-table-child", {
    hash: "LyteTreeTableChildComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
