// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/ac919d7a745fe073a2ee0da048526ab1.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../component.js";

/**
 * Renders a keyboard-navigator
 * @component lyte-keyboard-navigator
 * @version 3.0.0
 * @utility click, focus, blur
 * @dependencies lyte-shortcut
 * /plugins/lyte-keyboardNavigator.js
 */
class LyteKeyboardNavigatorComponent extends Component {
 constructor() {
  super();
 }

 data(arg1) {
     return Object.assign(super.data({
         /**
          * @componentProperty {string} ltPropParent
          * @version 3.0.0
          * @default undefined
          */
         "ltPropParent": prop("string", { "default": undefined }),
         /**
          * @componentProperty {string} ltPropHighlighted
          * @version 3.0.0
          * @default undefined
          */
         "ltPropHighlighted": prop("string", { "default": undefined }),
         /**
          * @componentProperty {string} ltPropSelected
          * @version 3.0.0
          * @default undefined
          */
         "ltPropSelected": prop("string", { "default": undefined }),
         /**
          * @componentProperty {string} ltPropChildren
          * @version 3.0.0
          * @default undefined
          */
         "ltPropChildren": prop("string", { "default": undefined }),
         /**
          * @componentProperty {string} ltPropSkipElements
          * @version 3.0.0
          * @default undefined
          */
         "ltPropSkipElements": prop("string", { "default": undefined }),
         /**
          * @componentProperty {boolean} ltPropIfCycle
          * @version 3.0.0
          * @default false
          */
         "ltPropIfCycle": prop("boolean", { "default": false }),
         /**
          * @componentProperty {array} ltPropOptions
          * @version 3.0.0
          * @default []
          */
         "ltPropOptions": prop("array", {"default": undefined}),
         /**
          * @componentProperty {string} ltPropOrientation
          * @version 3.0.0
          * @default vertical
          */
         "ltPropOrientation": prop("string", { "default": "vertical" }),
         /**
          * @componentProperty {string} ltPropScope
          * @version 3.0.0
          * @default undefined
          */
         "ltPropScope": prop("string", { "default": undefined }),
         /**
          * @componentProperty {boolean} ltPropTriggerClick
          * @version 3.0.0
          * @default false
          */

         "ltPropChangeScope": prop("boolean", { "default": false }),
         /**
          * @componentProperty {boolean} ltPropTriggerClick
          * @version 3.0.0
          * @default false
          */

         "ltPropTriggerClick": prop("boolean", { "default": false }),
         /**
          * @componentProperty {string} customDropdownSel
          * @version 3.0.0
          * @default ""
          */
         "customDropdownSel": prop("string", { "default": "" }),
         /**
          * @componentProperty {string} customDropdownClass
          * @version 3.0.0
          * @default ""
          */
         "customDropdownClass": prop("string", { "default": "" }),
         /**
          * @componentProperty {string} dropdown
          * @version 3.0.0
          * @default ""
          */
         "dropdown":  prop("string", { "default": "" }),
         /**
          * @componentProperty {string} ltPropAriaExpand
          * @version 3.0.0
          * @default ""
          */
         "ltPropAriaExpand":  prop("string", { "default": undefined }),

         "ltPropIgnoreMouseMove":  prop("boolean", { "default": false }),

         "ltPropBaseScope":  prop("string", { "default": '' }),

         "parentObj": prop("object")
     }), arg1);		
 }

 didConnect() {
    this.$node.reset = function () {
		var parent = this.getData('ltPropParent');
        $L(parent).keyboardNavigator('reset');
    }
}

 didDestroy() {
     var parent = this.getData('parentObj');
     $L(parent).keyboardNavigator('destroy');

     this.setData('parentObj',null);

     delete this.$node.reset;
 }

 static actions(arg1) {
  return Object.assign(super.actions({
  }), arg1);
 }

 static methods(arg1) {
  return Object.assign(super.methods({
  }), arg1);
 }

 static observers(arg1) {
  return Object.assign(super.observers({
   parentObserver: function (changeObj) { 

    if ( changeObj && changeObj.oldValue ) {
        var oldParent = changeObj.oldValue;
        $L(oldParent).keyboardNavigator('destroy');
        this.setData('parentObj',null);
    }
        
    var parent = this.getData('ltPropParent');

    this.setData('parentObj', $L(parent));

    if (document.activeElement != parent) {
        $L(parent).focus();
    }

    $L(parent).keyboardNavigator({
            comp: this,
            focusableElement: this.getData('ltPropParent'),
            highlightValue: this.getData('ltPropHighlighted'),
            child: this.getData('ltPropChildren'),
            selectedClass: this.getData('ltPropSelected'),
            skipElements: this.getData('ltPropSkipElements'),
            ifCycle: this.getData('ltPropCycle'),
            options: this.getData('ltPropOptions'),
            orientation: this.getData('ltPropOrientation'),
            scope: this.getData('ltPropScope'),
            triggerClick: this.getData('ltPropTriggerClick'),
            customDropdownSel: this.getData('customDropdownSel'),
            customDropdownClass: this.getData('customDropdownClass'),
            dropdown: this.getData('dropdown'),
            onBeforeHighlight: this.getMethods('onBeforeHighlight'),
            onAfterHighlight: this.getMethods('onAfterHighlight'),
            changeScope: this.getData('ltPropChangeScope'),
            ariaExpand: this.getData('ltPropAriaExpand'),
            ignoreMouseMove: this.getData('ltPropIgnoreMouseMove'),
            baseScope: this.getData('ltPropBaseScope')
        }
    );

   }.observes('ltPropParent').on('didConnect')
  }), arg1);
 }

 _() {
  _;
 }
}

LyteKeyboardNavigatorComponent._template = "<template tag-name=\"lyte-keyboard-navigator\"> <lyte-yield yield-name=\"keyboard-navigator-content\"></lyte-yield> </template>";;
LyteKeyboardNavigatorComponent._dynamicNodes = [{"t":"i","p":[1],"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;

LyteKeyboardNavigatorComponent._observedAttributes = [
 "ltPropParent",
 "ltPropHighlighted",
 "ltPropSelected",
 "ltPropChildren",
 "ltPropSkipElements",
 "ltPropIfCycle",
 "ltPropOptions",
 "ltPropOrientation",
 "ltPropScope",
 "ltPropChangeScope",
 "ltPropTriggerClick",
 "customDropdownSel",
 "customDropdownClass",
 "dropdown",
 "ltPropAriaExpand",
 "ltPropIgnoreMouseMove",
 "ltPropBaseScope",
 "parentObj"
];

export { LyteKeyboardNavigatorComponent };
LyteKeyboardNavigatorComponent.register("lyte-keyboard-navigator", {
 hash: "LyteKeyboardNavigatorComponent_2",
 refHash: "C_lyte-ui-component 2_app_0"
});
