// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/b61e3199e77cae21a36061b4a5bbeafd.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../../node_modules/@slyte/core/index.js";
import { arrayUtils } from "../../../../node_modules/@slyte/component/index.js";
import { Component } from "../../../component.js";

class LyteNestedTableComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
			ltPropHeader : prop( 'array', { default : [] } ),
			ltPropContent : prop( 'array', { default : [] } ),
			ltPropGroupTitle : prop( 'string', { default : "" } ),
			ltPropCount : prop( 'number' ),

			isGroup : prop( 'boolean', { default : true } ),
			renderContent : prop( 'array', { default : [] } ),
			topLevel : prop( 'boolean', { default : false } ),
			intersection : prop( "object" ),
			ariaLabel : prop( 'string' ),
			item : prop( 'object' ),

			fixedColumnStatus : prop( 'object', { default : {}, watch : true } )
		}), arg1);		
	}

    check_fixed_column() {
		var groups = $L( this.$node ).children( 'thead,tbody' ),
		status = this.data.fixedColumnStatus,
		cls_name = 'lyteFixedColumn',
		fixed_cls = 'lyteTableFixed';
		
		Array.from( groups ).forEach( function( tbody ){
			Array.from( tbody.children ).forEach( function( row ){
				Array.from( row.children ).forEach( function( cell, index ){
					var __cur = status[ index ],
					fn = "removeClass",
					__style = {
						left : "",
						right : ""
					};
					
					if( __cur ){
						var inner_cls = fn;
	
						fn = "addClass";
						$L.extend( __style, __cur );
	
						if(  status[ 'fixed_' + index ] ){
							inner_cls = fn;
						}
	
						$L( cell )[ inner_cls ]( fixed_cls );
					}
	
					$L( cell )[ fn ]( cls_name ).css( __style );
				});
			});
		});
		
	}

    static actions(arg1) {
        return Object.assign(super.actions({

            rowclick : function( evt, __this, index ){
                var row_data = this.data.renderContent[ index ].data,
                cell = evt.target.closest( 'td' ),
                cell_index = Array.from( cell.parentNode.children ).indexOf( cell );

                this.throwEvent( 'nestedRowClick', row_data, __this, evt, this.data.ltPropHeader[ cell_index ].data );
                return false;
            },

            toggle : function( evt ){
                if( evt ){
                    if( /^32|13$/.test( evt.which ) ){
                        evt.preventDefault();
                    } else {
                        return;
                    }
                }
                this.toggle();
            }
        }), arg1);
    }

    init() {
		window.requestAnimationFrame( function(){
			window.requestAnimationFrame( function(){
				var cells,
					col_grps = this.$node.querySelector( 'colgroup' ).children;
				if( this.data.topLevel ){
					cells = this.$node.closest( 'lyte-table-structure' ).querySelector( 'lyte-thead' ).children[ 0 ];
				} else {
					cells = this.$node.parentNode.closest( 'lyte-nested-table' ).querySelector( 'colgroup' );
				}
				
				Array.from( cells.children ).forEach( function( item, index ){
					var __col = col_grps[ index ],
						__width = __col.style.width;
					if(__width.indexOf( 'px' ) == -1 ){
						__col.style.width = item.style.width;
					}
				} )
			}.bind( this ));
		}.bind( this ));
	}

    didConnect() {
		if( this.data.topLevel ){
			// var scroll_elem = this.$node.closest( '.lyteTableScroll' );
			this.setData( 'intersection', { obs : new IntersectionObserver( this.intersection.bind( this ), { threshold : [ 0.01 ], root : document } ) } );
		} else if( this.data.intersection ){
			this.add_to_inter();
			this.update_aria();
		}

		this.$node.toggle = this.toggle.bind( this );
	}

    update_aria(ns) {
		this.setData( 'ariaLabel', this.data.ltPropGroupTitle +  " Row group " + ( ns || "expanded" ) );
	}

    toggle() {
		var cls_name = "lyteListviewNestedTableClosed",
		anime_class = "lyteListviewNestedAnimationHappening",
		__this = this,
		$node = $L( this.$node ).parent(),
		height = __this.$node.offsetHeight,
		thead = __this.$node.querySelector( "thead" ),
		tbody = thead.nextElementSibling;

		if( $node.hasClass( anime_class ) ){
			return;
		}

		if( $node.hasClass( cls_name ) ){
			$L( tbody ).css( "display", "" );
			$node.css({
				height : height + 'px'
			}).removeClass( cls_name ).addClass( anime_class ).on( 'transitionend', function(){
				$node.removeClass( anime_class ).off( 'transitionend' ).css( "height", "" );
				__this.update_aria();
			});

			window.requestAnimationFrame( function(){
				window.requestAnimationFrame( function(){
					var new_hgt = __this.$node.offsetHeight;
					$node.css({
						height : new_hgt + 'px'
					} );

					if( new_hgt == height ){
						$node.trigger( 'transitionend' );
					}
				});
			});
		} else {
			var th_height = thead.offsetHeight;

			$node.css({
				height : height + 'px'
			}).addClass( cls_name + " " + anime_class ).on( 'transitionend', function(){ 
				$node.removeClass( anime_class ).off( 'transitionend' ).css( "height", "" );
				$L( tbody ).css( "display", "none" );
				__this.update_aria( 'collapsed' );
			});

			window.requestAnimationFrame( function(){
				window.requestAnimationFrame( function(){
					$node.css({
						height : th_height + 'px'
					} );

					if( th_height == height ){
						$node.trigger( 'transitionend' );
					}
				} );
			} );
		}
	}

    didDestroy() {
		var obs = this.data.intersection,
		span = $L( this.$node.children ).get( -1 );
		obs.obs.unobserve( span );

		this.data.intersection = void 0;
	}

    intersection(intersections) {
		intersections.forEach( function( item ){
			if( item.intersectionRatio ){
				var target = item.target,
				table = target.parentNode,
				root_bound = item.rootBounds,
				__this = table.component;

				if( !__this.add_more_data() ){
					__this.check_nested_end( root_bound, target );
				}
			}
		} ); 
	}

    check_nested_end(root_bound, target) {
		var __this = this;
		$L.fastdom.measure( function(){
			var bcr = target.getBoundingClientRect(),
			is_vert_outside = root_bound.top > bcr.bottom || root_bound.bottom < bcr.top,
			is_hori_outside = root_bound.left > bcr.right || root_bound.right < bcr.left;

			if( !( is_vert_outside || is_hori_outside ) ){
				$L.fastdom.mutate( function(){
					if( !__this.add_more_data() ){
						__this.check_nested_end( root_bound, target );
					}
				} )
			}
		} )
	}

    add_to_inter() { 
		var  obs = this.data.intersection,
		span = $L( this.$node.children ).get( -1 );
		obs.obs.observe( span );
	}

    add_more_data() {
		var __data = this.data,
		renderContent = __data.renderContent,
		content = __data.ltPropContent || [],
		__len = renderContent.length;

		if( __len == content.length ){
			return true;
		}

		arrayUtils( renderContent, 'push', content.slice( __len, __len + 10 ) );
		!__data.isGroup && this.check_fixed_column();
	}

    static observers(arg1) {
        return Object.assign(super.observers({
            fixed_obs : function( arg ){

                // if( this.data.isGroup ){
                // 	return;
                // }

                this.check_fixed_column();
            }.observes( 'fixedColumnStatus.*' ),

            obs : function(){
                this.add_to_inter();
            }.observes( 'intersection' ),

            content_obs : function( arg ){
                this.setData( 'renderContent', [] );
                this.add_more_data();
            }.observes( 'ltPropContent' )
        }), arg1);
    }

    _() {
        _;
    }
}

LyteNestedTableComponent._template = "<template tag-name=\"lyte-nested-table\"> <colgroup> <template items=\"{{ltPropHeader}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"><col style=\"{{listStyle(item.data,item.data.width)}}\"></template> </colgroup> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropGroupTitle}}\" is=\"case\" lc-id=\"lc_id_0\" depth=\"1\"><table><thead> <tr> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropHeader}}\" item=\"cell\" index=\"index\" _new=\"true\" depth=\"3\"></td> </tr> </thead></table></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{isGroup}}\" is=\"case\" lc-id=\"lc_id_0\" depth=\"1\"><table><tbody> <tr is=\"for\" lyte-for=\"true\" items=\"{{renderContent}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody></table></template><template default=\"\" depth=\"1\"><table><tbody> <tr is=\"for\" lyte-for=\"true\" items=\"{{renderContent}}\" item=\"item\" index=\"index\" _new=\"true\" depth=\"2\"></tr> </tbody></table></template></template><span class=\"lyteListviewIntersectionspan\"></span> </template>";;
LyteNestedTableComponent._dynamicNodes = [{"t":"a","p":[1,1]},{"t":"f","p":[1,1],"dN":[{"t":"a","p":[0],"a":{"style":{"name":"style","helperInfo":{"name":"listStyle","args":["item.data","item.data.width"]}}}}],"in":2,"sibl":[1]},{"t":"s","p":[3],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0,1,1],"cn":"lc_id_0"},{"t":"f","p":[0,1,1],"dN":[{"t":"a","p":[0]},{"t":"s","p":[0,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"i","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[0,1],"cn":"default"},{"t":"tX","p":[0,3,0],"cn":"default"},{"t":"tX","p":[0,7,0],"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr><th class=\"{{cell.data.class}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{index}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-yield yield-name=\"{{cell.data.groupYield}}\" row-data=\"{{item}}\" cell-data=\"{{cell.data}}\" cell-index=\"{{index}}\"></lyte-yield></template><template default=\"\"><div class=\"lyteNestedTableMainThWrapper\"> <span tabindex=\"0\" role=\"button\" aria-live=\"polite\" aria-label=\"{{ariaLabel}}\" onkeydown=\"{{action('toggle',event)}}\" onclick=\"{{action('toggle')}}\" class=\"lyteListviewNestedTableHandler\"></span> <span class=\"lyteListviewNestedTableHeading lyteTextEllipsisNode\">{{ltPropGroupTitle}}</span> <span class=\"lyteListviewNestedTableCount lyteTextEllipsisNode lyteBracket\">(</span> <span class=\"lyteListviewNestedTableCount lyteTextEllipsisNode\">{{ltPropCount}}</span> <span class=\"lyteListviewNestedTableCount lyteTextEllipsisNode lyteBracket\">)</span> </div></template></template></th></tr></tbody></table></template>","tagName":"TR","in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"s","p":[4],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0,1],"cn":"lc_id_0"},{"t":"f","p":[0,1],"dN":[{"t":"a","p":[0,1]},{"t":"a","p":[0,1,1,1],"a":{"style":{"name":"style","helperInfo":{"name":"concat","args":["'--lyte-listview-nested-level:'","item.level"]}}}},{"t":"r","p":[0,1,1,1,1],"dN":[],"in":2,"sibl":[1]},{"t":"a","p":[0,1,1,1,3]},{"t":"f","p":[0,1,1,1,3],"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"r","p":[0],"dN":[],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"sibl":[0]},{"t":"s","p":[2],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"r","p":[0],"dN":[],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"]}],"in":1,"sibl":[0]},{"t":"cD","p":[0,1,1,1],"in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody><tr> <td colspan=\"{{ltPropHeader.length}}\" class=\"lyteListviewNestedGroupHolder\"> <div class=\"lyteListviewTableHolder\"> <lyte-nested-table item=\"{{item}}\" intersection=\"{{intersection}}\" lt-prop-count=\"{{item.count}}\" lt-prop-group-title=\"{{item.name}}\" lt-prop-column-title=\"{{item.column}}\" lt-prop-header=\"{{ltPropHeader}}\" lt-prop-content=\"{{item.rows}}\" is-group=\"{{item.isGroup}}\" fixed-column-status=\"{{fixedColumnStatus}}\" style=\"--lyte-listview-nested-level:{{item.level}}\"> <template is=\"registerYield\" yield-name=\"yield\" from-parent=\"true\"></template> <template items=\"{{ltPropHeader}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.data.yield}}\" is=\"case\" lc-id=\"lc_id_0\"><template is=\"registerYield\" yield-name=\"{{item.data.yield}}\" from-parent=\"true\"></template></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.data.groupYield}}\" is=\"case\" lc-id=\"lc_id_0\"><template is=\"registerYield\" yield-name=\"{{item.data.groupYield}}\" from-parent=\"true\"></template></template></template></template> </lyte-nested-table> </div> </td> </tr></tbody></table></template>","tagName":"TBODY","in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[0,1],"cn":"default"},{"t":"f","p":[0,1],"dN":[{"t":"a","p":[0]},{"t":"a","p":[0,1]},{"t":"f","p":[0,1],"dN":[{"t":"a","p":[0,1]},{"t":"i","p":[0,1],"in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"3\"><table><tbody><tr><td> <lyte-yield yield-name=\"{{expHandlers(cell.data.yield,'||',&quot;yield&quot;)}}\" row-data=\"{{item.data}}\" cell-data=\"{{cell.data}}\" actual-index=\"{{index}}\" cell-index=\"{{cellIndex}}\" row-index=\"{{index}}\"></lyte-yield> </td></tr></tbody></table></template>","tagName":"TR","in":0}],"dc":[0],"hc":true,"trans":true,"actualTemplate":"<template is=\"for\" depth=\"2\"><table><tbody><tr onclick=\"{{action('rowclick',event,this,index)}}\"> <td is=\"for\" lyte-for=\"true\" items=\"{{ltPropHeader}}\" item=\"cell\" index=\"cellIndex\" _new=\"true\" depth=\"3\"></td> </tr></tbody></table></template>","tagName":"TBODY","in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[1,0]}];;

LyteNestedTableComponent._observedAttributes = [
    "ltPropHeader",
    "ltPropContent",
    "ltPropGroupTitle",
    "ltPropCount",
    "isGroup",
    "renderContent",
    "topLevel",
    "intersection",
    "ariaLabel",
    "item",
    "fixedColumnStatus"
];

export { LyteNestedTableComponent };

LyteNestedTableComponent.register("lyte-nested-table", {
    hash: "LyteNestedTableComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
