// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/c080c974661e80e9c55ef11ed3dd1723.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../component.js";
import $L from "../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";

/**
 * Lyte number is used to get number input from user
 * @component lyte-number
 * @version 2.2.3
 * @utility focus,blur,click,select
 * @methods beforeRender,afterRender,onValueChange,onBeforePaste,onFocus,onBlur
 */


// todo ==> prefix / suffix add, thousand separator

class LyteNumberComponent extends Component {
    constructor() {
        super();
        this._lyteUtilFunctions = [ "focus", "blur", "click", "select" ];
    }

    didConnect() {
		[ 'focus', 'blur', 'click', 'select' ].forEach( function( item ){
	        this.$node[ item ] = function( arg ){
	            this.getInput()[item]( arg );
	        }.bind( this )      
	    }.bind( this ) ) 
		if( this.data.ltPropAutofocus ){
			this.$node.focus();
		}
		this.getMethods( 'afterRender' ) && this.executeMethod( 'afterRender', this.$node );
	}

    init() {
		this.getMethods( 'beforeRender' ) && this.executeMethod( 'beforeRender', this.$node );
	}

    didDestroy() {
		clearTimeout( this._time );
	}

    data(arg1) {
		var default_values = _lyteUiUtils.getDefault( 'lyte-number' );
		return Object.assign(super.data({
			/**
			 * @componentProperty {string} ltPropValue=''
			 * @version 2.2.3
			 */

			ltPropValue : prop( 'string', { default : default_values.string || '' } ),
			/**
			 * @componentProperty {number} ltPropMaxlength
			 * @minValue 0
			 * @version 2.2.3
			 */
			ltPropMaxlength : prop( 'number', { default : default_values.maxlength } ),
			/**
			 * @componentProperty {string} ltPropName
			 * @version 2.2.3
			 */
			ltPropName : prop( 'string', { default : default_values.name } ),
			/**
			 * @componentProperty {string} ltPropClass
			 * @version 2.2.3
			 */
			ltPropClass : prop( 'string', { default : default_values.class } ),
			/**
			 * @componentProperty {string} ltPropId
			 * @version 2.2.3
			 */
			ltPropId : prop( 'string', { default : default_values.id } ),
			/**
			 * @componentProperty {string} ltPropPlaceholder=''
			 * @version 2.2.3
			 */
			ltPropPlaceholder:prop('string',{default : default_values.placeholder || ''}),
			/**
			 * @componentProperty {boolean} ltPropIgnoreSymbols=false
			 * @version 2.2.3
			 */
			ltPropIgnoreSymbols : prop( 'boolean', { default : default_values.ignoreSymbols || false } ),
			/**
			 * @componentProperty {number} ltPropStep=1
			 * @version 2.2.3
			 */
			ltPropStep : prop( 'number', { default : default_values.step || 1 } ),
			/**
			 * @componentProperty {boolean} ltPropInverse=false
			 * @version 2.2.3
			 */
			ltPropInverse : prop( 'boolean', { default : default_values.inverse || false } ),
			/**
			 * @componentProperty {boolean} ltPropWheel=true
			 * @version 2.2.3
			 */
			ltPropWheel : prop( 'boolean', { default : default_values.wheel || true } ),
			/**
			 * @componentProperty {number} ltPropMax
			 * @version 2.2.3
			 */
			ltPropMax : prop( 'number', { default : default_values.max } ),
			/**
			 * @componentProperty {number} ltPropMin
			 * @version 2.2.3
			 */
			ltPropMin : prop( 'number', { default : default_values.min } ),
			/**
			 * @componentProperty {boolean} ltPropAutofocus=false
			 * @version 2.2.3
			 */
			ltPropAutofocus : prop( 'boolean', { default : default_values.autofocus || false } ),
			/**
			 * @componentProperty {boolean} ltPropDisabled=false
			 * @version 2.2.3
			 */
			ltPropDisabled : prop( 'boolean', { default : default_values.disabled || false } ),
			/**
			 * @componentProperty {boolean} ltPropReadonly=false
			 * @version 2.2.3
			 */
			ltPropReadonly : prop( 'boolean', { default : default_values.readonly || false } ),
			/**
			 * @componentProperty {string} ltPropStyle=''
			 * @version 2.2.3
			 */
			ltPropStyle : prop( 'string', { default : default_values.style || '' } ),
			/**
			 * @componentProperty {string} ltPropInputTitle=''
			 * @version 2.2.3
			 */
			ltPropInputTitle : prop( 'string', { default : default_values.inputTitle } ),
			/**
			 * @componentProperty {string} ltPropPattern='.+'
			 * @version 2.2.3
			 */
			ltPropPattern : prop( 'string', { default : default_values.pattern || '.+' } ),
			/**
			 * @componentProperty {number} ltPropUpdateDelay=250
			 * @version 2.2.3
			 */
			ltPropUpdateDelay : prop( 'number', { default : default_values.updateDelay == "undefined" ? void 0 : ( default_values.updateDelay == void 0 ? 250 : default_values.updateDelay ) } ),
			/**
			 * @componentProperty {string} ltPropLabel=''
			 * @version 2.2.3
			 */
			ltPropLabel : prop( 'string', { default : default_values.label || '' } ),
			/**
			 * @componentProperty {on | off} ltPropAutocomplete=off
			 * @version 2.2.3
			 */
			ltPropAutocomplete : prop( 'string', { default : default_values.autocomplete || 'off' } ),
			/**
			 * @componentProperty {box | flat} ltPropAppearance=box
			 * @version 2.2.3
			 */
			ltPropAppearance : prop( 'string', { default : default_values.appearance || 'box' } ),
			/**
			 * @componentProperty {vertical | horizontal} ltPropDirection=vertical
			 * @version 2.2.3
			 */
			ltPropDirection : prop( 'string', { default : default_values.direction || 'vertical' } ),
			/**
			 * @componentProperty {boolean} ltPropValidateOnInput=false
			 * @version 2.2.3
			 */
			ltPropValidateOnInput : prop( 'boolean', { default : default_values.validateOnInput || false } ),
			/**
			 * @componentProperty {boolean} ltPropAutoUpdate=true
			 * @version 2.2.3
			 */
			ltPropAutoUpdate : prop( 'boolean', { default : default_values.autoUpdate == false ? false : true } ),
			/**
			 * @componentProperty {string} ltPropDecimal=.
			 * @version 2.2.8
			 */
			ltPropDecimal : prop( 'string', { default : default_values.decimal || '.' } ),
			/**
			 * @componentProperty {boolean} ltPropIncrement=true
			 * @version 2.2.9
			 */
			ltPropIncrement : prop( 'boolean', { default : default_values.increment == false ? false : true } ),

			// aria
			/**
			 * @componentProperty {boolean} ltPropAria=false
			 * @version 3.1.0
			 */
            ltPropAria : prop( 'boolean', { default : default_values.aria || false } ),
			/**
			 * @componentProperty {object} ltPropAriaAttributes
			 * @default {}
			 * @version 3.1.0
			 */            
            ltPropAriaAttributes : prop( 'object', { default : default_values.ariaAttributes || {}, watch : true } ),

            /**
			 * @componentProperty {object} ltPropRemoveAtCursor=false
			 * @version 2.2.15
			 */    
            ltPropRemoveAtCursor : prop( 'boolean', { default : default_values.removeAtCursor || false } ),

			//test
			/**
			 * @componentProperty {object} ltPropFireOnInit=false
			 * @version 2.2.3
			 */    
			ltPropFireOnInit : prop( 'boolean', { default : default_values.fireOnInit || false } ),
			/**
			 * @componentProperty {object} ltPropValidateOnEmpty=true
			 * @version 2.2.3
			 */    
			ltPropValidateOnEmpty : prop( 'boolean', { default : default_values.validateOnEmpty == false ? false :true } ),
			/**
			 * @componentProperty {string} ltPropRestrict
			 * @version 3.52.0
			 */ 
			ltPropRestrict : prop( 'string', { default : default_values.restrict } ),
			/**
			 * @componentProperty {string} ltPropInputWrapperClass
			 * @version 3.57.0
			 */ 
			ltPropInputWrapperClass : prop( 'string', { default : default_values.wrapperClass || "" } ),
			/**
			 * @componentProperty {string} ltPropTabIndex="0"
			 * @version 3.80.0
			 */ 
			ltPropTabIndex : prop( 'string', { default : default_values.tabIndex || "0" } ),
			/**
			 * @componentProperty {string} ltPropValidation=""
			 * @version 3.86.0
			 */ 
			ltPropValidation : prop( 'string', { default : default_values.validation || "" } ),
			/**
			 * @componentProperty {boolean} ltPropMandatory=false
			 * @version 3.86.0
			 */ 
			ltPropMandatory : prop( 'boolean',  { default : default_values.mandatory || false } ),
			/**
			 * @componentProperty {boolean} ltPropTriggerValidation=false
			 * @version 3.86.0
			 */ 
			ltPropTriggerValidation : prop( 'boolean', { default : default_values.triggerValidation || false } ),
			/**
			 * @typedef errorDef
			 * @property {string} minmax
			 * @property {string} mandatory
			 * @property {string} maxlength
			 * @property {string} pattern
			 */

			/**
			 * @componentProperty {errorDef} ltPropErrorMessage="{}"
			 * @version 3.86.0
			 */ 
			ltPropErrorMessage : prop( 'object', { default : default_values.errorMessage || {} } ),
			/**
			 * @componentProperty {string} ltPropDefaultErrorMessage="Invalid input"
			 * @version 3.80.0
			 */ 
			ltPropDefaultErrorMessage : prop( 'string', { default : default_values.defaultErrorMessage || "Invalid input" } ),

			// ltPropAutoCorrectDuration : Lyte.attr( 'number', { default : void 0 } ),
			/**
			 * @componentProperty {boolean} ltPropError=false
			 * @version 3.80.0
			 */ 
			ltPropError : prop( 'boolean', { default : default_values.error || false } ),
			/**
			 * @componentProperty {boolean} ltPropClearValidation=false
			 * @version 3.80.0
			 */ 
			ltPropClearValidation : prop( 'boolean', { default : default_values.clearValidation || false } ),
			/**
			 * @componentProperty {string} ltPropDataTabindex="0"
			 * @version 3.91.1
			 */ 
			ltPropDataTabindex : prop( 'string', { default : default_values.dataTabindex || "0" } ),
			/**
			 * @componentProperty {boolean} ltPropControls=false
			 * @version 3.91.1
			 */ 
			ltPropControls : prop( 'boolean', { default : default_values.controls || false } ),
			/**
			 * @componentProperty {boolean} ltPropCyclic=false
			 * @version 3.102.0
			 */ 
			ltPropCyclic : prop( 'boolean', { default : default_values.cyclic || false } ),
			ltPropUnit : prop('string', { default : default_values.unit || undefined } ),
			ltPropDigits : prop( 'number', { default : default_values.digits || undefined } ),
			ltPropSuffix : prop('boolean', { default : default_values.suffix || false } ),
			ltPropControlsYield : prop('boolean', { default : default_values.controlsYield || false } ),
			ltPropControlsClass : prop('string', { default : default_values.controlsClass || "lyteNumberControls" } ),
			ltPropYield : prop('boolean', { default : default_values.yield || false } ),
			ltPropControlsAriaLabel : prop('object', { default : default_values.counterAria || {} } ),

			/**
			 * @componentProperty {boolean} ltPropPreventObserver=false
			 * @version 3.106.0
			 */

			ltPropPreventObserver : prop( 'boolean', { default : default_values.preventObserver || false } ),

			errorMessage : prop( 'string', { default : "" } ),
			upArrowClass : prop('string', { default : "" } ),
			downArrowClass : prop('string', { default : "" } )
		}), arg1);		
	}

    getDecimal(value, revert) {
		var decimal = this.data.ltPropDecimal;
		return decimal != '.' ? (revert ? value.replace( '.', decimal ) : value.replace( decimal, '.' )) : value;
	}

    increment(_this, isIncrement) {
		var value = _this.value || '0',
		newVal = ( Number( this.getDecimal( value ) ) + this.data.ltPropStep * ( isIncrement ? 1 : -1 ) ).toString(),
		cpyVal = newVal,
		data = this.data,
		digits = data.ltPropDigits,
		roundToFixed = function( value ){
			return parseFloat( Number(value).toFixed( digits )).toString();
		};

		if( digits != undefined ){
			newVal = cpyVal = roundToFixed( newVal );
		}

		newVal = cpyVal = this.getDecimal( newVal, true );

		if( newVal == Infinity || newVal == -Infinity ){
			return;
		}

		var validation = data.ltPropValidation;

		if( !validation ){
			newVal = this.maxLenValidation( newVal );
			newVal = this.maxMinCheck( newVal, true );
		}

		if (data.ltPropCyclic && cpyVal != newVal) {
			var max = data.ltPropMax,
				min = data.ltPropMin;
			if (max != undefined && min != undefined) {
				var alteredCpy = parseFloat(this.getDecimal(cpyVal));
				newVal = this.getDecimal(roundToFixed(alteredCpy > max ? min + alteredCpy - max : max - min + alteredCpy), true);
			}
		}

		_this.value = newVal;
		_this.selectionEnd = newVal.length;
		_this.selectionStart = newVal.length;
		data.ltPropAutoUpdate && this.valueUpdate( _this, validation == 'input' );
	}

    check_restrict(text, frm_paste) {
		var restrict = this.data.ltPropRestrict;

		if( restrict ){
			var regex = new RegExp( restrict, 'g' );
			if( frm_paste ){
				var __index = regex.exec( text );
				if( __index ){
					return text.slice( 0, __index.index );
				}
			} else {
				return text.replace( regex, "" );
			}
		}

		return text;
	}

    static actions(arg1) {
        return Object.assign(super.actions({

            change_value : function( evt, __this ){

                if( evt.target == __this ){
                    return;
                }

                var is_up = $L( evt.target ).closest( '.lyteNumberUpArrow' ).length,
                __data = this.data,
                readonly = __data.ltPropReadonly,
                disabled = __data.ltPropDisabled,
                keydown = evt.type == "keydown",
                click = evt.type == "click",
                input = this.getInput(),
                rptInc = function () {
                    this.increment(input, is_up);
                }.bind(this);

                if (disabled || (keydown && !(evt.key == 'Enter' || evt.key == ' '))){
                    return;
                } else if( !readonly ){
                    if (click || keydown) {
                        this.increment(input, is_up);
                    } else {
                        evt.preventDefault();
                        this.getInput().focus();
                        clearInterval(this.controlsTime);
                        this.removeControlsRef = this.removeControls.bind(this);

                        var ns = "addEventListener" + ( _lyteUiUtils.isWidget ? "Global" : "" );

                        document[ ns ]("mouseup", this.removeControlsRef);
                        document[ ns ]("touchend", this.removeControlsRef);
                        this.controlsTime = setInterval(rptInc, 500);
                    }
                }

                if (keydown) {
                    this._valueUpdate(input.value);
                    evt.preventDefault();
                }

                click && this.setData( "ltPropAutofocus", true );
            },

            keydown : function( evt, _this ){
                var keyCode = evt.which || evt.keyCode, prevent,
                start = _this.selectionStart,
                end = _this.selectionEnd,
                data = this.data,
                inc = data.ltPropIncrement,
                max = data.ltPropMax,
                min = data.ltPropMin;

                if( keyCode == 8 ){
                    if( start == end && end == 0 ){
                        return;
                    }
                    this._prevent = true;
                } else if( inc && keyCode == 38 ){
                    this.increment( _this, !data.ltPropInverse );
                    prevent = true;
                } else if( inc && keyCode == 40 ){
                    this.increment( _this, data.ltPropInverse );
                    prevent = true;
                } else if ( inc && keyCode == 35 && max != undefined ) {
					this._valueUpdate( max );
                } else if ( inc && keyCode == 36 && min != undefined ) {
					this._valueUpdate( min );
                }
                if( prevent ){
                    evt.preventDefault();
                }
            },

            wheel : function( evt, _this ){
                var __data = this.data;
                if( __data.ltPropWheel && this._focused && !__data.ltPropDisabled && !__data.ltPropReadonly ){
                    var __inverse = __data.ltPropInverse;

                    if( evt.deltaY > 10 ){
                        this.increment( _this, !__inverse );
                    } else if( evt.deltaY < -10 ){
                        this.increment( _this, __inverse );
                    }
                    evt.preventDefault();
                }
            },

            paste : function( evt, _this ){
                evt.preventDefault();

                var clip = evt.clipboardData || window.clipboardData,
                oldValue = _this.value,
                start = _this.selectionStart,
                end = _this.selectionEnd,
                __data = this.data,
                decimal = __data.ltPropDecimal,
                pasteText = this.check_restrict( clip.getData( 'text' ).trim().replace(/^(\'|\")|(\'|\")$/, '').replace( new RegExp('\[\^0-9e\\+\\-\\' + decimal + '\]', 'g' ), '' ), true ),
                newVal = oldValue.slice( 0, start ) + pasteText + oldValue.slice( end ),
                remove = this.validate( newVal ),
                num_value = Number( oldValue ),
                paste_len = pasteText.length,
                __inf = Infinity,
                validation = __data.ltPropValidation;

                if( remove ){
                    newVal = this.convertToNo( newVal );
                }
                if( num_value == __inf || num_value == -__inf ){
                    return;
                }
                if( this.getMethods( 'onBeforePaste' ) ){
                    var ret = this.executeMethod( 'onBeforePaste', _this.value, clip.getData( 'text' ), newVal );
                    if( ret != undefined ){
                        newVal = ret;
                        if( ret == false ){
                            return;
                        }
                    } 
                }
                if( num_value == __inf || num_value == -__inf ){
                    return;
                }

                if( !validation ){
                    newVal = this.maxLenValidation( newVal );
                    if( newVal || ( !newVal && __data.ltPropValidateOnEmpty ) ){
                        newVal = this.maxMinCheck( newVal, true );
                    }
                }

                _this.value = newVal;
                _this.selectionStart = end + paste_len;
                _this.selectionEnd = end + paste_len;

                __data.ltPropAutoUpdate && this._valueUpdate( _this.value, validation == 'input' );

                $L.fastdom.measure( this.scroll_to_view.bind( this, _this ) );
            },

            input : function( evt, _this ){
                var prevent;
                if( this._prevent ){
                    delete this._prevent;
                    this.data.ltPropAutoUpdate && this.valueUpdate( _this );
                    prevent = true;
                }
                var value = _this.value,
                oldValue = value,
                start = _this.selectionStart,
                end = _this.selectionEnd,
                remove = !prevent && this.validate( value ),
                minLenCheck,
                validation = this.data.ltPropValidation;

                if( remove ){
                    var __allow = true;
                    if( /insertText/i.test( evt.inputType ) ){
                        var evt_data = evt.data,
                        newvalueISNo = parseFloat( evt_data ),
                        __length = evt_data.length;

                        if( !isNaN( newvalueISNo ) ){
                            newvalueISNo = newvalueISNo.toString(); 
                            value = value.slice( 0, end - __length ) + newvalueISNo + value.slice( end );
                            end = end - __length + 1 + newvalueISNo.length;
                        } else {
                            value = value.slice( 0, end - __length ) + value.slice( end );
                            end = end - __length + 1;
                        }

                        if( __allow = this.validate( value ) ){
                            end--;
                        }
                    }

                    if( __allow ){
                        value = value.slice( 0, start - __length ) + value.slice( end );
                    }
                } else {
                    value = this.check_restrict( value );

                    if( !validation ){
                        var nw = this.maxLenValidation( value, start, end );
                        if( nw != value ){
                            if( this.data.ltPropRemoveAtCursor ){
                                minLenCheck = true;
                            }
                            value = nw;
                        }
                        nw = this.maxMinCheck( value, this.data.ltPropValidateOnInput );
                        if( nw != value ){
                            value = nw;
                            minLenCheck = false;
                        }
                    }
                }
                _this.value = value;
                if( remove ){
                    _this.selectionStart = Math.min( end - 1, value.length );
                    _this.selectionEnd = Math.min( end - 1, value.length )
                } else {
                    if( oldValue != value ){
                        if( minLenCheck ){
                            _this.selectionStart = _this.selectionEnd = start - 1;
                        } else {
                            _this.selectionStart = _this.selectionEnd = value.length;
                        }
                    } else {
                        _this.selectionStart = Math.min( start, value.length );
                        _this.selectionEnd = Math.min( end, value.length );
                    }
                }

                this.data.ltPropAutoUpdate && this.valueUpdate( _this, validation == 'input' );
            },

            focus : function( evt, _this ){

                if( this.__ignore_focus ){
                    return;
                }

                this._focused = true;
                this.$node.classList.add( 'lyteInputFocus' );
                this.getMethods( 'onFocus' ) && this.executeMethod( 'onFocus', evt, this.$node );
            },

            blur : function( evt, _this ){
                if( this.__ignore_focus ){
                    return;
                }


                delete this._focused; 
                this.$node.classList.remove( 'lyteInputFocus' );

                var validation = this.data.ltPropValidation;
                
                if( !validation && ( _this.value || ( !_this.value && this.data.ltPropValidateOnEmpty ) ) ){
                    _this.value = this.maxMinCheck( _this.value, true );
                }

                this._valueUpdate( _this.value, true );

                if( validation == "blur" ){
                    this.do_validation();
                }

                this.getMethods( 'onBlur' ) && this.executeMethod( 'onBlur', evt, this.$node );
            }
        }), arg1);
    }

    valueUpdate(_this, validate) {
		_this = _this || this.getInput();

		var __fn = function(){
			this._valueUpdate( _this.value );
			validate && this.do_validation();
		}.bind( this ),
		delay = this.data.ltPropUpdateDelay;

		if( delay == undefined ){
			__fn();
		} else {
			clearTimeout( this._time );
			this._time = setTimeout( __fn, delay );
		}
	}

    _valueUpdate(value, force, isScriptChange) {

		var __old = value,
		cb = "onBeforeValueUpdate",
		new_value;

		if( force ){
			var endDotRegex = new RegExp( "\\" + this.data.ltPropDecimal + "$" );
			if( /e$/i.test( value ) ){
				value = value.replace( /e$/, '' );
			}
			if( endDotRegex.test( value ) ){
				value = value.replace( endDotRegex, '' );
			}
		}

		/**
		 * @method onBeforeValueUpdate
		 * @version 3.80.0
		 */

		if( this.getMethods( cb ) && ( new_value = this.executeMethod( cb, __old, value, this.$node, this.data.ltPropValue ) ) == false ){
			return;
		}

		if( typeof new_value == 'string' ){
			value = new_value;
		}

		if( value != this.data.ltPropValue ){
			isScriptChange && (this._scriptChange = true);
			this._allowCallback = true;
			this.$node.ltProp( 'value', value );
		} else {
			var inputElem = this.$node.querySelector('input');
			if( value != inputElem.value ){
				inputElem.value = value;
			}
		}
	}

    validate(value) {
		var numberRegex = /\d+/,
		eRegex = /e/ig,
		plusOrMinus = /(\+|\-)/g,
		decimal = this.data.ltPropDecimal,
		dotRegex = new RegExp("\\" + decimal, 'g'),
		eRegexIndex,
		remove;

		if( eRegex.test( value ) ){
			if( value.match( eRegex ).length > 1 ){
				remove = true;
			} if( /^e/i.test( value ) ){
				remove = true;
			} else {
				eRegexIndex = /e/ig.exec( value ).index;
				value = value.replace( eRegex, '' );
			}
		} 
		if( plusOrMinus.test( value ) ) {
			if( value.match( plusOrMinus ).length > 1 ){
				remove = true;
			} else if( !/^(\+|\-)/g.test( value ) ){
				remove = true;
			} else {
				value = value.replace( plusOrMinus, '' );
			}
		} 
		if( dotRegex.test( value ) ){
			if( value.match( dotRegex ).length > 1 || ( eRegexIndex != undefined && eRegexIndex < value.indexOf( decimal ) ) ){
				remove = true;
			} else if( ( /^(\+|\-)/g.test( value ) && value.indexOf( decimal ) == 1 && eRegexIndex == 2  ) || ( !/^(\+|\-)/g.test( value ) && value.indexOf( decimal ) == 0 && eRegexIndex == 1 ) ){ 
 				remove = true;
			} else {
				value = value.replace( dotRegex, '' );
			}
		}
		if( !remove && value.length ){
			remove = !/^\d+$/.test( value );
		}
		return remove;
	}

    digitsLengthValidation(value) {//here string is manupulated to allow .00
		var digits = this.data.ltPropDigits;
		if( digits != undefined ){
			var decimal = this.data.ltPropDecimal,
				separator = value.split( decimal );
			return separator[0] + (separator[1] != undefined ? decimal + separator[1].slice(0, digits) : '');
		}
		return value;
	}

    maxLenValidation(value, start, end) {
		var maxLen = this.data.ltPropMaxlength,
		length = value.length, dotIndex,
		decimal = this.data.ltPropDecimal,
		removeAt = this.data.ltPropRemoveAtCursor;

		value = this.digitsLengthValidation( value );
		if( maxLen == undefined ){
			return value;
		}
		if( !this.data.ltPropIgnoreSymbols ){
			maxLen = maxLen + ( /^(\+|\-)/.test( value ) ? 1 : 0 );
			if( new RegExp("\\" + decimal ).test( value ) ){
				dotIndex = new RegExp("\\" + decimal, 'g').exec( value ).index;
				maxLen++;
			}
			if( /e/i.test( value ) ){
				var index = /e/i.exec( value ).index,
				power = value.slice( index + 1 ),
				parsedPow = parseInt( power );
				if( power.length ){
					length--;
					//if( dotIndex != undefined ){
						var lenAfterDot = dotIndex != undefined ? value.slice( dotIndex + 1, index ).length : 0;
						if( lenAfterDot <= parsedPow ){
							if( dotIndex != undefined ){
								length--;
								maxLen--;
							}
							length -= lenAfterDot;
						} 
						length -= power.length;
						length += parsedPow;
						if( length > maxLen ){
							if( parsedPow > ( length - maxLen ) ){
								return value.slice( 0, index ) + 'e' + ( parsedPow - ( length - maxLen ) );
							} else {
								if( removeAt && start != undefined && value.length > maxLen ){
									value = value.split( /e/i )[ 0 ];
									value = value.slice( 0, start - 1 ) + value.slice( end );
								} else {
									value = value.split( /e/i )[ 0 ].match( new RegExp('.{0,' + maxLen + '}') )[ 0 ];
								}

							}
						} 
						return value;
					// }
				} else {
					maxLen++;
				}
			} 
		}
		if( removeAt && start != undefined && value.length > maxLen ){
			return value.slice( 0, start - 1 ) + value.slice( end );
		}
		return value.match( new RegExp('.{0,' + maxLen + '}') )[ 0 ];
	}

    maxMinCheck(value, allowMin) {
		var initial = value,
		max = this.data.ltPropMax,
		min = this.data.ltPropMin,
		parsedValue = Number( value.replace( new RegExp("\\" + this.data.ltPropDecimal ), '.' ) ),
		maxMinVal = this.getMethods( 'onBeforeMaxMinValidation' )

		if( max == undefined && min == undefined ){
			return value;
		} 
		if( max != undefined ){
			if( max < parsedValue ){
				value = max;
				/**
				 * @method onBeforeMaxMinValidation
				 * @version 2.2.9
				 */				
				var callbackvalue = maxMinVal ? this.executeMethod( 'onBeforeMaxMinValidation', 'max', initial, value, this.$node ) : value;
				value = callbackvalue == undefined ? value : callbackvalue;
			}
		}  
		if( allowMin && min != undefined  ){
			if( min > parsedValue ){
				value = min;
				value = min;
				var callbackvalue = maxMinVal ? this.executeMethod( 'onBeforeMaxMinValidation', 'min', initial, value, this.$node ) : value;
				value = callbackvalue == undefined ? value : callbackvalue;
			} else if( parsedValue == 0 && value == '' ){
				value = Math.max( parsedValue, min );
				var callbackvalue = maxMinVal ? this.executeMethod( 'onBeforeMaxMinValidation', 'min', initial, value, this.$node ) : value;
				value = callbackvalue == undefined ? value : callbackvalue;
			}
		}
		return value.toString().replace( /\./, this.data.ltPropDecimal );	
	}

    removeTwice(text, regex) {
		var split = text.split( regex );
		if( split.length > 1 ){
			var pop = split.shift(),
			eVal = regex.exec( text )[ 0 ];
			text = pop + eVal + split.join( '' );
		}

		return text;
	}

    convertToNo(newVal) {
		var decimal = this.data.ltPropDecimal,
		dotRegex = new RegExp("\\" + decimal ),
		eRegex = /e/i,
		dotIndex,
		eIndex;

		newVal = this.removeTwice( newVal, dotRegex );
		
		newVal = this.removeTwice( newVal, eRegex );
		newVal = this.removeTwice( newVal, /\+|\-/i );

		if( dotRegex.test( newVal ) && eRegex.test( newVal ) ){
			dotIndex = dotRegex.exec( newVal ).index;
			eIndex = eRegex.exec( newVal ).index;
			if( dotIndex > eIndex || Math.abs( dotIndex - eIndex ) == 1 ){
				newVal = newVal.replace( dotIndex > eIndex ? eRegex : dotRegex, '' );
			}
		}

		if( /^e/i.test( newVal ) ){
			newVal = newVal.replace( /e/i, '' );
		}

		if( /\+|\-/.test( newVal ) && !/^(\+|\-)/.test( newVal ) ) {
			if( ( eIndex != undefined && /\+|\-/.exec( newVal ).index != eIndex + 1 ) || eIndex == undefined ){
				newVal = newVal.replace( /\+|\-/g, '' );
			}
		}

		return newVal;	
	}

    error_decision(value, msg, type) {
       this.setData({
           errorMessage : msg,
           ltPropError : !!msg

       });

       $L( this.$node )[ ( msg ? 'add' : 'remove' ) + 'Class' ]( 'lyteErrorInput' );

       if( msg ){
           /**
            * @method onError
            * @version 3.86.0
            */
           var cb = "onError";
           this.getMethods( cb ) && this.executeMethod( cb, type, value, this.$node );
       } else if( msg == "" ){
           /**
            * @method onValidInput
            * @version 3.86.0
            */
           var cb = "onValidInput";
           this.getMethods( cb ) && this.executeMethod( cb, value, this.$node );
       }
    }

    do_validation() {
       var __data = this.data,
       message = __data.ltPropErrorMessage || {},
       def_message = __data.ltPropDefaultErrorMessage,
       value = __data.ltPropValue || "",
       validateEmpty = __data.ltPropValidateOnEmpty,
       __allow = value || validateEmpty,
       maxMin = __allow ? this.maxMinCheck( value, true ) : value,
       final_msg,
       __pattern = __data.ltPropPattern,
       // __dur = __data.ltPropAutoCorrectDuration,
       __fn = this.error_decision.bind( this, value );

       if( value != maxMin ){
           return __fn( message.minmax || def_message, 'maxmin' );
       }

       var maxlen = this.maxLenValidation( value );

       if( maxlen != value ){
           return __fn( message.maxlength || def_message, 'maxlength' );
       }

       if( __pattern && __allow ){
           var rgx = new RegExp( __pattern, 'g' );

           if( !rgx.test( value ) ){
               return __fn( message.pattern || def_message, 'pattern' );
           }
       }

       if( __data.ltPropMandatory && !value ){
           return __fn( message.mandatory || def_message, 'mandatory' );
       }

       return __fn( "" );
    }

    scroll_to_view(input) {

       this.__ignore_focus = true;

       input.blur();
       input.focus();

       delete this.__ignore_focus;
    }

    removeControls() {
        clearInterval(this.controlsTime);

       var ns = "removeEventListener" + ( _lyteUiUtils.isWidget ? "Global" : "" );

        document[ ns ]("mouseup", this.removeControlsRef);
        document[ ns ]("touchend", this.removeControlsRef);
        delete this.removeControlsRef;
        delete this.controlsTime;
    }

    getInput() {
        return Array.from(Array.from(this.$node.children).filter(function(item){
           return $L(item).hasClass('lyteField')
       })[0].children).filter(function(item){
           return item.tagName == 'INPUT';
       })[0]
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            valueObs : function( arg ){
                var data = this.data,
					isScriptChange = this._scriptChange;
                if( data.ltPropControls ){
                    this.setData( 'upArrowClass', (data.ltPropValue == data.ltPropMax) ? 'lyteNumberArrowDisabled' : '' );
                    this.setData( 'downArrowClass', (data.ltPropValue == data.ltPropMin) ? 'lyteNumberArrowDisabled' : '' );
                }

				delete this._scriptChange;

                if( this._init ){
					delete this._allowCallback;
                    return;
                }
                var isInit = !arg;
                if( !arg ){
                    arg = { newValue : this.data.ltPropValue || '' };
                } else {
                    arg.newValue = arg.newValue || "";
                }
                if( this._allowCallback ){
                    delete this._allowCallback;
                    this.getMethods( 'onValueChange' ) && this.executeMethod( 'onValueChange', arg, this.$node, isScriptChange ? 'script' : 'input' );
                    return;
                }
                var newVal = arg.newValue, remove;

                if( isInit && !this.data.ltPropFireOnInit ){
                    this._init = true;
                }

                remove = this.validate( newVal );
                if( remove ){
                    this._valueUpdate( '', undefined, true );
                    delete this._init;
                    return
                }
                newVal = this.maxLenValidation( newVal );
                if( ( this.data.ltPropValidateOnEmpty && !arg.newValue ) || arg.newValue ){
                    newVal = this.maxMinCheck( newVal, true );
                }
                if( arg.newValue != newVal ){
                    this._valueUpdate( newVal, undefined, true );
                } else if( !this._init ) {
                    if( isInit && arg.newValue == newVal ){
                        return;
                    }
                    this.getMethods( 'onValueChange' ) && this.executeMethod( 'onValueChange', arg, this.$node, 'script' );
                }
                delete this._init;

            }.observes( 'ltPropValue' ).on( 'didConnect' ),

            maxMinOns : function(){

                if( this.data.ltPropPreventObserver ){
                    return;
                }

                this._valueUpdate( this.maxMinCheck( this.data.ltPropValue || '', true ) );
            }.observes( 'ltPropMax', 'ltPropMin' ),

            wrp_class_obs : function( arg ){
                var __data = this.data;
                arg = arg || { newValue : __data.ltPropInputWrapperClass };

                $L( this.$node ).addClass( arg.newValue ).removeClass( arg.oldValue || "" );

            }.observes( 'ltPropInputWrapperClass' ).on( 'didConnect' ),

            appearanceObs : function(){
                var __remove = "remove",
                __add = "add",
                __classList = this.$node.classList;

                if( /box/i.test( this.data.ltPropAppearance ) ){
                    __remove = "add";
                    __add = "remove";
                }

                __classList[ __add ]( 'lyteInput' );
                __classList[ __remove ]( 'lyteInputBox' );

            }.observes( 'ltPropAppearance' ).on( 'didConnect' ),

            control_obs : function( arg ){
                arg = arg || { newValue : this.data.ltPropControls };
                this.$node.classList[ arg.newValue ? 'add' : "remove" ]( 'lyteNumberWithArrowKeys' );
            }.observes( 'ltPropControls' ).on( "didConnect" ),

            directionObs : function(){
                var __remove = "remove",
                __add = "add",
                __classList = this.$node.classList;

                if( /vertical/i.test( this.data.ltPropDirection ) ){
                    __add = "remove";
                    __remove = "add";
                }

                __classList[ __add ]( 'horizontal' );
                __classList[ __remove ]( 'vertical' );
            }.observes( 'ltPropDirection' ).on( 'didConnect' ),

            disAbs : function(){
                this.$node.classList[ this.data.ltPropDisabled ? 'add' : 'remove' ]( 'lyteNumberDisabled' );
            }.observes( 'ltPropDisabled' ).on( 'didConnect' ),

            readAbs : function(){
                this.$node.classList[ this.data.ltPropReadonly ? 'add' : 'remove' ]( 'lyteNumberReadonly' );
            }.observes( 'ltPropReadonly' ).on( 'didConnect' ),

            focus_obs : function( arg ){
                if( arg.newValue ){
                    this.$node.focus();
                    this.setData( arg.item, !1 );

                    var __input = this.getInput();
                    __input.selectionStart = __input.selectionEnd = __input.value.length;
                }
            }.observes( 'ltPropAutofocus' ),

            attrObs : function( arg ){
               this.data.ltPropAria && _lyteUiUtils.setAttribute( this.$node.querySelector( 'input' ), this.data.ltPropAriaAttributes || {}, arg ? arg.oldValue : {} )
            }.observes( 'ltPropAriaAttributes', 'ltPropAriaAttributes.{}', 'ltPropType' ).on( 'didConnect' ),

            single_obs : function( arg ){
               if( !arg.path ){
                   return;
               }
               var key = arg.path.replace( /^\./, '' ),
               newValue = arg.newValue,
               data = this.data;

               if( data.ltPropAria ){
                   var obj = {};
                   obj[ key ] = newValue;

                   _lyteUiUtils.setAttribute( $L( 'input', this.$node ).get( 0 ), obj, {} );
               }

            }.observes( 'ltPropAriaAttributes.*' ),

            trigger_obs : function( arg ){
               if( arg.newValue ){
                   var __item = arg.item;

                   if( /clear/i.test( __item ) ){
                       this.error_decision();
                   } else {
                       this.do_validation();
                   }
                   this.setData( __item, !1 );
               }
            }.observes( 'ltPropTriggerValidation', 'ltPropClearValidation' )
        }), arg1);
    }

    _() {
        _;
    }
}

LyteNumberComponent._template = "<template tag-name=\"lyte-number\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropLabel}}\" is=\"case\" lc-id=\"lc_id_0\"> <label for=\"{{ltPropId}}\" class=\"lyteLabel\">{{ltPropLabel}}</label> </template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropYield}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-yield yield-name=\"number\"></lyte-yield></template></template><div class=\"lyteField {{if(ltPropSuffix,'lyteNumberSuffix','')}}\"> <input type=\"text\" name=\"{{ltPropName}}\" class=\"{{ltPropClass}}\" id=\"{{ltPropId}}\" placeholder=\"{{ltPropPlaceholder}}\" value=\"{{lbind(ltPropValue)}}\" onkeydown=\"{{action('keydown',event,this)}}\" onpaste=\"{{action('paste',event,this)}}\" oninput=\"{{action('input',event,this)}}\" onfocus=\"{{action('focus',event,this)}}\" onblur=\"{{action('blur',event,this)}}\" onwheel=\"{{action('wheel',event,this)}}\" disabled=\"{{ltPropDisabled}}\" readonly=\"{{ltPropReadonly}}\" style=\"{{ltPropStyle}}\" title=\"{{ltPropInputTitle}}\" pattern=\"{{ltPropPattern}}\" autocomplete=\"{{ltPropAutocomplete}}\" tabindex=\"{{ltPropTabIndex}}\" data-tabindex=\"{{ltPropDataTabindex}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(ltPropUnit,'&amp;&amp;',expHandlers(ltPropSuffix,'!'))}}\" is=\"case\" lc-id=\"lc_id_0\"><span class=\"lyteNumberUnitElem\">{{ltPropUnit}}</span></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSuffix}}\" is=\"case\" lc-id=\"lc_id_0\"><div class=\"lyteNumberSuffixDummyWrapper\"> <span class=\"lyteNumberSuffixDummyElem\">{{ltPropValue}}</span> <span class=\"lyteNumberUnitElem\">{{ltPropUnit}}</span> </div></template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropControls}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteNumberArrowContainer\" onclick=\"{{action('change_value',event,this)}}\" onmousedown=\"{{action('change_value',event,this)}}\" ontouchstart=\"{{action('change_value',event,this)}}\" onkeydown=\"{{action('change_value',event,this)}}\"> <div class=\"lyteNumberUpArrow {{ltPropControlsClass}} {{upArrowClass}}\" tabindex=\"0\" role=\"spinbutton\" aria-label=\"{{ltPropControlsAriaLabel.inc}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropControlsYield}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-yield yield-name=\"upArrow\" lt-prop-disabled=\"{{if(upArrowClass,true,false)}}\" lt-prop-value=\"{{ltPropValue}}\"></lyte-yield></template></template></div> <div class=\"lyteNumberDownArrow {{ltPropControlsClass}} {{downArrowClass}}\" tabindex=\"0\" role=\"spinbutton\" aria-label=\"{{ltPropControlsAriaLabel.dec}}\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropControlsYield}}\" is=\"case\" lc-id=\"lc_id_0\"><lyte-yield yield-name=\"downArrow\" lt-prop-disabled=\"{{if(downArrowClass,true,false)}}\" lt-prop-value=\"{{ltPropValue}}\"></lyte-yield></template></template></div> </div> </template></template></div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(ltPropError,'&amp;&amp;',errorMessage)}}\" is=\"case\" lc-id=\"lc_id_0\"> <span class=\"lyteNumberErrorMessage\">{{errorMessage}}</span> </template></template></template>";;
LyteNumberComponent._dynamicNodes = [{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":5,"sibl":[4]},{"t":"s","p":[2],"c":{"lc_id_0":{"dN":[{"t":"i","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":4,"sibl":[3]},{"t":"a","p":[3]},{"t":"a","p":[3,1],"a":{"style":{"name":"style","dynamicValue":"ltPropStyle"}}},{"t":"s","p":[3,3],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":3,"sibl":[2]},{"t":"s","p":[3,4],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[0,1,0],"cn":"lc_id_0"},{"t":"tX","p":[0,3,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1]},{"t":"s","p":[3,5],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"a","p":[1,1],"cn":"lc_id_0"},{"t":"s","p":[1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"i","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"a","p":[1,3],"cn":"lc_id_0"},{"t":"s","p":[1,3,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0],"cn":"lc_id_0"},{"t":"i","p":[0],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"s","p":[5],"c":{"lc_id_0":{"dN":[{"t":"tX","p":[1,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":0},{"type":"dc","trans":true,"hc":true,"p":[4,1]}];;

LyteNumberComponent._observedAttributes = [
    "ltPropValue",
    "ltPropMaxlength",
    "ltPropName",
    "ltPropClass",
    "ltPropId",
    "ltPropPlaceholder",
    "ltPropIgnoreSymbols",
    "ltPropStep",
    "ltPropInverse",
    "ltPropWheel",
    "ltPropMax",
    "ltPropMin",
    "ltPropAutofocus",
    "ltPropDisabled",
    "ltPropReadonly",
    "ltPropStyle",
    "ltPropInputTitle",
    "ltPropPattern",
    "ltPropUpdateDelay",
    "ltPropLabel",
    "ltPropAutocomplete",
    "ltPropAppearance",
    "ltPropDirection",
    "ltPropValidateOnInput",
    "ltPropAutoUpdate",
    "ltPropDecimal",
    "ltPropIncrement",
    "ltPropAria",
    "ltPropAriaAttributes",
    "ltPropRemoveAtCursor",
    "ltPropFireOnInit",
    "ltPropValidateOnEmpty",
    "ltPropRestrict",
    "ltPropInputWrapperClass",
    "ltPropTabIndex",
    "ltPropValidation",
    "ltPropMandatory",
    "ltPropTriggerValidation",
    "ltPropErrorMessage",
    "ltPropDefaultErrorMessage",
    "ltPropError",
    "ltPropClearValidation",
    "ltPropDataTabindex",
    "ltPropControls",
    "ltPropCyclic",
    "ltPropUnit",
    "ltPropDigits",
    "ltPropSuffix",
    "ltPropControlsYield",
    "ltPropControlsClass",
    "ltPropYield",
    "ltPropControlsAriaLabel",
    "ltPropPreventObserver",
    "errorMessage",
    "upArrowClass",
    "downArrowClass"
];

/**
 * @syntax nonYielded
 * <lyte-number></lyte-number>
 */

export { LyteNumberComponent };

LyteNumberComponent.register("lyte-number", {
    hash: "LyteNumberComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
