// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/d8426678cee8e139c99b9d2de7cf2a67.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../component.js";

class LyteSkipToComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
			ltPropData: prop('array', { default: [{}] })
		}), arg1);
	}

    init() {
		var _this = this;
		document.addEventListener("keydown", (event) => {
			if (event.key == "Enter") {
				if (document.activeElement) {
					var currentIndex = parseInt($L(document.activeElement).attr('dom-index'))
					if (currentIndex >= 0) {
						var currentObj = _this.getData('ltPropData')[currentIndex]
						if (currentObj.target) {
							$L(currentObj.target)[0].focus();
						}
						else if (currentObj.container) {
							let initializeFocusEle = _lyteUiUtils.getFirstFocusableElem(currentObj.container);

							if (initializeFocusEle) {
								initializeFocusEle.focus();
							}

						}
					}
				}
			}
		});
	}

    static actions(arg1) {
        return Object.assign(super.actions({
            // Functions for event handling
        }), arg1);
    }

    static methods(arg1) {
        return Object.assign(super.methods({
            // Functions which can be used as callback in the component.
        }), arg1);
    }

    _() {
        _;
    }
}

LyteSkipToComponent._template = "<template tag-name=\"lyte-skip-to\"> <template items=\"{{ltPropData}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <div class=\"lyteSkipToClass\" tabindex=\"0\" dom-index=\"{{index}}\">{{item.message}}</div> </template> </template>";;
LyteSkipToComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"f","p":[1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,0]}]}];;
LyteSkipToComponent._observedAttributes = ["ltPropData"];
export { LyteSkipToComponent };
LyteSkipToComponent.register("lyte-skip-to", {
    hash: "LyteSkipToComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
