// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/dfcd4e44d8b31dc6ca048ee40086e7ab.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../../node_modules/@slyte/core/index.js";
import $L from "../../../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";
import { Component } from "../../../component.js";
import { objectUtils } from "../../../../node_modules/@slyte/component/index.js";

class LyteListFilterComponent extends Component {
    constructor() {
        super();
    }

    init() {
		this.data.ltPropModalProperties != void 0 ? this.setData( 'ltPropModal' , this.data.ltPropModalProperties ) : null;
		
		var data = this.data,
		fields = data.ltPropFields,
		final = [],
		obj = {};
		
		fields.forEach( function( item ){
			var __data = item.data,
				selected = data.ltPropConditions[item.data.prop];
			if( __data.dataType ){
				final.push({
					data : __data,
					render : window._lyteUiUtils.lyteUiIsEmpty( selected ),
					selected : selected,
					temp : {},
					reset : false,
					class : selected && selected.isValid ? 'lyteListFilterApplied' : ''
				});

				obj[ __data.prop ] = {};
			}
		});

		this.data.ltPropHiddenRows.forEach( function(item){
			final.splice( item.index, undefined, {
				data : item.data,
				render : window._lyteUiUtils.lyteUiIsEmpty( data.ltPropConditions[item.data.prop] ),
				selected : data.ltPropConditions[item.data.prop],
				temp : {},
				reset : false,
				class : data.ltPropConditions[item.data.prop] && data.ltPropConditions[item.data.prop].isValid ? 'lyteListFilterApplied' : ''
			} )
		} );
			
		this.$node.reset = function( key ){
			this.reset( key == void 0 ? undefined : this.data.fields.filter( function( item ){
			    return key == item.data.prop   
			}) );
		}.bind( this );

		if( Object.keys(this.data.ltPropConditions).length == 0 ){
			this.setData( "ltPropConditions" , obj );
		}

		this.setData({
			fields : final,
			noResult : fields.length ? 'lyteSearchHidden' : ''
		});

		this.setData('finalStatus', this.data.ltPropAny);
	}

    data(arg1) {

		var __array = "array",
		__string = "string",
		__boolean = "boolean";

		return Object.assign(super.data({
			ltPropModal : prop( __string, { default : '{"wrapperClass" : "lyteListModalWrapper", "maxHeight" : "70%", "animation" : "slideFromRight","offset" : { "right" : 0 } }' } ),

			ltPropFields : prop( __array, { default : [] , watch : true } ),
			ltPropContent : prop( __array, { default : [] } ),

			ltPropShow : prop( __boolean, { default : false } ),

			ltPropAny : prop( __boolean ),
			ltPropReset : prop( __boolean, { default : false } ),

			ltPropFormat : prop( __string, { default : "MM-DD-YYYY" } ),
			ltPropTimeFormat : prop( __string, { default : undefined } ),

			ltPropConditions : prop( 'object', { default : {}, watch : true } ),
			ltPropNoResult : prop( __string, { default : window._lyteUiUtils.i18n( 'no.results.found', void 0, 'No Results Found' ) } ),
			ltPropFilteredContent : prop( __array , { default : [] } ),
			ltPropModalProperties : prop( 'object' , { default : undefined } ),
			ltPropFilterElementProperties : prop( 'object', { default : {} } ),
			ltPropHiddenRows : prop( 'array', { default : [] } ),

			fields : prop( __array, { default : [] } ),		

			finalStatus : prop( __boolean, { default : true } ),
			noResult : prop( __string, { default : '' } )
		}), arg1);		
	}

    execute(cb) {
		return this.getMethods( cb ) && this.executeMethod.apply( this, arguments );
	}

    static methods(arg1) {
        return Object.assign(super.methods({

            search : function( arr ){
                this.setData( 'noResult', arr.length ? 'lyteSearchHidden' : '' );
            },

            customReset : function( data, condition, node ){
                return this.execute( 'onCustomFilterReset', data, condition, node );
            },

            fetchMoreData : function( first, data, picklistOptions ){
                return this.execute( 'fetchMoreData', first, data, picklistOptions );
            },

            picklistData : function( value, cell_data, old_value ){
                return this.execute( 'onPicklistConstruct', value, cell_data, old_value );
            },	

            beforeOpen : function(){
                var __elem = arguments[ 1 ],
                index = parseInt( __elem.getAttribute( 'index' ) ),
                __data = this.data.fields[ index ];

                objectUtils( __data, 'add', 'render', true );
                return this.execute( 'onBeforeFilterOpen', __data.data, __data.selected, __elem.querySelector( '.lyteFilterElement' ) );
            },

            reset : function( item ){
                objectUtils( item, 'add', 'value', '' );
            },

            beforeClose : function(){
                var elem = arguments[ 2 ].$node.querySelector( '.lyteAccordionActive .lyteFilterElement' ),
                __data = ( ( elem || {} ).component || {} ).data;

                __data && this.execute( 'onBeforeFilterClose', __data.ltPropData, __data.ltPropCondition, elem );

                if( __data && !__data.ltPropTempCondition.isValid ){
                    elem.ltProp( 'reset', true );
                }
            }
        }), arg1);
    }

    first_extn(obj) {
		return $L.extend( true, {}, obj );
	}

    reset(fields) {
		var __data = this.data,
		conditions = __data.ltPropConditions,
		Lc = objectUtils,
		__this = this;

		this.setData( 'ltPropFilteredContent' , undefined );

		fields = fields || __data.fields;

		fields.forEach( function( item ){
			var f_data = item.data,
			__prop = f_data.prop;

			Lc( item, 'add', 'reset', true );
			Lc( conditions, 'add', __prop, {} );
		});

		this.setData( 'ltPropAny', __data.finalStatus );
	}

    apply() {
		var __data = this.data,
		conditions = __data.ltPropConditions,
		fields = __data.fields,
		Lc = objectUtils,
		__this = this;

		fields.forEach( function( item ){
			var f_data = item.data,
			extended = __this.first_extn( item.temp );
			extended.filterName = f_data.filterName;

			Lc( conditions , 'add', f_data.prop, extended );
			
			if( conditions[f_data.prop].isValid ){
				Lc( item, 'add', 'class',  'lyteListFilterApplied');
			}else{
				Lc( item, 'add', 'class',  '');
			}
			// Lc( conditions, 'add', f_data.prop, __this.first_extn( item.selected ) );
		});

		this.setData( 'finalStatus', __data.ltPropAny );

		var cb = "onModalFilterApply";
		if( this.getMethods(cb) ){
			setTimeout(function(){
				this.executeMethod( cb , conditions );
			}.bind(this), 0);
		}
	}

    cancel() {
		var __data = this.data,
		conditions = __data.ltPropConditions,
		fields = __data.fields,
		Lc = objectUtils,
		__this = this;

		fields.forEach( function( item ){
			Lc( item, 'add', 'selected', __this.first_extn( conditions[ item.data.prop ] ) );
		});
	}

    callPicklistContruct(index) {
		let data = this.data.fields;

		objectUtils(data[index], 'add', 'reset', true);
	}

    static actions(arg1) {
        return Object.assign(super.actions({
            reset : function(){
                let ret = true;
                this.reset();
                if( this.getMethods('onReset') ){
                    ret = this.executeMethod('onReset');
                }
                this.setData( 'ltPropShow', !ret );
            },

            apply : function(){
                let ret = true;
                this.apply();
                if( this.getMethods('onApply') ){
                    ret = this.executeMethod('onApply');
                }
                this.setData( 'ltPropShow', !ret );
            },

            cancel : function(){
                let ret = true;
                if( this.getMethods('onCancel') ){
                    ret = this.executeMethod('onCancel');
                }
                this.setData( 'ltPropShow', !ret );
            }
        }), arg1);
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            cond_obs : function(){

                var cb = "onConditionChange";

                if( this.getMethods( cb ) ){
                    clearTimeout( this.__timeout );
                    this.__timeout = setTimeout( function(){
                        var __data = this.data;
                        this.executeMethod( cb, __data.ltPropConditions, __data.ltPropAny , true, this.$node );
                    }.bind( this ), 0 );
                }
            }.observes( 'ltPropConditions.*', 'finalStatus' ),

            reset_obs : function( arg ){
                if( arg.newValue ){
                    this.reset();
                    this.setData( arg.item, false );
                }
            }.observes( 'ltPropReset' ),

            show_obs : function( arg ){
                if( !arg.newValue ){
                    this.cancel();
                } 
            }.observes( 'ltPropShow' )
        }), arg1);
    }

    _() {
        _;
    }
}

LyteListFilterComponent._template = "<template tag-name=\"lyte-list-filter\"> <lyte-modal lt-prop=\"{{ltPropModal}}\" lt-prop-show=\"{{lbind(ltPropShow)}}\"> <template is=\"registerYield\" yield-name=\"modal\"> <lyte-modal-header> <div class=\"lyteFilterHeader\"> <div class=\"lyteFilterName\">Filter</div> </div> <div class=\"lyteFilterType\"> <lyte-radiobutton lt-prop-type=\"primary\" lt-prop-name=\"filter_type\" lt-prop-label=\"Any of these Condition(s)\" lt-prop-value=\"any_of_these\" lt-prop-checked=\"{{lbind(ltPropAny)}}\"></lyte-radiobutton> <lyte-radiobutton lt-prop-type=\"primary\" lt-prop-name=\"filter_type\" lt-prop-label=\"All of these Condition(s)\" lt-prop-value=\"all_of_these\" lt-prop-checked=\"{{expHandlers(ltPropAny,'!')}}\"></lyte-radiobutton> </div> <lyte-search lt-prop-appearance=\"box\" lt-prop-query-selector=\"{&quot;scope&quot; : &quot;.lyteFilterContent&quot;,&quot;search&quot; : &quot;lyte-accordion-header&quot;,&quot;target&quot; : &quot;lyte-accordion-item&quot;}\" lt-prop-search-delay=\"\" on-search=\"{{method('search')}}\"></lyte-search> </lyte-modal-header> <lyte-modal-content class=\"lyteFilterContent\"> <lyte-accordion lt-prop-dynamic=\"true\" on-before-open=\"{{method('beforeOpen')}}\" on-before-close=\"{{method('beforeClose')}}\"> <template is=\"registerYield\" yield-name=\"yield\"> <template items=\"{{fields}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <lyte-accordion-item index=\"{{index}}\"> <lyte-accordion-header class=\"{{expHandlers(item.class,'||',item.temp.class)}}\"> {{expHandlers(item.data.filterName,'||',item.data.name)}} <lyte-icon class=\"lyteAccordionArrow\"></lyte-icon> </lyte-accordion-header> <lyte-accordion-body> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.render}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"component\" class=\"lyteFilterElement\" component-name=\"{{concat('lyte-',item.data.dataType,'-filter')}}\" lt-prop-condition=\"{{lbind(item.selected)}}\" lt-prop-temp-condition=\"{{lbind(item.temp)}}\" lt-prop-render-content=\"{{lbind(ltPropRenderContent)}}\" lt-prop-format=\"{{ltPropFormat}}\" lt-prop-picklist-options=\"{{item.picklistOptions}}\" lt-prop-min=\"{{item.min}}\" lt-prop-max=\"{{item.max}}\" lt-prop-data=\"{{item.data}}\" lt-prop-filter-element-properties=\"{{ltPropFilterElementProperties}}\" on-picklist-construct=\"{{method('picklistData')}}\" fetch-more-data=\"{{method('fetchMoreData')}}\" lt-prop-reset=\"{{lbind(item.reset)}}\" on-custom-filter-reset=\"{{method('customReset')}}\"> <template is=\"registerYield\" yield-name=\"lyte-custom-filter\" from-parent=\"\"></template> </template> </template></template></lyte-accordion-body> </lyte-accordion-item> </template> </template> </lyte-accordion> <div class=\"lyteFilterNoResult {{noResult}}\">{{ltPropNoResult}}</div> </lyte-modal-content> <lyte-modal-footer> <span class=\"lyteFilterReset\" onclick=\"{{action('reset')}}\">Reset</span> <div class=\"lyteListButtons\"> <lyte-button class=\"lyteFilterCancel\" onclick=\"{{action('cancel')}}\"> <template is=\"registerYield\" yield-name=\"text\"> Cancel </template> </lyte-button> <lyte-button lt-prop-appearance=\"primary\" class=\"lyteFilterApply\" onclick=\"{{action('apply')}}\"> <template is=\"registerYield\" yield-name=\"text\"> Apply </template> </lyte-button> </div> </lyte-modal-footer> </template> </lyte-modal> </template>";;
LyteListFilterComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"r","p":[1,1],"dN":[{"t":"a","p":[1,3,1]},{"t":"cD","p":[1,3,1],"in":11,"sibl":[10]},{"t":"a","p":[1,3,3]},{"t":"cD","p":[1,3,3],"in":10,"sibl":[9]},{"t":"a","p":[1,5]},{"t":"cD","p":[1,5],"in":9,"sibl":[8]},{"t":"cD","p":[1],"in":8,"sibl":[7]},{"t":"a","p":[3,1]},{"t":"r","p":[3,1,1],"dN":[{"t":"a","p":[1]},{"t":"f","p":[1],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"tX","p":[1,1,1]},{"t":"cD","p":[1,1,3],"in":4,"sibl":[3]},{"t":"cD","p":[1,1],"in":3,"sibl":[2]},{"t":"s","p":[1,3,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cM","p":[1],"dN":[{"t":"r","p":[1],"dN":[]}],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":2,"sibl":[1]},{"t":"cD","p":[1,3],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[4,3,2,1,0],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":7,"sibl":[6]},{"t":"cD","p":[3,1],"in":6,"sibl":[5]},{"t":"a","p":[3,3]},{"t":"tX","p":[3,3,0]},{"t":"cD","p":[3],"in":5,"sibl":[4]},{"t":"a","p":[5,1]},{"t":"a","p":[5,3,1]},{"t":"r","p":[5,3,1,1],"dN":[],"in":4,"sibl":[3]},{"t":"cD","p":[5,3,1],"in":3,"sibl":[2]},{"t":"a","p":[5,3,3]},{"t":"r","p":[5,3,3,1],"dN":[],"in":2,"sibl":[1]},{"t":"cD","p":[5,3,3],"in":1,"sibl":[0]},{"t":"cD","p":[5],"in":0}],"dc":[11,10,9,8,7,6,5,3,1,0],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0},{"type":"dc","trans":true,"hc":true,"p":[1,0]}];;

LyteListFilterComponent._observedAttributes = [
    "ltPropModal",
    "ltPropFields",
    "ltPropContent",
    "ltPropShow",
    "ltPropAny",
    "ltPropReset",
    "ltPropFormat",
    "ltPropTimeFormat",
    "ltPropConditions",
    "ltPropNoResult",
    "ltPropFilteredContent",
    "ltPropModalProperties",
    "ltPropFilterElementProperties",
    "ltPropHiddenRows",
    "fields",
    "finalStatus",
    "noResult"
];

export { LyteListFilterComponent };

LyteListFilterComponent.register("lyte-list-filter", {
    hash: "LyteListFilterComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
