// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/f38aa5317aa96e9b44ba1c208e55c177.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../../node_modules/@slyte/component/index.js";
import $L from "../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";

class LyteSpeedDialComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
            ltPropSpeedShow : prop('boolean',{default:false}),
			ltPropIcons : prop('array',{default:[]})
		}), arg1);		
	}

    static actions(arg1) {
        return Object.assign(super.actions({
                iconClick : function (node){
                     if(this.getMethods('onItemClick')){
                        this.executeMethod('onItemClick',this.getData('ltPropIcons')[node.id]);
                     }
                },
                speedIconClick : function(node){
                    if($L('.lyteSpeedDialMainIcon')[0].classList.contains('lyteSpeedDialButtonIconRotated')){
                        $L('.lyteSpeedDialMainIcon').removeClass('lyteSpeedDialButtonIconRotated');
                    }
                    this.speedShowFunction(node);
                },
                onOver : function(node){
                    if(this.getMethods('onOver')){
                        this.executeMethod('onOver',this.getData('ltPropIcons')[node.id]);
                    }
                }
        }), arg1);
    }

    static methods(arg1) {
        return Object.assign(super.methods({

        }), arg1);
    }

    didConnect() {
		$L('.lyteSpeedDial')[0].style.top=this.getData('ltPropSpeedButton').offset[0]+'px';
		$L('.lyteSpeedDial')[0].style.left=this.getData('ltPropSpeedButton').offset[1]+'px';
	}

    didDestroy() {
		if(this.getMethods("onBeforeDestroy")){
			this.executeMethod("onBeforeDestroy",this);
		}	
	}

    circleDataAdd(preference, fullCircleRadius, halfCircleRadius, quaterCircleradius) {
		if(preference=='circle'){
			this.$node.style.setProperty('--lyte-speed-dial-negative-deg',90);
			this.$node.style.setProperty('--lyte-speed-dial-circle-total-degree',360);
			
			this.$node.style.setProperty('--lyte-speed-dial-circle-radius',fullCircleRadius);
		}else if(preference=='topLeftCorner'){
			this.$node.style.setProperty('--lyte-speed-dial-circle-total-degree',90);
			this.$node.style.setProperty('--lyte-speed-dial-circle-radius',quaterCircleradius);
			this.$node.style.setProperty('--lyte-speed-dial-negative-deg',0);
		}else if(preference=='topRightCorner'){
			this.$node.style.setProperty('--lyte-speed-dial-circle-total-degree',90);
			this.$node.style.setProperty('--lyte-speed-dial-circle-radius',quaterCircleradius);
			this.$node.style.setProperty('--lyte-speed-dial-negative-deg',90);
		}else if(preference=='bottomLeftCorner'){
			this.$node.style.setProperty('--lyte-speed-dial-circle-total-degree',90);
			this.$node.style.setProperty('--lyte-speed-dial-circle-radius',quaterCircleradius);
			this.$node.style.setProperty('--lyte-speed-dial-negative-deg',-90);
		}else if(preference=='bottomRightCorner'){
			this.$node.style.setProperty('--lyte-speed-dial-circle-total-degree',90);
			this.$node.style.setProperty('--lyte-speed-dial-circle-radius',quaterCircleradius);
			this.$node.style.setProperty('--lyte-speed-dial-negative-deg',-180);
		}else{
			this.$node.style.setProperty('--lyte-speed-dial-circle-total-degree',180);
			this.$node.style.setProperty('--lyte-speed-dial-circle-radius',halfCircleRadius);
			if(preference==='leftSemiCircle'){
				 this.$node.style.setProperty('--lyte-speed-dial-negative-deg',90);
			}else if(preference==='rightSemiCircle'){
				this.$node.style.setProperty('--lyte-speed-dial-negative-deg',-90);
			}else if(preference==='topSemiCircle'){
				this.$node.style.setProperty('--lyte-speed-dial-negative-deg',180);
			}else if(preference==='bottomSemiCircle'){
				this.$node.style.setProperty('--lyte-speed-dial-negative-deg',0);
			}
		}
	}

    speedShowFunction(node) {
		if(this.$node.getData('ltPropSpeedShow')){
			var onBeforeCloseVal = this.executeMethod('onBeforeClose');
			if(onBeforeCloseVal !==false){
				this.$node.setData('ltPropSpeedShow',false);
			}
			
		}else{
			if(this.getMethods('onBeforeShow')){
				var onBeforeShowVal = this.executeMethod('onBeforeShow');
			}
			if(onBeforeShowVal!==false){
				$L('.lyteSpeedDialMainIcon').addClass('lyteSpeedDialButtonIconRotated');				
				let top = this.getData('ltPropSpeedButton').offset[0];
				let left = this.getData('ltPropSpeedButton').offset[1];
	
				this.$node.style.setProperty('--lyte-speed-dial-icon-count',this.getData('ltPropIcons').length-1);
				
				let availablePositions = this.getActualPosition(500,50,[top,left],this.getData('ltPropPlacement'));
				
				// if(availablePositions!==this.getData('ltPropPlacement')){
				// 	console.log(" not available",this.getData('ltPropPlacement'));
				// }
				let lyteSpeedDial  = $L('.lyteSpeedDial');
				lyteSpeedDial.removeClass('lyteSpeedDialIconLeftAlign','lyteSpeedDialIconTopAlign','lyteSpeedDialIconRightAlign','lyteSpeedDialIconBottomAlign');
				$L('.lyteSpeedDial').removeClass('lyteSpeedDialIconCircleAlign');
					if(availablePositions==='right'){
						lyteSpeedDial.addClass('lyteSpeedDialIconRightAlign');			
					 }else if(availablePositions==='left'){
						lyteSpeedDial.addClass('lyteSpeedDialIconLeftAlign');
						
					 }else if(availablePositions==='top'){
				
						lyteSpeedDial.addClass('lyteSpeedDialIconTopAlign');
						
					 }else if(availablePositions==='bottom'){
						
						lyteSpeedDial.addClass('lyteSpeedDialIconBottomAlign');
					
					 }else if(availablePositions==='circle'){
						this.$node.style.setProperty('--lyte-speed-dial-icon-count',this.getData('ltPropIcons').length);
						$L('.lyteSpeedDial').addClass('lyteSpeedDialIconCircleAlign');
					 }else{
						$L('.lyteSpeedDial').addClass('lyteSpeedDialIconCircleAlign');
					 }	
					 this.$node.setData('ltPropSpeedShow',true);

			}			
		}
	}

    getActualPosition(popLength, popHeight, obj, preference) {
		        let width = this.getData('ltPropIconSize');
				let heigth = this.getData('ltPropIconSize');
		        popLength = this.getData('ltPropIcons').length *(20+width);
				let fullCircleRadius = popLength/(2*Math.PI);
				let halfCircleRadius = (popLength)/Math.PI;
				let quaterCircleradius = halfCircleRadius*2;

				fullCircleRadius = fullCircleRadius+(width/2);
				if(fullCircleRadius < 70){
					fullCircleRadius = 70;
				}
				
		

		let wlength   = window.innerWidth;
		let wHeight   = window.innerHeight;

		var tarTop    = obj[0]-(heigth/2);
		var tarLeft   = obj[1]-(width/2);

		var tarRight  = wlength-(tarLeft+width);
		var tarBottom = wHeight-(tarTop+heigth);

		
		let results = {			
			top               : tarTop>=popLength ? true : false,    		   
			left              : tarLeft>=popLength ? true : false,   
			right             : tarRight>=popLength ? true : false, 
			bottom            : tarBottom>=popLength ? true : false,  
	        circle            : tarBottom>=fullCircleRadius && tarLeft>=fullCircleRadius && tarRight>=fullCircleRadius && tarTop>=fullCircleRadius ? true : false,
		   
			leftSemiCircle    : tarBottom>=halfCircleRadius && tarLeft>=halfCircleRadius && tarTop>=halfCircleRadius ? true : false,
			rightSemiCircle   : tarBottom>=halfCircleRadius && tarRight>=halfCircleRadius && tarTop>=halfCircleRadius ? true : false,
			topSemiCircle     : tarLeft>=halfCircleRadius && tarRight>=halfCircleRadius && tarTop>=halfCircleRadius ? true : false,
			bottomSemiCircle  : tarBottom>=halfCircleRadius && tarLeft>=halfCircleRadius && tarRight>=halfCircleRadius ? true : false,
		     
			topLeftCorner     : tarBottom>=quaterCircleradius && tarRight>=quaterCircleradius ? true : false,
			topRightCorner    : tarBottom>=quaterCircleradius && tarLeft>=quaterCircleradius ? true : false,
			bottomLeftCorner  : tarTop>=quaterCircleradius && tarRight>=quaterCircleradius ? true : false,
			bottomRightCorner : tarTop>=quaterCircleradius && tarLeft>=quaterCircleradius ? true : false
		
		}		
			if(results.hasOwnProperty(preference) && results[preference]){
				if(preference==='circle' || preference==='leftSemiCircle' || preference==='rightSemiCircle' || preference==='topSemiCircle' || preference==='bottomSemiCircle' || preference==='topLeftCorner' || preference==='topRightCorner' || preference==='bottomLeftCorner' || preference==='bottomRightCorner'){
				   this.circleDataAdd(preference,fullCircleRadius,halfCircleRadius,quaterCircleradius);
				}
				return preference;
			}
			for(var key in results){
				if(results[key]){
					if(key==='circle' || key==='leftSemiCircle' || key==='rightSemiCircle' || key==='topSemiCircle' || key==='bottomSemiCircle' || key==='topLeftCorner' || key==='topRightCorner' || key==='bottomLeftCorner' || key==='bottomRightCorner'){
						this.circleDataAdd(preference,fullCircleRadius,halfCircleRadius,quaterCircleradius);
					 }
					return key;
				}
			}
		return "bottom";	
	}

    static observers(arg1) {
        return Object.assign(super.observers({
            _placementChange : function(){
              this.setData('ltPropSpeedShow',false);
                  this.speedShowFunction();
            }.observes('ltPropPlacement')
        }), arg1);
    }

    _() {
        _;
    }
}

LyteSpeedDialComponent._template = "<template class=\"lyteSpeedDialComp\" tag-name=\"lyte-speed-dial\"> <div class=\"lyteSpeedDial\" style=\"--lyte-speed-dial-icon-size : {{ltPropIconSize}}px;\"> <div class=\"lyteSpeedDialIconWrapper lyteSpeedDialButtonIcon\"> <img class=\"lyteSpeedDialMainIcon\" src=\"{{ltPropSpeedButton.src}}\" alt=\"{{ltPropSpeedButton.name}}\" onclick=\"{{action('speedIconClick',this)}}\"> </div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropSpeedShow}}\" is=\"case\" lc-id=\"lc_id_0\"> <template items=\"{{ltPropIcons}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"> <div id=\"{{index}}\" class=\"lyteSpeedDialIconWrapper\" lt-prop-title=\"{{if(ifEquals(ltPropTooltipShow,true),item.name,'')}}\" lt-prop-tooltip-config=\"{{ltPropTooltipConfig}}\" style=\"--lyte-speed-dial-icon-index: {{index}}\" onmouseover=\"{{action('onOver',this)}}\" onclick=\"{{action('iconClick',this)}}\"> <img src=\"{{item.src}}\" class=\"lyteSpeedDialIcon\" alt=\"{{item.name}}\"> </div> </template> </template></template></div> </template>";;
LyteSpeedDialComponent._dynamicNodes = [{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"concat","args":["'--lyte-speed-dial-icon-size : '","ltPropIconSize","'px;'"]}}}},{"t":"a","p":[1,1,1]},{"t":"s","p":[1,3],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"f","p":[1],"dN":[{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"concat","args":["'--lyte-speed-dial-icon-index: '","index"]}}}},{"t":"a","p":[1,1]}],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"]}];;
LyteSpeedDialComponent._observedAttributes = ["ltPropSpeedShow", "ltPropIcons"];
export { LyteSpeedDialComponent };

LyteSpeedDialComponent.register("lyte-speed-dial", {
    hash: "LyteSpeedDialComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
