// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/f738a93a5b8993932468c58062993791.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../component.js";
import $L from "../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";

/**
 * This component is used to select user desired items from the list of available items
 * @component  lyte-column-chooser
 * @version 3.64.0
 * @dependency lyte-checkbox
 *  /components/lyte-checkbox
 *  /theme/compiledCSS/default/ltr/lyte-ui-checkbox.css 
 */

class LyteColumnChooserComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
			/**
			 * @componentProperty {array} ltPropHeader
			 * @default []
			 */
			ltPropHeader : prop( 'array', { default : [] } ),
			ltPropSortable : prop( 'boolean', { default : false } ),
			ltPropYield : prop( 'boolean', { default : false } )
		}), arg1);		
	}

    add2Sort(__src, __target, __grp) {
		var __newelem = __grp.children[ __target ];
		__grp.addToSortable( __newelem );
	}

    execute(cb, args) {
		if( this.getMethods( cb ) ){
			args = Array.from( args );
			args.unshift( cb );
			return this.executeMethod.apply( this, args );
		}
	}

    switch(__old, __new) {
		var __elem = this.$node.getElementsByClassName( 'lyteSortableParent' )[ 0 ];

		if( __elem ){
			var place_index,
			__children = Array.from( __elem.children ).filter( function( item, index ){
				if( item.classList.contains( 'lyteSortablePlaceholder' ) ){
					place_index = index;
					return false;
				}
				return true;
			}),
			__src = __children[ __old ],
			__target = __children[ __new ],
			__header = this.data.ltPropHeader,
			__len = __header.length,
			__max = Math.max( __old, __new ),
			__count = 0,
			__over = 0;

			for( var i = 0; i < __len; i++ ){
				var __cur = __header[ i ];
				if( __cur.columnChooser ){
					if( __count == __old ){
						__src.setAttribute( "parent_index", i );
						__over++;
					} else if( __count == __new ){
						__target.setAttribute( "parent_index", i );
						__over++;
					}

					__count++;

					if( __over == 2 ){
						break;
					}
				}
			}

			if( this.getMethods( 'onBeforeDrop' ) ){
				this.executeMethod( 'onBeforeDrop', __src, __target, __elem, __old, __new );
			}

			window.requestAnimationFrame( this.add2Sort.bind( this, __old, __new > place_index ? __new + 1 : __new, __elem ) );

			return false;
		}
	}

    checked() {
		this.execute( 'onChecked', arguments );
	}

    unchecked() {
		this.execute( 'onUnchecked', arguments );
	}

    static methods(arg1) {
        return Object.assign(super.methods({
            checked : function(){
                /**
                  * @method onChecked
                  * @version 3.64.0
                  */
                this.execute( "onChecked", arguments );
            },

            unchecked : function(){
                /**
                  * @method onChecked
                  * @version 3.64.0
                  */
                this.execute( "onUnchecked", arguments );
            }
        }), arg1);
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            sort_obs : function( arg ){

                var __this = this,
                __data = __this.data;

                if( !__data.ltPropYield ){
                    var __elem = $L( '.lyteColumnChooser', __this.$node );
                    if( __data.ltPropSortable ){
                        __elem.sortable({
                            onSelect : function(){
                                return __this.execute( "onBeforeSelect", arguments );
                            },
                            onBeforeDrop : function( __checkbox, __grp, __place, __src, __target ){
                                if( __src != __target ){

                                    if( !__grp.classList.contains( 'lyteColumnChooser' ) ){
                                        __grp = __grp.parentNode;
                                    }

                                    window.requestAnimationFrame( __this.add2Sort.bind( __this, __src, __target, __grp ) );

                                    return __this.execute( "onBeforeDrop", arguments );
                                }
                            },
                            onDrop : function(){
                                __this.execute( "onDrop", arguments );
                            }
                        })
                    } else if( arg ) {
                        __elem.sortable( "destroy" );
                    }
                } else if( !arg ) {

                    var $node = this.$node;

                    $node.switchPositions = this.switch.bind( this );
                    $node.checked = this.checked.bind( this );
                    $node.unchecked = this.unchecked.bind( this );

                    var cb = "onRender";
                    this.getMethods( cb ) && this.executeMethod( cb, this.$node );
                }
            }.observes( 'ltPropSortable' ).on( 'didConnect' )
        }), arg1);
    }

    _() {
        _;
    }
}

LyteColumnChooserComponent._template = "<template tag-name=\"lyte-column-chooser\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropYield}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-yield yield-name=\"columnChooser\" class=\"lyteColumnChooser\" lt-prop-header=\"{{ltPropHeader}}\"></lyte-yield> </template><template default=\"\"> <div class=\"lyteColumnChooser\"> <template items=\"{{ltPropHeader}}\" item=\"item\" index=\"index\" is=\"for\" _new=\"true\"><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{item.columnChooser}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-checkbox lt-prop-checked=\"{{lbind(item.columnChooser.selected)}}\" lt-prop-label=\"{{expHandlers(item.name,'||',item.children[0].name)}}\" lt-prop-disabled=\"{{item.columnChooser.disabled}}\" on-checked=\"{{method('checked',item,index)}}\" on-unchecked=\"{{method('unchecked',item,index)}}\" parent_index=\"index_{{index}}\"></lyte-checkbox> </template></template> </template> </div> </template></template></template>";;
LyteColumnChooserComponent._dynamicNodes = [{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"i","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1,1],"cn":"default"},{"t":"f","p":[1,1],"dN":[{"t":"s","p":[0],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0}],"dc":[0],"hc":true,"trans":true,"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;
LyteColumnChooserComponent._observedAttributes = ["ltPropHeader", "ltPropSortable", "ltPropYield"];

/**
 * @syntax Non Yielded
 * <lyte-column-chooser></lyte-column-chooser>
 */

export { LyteColumnChooserComponent };

LyteColumnChooserComponent.register("lyte-column-chooser", {
    hash: "LyteColumnChooserComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
