// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/f8da6eb08e4c519c2b033b0d52a42d84.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../../node_modules/@slyte/component/index.js";

class LyteNoticeComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
			'ltPropNoticeType' : prop('string')
		}), arg1);		
	}

    static actions(arg1) {
        return Object.assign(super.actions({
            // Functions for event handling
        }), arg1);
    }

    static methods(arg1) {
        return Object.assign(super.methods({
            // Functions which can be used as callback in the component.
        }), arg1);
    }

    _() {
        _;
    }
}

LyteNoticeComponent._template = "<template tag-name=\"lyte-notice\"> <div class=\"lyteNotice lyte{{lyteUiCapitalizeName(ltPropNoticeType)}}Notice\"> <p class=\"lyteNoticeContent\"> <lyte-yield yield-name=\"notice\"></lyte-yield> </p> </div> </template>";;
LyteNoticeComponent._dynamicNodes = [{"t":"a","p":[1]},{"t":"i","p":[1,1,1],"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;
LyteNoticeComponent._observedAttributes = ["ltPropNoticeType"];
export { LyteNoticeComponent };
LyteNoticeComponent.register("lyte-notice", {
    hash: "LyteNoticeComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
