// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/fe68d17abc2def6a0be586941efed9b0.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../../node_modules/@slyte/core/index.js";
import { Component,LyteUiComponentComponentRegistry } from "../../component.js";
import $L from "../../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js";
import { arrayUtils } from "../../../node_modules/@slyte/component/index.js";

/*------------------------   NOTES   ------------------------*/
/*
  Things needed to document:
  1. No fill color will return transparent as color value on select. - done
*/

if(!window.ColorPicker_Util){
    window.ColorPicker_Util = {
		component : null,
		baseConverter : function(numberToConvert, oldBase, newBase) {
	        if (newBase == 10) {
	            return parseInt(numberToConvert, 16);
	        }
	        if (newBase == 16) {
	            return ('0'+parseInt(numberToConvert).toString(16)).slice(-2);
	        }
	        numberToConvert = numberToConvert + "";
	        numberToConvert = numberToConvert.toUpperCase();
	        var listOfCharacters = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
	        var dec = 0;
	        for (var i = 0; i <= numberToConvert.length; i++) {
	            dec += (listOfCharacters.indexOf(numberToConvert.charAt(i))) * (Math.pow(oldBase, (numberToConvert.length - i - 1)));
	        }
	        numberToConvert = "";
	        var magnitude = Math.floor((Math.log(dec)) / (Math.log(newBase)));
	        for (var i = magnitude; i >= 0; i--) {
	            var amount = Math.floor(dec / Math.pow(newBase, i));
	            numberToConvert = numberToConvert + listOfCharacters.charAt(amount);
	            dec -= amount * (Math.pow(newBase, i));
	        }
	        if (numberToConvert.length == 0){
	            numberToConvert = 0;
	        }
	        if (!numberToConvert){
	            numberToConvert = 0;
	        }
	        return numberToConvert;
	    },
	    convert3DigitTo6DigitRgbCode : function(rgbColor){
	    	return (rgbColor[0]+rgbColor[0]+rgbColor[1]+rgbColor[1]+rgbColor[2]+rgbColor[2]);
	    },
	    getHsvByRgbCode : function(rgbColor) {
	        var rgbColor = rgbColor.replace('#', ''),
	        red = window.ColorPicker_Util.baseConverter(rgbColor.substr(0, 2), 16, 10),
	        green = window.ColorPicker_Util.baseConverter(rgbColor.substr(2, 2), 16, 10),
	        blue = window.ColorPicker_Util.baseConverter(rgbColor.substr(4, 2), 16, 10),
	        maxValue, minValue, hue, saturation, valueBrightness;
	        if (red == 0 && green == 0 && blue == 0) {
	            var returnArray = {};
	            returnArray.hue = 0;
	            returnArray.saturation = 0;
	            returnArray.brightness = 0;
	            return returnArray;
	        }
	        red = red / 255;
	        green = green / 255;
	        blue = blue / 255;
	        maxValue = Math.max(red, green, blue);
	        minValue = Math.min(red, green, blue);
	       	hue = 0;
	        if (maxValue == minValue) {
	            hue = 0;
	            saturation = 0;
	        } else {
	            if (red == maxValue) {
	                hue = (green - blue) / (maxValue - minValue) / 1;
	            } else if (green == maxValue) {
	                hue = 2 + (blue - red) / 1 / (maxValue - minValue) / 1;
	            } else if (blue == maxValue) {
	                hue = 4 + (red - green) / (maxValue - minValue) / 1;
	            }
	            saturation = (maxValue - minValue) / maxValue;
	        }
	        hue = hue * 60;
	        valueBrightness = maxValue;
	        if (hue < 0){
	            hue += 360;
	        }
	        var returnArray = {};
	        returnArray.hue = hue;
	        returnArray.saturation = saturation;
	        returnArray.brightness = valueBrightness;
	        return returnArray;
	    },
	    getRgbCodeByRgbColors : function(red, green, blue) {
	        var red = window.ColorPicker_Util.baseConverter(red, 10, 16),
	        green = window.ColorPicker_Util.baseConverter(green, 10, 16),
	        blue = window.ColorPicker_Util.baseConverter(blue, 10, 16),
	        rgbColor;
	        red = red + "";
	        green = green + "";
	        blue = blue + "";
	        while (red.length < 2) {
	            red = "0" + red;
	        }
	        while (green.length < 2) {
	            green = "0" + green;
	        }
	        while (blue.length < 2) {
	            blue = "0" + "" + blue;
	        }
	        rgbColor = red + "" + green + "" + blue;
	        return rgbColor.toUpperCase();
	    },
	    getRgbColorsByRgbCode : function(rgbCode) {
	        var retArray = {};
	        retArray.red = window.ColorPicker_Util.baseConverter(rgbCode.substr(0, 2), 16, 10);
	        retArray.green = window.ColorPicker_Util.baseConverter(rgbCode.substr(2, 2), 16, 10);
	        retArray.blue = window.ColorPicker_Util.baseConverter(rgbCode.substr(4, 2), 16, 10);
	        return retArray;
	    },
	    getRgbColorsByHsv : function(hue, saturation, valueBrightness) {
	    	if (hue == 360){
	            hue = 0;
	        }
	        var Hi = Math.floor(hue / 60),
	        red,green,blue,f,p,q,t;
	        // if (hue == 360){
	        //     hue = 0;
	        // }
	        f = hue / 60 - Hi;
	        if (saturation > 1){
	            saturation /= 100;
	        }
	        if (valueBrightness > 1){
	            valueBrightness /= 100;
	        }
	        p = (valueBrightness * (1 - saturation));
	        q = (valueBrightness * (1 - (f * saturation)));
	        t = (valueBrightness * (1 - ((1 - f) * saturation)));
	        switch (Hi) {
	        case 0:
	            red = valueBrightness;
	            green = t;
	            blue = p;
	            break;
	        case 1:
	            red = q;
	            green = valueBrightness;
	            blue = p;
	            break;
	        case 2:
	            red = p;
	            green = valueBrightness;
	            blue = t;
	            break;
	        case 3:
	            red = p;
	            green = q;
	            blue = valueBrightness;
	            break;
	        case 4:
	            red = t;
	            green = p;
	            blue = valueBrightness;
	            break;
	        default:
	            red = valueBrightness;
	            green = p;
	            blue = q;
	            break;
	        }
	        if (saturation == 0) {
	            red = valueBrightness;
	            green = valueBrightness;
	            blue = valueBrightness;
	        }
	        red *= 255;
	        green *= 255;
	        blue *= 255;
	        red = Math.round(red);
	        green = Math.round(green);
	        blue = Math.round(blue);
	        return {
	            red: red,
	            green: green,
	            blue: blue
	        }
	    },
	    getRgbCodeByHsv : function(hue, saturation, valueBrightness) {
	        while (hue >= 360){
	            hue -= 360;
	        }
	        var colors = window.ColorPicker_Util.getRgbColorsByHsv(hue, saturation, valueBrightness);
	        return window.ColorPicker_Util.getRgbCodeByRgbColors(colors.red, colors.green, colors.blue);
	    },
	    getCmykByRgbColors: function(rgb){
				var c = 1 - rgb[0] / 255;
			    var m = 1 - rgb[1] / 255;
			    var y = 1 - rgb[2] / 255,k;

			    var min_cmy = Math.min(c, m, y);
			    c = (c - min_cmy) / (1 - min_cmy);
			    m = (m - min_cmy) / (1 - min_cmy);
			    y = (y - min_cmy) / (1 - min_cmy);
			    k = min_cmy;

			    c = isNaN( c )? 0 :c;
			    m = isNaN( m )? 0 :m;
			    y = isNaN( y )? 0 :y;

				return [c, m, y, k];
			},
			getRgbColorsByCmyk: function(c, m, y, k){
				if (c > 1){
		            c /= 100;
		        }
		        if (m > 1){
		            m /= 100;
		        }
		        if (y > 1){
		            y /= 100;
		        }
		        if (k > 1){
		            k /= 100;
		        }
				var red = 255 * (1-c) * (1-k),   //The red (R) color is calculated from the cyan (C) and black (K) colors
				green = 255 * (1-m) * (1-k),   //The green color (G) is calculated from the magenta (M) and black (K) colors:
				blue = 255 * (1-y) * (1-k);	  //The blue color (B) is calculated from the yellow (Y) and black (K) colors:

				red = Math.round(red);
		        green = Math.round(green);
		        blue = Math.round(blue);
		        return {
		            red: red,
		            green: green,
		            blue: blue
		        }
			},
			getRgbCodeByCmyk: function(c, m, y, k){
				var colors = window.ColorPicker_Util.getRgbColorsByCmyk(c, m, y, k);
				return window.ColorPicker_Util.getRgbCodeByRgbColors(colors.red, colors.green, colors.blue);
			},
			getDecimalToHexAlphaCode: function(alpha){
				if(alpha > 100){
					alpha = 100;
				}
				if(alpha < 0){
					alpha = 0;
				}
				alpha = Math.round(alpha * 255 /100);
				return window.ColorPicker_Util.baseConverter(alpha, 10, 16);
				// return (alpha + 0x10000).toString(16).substr(-2).toUpperCase();
				// for (var i = 0.1; i >= 0; i -= 0.01) {
				//     i = Math.round(i * 100) / 100; 
				//     var alpha = Math.round(i * 255);
				//     var hex = (alpha + 0x10000).toString(16).substr(-2).toUpperCase();
				//     var perc = Math.round(i * 100);
				//     console.log(perc + "% — " + hex + " (" + alpha + ")");
				// }
			},
			getHexToDecimalAlphaCode: function(alpha){
				var alpha = window.ColorPicker_Util.baseConverter(alpha, 16, 10);
				if(alpha > 255){
					alpha = 255;
				}
				if(alpha < 0){
					alpha = 0;
				}
				return Math.round(alpha * 100 / 255);
			},
	    getLeftPos : function(el) {
	        if (document.getBoxObjectFor) {
	            if (el.tagName != 'INPUT' && el.tagName != 'SELECT' && el.tagName != 'TEXTAREA'){
	                return document.getBoxObjectFor(el).x
	            }
	        }
	        var ret = el.offsetLeft;
	        while ((el = el.offsetParent) != null) {
	            if (el.tagName != 'HTML') {
	                ret += el.offsetLeft;
	                if (document.all){
	                    ret += el.clientLeft;
	                }
	            }
	        }
	        return ret;
	    },
	    getTopPos : function(el) {
	        if (document.getBoxObjectFor) {
	            if (el.tagName != 'INPUT' && el.tagName != 'SELECT' && el.tagName != 'TEXTAREA'){
	                return document.getBoxObjectFor(el).y
	            }
	        }
	        var ret = el.offsetTop;
	        while ((el = el.offsetParent) != null) {
	            if (el.tagName != 'HTML') {
	                ret += (el.offsetTop - el.scrollTop);
	                if (document.all){
	                    ret += el.clientTop;
	                }
	            }
	        }
	        return ret;
	    },
		getColorPicker : function(arg){
			var ele = arg.closest('colorpicker-ui')/*arg.closest('lyte-colorpicker')*/;
			// console.log(arg);
			// console.log(ele);
			// if(!ele){
			// 	var components = LytePopup.components;
			// 	for(var i =components.length - 1; i>=0; i--){
			// 		if(components[i].$node.parentElement.tagName == 'LYTE-COLORPICKER' && components[i].childComp.style.visibility == 'visible'){
			// 			ele = components[i].childComp;
			// 			break;
			// 		}
			// 	}
			// 	if(ele){
			// 		var div = ele.querySelector('.popoverWrapper');
			// 		if(div && div.classList.length > 1){
			// 			var className = div.className.split(" ")[1];
			// 			var colorpickers = document.querySelectorAll('lyte-colorpicker');
			// 			for(var i = 0 ; i<colorpickers.length ; i++){
			// 				if(colorpickers[i].component.getData('cpWrapperClass') === className){
			// 					ele = colorpickers[i];
			// 					break;
			// 				}
			// 			}
			// 		}
			// 	}
			// }
			return ele;
		},
		splitIntoSubArray : function(arr, count) {
			var newArray = [];
			while (arr.length > 0) {
		    	newArray.push(arr.splice(0, count)); 
			}
			return newArray;
		}
	};
}

if(!window.ColorPicker_EventUtil){
    window.ColorPicker_EventUtil = {
		__stopPropagation : false,
		__eventBound : false,
		__initOpacityMove: function(e,_this) {
			if(_this){
				if(e.type == "mousedown"){
					window._lyteUiUtils.addEventListenerGlobal('mousemove',window.ColorPicker_EventUtil.__mm = function(event){
						window.ColorPicker_EventUtil.__moveOnOpacityTrack(event,_this);
					});
					window._lyteUiUtils.addEventListenerGlobal('mouseup',window.ColorPicker_EventUtil.__mu = function(event){
						window.ColorPicker_EventUtil.__endDrag(event,_this);
					});
				}
				if(e.type == "touchstart"){
					window._lyteUiUtils.addEventListenerGlobal('touchmove',window.ColorPicker_EventUtil.__tm = function(event){
						window.ColorPicker_EventUtil.__moveOnOpacityTrack(event,_this);
					});
					window._lyteUiUtils.addEventListenerGlobal('touchend',window.ColorPicker_EventUtil.__te = function(event){
						window.ColorPicker_EventUtil.__endDrag(event,_this);
					});
				}
				_this.setData('opacityStatus', 1);
				_this.setData('calledForHueMove',true);
		        // _this.setData('poxYHue', window.ColorPicker_Util.getTopPos(_this.getData('divElHueBar')));
		        var opacitySlider = _this.$node.querySelector('.opacityslider__track');
		        _this.setData('opacityTrackSize', opacitySlider.offsetWidth /*_this.getData('cpInline') ? opacitySlider.offsetWidth : _this.childComp.querySelector('.opacityslider__track').offsetWidth*/);
		        _this.setData('posXOpacity', window.ColorPicker_Util.getLeftPos(opacitySlider /*_this.getData('cpInline') ? opacitySlider : _this.childComp.querySelector('.opacityslider__track')*/));
		        window.ColorPicker_EventUtil.__moveOnOpacityTrack(e,_this);
		        return false;
			}
	    },

		__initHueMove: function(e,_this) {
			if(_this){
				if(e.type == "mousedown"){
					window._lyteUiUtils.addEventListenerGlobal('mousemove',window.ColorPicker_EventUtil.__mm = function(event){
						window.ColorPicker_EventUtil.__moveOnHueBar(event,_this);
					});
					window._lyteUiUtils.addEventListenerGlobal('mouseup',window.ColorPicker_EventUtil.__mu = function(event){
						window.ColorPicker_EventUtil.__endDrag(event,_this);
					});
				}
				if(e.type == "touchstart"){
					window._lyteUiUtils.addEventListenerGlobal('touchmove',window.ColorPicker_EventUtil.__tm = function(event){
						window.ColorPicker_EventUtil.__moveOnHueBar(event,_this);
					});
					window._lyteUiUtils.addEventListenerGlobal('touchend',window.ColorPicker_EventUtil.__te = function(event){
						window.ColorPicker_EventUtil.__endDrag(event,_this);
					});
				}
				_this.setData('setPrevCmyk',false);
				_this.setData('hueStatus', 1);
				_this.setData('calledForHueMove',true);
		        // _this.setData('poxYHue', window.ColorPicker_Util.getTopPos(_this.getData('divElHueBar')));
		        var colorSlider = _this.$node.querySelector('.colorSlider_hueBar');
		        _this.setData('hueBarSize', colorSlider.offsetWidth /*_this.getData('cpInline') ? colorSlider.offsetWidth : _this.childComp.querySelector('.colorSlider_hueBar').offsetWidth*/);
		        _this.setData('poxXHue', window.ColorPicker_Util.getLeftPos(colorSlider /*_this.getData('cpInline') ? colorSlider : _this.childComp.querySelector('.colorSlider_hueBar')*/));
		        window.ColorPicker_EventUtil.__moveOnHueBar(e,_this);
		        return false;
			}
	    },

		__initPaletteMove : function(e,_this) {
	    	if(_this){
	    		_this.setData('setPrevCmyk',false);
	    		if(e.type == "mousedown"){
					var popOver = _this.$node.closest('lyte-wormhole');
					if(!_this.getData('cpInline') && popOver && popOver.component.parent && popOver.component.parent.closest('lyte-colorpicker').getData('ltPropCloseOnBodyClick')){
						popOver.component.parent.setData('ltPropCloseOnBodyClick',false);
					}
					window._lyteUiUtils.addEventListenerGlobal('mousemove',window.ColorPicker_EventUtil.__mm = function(event){
						window.ColorPicker_EventUtil.__moveOnPalette(event,_this);
					});
					window._lyteUiUtils.addEventListenerGlobal('mouseup',window.ColorPicker_EventUtil.__mu = function(event){
						window.ColorPicker_EventUtil.__endDrag(event,_this);
					});
				}
				if(e.type == "touchstart"){
					var popOver = _this.$node.closest('lyte-wormhole');
					if(!_this.getData('cpInline') && popOver && popOver.component.parent && popOver.component.parent.closest('lyte-colorpicker').getData('ltPropCloseOnBodyClick')){
						popOver.component.parent.setData('ltPropCloseOnBodyClick',false);
					}
					window._lyteUiUtils.addEventListenerGlobal('touchmove',window.ColorPicker_EventUtil.__tm = function(event){
						window.ColorPicker_EventUtil.__moveOnPalette(event,_this);
					});
					window._lyteUiUtils.addEventListenerGlobal('touchend',window.ColorPicker_EventUtil.__te = function(event){
						window.ColorPicker_EventUtil.__endDrag(event,_this);
					});
				}
	    		var posdivElPalette = _this.getData('posdivElPalette');
		    	var circleOffsetBecauseOfWinWidget = _this.getData('circleOffsetBecauseOfWinWidget');
		    	var divElPalette =  _this.$node.querySelector('#colorDiv') /*_this.getData('cpInline') ? _this.$node.querySelector('#colorDiv') : _this.childComp.querySelector('#colorDiv')*/;
		    	var divElPaletteOffset = divElPalette.getBoundingClientRect();
		    	var circleOffsetSize = _this.getData('circleOffsetSize');
		        if (document.all){
		            e = window.event || window.event;
		        }
		        _this.__ffHackWinWidget();
		        posdivElPalette.x = divElPaletteOffset.left /*window.ColorPicker_Util.getLeftPos(divElPalette)*/ + circleOffsetBecauseOfWinWidget;
		        posdivElPalette.y = divElPaletteOffset.top /*window.ColorPicker_Util.getTopPos(divElPalette)*/ + circleOffsetBecauseOfWinWidget;
		        _this.setData('paletteSize',{'height' : divElPalette.offsetHeight, 'width' : divElPalette.offsetWidth});
		        _this.setData('posdivElPalette',posdivElPalette);
		        _this.setData('dragStatus', 1);
		        _this.setData('paletteMaxX', (divElPalette.offsetWidth - circleOffsetSize));
		        _this.setData('paletteMaxY', (divElPalette.offsetHeight - circleOffsetSize));
		        window.ColorPicker_EventUtil.__moveOnPalette(e,_this);
		        posdivElPalette = null;
		        divElPalette = null;
		        return false;
	    	}
	    },

	    __moveOnPalette : function(e,_this) {
			
	    	if(_this){
				
	    		if(e.type == "mousedown" || e.type == "mousemove"){
	    			e.preventDefault();
	    		}
		    	var posdivElPalette = _this.getData('posdivElPalette');
		    	var circleOffsetSize = _this.getData('circleOffsetSize');
		    	var divElPaletteCircle = _this.$node.querySelector('.colorSlider_palette_circle') /*_this.getData('cpInline') ? _this.$node.querySelector('.colorSlider_palette_circle') : _this.childComp.querySelector('.colorSlider_palette_circle')*/;
		        if (_this.getData('dragStatus') != 1){
		            return;
		        }
		        if (_this.getData('clickOnPaletteInProgress')){
		            return;
		        }
		        _this.setData('clickOnPaletteInProgress', true);
		        _this.setData('dragged',true);
		        if (document.all){
		            e = event;
		        }
		        var leftEl = posdivElPalette.x;
		        var topEl = posdivElPalette.y;
		        var left,top;
		        if(e.type == "mousedown" || e.type == "mousemove"){
		        	left = e.clientX/* + document.documentElement.scrollLeft*/ - leftEl - circleOffsetSize;
		        	top = e.clientY/* + document.documentElement.scrollTop*/ - topEl - circleOffsetSize;
		        }
		        if(e.type == "touchstart" || e.type == "touchmove"){
		        	left = e.touches[0].clientX/* + document.documentElement.scrollLeft*/ - leftEl - circleOffsetSize;
		        	top = e.touches[0].clientY/* + document.documentElement.scrollTop*/ - topEl - circleOffsetSize;
		        }
		        if (left < circleOffsetSize * -1){
		            left = circleOffsetSize * -1;
		        }
		        if (top < circleOffsetSize * -1){
		            top = circleOffsetSize * -1;
		        }
		        if (left > _this.getData('paletteMaxX')){
		            left = _this.getData('paletteMaxX');
		        }
		        if (top > _this.getData('paletteMaxY')){
		            top = _this.getData('paletteMaxY');
		        }
		        // if(_this.getData('cpInline')){
		        // 	var xscroll = window.pageXOffset;
          //           var yscroll = window.pageYOffset;
          //           left += xscroll;
          //           top += yscroll;
		        // }
		        divElPaletteCircle.style.left = left + 'px';
		        divElPaletteCircle.style.top = top + 'px';
		        _this.setData('currentSaturation', Math.round(((left + circleOffsetSize) / _this.getData('paletteSize').width) * 100));
		        _this.setData('currentBrightness', 100 - Math.round(((top + circleOffsetSize) / _this.getData('paletteSize').height) * 100));
		        _this.__setCurrentRgbCode();
		        // _this.__setBgColorPreviewDiv();
		        _this.__updateRgbInForm();
		        _this.__changeColorInOpacityTrack();
				_this.__setOpacityColorPreviewDiv();
		        _this.executeOnChange();
		        _this.setData('clickOnPaletteInProgress', false);
		        posdivElPalette = null;
		        divElPaletteCircle = null;
		        circleOffsetSize = null;
	    	}
	    },

	    __moveOnHueBar: function(e,_this) {
	    	if(_this){
	    		if(e.type == "mousedown" || e.type == "mousemove"){
	    			e.preventDefault();
	    		}
		        if (_this.getData('hueStatus') != 1){
		            return;
		        }
		        _this.setData('dragged',true);
		        if (document.all){
		            e = event;
		        }
		        var sliderOffset = _this.getData('sliderOffset');
		        var leftPos = _this.getData('poxXHue');
		        var diff;
		        if(e.type == "mousedown" || e.type == "mousemove"){
		        	diff = e.clientX + document.documentElement.scrollLeft - leftPos - sliderOffset;
		        }
		        if(e.type == "touchstart" || e.type == "touchmove"){
		        	diff = e.touches[0].clientX + document.documentElement.scrollLeft - leftPos - sliderOffset;
		        }
		        if (diff > _this.getData('hueBarSize') - sliderOffset){
		            diff = _this.getData('hueBarSize') - sliderOffset;
		        }
		        if (diff < 0 - sliderOffset){
		            diff = 0 - sliderOffset;
		        }
		        var ele = _this.$node /*_this.getData('cpInline') ? _this.$node : _this.childComp*/;
	        	ele.querySelector('.colorSlider_sliderHandle').style.left = diff + 'px';
		        var hue = Math.round((_this.getData('hueBarSize') - (diff + sliderOffset)) * (360 / _this.getData('hueBarSize')));
		        if (hue >= 360){
		            hue = 360;
		        }
		        _this.setData('currentHue', hue);
		        _this.__setCurrentRgbCode();
		        _this.__setPaletteBgColor();
		        // _this.__setBgColorPreviewDiv();
		        _this.__updateRgbInForm();
		        _this.__changeColorInOpacityTrack();
				_this.__setOpacityColorPreviewDiv();
		        _this.executeOnChange();
	    	}
	    },

	    __moveOnOpacityTrack: function(e,_this) {
	    	if(_this){
	    		if(e.type == "mousedown" || e.type == "mousemove"){
	    			e.preventDefault();
	    		}
		        if (_this.getData('opacityStatus') != 1){
		            return;
		        }
		        // _this.setData('dragged',true);
		        if (document.all){
		            e = event;
		        }
		        var leftPos = _this.getData('posXOpacity');
		        var sliderOffset = _this.getData('sliderOffset');
		        var diff;
		        if(e.type == "mousedown" || e.type == "mousemove"){
		        	diff = e.clientX + document.documentElement.scrollLeft - leftPos - sliderOffset;
		        }
		        if(e.type == "touchstart" || e.type == "touchmove"){
		        	diff = e.touches[0].clientX + document.documentElement.scrollLeft - leftPos - sliderOffset;
		        }
		        if (diff > _this.getData('opacityTrackSize') - sliderOffset){
		            diff = _this.getData('opacityTrackSize') - sliderOffset;
		        }
		        if (diff < 0 - sliderOffset){
		            diff = 0 - sliderOffset;
		        }
		        var diffPercent = parseInt(((diff + sliderOffset) / _this.getData('opacityTrackSize')) * 100);
		        var ele = _this.$node/*_this.getData('cpInline') ? _this.$node : _this.childComp*/;
		        ele.querySelector('.opacityslider__circlethumb').style.left = diff + 'px';
				_this.setData("opacity", diffPercent);
		        _this.__setOpacityColorPreviewDiv();
		        _this.__updateOpacityInForm();
		        _this.executeOnChange();
	    	}
	    },

	    __endDrag : function(e,_this) {
	    	if(_this && _this.$node){
				e.preventDefault();
	    		if(_this.getData('dragStatus') == 0 && _this.getData('hueStatus') == 0 && _this.getData('opacityStatus') == 0){
		    		return;
		    	}
	    		if(e.type == "mouseup"){
					window._lyteUiUtils.removeEventListenerGlobal('mousemove',window.ColorPicker_EventUtil.__mm);
					window._lyteUiUtils.removeEventListenerGlobal('mouseup',window.ColorPicker_EventUtil.__mu);
					window.ColorPicker_EventUtil.__mm = undefined;
					window.ColorPicker_EventUtil.__mu = undefined;
				}
				if(e.type == "touchend"){
					window._lyteUiUtils.removeEventListenerGlobal('touchmove',window.ColorPicker_EventUtil.__tm);
					window._lyteUiUtils.removeEventListenerGlobal('touchend',window.ColorPicker_EventUtil.__te);
					window.ColorPicker_EventUtil.__tm = undefined;
					window.ColorPicker_EventUtil.__te = undefined;
				}
		    	window.ColorPicker_EventUtil.__stopPropagation = true;
		        if (_this.getData('dragStatus') == 1) {
		            // _this.__updateHsvInForm();
		            _this.__updateRgbInForm();
		        }
		        _this.setData('dragStatus', 0);
		        _this.setData('hueStatus', 0);
		        _this.setData('opacityStatus',0);
		        if(_this.getData('dragged')){
		        	_this.setData('dragged', false);
		        }
		        if(_this.getData('counter') != 0){
		        	_this.setData('counter',0);
		        }
		        if(_this.getData('restrictOnChange')){
		        	_this.setData('restrictOnChange',false);
		        }
		        var popup = _this.getData('cpInline') ? _this.$node.closest('.lytePopover') : null;
		        if(popup){
		        	popup._callee.ltProp('stopClick', true);
		        }

	    	}
	    	window.ColorPicker_Util.component = null;
			
			var popOver = _this.$node.closest('lyte-wormhole');
			if(!_this.getData('cpInline') && popOver && popOver.component.parent && popOver.component.parent.closest('lyte-colorpicker').getData('ltPropCloseOnBodyClick')){
				setTimeout(function(){
					popOver.component.parent.setData('ltPropCloseOnBodyClick',true);
				},10)
				
			}
	    }
	};
}

class ColorpickerUiComponent extends Component {
    constructor() {
        super();
    }

    data(arg1) {
		return Object.assign(super.data({
			//Component Variables
			"cpDraggable" : prop("boolean",{"default" : false}),
			"cpCloseOnBodyClick" : prop("boolean",{"default" : true}),
			"cpCloseOnSelection" : prop("boolean",{"default" : false}),
			"cpAdvancedColorButton" : prop("boolean",{"default" : true}),
			"cpAppearance" : prop("string",{"default" : "callout"}),
			"cpAccent" : prop('boolean',{"default": false}),
			"cpAccentColor" : prop('array',{"default": []}),
			// "cpColorPreview" : Lyte.attr("boolean",{"default" : true}),
			"cpStandardColors" : prop("boolean",{"default" : true}),
			"cpDefaultColorEnable" : prop("boolean",{"default" : false}),
			"cpUsedColors" : prop("boolean",{"default" : true}),
			"cpBasicColorPicker" : prop("boolean",{"default" : true}),
			"cpSelectedColor" : prop("object",{"default" : null}),
			"cpDynamicColor" : prop("object",{"default" : null}),
			"cpOriginElement" : prop("string",{"default" : ""}),
			"cpWrapperClass" : prop("string",{"default":"lyteColorPickerPopover"}),
			"cpCurrentRgbCode" : prop("string",{ "default" : "90c3d4"}),
			"cpBoundary" : prop("object",{"default" : null}),
			"cpScrollable":prop("boolean",{"default": true}),
			"cpYield":prop("boolean",{"default":false}),
			"cpNoFillButton":prop("boolean",{"default":false}),
			"cpNoFillLabel":prop("string",{"default":""}),
			"cpPaletteLabel": prop("string",{"default":""}),
			"cpInline" : prop("boolean",{"default" : false}),
			"cpAdvancedNoFillButton" : prop("boolean",{"default": false}),
			"cpColorFormats" : prop("array"),
			"cpDoneBtnText" : prop("string",{"default" : "lyte.colorpicker.done"}),
			"cpCancelBtnText" : prop("string",{"default" : "lyte.colorpicker.cancel"}),
			"cpBackBtnText" : prop("string",{"default" : "lyte.colorpicker.back"}),
			"cpApplyBtnText" : prop("string",{"default" : "lyte.colorpicker.apply"}),
			"cpStandardColorArray" : prop("array",{"default" : []}),
			"cpDefaultColor" : prop("string",{'default' : "rgb(0,0,0)"}),
			"cpDuration" : prop("number",{"default":600}),
			"cpBoardColor" : prop("string"),
			"cpAvailableColors" : prop("array",{"default" : []}),
			"cpMoreColorOptionSelected" : prop("boolean",{"default" : false}),
			"cpOpacityLabel" : prop("string"),
			"cpUsedColorsNumber" : prop("number",{"default":10}),
			"cpDropBoxClass" : prop("string",{"default":''}),
			"cpDisplayPreviousColor" : prop("boolean",{"default" : false}),
			//Local variables
			"stopObserverOnBoardColorChange" : prop("boolean",{"default" : false}),
			

			"usedColors" : prop("array"),
			
			"polygonPoints" : prop("string",{"default" : "4.4,10.3 3.7,9.6 6.7,6.5 3.6,3.4 4.4,2.7 8.1,6.5"}),
			"noFillPath" : prop("string",{"default" : "M8,0C3.6,0,0,3.6,0,8s3.6,8,8,8s8-3.6,8-8S12.4,0,8,0z M8,1c1.8,0,3.3,0.7,4.6,1.7l-9.9,9.9C1.7,11.3,1,9.8,1,8 C1,4.1,4.1,1,8,1z M8,15c-1.8,0-3.3-0.7-4.6-1.7l9.9-9.9C14.3,4.7,15,6.2,15,8C15,11.9,11.9,15,8,15z"}),
			"moreColorsPath" : prop("string", {"default" : "M299.7-0.2C128.3-0.2-7,134,0.2,305.2C7.4,476.9,138,577.3,259.7,595.9c88.2,13.4,107.8-45.1,87.3-67.3 c-35.5-38.3-22.6-68.6-7.3-82.5c17.4-15.7,50.3-17.3,77.3-13.8c65.5,8.5,175-40,182.2-144.8C611,116.5,471.1-0.2,299.7-0.2z M118.1,299.7c-27.7,0-50.1-22.4-50.1-50.1c0-27.7,22.4-50.1,50.1-50.1c27.7,0,50.1,22.4,50.1,50.1 C168.2,277.3,145.8,299.7,118.1,299.7z M217.6,166.6c-27.7,0-50.1-22.4-50.1-50.1c0-27.7,22.4-50.1,50.1-50.1 c27.7,0,50.1,22.4,50.1,50.1C267.7,144.1,245.2,166.6,217.6,166.6z M384.9,166.9c-27.7,0-50.1-22.4-50.1-50.1 c0-27.7,22.4-50.1,50.1-50.1c27.7,0,50.1,22.4,50.1,50.1C435,144.4,412.6,166.9,384.9,166.9z M485.2,299.9 c-27.7,0-50.1-22.4-50.1-50.1c0-27.7,22.4-50.1,50.1-50.1c27.7,0,50.1,22.4,50.1,50.1C535.2,277.5,512.8,299.9,485.2,299.9z"}),
			

			//AdvColorPicker Local Variables
			"currentHue" : prop("number",{ "default" : 0}),
			"currentBrightness" : prop("number",{ "default" : 100}),
			"currentSaturation" : prop("number",{ "default" : 100}),
			"paletteSize" : prop("object"/*,{ "default" : {'height':160,'width':260}}*/),
			"circleOffsetSize" : prop("number",{ "default" : 6}),
			"circleOffsetBecauseOfWinWidget" : prop("number",{ "default" : 0}),
			"posdivElPalette" : prop("object",{ "default" : {}}),
			"dragStatus" : prop("number",{ "default" : 0}),
			"paletteMaxX" : prop("number",{ "default" : 0}),
			"paletteMaxY" : prop("number",{ "default" : 0}),
			"clickOnPaletteInProgress" : prop("boolean",{ "default" : false}),
			"component" : prop("object",{"default" : null}),
			"hueStatus" : prop("number",{ "default" : 0}),
			"poxYHue" : prop("number", { "default" : 0}),
			"poxXHue" : prop("number", { "default" : 0}),
			"dropButtonValue" : prop("string",{"default" : "HEX"}),
			"calledForHueMove" : prop("boolean",{"default":false}),
			"changedColor" : prop("object"),
			"tooltipTextChanged" : prop('boolean',{'default' : false}),
			"inputChangeEvnt" : prop("boolean",{"default" : false}),
			"hueBarSize" : prop("number"),
			"opacityStatus" : prop("number",{"default":0}),
			"opacityTrackSize" : prop("number"),
			"posXOpacity" : prop("number", { "default" : 0}),
			"opacity" : prop("number"),
			"sliderOffset" : prop("number",{"default" : 4}),
			"dragged" : prop("boolean", {"default" : false}),
			"counter" : prop("number"),
			"cmykColors" : prop("array",{"default":[]}),
			"setPrevCmyk" : prop("boolean",{"default":"false"}),
			"restrictOnChange" : prop("boolean",{"default": false}),
			"prevColor" : prop("object", {"default":{}}),
			"cpShowHiddenColor" : prop('boolean',{"default": false}),
			"colorFormat" : prop("object", {"default" : {"R" : "Red", "G" : "Green", "B" : "Blue", "H" : "Hue", "S" : "Saturation", "V" : "Brightness", "C" : "Cyan", "M" : "Magenta", "Y" : "Yellow", "K" : "Key", "A" : "Opacity"}})
		}), arg1);		
	}

    didConnect() {
		var self = this;
		this.$node.selectColor = function( selectedColor ){
			if(selectedColor){
				if( !(selectedColor.rgb && selectedColor.rgb.length) && !(selectedColor.rgba && selectedColor.rgba.length) && !(selectedColor.hex && selectedColor.hex.length)  &&  !(selectedColor['hex-alpha'] && selectedColor['hex-alpha'].length) && !(selectedColor.cmyk && selectedColor.cmyk.length)   ){
					self.nofillFunc(undefined);
					return;
				}
			}else if(!selectedColor){
				self.nofillFunc(undefined);
				return;
			}
			self.setData('cpSelectedColor',selectedColor);
			self.setData('cpDynamicColor',selectedColor);
			var usedColor  = self.getData('usedColors');
			var selectedColor = self.getData('cpSelectedColor');
			if(!selectedColor.nofill && (!usedColor[0] || !usedColor[0].hex || usedColor[0].hex !== selectedColor.hex)){
				var rgbCode = self.getData('cpCurrentRgbCode');
				var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(rgbCode);
				var cmykColors = self.getData('cmykColors');
				var type = self.getData('opacity') < 100 ? 'advanced' : 'basic'; 
				var selectedColor = {
					"hex" : "#"+rgbCode,
					"rgb" : "rgb("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+")",
					"opacity" : self.getData('opacity'),
					"hex-alpha" : "#"+rgbCode+window.ColorPicker_Util.getDecimalToHexAlphaCode(self.getData('opacity')),
					"rgba" : "rgba("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+", "+(self.getData('opacity')/100)+")",
					"cmyk" : cmykColors,
					"type" : self.getData('cpSelectedColor') ? self.getData('cpSelectedColor').type : type,
					"nofill" : self.getData('cpSelectedColor') ? self.getData('cpSelectedColor').nofill : null
				};
				self.setData('cpSelectedColor',selectedColor);
				self.setData('cpDynamicColor',selectedColor);
				self.addedUsedColor();
			}
		}

		
		var focused_container = null;
		this.$node.addEventListener('keydown',function(event){
			var focusedItem = document.activeElement;
			var key = event.keyCode;
			if(key === 9){
				if($L('.lyteColorPicker__colorpan[tabindex="0"]',self.$node)[0]){
					$L('.lyteColorPicker__colorpan[tabindex="0"]',self.$node)[0].setAttribute('tabindex',-1);
				}
				
			}else if(key === 8){
				if(focusedItem && focusedItem.classList.contains('lyteColorPicker__colorpan')){
					focusedItem.setAttribute('tabindex',-1);
				}else if(focusedItem.classList.contains('cPDropDown')){
					focusedItem.component.data.ltPropAriaButton = 'color Format';
				}
			}else if(key === 13){
				if(focusedItem.classList.contains('lyteColorPicker__morecolorbutton')){
					focusedItem.click();
					self.$node.focus();
				}else if(focusedItem.classList.contains('LCP_predefinedPallete') || focusedItem.classList.contains('LCP_standardColorCont') || focusedItem.classList.contains('LCP_defaultColorCont') || focusedItem.classList.contains('usedColor__container')){
					focused_container = focusedItem;
					var focused_color = focusedItem.querySelector('[tabindex="0"]') || focusedItem.querySelector('.lyteColorPicker__colorpan');
					focused_color.setAttribute('tabindex',0);
					focused_color.focus();
				}else if(focusedItem.classList.contains('lyteColorPicker__colorpan')){
					focusedItem.click();
				}else if(focusedItem.classList.contains('LCP_backbtn')){
					focusedItem.click();
				}
			}else if(key === 27){
				if(focusedItem.classList.contains('lyteColorPicker__colorpan') && focused_container){
					focusedItem.setAttribute('tabindex',-1);
					focused_container.focus();
				}
			}else if(key === 37){
				if(focusedItem.classList.contains('lyteColorPicker__colorpan')){
					var col = parseInt(focusedItem.dataset.col);
					var left_color = focusedItem.parentNode.querySelector('[data-col="' + ( col - 1 ) + '"]');
					if(left_color){
						focusedItem.setAttribute('tabindex',-1);
						left_color.setAttribute('tabindex',0);
						left_color.focus();
					}
				}else if(focusedItem.classList.contains('lyteColorPicker__palettecontainer')){
					var focused_color = focusedItem.querySelector('[tabindex="0"]') || focusedItem.querySelector('.lyteColorPicker__colorpan');
					focused_color.setAttribute('tabindex',0);
					focused_color.focus();
				}
			}else if(key === 39){
				if(focusedItem.classList.contains('lyteColorPicker__colorpan')){
					var col = parseInt(focusedItem.dataset.col);
					var right_color = focusedItem.parentNode.querySelector('[data-col="' + ( col + 1 ) + '"]');
					if(right_color){
						focusedItem.setAttribute('tabindex',-1);
						right_color.setAttribute('tabindex',0);
						right_color.focus();
					}
				}else if(focusedItem.classList.contains('lyteColorPicker__palettecontainer')){
					var focused_color = focusedItem.querySelector('[tabindex="0"]') || focusedItem.querySelector('.lyteColorPicker__colorpan');
					focused_color.setAttribute('tabindex',0);
					focused_color.focus();
				}
			}else if(key == 38){
				if(focusedItem.classList.contains('lyteColorPicker__colorpan')){
					var row = parseInt(focusedItem.dataset.row);
					var col = parseInt(focusedItem.dataset.col);
					row -= 1;
					var top_color = focusedItem.parentNode.parentNode.querySelector('[data-row="' + ( row ) + '"][ data-col="' + col + '"]');
					if(top_color){
						var isVisible = window.getComputedStyle(top_color.parentNode).display == 'none' ? false : true;
						while(!isVisible){
							row -= 1;
							top_color = top_color.parentNode.parentNode.querySelector('[data-row="' + ( row ) + '"][ data-col="' + col + '"]');
							isVisible = window.getComputedStyle(top_color.parentNode).display == 'none' ? false : true;
						}
						focusedItem.setAttribute('tabindex',-1);
						top_color.setAttribute('tabindex',0);
						top_color.focus();
					}
				}else if(focusedItem.classList.contains('lyteColorPicker__palettecontainer')){
					var focused_color = focusedItem.querySelector('[tabindex="0"]') || focusedItem.querySelector('.lyteColorPicker__colorpan');
					focused_color.setAttribute('tabindex',0);
					focused_color.focus();
				}
			}else if( key === 40){
				if(focusedItem.classList.contains('lyteColorPicker__colorpan')){
				var row = parseInt(focusedItem.dataset.row);
				var col = parseInt(focusedItem.dataset.col);
				row += 1;
				var down_color = focusedItem.parentNode.parentNode.querySelector('[data-row="' + ( row ) + '"][ data-col="' + col + '"]');
				if(down_color){
					var isVisible = window.getComputedStyle(down_color.parentNode).display == 'none' ? false : true;
					while(!isVisible){
						row += 1;
						down_color = down_color.parentNode.parentNode.querySelector('[data-row="' + ( row ) + '"][ data-col="' + col + '"]');
						isVisible = window.getComputedStyle(down_color.parentNode).display == 'none' ? false : true;
					}
					focusedItem.setAttribute('tabindex',-1);
					down_color.setAttribute('tabindex',0);
					down_color.focus();
				}
			}else if(focusedItem.classList.contains('lyteColorPicker__palettecontainer')){
				var focused_color = focusedItem.querySelector('[tabindex="0"]') || focusedItem.querySelector('.lyteColorPicker__colorpan');
				focused_color.setAttribute('tabindex',0);
				focused_color.focus();
			}
			}
		})
		// shortcut.register( 'tab' , function() {
		// 	var focusedItem = document.activeElement;
			
		// });
		// shortcut.register( 'enter' , function() {
			
		// 	var focusedItem = document.activeElement;
			
		// });
		// shortcut.register( 'esc' , function() {
		// 	var focusedItem = document.activeElement;
			
		// });

	this.setData('component',this);

	}

    isCurrentRgbCodeDifferent() {
		var boardColor = this.getData('cpBoardColor'),
			currentRgbCode = this.getData('cpCurrentRgbCode'), code;
		if(/#/.test(boardColor)){
			if(boardColor.length == 9){
				code = boardColor.substring(1,7);
			}
			if(boardColor.length == 7){
				code = boardColor.substring(1);
			}
			else if(boardColor.length == 4){
				code = window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(boardColor.substring(1));
			}
		}
		else if(/rgba/.test(boardColor)){
			var rgbColor = boardColor.substring(5,boardColor.length-1).split(",");
			code = window.ColorPicker_Util.getRgbCodeByRgbColors(rgbColor[0],rgbColor[1],rgbColor[2]);
		}
		else if(/rgb/.test(boardColor)){
			var rgbColor = boardColor.substring(4,boardColor.length-1).split(",");
			code = window.ColorPicker_Util.getRgbCodeByRgbColors(rgbColor[0],rgbColor[1],rgbColor[2]);
		}

		if(code && code.toUpperCase() == currentRgbCode.toUpperCase()){
			return false;
		}
		return true;
	}

    addPasteEvent() {
		var comp = this.$node,
			format = this.getData('dropButtonValue'),
			self = this;
		if(format === window._lyteUiUtils.i18n("HEX")){
			comp.querySelector('#lyteCPShowValue input').addEventListener('paste', function(e){
				self.pasteFunc(e, "HEX");
			});
			comp.querySelector('#lyteCP__A input').addEventListener('paste', function(e){
				self.pasteFunc(e, "A");
			});
		}
		else if(format === window._lyteUiUtils.i18n("RGB")){
			comp.querySelector('#lyteCP__R input').addEventListener('paste', function(e){
				self.pasteFunc(e, "R");
			});
			comp.querySelector('#lyteCP__G input').addEventListener('paste', function(e){
				self.pasteFunc(e, "G");
			});
			comp.querySelector('#lyteCP__B input').addEventListener('paste', function(e){
				self.pasteFunc(e, "B");
			});
			comp.querySelector('#lyteCP__A input').addEventListener('paste', function(e){
				self.pasteFunc(e, "A");
			});
		}
		else if(format === window._lyteUiUtils.i18n("HSV")){
			comp.querySelector('#lyteCP__H input').addEventListener('paste', function(e){
				self.pasteFunc(e, "H");
			});
			comp.querySelector('#lyteCP__S input').addEventListener('paste', function(e){
				self.pasteFunc(e, "S");
			});
			comp.querySelector('#lyteCP__V input').addEventListener('paste', function(e){
				self.pasteFunc(e, "V");
			});
			comp.querySelector('#lyteCP__A input').addEventListener('paste', function(e){
				self.pasteFunc(e, "A");
			});
		}
		else if(format === window._lyteUiUtils.i18n("CMYK")){
			comp.querySelector('#lyteCP__C input').addEventListener('paste', function(e){
				self.pasteFunc(e, "C");
			});
			comp.querySelector('#lyteCP__M input').addEventListener('paste', function(e){
				self.pasteFunc(e, "M");
			});
			comp.querySelector('#lyteCP__Y input').addEventListener('paste', function(e){
				self.pasteFunc(e, "Y");
			});
			comp.querySelector('#lyteCP__K input').addEventListener('paste', function(e){
				self.pasteFunc(e, "K");
			});
		}

	}

    initializeAdvColorPicker(event, ele, restrictCmykChange) {
		this.setData('dropButtonValue', this.getData('cpColorFormats')[0]);
		var imageObj = ele.querySelector('#lyteCPImgDiv')
		var divElPalette = ele.querySelector('#colorDiv');
		var divElPaletteCircle = ele.querySelector('.colorSlider_palette_circle');
		var cpDiv = ele.querySelector('.lyteColorPicker--advanced');
		if(!cpDiv.addedEvents){
			ele.querySelector('.colorSlider_hue').addEventListener('mousedown', function(event){
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
				return window.ColorPicker_EventUtil.__initHueMove(event,window.ColorPicker_Util.component);
			});
			divElPalette.addEventListener('mousedown',function(event){
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
				return window.ColorPicker_EventUtil.__initPaletteMove(event,window.ColorPicker_Util.component);
			});
			ele.querySelector('.colorSlider_hueBar').addEventListener('mousedown', function(event) {
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
	            window.ColorPicker_EventUtil.__moveOnHueBar(event,window.ColorPicker_Util.component);
	        });
	        ele.querySelector('.colorSlider_hue').addEventListener('touchstart', function(event){
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
				return window.ColorPicker_EventUtil.__initHueMove(event,window.ColorPicker_Util.component);
			});
			divElPalette.addEventListener('touchstart',function(event){
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
				return window.ColorPicker_EventUtil.__initPaletteMove(event,window.ColorPicker_Util.component);
			});
			ele.querySelector('.colorSlider_hueBar').addEventListener('touchstart', function(event) {
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
	            window.ColorPicker_EventUtil.__moveOnHueBar(event,window.ColorPicker_Util.component);
	        });
			//for opacityslider circle
			ele.querySelector('.lyteColorPicker__opacityslider').addEventListener('mousedown', function(event){
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
				return window.ColorPicker_EventUtil.__initOpacityMove(event,window.ColorPicker_Util.component);
			});
			ele.querySelector('.lyteColorPicker__opacityslider').addEventListener('touchstart', function(event){
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
				return window.ColorPicker_EventUtil.__initOpacityMove(event,window.ColorPicker_Util.component);
			});
			//for opacityslider track
			ele.querySelector('.opacityslider__track').addEventListener('mousedown', function(event) {
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
	            window.ColorPicker_EventUtil.__moveOnOpacityTrack(event,window.ColorPicker_Util.component);
	        });
	        ele.querySelector('.opacityslider__track').addEventListener('touchstart', function(event) {
				window.ColorPicker_Util.component = window.ColorPicker_Util.getColorPicker(event.target).component;
	            window.ColorPicker_EventUtil.__moveOnOpacityTrack(event,window.ColorPicker_Util.component);
	        });
	        cpDiv.addedEvents = true;
		}
        this.setData('paletteSize',{'height' : divElPalette.offsetHeight, 'width' : divElPalette.offsetWidth});
        this.setData('hueBarSize',ele.querySelector('.colorSlider_hueBar').offsetWidth);
        this.setData('opacityTrackSize',ele.querySelector('.opacityslider__track').offsetWidth);
        if(this.getData('opacity') === undefined || (this.getData('opacity') && this.getData('opacity') < 0)){
        	this.setData('opacity',100);
        }
		this.selectedColor();
        if(this.getData('cpBoardColor')){
        	this.__setCurrentRgbCodeFromBoardColor();
        }
        if(this.getData('inputChangeEvnt')){
        	this.setData('inputChangeEvnt',false);
        }
		this.setRgbColor("#"+this.getData("cpCurrentRgbCode"),null,restrictCmykChange);
		ele = null;
		divElPalette = null;
		divElPaletteCircle = null;
		imageObj = null;
	}

    initializeBasicColorPicker(event, ele) {
		var liELes = ele.querySelectorAll('.default__colors > li');
		for(var v =0; v<liELes.length ; v++){
			liELes[v].style.background = liELes[v].textContent;
		}
		if(this.getData('cpStandardColors')){
			liELes = ele.querySelectorAll('.standard__colors > li');
			for(var v =0; v<liELes.length ; v++){
				liELes[v].style.background = liELes[v].textContent;
			}
		}
		if(this.getData('cpDefaultColor') && this.getData('cpDefaultColorEnable')){
			liELes = ele.querySelectorAll('.cpDefault__colors');
			liELes[0].style.background = liELes[0].textContent;
		}
		if(this.getData('cpUsedColors')){
			liELes = ele.querySelectorAll('.used__colors > li');
			for(var v =0; v<liELes.length ; v++){
				liELes[v].style.background = liELes[v].textContent;
			}
		}
		if(this.getData('cpColorPreview')){
			if(this.getData('cpCurrentRgbCode')){
				var color = window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData('cpCurrentRgbCode'));
			}
		}
		var _this = this;
		$L.fastdom.mutate(function(){
			var selectColor = _this.getData('cpSelectedColor');
			var prev_element = $L('.lyteCPSelectedColor');
			if(prev_element.length){
				prev_element.removeClass('lyteCPSelectedColor');
			}
			if(selectColor && !selectColor.nofill && Object.keys(selectColor).length){
				var ele;
				if(_this.getData('cpAccent')){
					ele = _this.$node.querySelector('[data-color="'+selectColor.hex+'"]');
				}else{
					ele = _this.$node.querySelector('[lt-prop-title="'+selectColor.hex+'"]');
				}
				var alpha = window.ColorPicker_Util.getDecimalToHexAlphaCode(_this.getData('opacity'));
				
				if(selectColor.hex + alpha !== selectColor['hex-alpha'] || !ele){	
					ele = _this.$node.querySelector('[lt-prop-title="'+selectColor['hex-alpha'].toUpperCase()+'"]');
				}
				else if (selectColor.hex + alpha === selectColor['hex-alpha'] && _this.getData('opacity') < 100){
					ele = _this.$node.querySelector('[lt-prop-title="'+selectColor['hex-alpha'].toUpperCase()+'"]');
				}
				
				ele && ele.classList.add('lyteCPSelectedColor');
			}
		})
		
		liELes = null;
		if(!event) {
			this.selectedColor();
		}
	}

    executeOnBeforeOpen() {
		if(this.getMethods('onBeforeOpen')){
			this.executeMethod('onBeforeOpen');
		}
	}

    addedUsedColor() {
		var colors = this.getData('usedColors');
		var color = this.getData('cpSelectedColor').rgba;
		if(colors.indexOf(color) != -1){		
			this.$addon.arrayUtils(this.getData('usedColors'),'removeAt',colors.indexOf(color),1);
		}
		else{
			this.$addon.arrayUtils(this.getData('usedColors'),'removeAt',this.getData('cpUsedColorsNumber') - 1,1);
		}
		this.$addon.arrayUtils(this.getData('usedColors'),'insertAt',0,color);
		var _this = this;
		$L.fastdom.mutate(function(){
			setTimeout(function(){
				var selectColor = _this.getData('cpSelectedColor');
				var alpha = window.ColorPicker_Util.getDecimalToHexAlphaCode(_this.getData('opacity'));
				if(selectColor && !selectColor.nofill && Object.keys(selectColor).length){
					var ele;
					var color;
					if(_this.getData('cpAccent')){
						ele = _this.$node.querySelector('[data-color="'+selectColor.hex+'"]');
						color = ele ? ele.dataset.color : '';
					}else{
						ele = _this.$node.querySelector('[lt-prop-title="'+selectColor.hex+'"]');
						color = ele ? ele.getAttribute('lt-prop-title') : '';
					}
					var prev_element = _this.$node.querySelector('.lyteCPSelectedColor');
						prev_element && prev_element.classList.remove('lyteCPSelectedColor');
					if(!ele || color + alpha !== selectColor['hex-alpha']  ){
					
						ele = _this.$node.querySelector('[lt-prop-title="'+selectColor['hex-alpha'].toUpperCase()+'"]');

					}else if (color + alpha === selectColor['hex-alpha']  &&  _this.getData('opacity') < 100 ){
						ele = _this.$node.querySelector('[lt-prop-title="'+selectColor['hex-alpha'].toUpperCase()+'"]');
					}
					ele && ele.classList.add('lyteCPSelectedColor');
				}
				
			},100)
			
		});
		var node = this.$node.querySelectorAll(".usedColor__container li")[0] /*(this.getData('cpInline') ? this.$node : this.childComp).querySelectorAll(".usedColor__container li")[0]*/;
		if(node){
			node.style.background = color;
		}
		// this.setData('usedColors',colors);
		if(this.getData('calledForHueMove')){
			this.setData('calledForHueMove',false);
		}
		
		

	}

    executeOnSelect(event, close, element) {
		var colors = this.getData('usedColors');
		if(this.data.cpSelectedColor){
			this.data.cpSelectedColor.nofill =  false;
		} 
		this.__setSelectedColor(element);
		this.addedUsedColor();
		if(this.getMethods("onSelect") && event){
			this.executeMethod("onSelect",event,close,this.getData('calledForHueMove'));	
		}
		
	}

    executeOnOpen() {
		if(this.getMethods('onOpen')){
			this.executeMethod('onOpen');
		}
	}

    executeOnClose(event) {
		if(this && this.inputEle){
			// var inputEle = this.inputEle;
			this.inputEle = null;
			// inputEle.remove();
		}
		if(this.$node.closest('lyte-wormhole')){
			this.$node.closest('lyte-wormhole')._callee.parentElement.setData('ltPropShow',false);
		}
		// if(this.getMethods('onClose')){
		// 	this.executeMethod('onClose',event);
		// }
	}

    executeOnCancelButton(event) {
		if(this.getMethods('onCancelButton')){
			this.executeMethod('onCancelButton',event);
		}
	}

    executeOnChange() {
		// var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData('cpCurrentRgbCode'));
  //       var selectedColor = {
  //       	"hex" : "#"+this.getData('cpCurrentRgbCode'),
  //       	"rgb" : "rgb("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+")",
  //       	"opacity" : this.getData('opacity'),
  //       	"hex-alpha" : "#"+this.getData('cpCurrentRgbCode')+window.ColorPicker_Util.getDecimalToHexAlphaCode(this.getData('opacity')),
  //       	"rgba" : "rgba("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+", "+(this.getData('opacity')/100)+")"
  //       };
  //       this.setData("cpSelectedColor",selectedColor);
  		this.__setSelectedColor();
  		if( !(this.isColorValueEqual()) ){
  			this.setData('prevColor', this.getData('cpSelectedColor'));
  			if(this.getMethods('onChange')){
				this.executeMethod('onChange',window.event,this.getData('cpSelectedColor'),this);
			}
  		}
	}

    isColorValueEqual() {
		var prevColor = this.getData('prevColor'),
			currColor = this.getData('cpSelectedColor');
		if(Object.keys(prevColor).length === Object.keys(currColor).length){
			for(window.key in prevColor){
				if(prevColor[window.key] != currColor[window.key]){
					return false;
				}
			}
			return true;
		}
		return false;
	}

    validateInputValue(type, lyteInp) {
		var value = lyteInp.getData('ltPropValue');
		if(type === "HEX"){
			if(value.length == 4 || value.length == 7){
				if(lyteInp.classList.contains('lyteCp__Invalid')){
					lyteInp.classList.remove('lyteCp__Invalid');
				}
			}
			else{
				lyteInp.classList.add('lyteCp__Invalid');
				if(this.getMethods('onInputError')){
					this.executeMethod('onInputError', type == "HEX" ? "HEX" : this.getData('colorFormat')[type]);
				}
			}
		}
		else if(type == "R" || type == "G" || type == "B" || type == "H" || type == "S" || type == "V" || type == "C" || type == "M" || type == "Y" || type == "K" || type == "A"){
			if(value.length > 0 && value.length < 4){
				if(lyteInp.classList.contains('lyteCp__Invalid')){
					lyteInp.classList.remove('lyteCp__Invalid');
				}
			}
			else{
				lyteInp.classList.add('lyteCp__Invalid');
				if(this.getMethods('onInputError')){
					this.executeMethod('onInputError', type == "HEX" ? "HEX" : this.getData('colorFormat')[type]);
				}
			}
		}

	}

    /*----------------- Start of Utility Methods for Color Picker ---------------*/

    __setSelectedColor(element) {
		var rgbCode = this.getData('cpCurrentRgbCode');
		var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(rgbCode);
		var cmykColors = this.getData('cmykColors');
		if(this.getData('dropButtonValue') != "CMYK"){
			var rgbCol = window.ColorPicker_Util.getRgbColorsByHsv(this.getData('currentHue'), this.getData('currentSaturation') / 100, this.getData('currentBrightness') / 100);
    		var derivedCmykColors = window.ColorPicker_Util.getCmykByRgbColors([rgbCol.red,rgbCol.green,rgbCol.blue]);
    		if(!cmykColors.length){
    			cmykColors = derivedCmykColors
    			this.setData('cmykColors', cmykColors);
    		}
    		else{
    			var calculatedRgbCode1 = window.ColorPicker_Util.getRgbCodeByCmyk(derivedCmykColors[0], derivedCmykColors[1], derivedCmykColors[2], derivedCmykColors[3]);
    			var calculatedRgbCode2 = window.ColorPicker_Util.getRgbCodeByCmyk(cmykColors[0], cmykColors[1], cmykColors[2], cmykColors[3]);
    			if(calculatedRgbCode1 != calculatedRgbCode2 && calculatedRgbCode1 == rgbCode){
    				cmykColors = derivedCmykColors;
    				this.setData('cmykColors', cmykColors);
    			}
    		}
		}
		var type = this.getData('opacity') < 100 ? 'advanced' : 'basic'; 
        var selectedColor = {
        	"hex" : "#"+rgbCode,
        	"rgb" : "rgb("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+")",
        	"opacity" : this.getData('opacity'),
        	"hex-alpha" : "#"+rgbCode+window.ColorPicker_Util.getDecimalToHexAlphaCode(this.getData('opacity')),
        	"rgba" : "rgba("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+", "+(this.getData('opacity')/100)+")",
        	"cmyk" : cmykColors,
			"type" : this.getData('cpSelectedColor') ? this.getData('cpSelectedColor').type : type,
			"nofill" : this.getData('cpSelectedColor') ? this.getData('cpSelectedColor').nofill : null

        };
		if(this.getData('cpAccent') && element && element.classList.contains('lyteColorPickerAccentColor') && element.dataset.row){
			var accentColor = this.getData('cpAccentColor');
			var colorTone = accentColor[element.dataset.col].colorTone[element.dataset.row - 1];
			if(colorTone){
				selectedColor.colorTone = {
					toneType : colorTone.toneType,
					tonePercent : colorTone.tonePercent
				}
			}
		}
		var dropdown = $L('.cPDropDown',this.$node)[0];
		if(this.getData('cpDropdownType') && dropdown){
			this.setData('cpDropdownType',dropdown.getData('ltPropSelected'))
		}
        this.setData("cpSelectedColor",selectedColor);
		this.setData("cpDynamicColor",selectedColor);
	}

    setRgbColor(rgbColor, hue, restrictCmykChange) {
        var hsv = window.ColorPicker_Util.getHsvByRgbCode(rgbColor);
        this.setData('currentHue', hue ? hue : hsv.hue);
        this.setData('currentBrightness', hsv.brightness * 100);
        this.setData('currentSaturation', hsv.saturation * 100);
        this.__changeViewAfterColorChange(restrictCmykChange);
        this.__setSelectedColor();
    }

    __changeViewAfterColorChange(restrictCmykChange) {
        this.__setCurrentRgbCode();
        this.__setPaletteBgColor();
        this.__setSliderPos();
        // this.__setBgColorPreviewDiv();
        if(!this.getData('inputChangeEvnt')){
        	this.__updateRgbInForm(restrictCmykChange);
        }else{
        	this.setData('inputChangeEvnt',false);
        }       
        /*this.__updateHsvInForm();*/
        this.__setSmallCirclePosition();
        // this.__setOpacitySliderPos();
		this.__changeColorInOpacityTrack();
		this.__setOpacityColorPreviewDiv();
		this.__setOpacityColorDynamicDiv();
		this.__updateOpacityInForm();
    }

    __setCurrentRgbCodeFromBoardColor() {
    	var color = this.getData('cpBoardColor');
		if(/#/.test(color)){
			if(color.length == 9){
				this.setData('opacity', window.ColorPicker_Util.getHexToDecimalAlphaCode(color.substring(7)));
				this.setData('cpCurrentRgbCode',color.substring(1,7));
			}
			else if(color.length == 7){
				this.setData('cpCurrentRgbCode',color.substring(1));
			}
			else if(color.length == 4){
				this.setData('cpCurrentRgbCode',window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(color.substring(1)));
			}
		}
		else if(/rgba/.test(color)){
			var rgbColor = color.substring(5,color.length-1).split(",");
			this.setData('opacity',rgbColor[3]*100);
			this.setData('cpCurrentRgbCode',window.ColorPicker_Util.getRgbCodeByRgbColors(rgbColor[0],rgbColor[1],rgbColor[2]));
		}
		else if(/rgb/.test(color)){
			var rgbColor = color.substring(4,color.length-1).split(",");
			this.setData('cpCurrentRgbCode',window.ColorPicker_Util.getRgbCodeByRgbColors(rgbColor[0],rgbColor[1],rgbColor[2]));
		}
    }

    __setCurrentRgbCode(code) {
		var rgbCode = code ? code : window.ColorPicker_Util.getRgbCodeByHsv(this.getData('currentHue'), this.getData('currentSaturation') / 100, this.getData('currentBrightness') / 100);
		this.setData('stopObserverOnBoardColorChange', true);
		this.setData('cpBoardColor', '#'+rgbCode);
        this.setData('cpCurrentRgbCode', rgbCode);
        if(this.getData('stopObserverOnBoardColorChange')){
        	this.setData('stopObserverOnBoardColorChange', false);
        }
    }

    __setPaletteBgColor() {
        try {
        	// if(this.getData('cpInline')){
        		this.$node.querySelector('#colorDiv').style.backgroundColor = '#' + window.ColorPicker_Util.getRgbCodeByHsv(this.getData('currentHue'), 1, 1);
        	// }
        	// else{
         //    	this.childComp.querySelector('#colorDiv').style.backgroundColor = '#' + window.ColorPicker_Util.getRgbCodeByHsv(this.getData('currentHue'), 1, 1);
        	// }
        } catch (e) {}
    }

    __setSliderPos() {
        var leftPos = Math.round(this.getData('hueBarSize') - ((this.getData('currentHue') / 360) * this.getData('hueBarSize')));
        // if(this.getData('cpInline')){
        	this.$node.querySelector('.colorSlider_sliderHandle').style.left = leftPos - this.getData('sliderOffset') + "px";
        // }
        // else{
        // 	this.childComp.querySelector('.colorSlider_sliderHandle').style.left = leftPos - this.getData('sliderOffset') + "px";
        // }
    }

    __setSmallCirclePosition() {
    	var currentBrightness = this.getData('currentBrightness');
    	var currentSaturation = this.getData('currentSaturation');
    	var paletteSize = this.getData('paletteSize');
    	var circleOffsetSize = this.getData('circleOffsetSize');
    	var divElPaletteCircle = this.$node.querySelector('.colorSlider_palette_circle') /*this.getData('cpInline') ? this.$node.querySelector('.colorSlider_palette_circle') : this.childComp.querySelector('.colorSlider_palette_circle')*/;
        var leftPos = Math.round(currentSaturation * (paletteSize.width / 100)) - circleOffsetSize;
        var topPos = paletteSize.height - Math.round(currentBrightness * (paletteSize.height / 100)) - circleOffsetSize;
        divElPaletteCircle.style.left = leftPos + 'px';
        divElPaletteCircle.style.top = topPos + 'px';
        divElPaletteCircle.className = divElPaletteCircle.className.replace('colorSlider_palette_circleBlack', '');
        if (currentBrightness > 80) {
            divElPaletteCircle.className = divElPaletteCircle.className + 'colorSlider_palette_circleBlack';
        }
        divElPaletteCircle = null;
    }

    __ffHackWinWidget() {
    	var ele = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
        if (ele.querySelector('.lyteColorPicker--advanced').parentNode.className && ele.querySelector('.lyteColorPicker--advanced').parentNode.className.indexOf('windowContent') >= 0 && !document.all) {
            this.setData('circleOffsetBecauseOfWinWidget', 0);
        }
    }

    __updateRgbInForm(restrictCmykChange) {
    	var comp = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
    	var format = comp.querySelector('.selectFormat').querySelector('lyte-dropdown').getData('ltPropSelected');
    	this.setData('dropButtonValue',format);
    	if(format === window._lyteUiUtils.i18n("HEX")){

    		comp.querySelector('#lyteCPShowValue').setData("ltPropValue",("#" + this.getData("cpCurrentRgbCode")));
    		this.inputEle.value = ("#" + this.getData("cpCurrentRgbCode") + window.ColorPicker_Util.getDecimalToHexAlphaCode(this.getData('opacity')));
    		this.setData('counter', 1);
    		if(!restrictCmykChange && this.getData('setPrevCmyk')){
				this.setData('setPrevCmyk',false);
			}
    	}
    	else if(format === window._lyteUiUtils.i18n("RGB")){
    		var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData("cpCurrentRgbCode"));
    		comp.querySelector('#lyteCP__R').component.setData('ltPropValue',rgbColor.red+"");
    		comp.querySelector('#lyteCP__G').component.setData('ltPropValue',rgbColor.green+"");
    		comp.querySelector('#lyteCP__B').component.setData('ltPropValue',rgbColor.blue+"");
    		this.setData('counter', 3);
    		// comp.querySelector('#lyteCPShowValue_RGB').component.setData('ltPropValue',"rgb("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+")");
    		this.inputEle.value = "rgba("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+", "+(this.getData('opacity')/100)+")";
    		// this.getData('rgbFieldCode').component.$node.cp("value",("rgb("+rgbColor.red+","+rgbColor.green+","+rgbColor.blue+")"));
    		if(!restrictCmykChange && this.getData('setPrevCmyk')){
				this.setData('setPrevCmyk',false);
			}
    	}
    	else if(format === window._lyteUiUtils.i18n("HSV")){
    		var hue = Math.round(this.getData('currentHue'));
    		var saturation = Math.round(this.getData('currentSaturation'));
    		var valueBrightness = Math.round(this.getData('currentBrightness'));
    		comp.querySelector('#lyteCP__H').component.setData('ltPropValue',hue+"");
    		comp.querySelector('#lyteCP__S').component.setData('ltPropValue',saturation+"");
    		comp.querySelector('#lyteCP__V').component.setData('ltPropValue',valueBrightness+"");
    		this.setData('counter', 3);
    		this.inputEle.value = "hsv("+hue+"°,"+saturation+"%,"+valueBrightness+"%,"+(this.getData('opacity')/100)+")";
    		// this.getData('rgbFieldCode').component.$node.cp("value",("hsv("+hue+"°,"+saturation+"%,"+valueBrightness+"%)"));
    		if(!restrictCmykChange && this.getData('setPrevCmyk')){
				this.setData('setPrevCmyk',false);
			}
    	}
    	else if(format === window._lyteUiUtils.i18n("CMYK")){
    		var rgbColor = window.ColorPicker_Util.getRgbColorsByHsv(this.getData('currentHue'), this.getData('currentSaturation') / 100, this.getData('currentBrightness') / 100);
    		var cmykColors;
    		if(this.getData('setPrevCmyk') && this.getData('cmykColors').length == 4){
    			cmykColors = this.getData('cmykColors');
    		}
    		else{
    			cmykColors = window.ColorPicker_Util.getCmykByRgbColors([rgbColor.red,rgbColor.green,rgbColor.blue]);
    			if(!restrictCmykChange && this.getData('setPrevCmyk')){
    				this.setData('setPrevCmyk',false);
    			}
    		}
    		comp.querySelector('#lyteCP__C').component.setData('ltPropValue',Math.round(cmykColors[0]*100)+"");
    		comp.querySelector('#lyteCP__M').component.setData('ltPropValue',Math.round(cmykColors[1]*100)+"");
    		comp.querySelector('#lyteCP__Y').component.setData('ltPropValue',Math.round(cmykColors[2]*100)+"");
    		comp.querySelector('#lyteCP__K').component.setData('ltPropValue',Math.round(cmykColors[3]*100)+"");
    		this.setData('cmykColors', cmykColors);
    		this.setData('counter', 4);
    		this.inputEle.value = "cmyk("+Math.round(cmykColors[0]*100)+"%,"+Math.round(cmykColors[1]*100)+"%,"+Math.round(cmykColors[2]*100)+"%,"+Math.round(cmykColors[3]*100)+"%)";
    		// this.getData('rgbFieldCode').component.$node.cp("value",("cmyk("+Math.round(cmykColors[0]*100)+"%,"+Math.round(cmykColors[1]*100)+"%,"+Math.round(cmykColors[2]*100)+"%,"+Math.round(cmykColors[3]*100)+"%)"));
    	}
    }

    __setOpacitySliderPos() {
    	var ele = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
    	var leftPos = (this.getData('opacity')*this.getData('opacityTrackSize'))/100;
    	ele.querySelector('.opacityslider__circlethumb').style.left = leftPos - this.getData('sliderOffset') + "px";
    }

    __changeColorInOpacityTrack() {
    	var ele = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
    	ele.querySelector('.opacityslider__track').style.backgroundImage = "linear-gradient(to right, transparent, #"+this.getData('cpCurrentRgbCode')+")";
    }

    __setOpacityColorPreviewDiv() {
    	var ele = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
    	var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData("cpCurrentRgbCode"));
    	ele.querySelector('.previewDiv').style.backgroundColor = "rgba("+rgbColor.red+","+rgbColor.green+","+rgbColor.blue+","+(this.getData('opacity')/100)+")";
    }

    __setOpacityColorDynamicDiv() {
		if(this.data.cpDisplayPreviousColor){
			var ele = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
			var rgbColor =  window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData("cpCurrentRgbCode"));
			ele.querySelector('.dynamicDiv').style.backgroundColor = "rgba("+rgbColor.red+","+rgbColor.green+","+rgbColor.blue+","+(this.getData('opacity')/100)+")";
		}
    }

    __updateOpacityInForm() {
    	var ele = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
    	if(ele.querySelector('#lyteCP__A')){
    		ele.querySelector('#lyteCP__A').component.setData('ltPropValue',this.getData('opacity'));
    		if(this.getData('dropButtonValue') == window._lyteUiUtils.i18n("HEX")){
    			this.inputEle.value = ("#" + this.getData("cpCurrentRgbCode") + window.ColorPicker_Util.getDecimalToHexAlphaCode(this.getData('opacity')));
    		}
    		if(this.getData('dropButtonValue') == window._lyteUiUtils.i18n("RGB")){
    			var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData("cpCurrentRgbCode"));
    			this.inputEle.value = "rgba("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+", "+(this.getData('opacity')/100)+")";
    		}
    		if(this.getData('dropButtonValue') == window._lyteUiUtils.i18n("HSV")){
    			var hue = Math.round(this.getData('currentHue'));
	    		var saturation = Math.round(this.getData('currentSaturation'));
	    		var valueBrightness = Math.round(this.getData('currentBrightness'));
    			this.inputEle.value = "hsv("+hue+"°,"+saturation+"%,"+valueBrightness+"%,"+(this.getData('opacity')/100)+")";
    		}
    	}
    }

    defaultSelectedColor(event, color, element) {
		if(color == "rgba(0, 0, 0, 0)"){
			return;
		}
		var ele;
		var node = this.$node.querySelector('.lyteCPSelectedColor');
		if(node){
			node.classList.remove('lyteCPSelectedColor');
			node.setAttribute('aria-selected',false);
		}
		if(element){
			element.classList.add('lyteCPSelectedColor');
		element.setAttribute('aria-selected',true);
		}
		var rgbColor;
		if(/rgba/.test(color)){
			rgbColor = color.substring(5,color.length-1).split(",");
			this.setData('opacity',rgbColor[3]*100);
		}
		else{
			rgbColor = color.substring(4,color.length-1).split(",");
			this.setData('opacity',100);
		}
		this.__setCurrentRgbCode(window.ColorPicker_Util.getRgbCodeByRgbColors(rgbColor[0],rgbColor[1],rgbColor[2]));
		var hsv = window.ColorPicker_Util.getHsvByRgbCode('#'+this.getData('cpCurrentRgbCode'));
		this.setData('currentHue', hsv.hue);
		this.setData('currentBrightness', hsv.brightness * 100);
		this.setData('currentSaturation', hsv.saturation * 100);
		this.executeOnSelect(event,undefined,element);
		this.executeOnBasicColorChange(event);
	}

    executeOnBasicColorChange(event) {
		if( !(this.isColorValueEqual()) ){
  			this.setData('prevColor', this.getData('cpSelectedColor'));
  			if(this.getMethods('onBasicColorChange')){
				this.executeMethod('onBasicColorChange',event,this.getData('cpSelectedColor'),this);
			}
  		}
	}

    selectedColor() {
		var selectedColor = this.getData("cpSelectedColor");
		if(selectedColor &&  Object.keys(selectedColor).length) {
			var _this = this;
			var rgb; 
			var count = 1;
			var rgbColor;
			
			if(selectedColor.rgb && selectedColor.rgb.length) {
				rgb = selectedColor.rgb;
			}else if(selectedColor.rgba && selectedColor.rgba.length) {
				
				var rgbaString = selectedColor.rgba;
				var regex = /rgba\(\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(\d{1,3})\s*,\s*(0|1|0?\.\d+)\s*\)/;
				var color = rgbaString.match(regex);
				selectedColor.opacity = color[4]* 100;
				this.setData('opacity',color[4]*100);
				rgb = "rgb("+color[1]+", "+color[2]+", "+color[3]+")";
			}
			else if(selectedColor.hex && selectedColor.hex.length) {
			    var color = selectedColor.hex.slice(1);
				if(color.length == 3) {
					color = window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(color);
				}
				rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(color);
				
				rgb = "rgb("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+")";
			}
			else if(selectedColor['hex-alpha'] && selectedColor['hex-alpha'].length){
				var color = selectedColor['hex-alpha'].slice(1);
				if(color.length == 3) {
					color = window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(color);
				}
				rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(color);
				var opacity = window.ColorPicker_Util.getHexToDecimalAlphaCode(color.substr(6, 4));
				selectedColor.opacity = opacity;
				this.setData('opacity',opacity);
				rgb = "rgb("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+")";
			}
			else if(selectedColor.cmyk && selectedColor.cmyk.length) {
				var color = selectedColor.cmyk;
				rgbColor = window.ColorPicker_Util.getRgbColorsByCmyk(color[0],color[1],color[2],color[3]);
				rgb = "rgb("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+")";
			}
			if(selectedColor.opacity !== undefined){
				this.setData('opacity',selectedColor.opacity);
			}else{
				this.setData('opacity', 100);
			}

			if( !selectedColor.nofill && rgb && (selectedColor.type == "basic" || selectedColor.type == undefined)) {
				window.rgbCode =  rgb.replace(/[^\d,]/g, '').split(',');
				
				var color =  window.ColorPicker_Util.getRgbCodeByRgbColors(rgbCode[0],rgbCode[1],rgbCode[2]);
				var element;
				if(this.getData('cpAccent')){
					element = this.$node.querySelector('[data-color= "#' + color + '"]');
				}else{
					element = this.$node.querySelector('[lt-prop-title= "#' + color + '"]');
				}
				var alpha = window.ColorPicker_Util.getDecimalToHexAlphaCode(this.getData('opacity'));
				
				if(selectedColor['hex-alpha'] && selectedColor.hex + alpha !== selectedColor['hex-alpha'] || !element ){
					element = this.$node.querySelector('[lt-prop-title="' + selectedColor['hex-alpha'] +'"]') ;
					
				}else if(selectedColor.hex + alpha === selectedColor['hex-alpha'] && _this.getData('opacity') < 100 ){
					element = this.$node.querySelector('[lt-prop-title="' + selectedColor.hex + alpha.toUpperCase() +'"]') ;
				}

				if(element && !element.classList.contains('lyteCPSelectedColor')){
					var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(color+alpha);
					var opacity = selectedColor.opacity;
					
					if(opacity){
						color = "rgba("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+", "+this.getData('opacity')/100+")";

					}else{
						color = "rgb("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+")";
					}
					_this.defaultSelectedColor(null,color,element);
				}else if( !element ){
					var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(color);
					this.data.cpSelectedColor.rgba = "rgba("+rgbColor.red+", "+rgbColor.green+", "+rgbColor.blue+", "+(this.getData('opacity')/100)+")";
				}
				
			}else{
				if(selectedColor){
					rgb = selectedColor.rgb;
					window.rgbCode =  rgb.replace(/[^\d,]/g, '').split(',');
				}else{
					rgbCode = ('rgb(0,0,0)').replace(/[^\d,]/g, '').split(',');
				}
				
			}

			var last =  this.getData("cpCurrentRgbCode");

			this.setData("cpCurrentRgbCode",window.ColorPicker_Util.getRgbCodeByRgbColors(window.rgbCode[0],window.rgbCode[1],window.rgbCode[2]));
			this.setData('cpBoardColor',"#"+this.getData('cpCurrentRgbCode'));
			if(this.getData("cpCurrentRgbCode") != last && this.getData("cpBasicColorPicker")==false) {
				this.setRgbColor("#"+this.getData("cpCurrentRgbCode"),null,true);
			}
		}else {
			$L('.lyteCPSelectedColor',this.childComp).removeClass('lyteCPSelectedColor');
		}
	}

    pasteFunc(event, prop) {
		var input = event.target,
				start = input.selectionStart,
				end = input.selectionEnd,
				value = input.value;
		if(start == end){
			value = value + (event.clipboardData || window.clipboardData).getData('text').trim();
		}
		else{
			var paste = value.substring(0,start) + (event.clipboardData || window.clipboardData).getData('text').trim();
			var restMsg = value.substring(end);
			if(restMsg.length > 0){
				paste += restMsg;
			}
			value = paste;
		}
		if(prop === "HEX"){
			if( (value.length == 1 && value[0] != '#') || value.length > 7 || (value.length > 0 && (value[0] != '#' || value.substring(1).indexOf('#') != -1)) ){
				event.preventDefault();
				event.stopPropagation();
			}
		}
		else if(prop == "R" || prop == "G" || prop == "B"){
			if(value === "" || isNaN(value) || (value && (parseInt(value) < 0 || parseInt(value) >255))){
				event.preventDefault();
				event.stopPropagation();
			}
		}
		else if(prop == "H"){
			if(value === "" || isNaN(value) || (value && (parseInt(value) < 0 || parseInt(value) >360))){
				event.preventDefault();
				event.stopPropagation();
			}
		}
		else if(prop == "S" || prop == "V" || prop == "C" || prop == "M" || prop == "Y" || prop == "K" || prop == "A"){
			if(value === "" || isNaN(value) || (value && (parseInt(value) < 0 || parseInt(value) >100))){
				event.preventDefault();
				event.stopPropagation();
			}
		}
	}

    /*----------------- End of Utility Methods for Color Picker ---------------*/

    didDestroy() {
        if(this.tId){
    	    clearInterval(this.tId);
    	    this.tId = false;
        }
    }

    nofillFunc(event) {
		var selectedColor = {
			"hex" : '#000000',
			'rgb' : 'rgb(0,0,0)',
			'rgba' : 'rgba(0,0,0,0)',
			"opacity" : 0,
			'nofill' : true
		};
		var advancedInput = $L('#lyteCPShowValue',this.$node)[0];
		if(advancedInput){
			advancedInput.setData('ltPropValue','#000000');
		}
		$L('.lyteCPSelectedColor',this.childComp).removeClass('lyteCPSelectedColor');
		this.setData('cpSelectedColor',selectedColor);
		this.setData('cpDynamicColor',selectedColor);
		this.setData('opacity',0);
		if(this.getData('calledForHueMove')){
			this.setData('calledForHueMove',false);
		}
		if(this.getMethods("onSelect") && event){
			this.executeMethod("onSelect",event,undefined, this.getData('calledForHueMove'));	
		}
		
	}

    fromAdvCPtoBasic(event) {
		this.setData('cpMoreColorOptionSelected',false);
		var ele = this.$node;
		if(this.getData('cpInline')){
			this.setData('cpBasicColorPicker',true);
		}
		else{
			var popOver = ele.closest('.lytePopover') /*ele.querySelector('.lytePopover')*/;
			var content = popOver.querySelector('lyte-popover-content');
			var contentHeight = 0;
			var height = 0;
			this.setData('cpBasicColorPicker',true);
			popOver._callee.component.computeOffsetImpl();
		}
		this.$node.focus();
		this.initializeBasicColorPicker(event,ele);
		popOver = null;
	}

    goToAdvancedCP(event) {
		var ele = this.$node;
		var color = this.getData('cpSelectedColor');
		if(color != null){
			this.setData("cpCurrentRgbCode",color.hex ? color.hex.substring(1) : "FF0000");
			if('cmyk' in color){
				this.setData('cmykColors', color.cmyk);
				this.setData('setPrevCmyk',true);
			}
			else{
				this.setData('cmykColors',[])
			}
		}
		this.setData('cpMoreColorOptionSelected',true);

		if(this.getData('cpInline')){
			// ele = this.$node.querySelector('.lyteColorPicker');
			this.setData('cpBasicColorPicker',false);
		}
		else{
			// ele = document.querySelector('.'+this.getData('cpWrapperClass'));
			var popOver = ele.closest('.lytePopover') /*ele.querySelector('.lytePopover')*/;
			var popOverOffset = popOver.getBoundingClientRect();
			var height = popOverOffset.height;
			// this.setData('basicDimensions',{"width" : popOverOffset.width, "height" : height/*, "usedColorsHeight" : usedColorsHeight*/});
			
		
			this.setData('cpBasicColorPicker',false);
			popOver._callee.component.computeOffsetImpl();
			popOverOffset = null;
			height = null;
			popOver = null;
		}
		this.inputEle = ele.querySelector('#lyteCPHiddenInput');
		// this.inputEle.id = "lyteCPHiddenInput";
		// document.body.append(this.inputEle);
		// this.inputEle.style.top = "-100px";
		// this.inputEle.style.position = "absolute";
		this.initializeAdvColorPicker(event,ele,true);
		ele = null;
		color = null;
	}

    static methods(arg1) {
        return Object.assign(super.methods({
            inputValueChanged : function(type,changedProp,lyteInp){		
                var comp = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
                if(type === "HEX"){
                    if(/^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/i.test(changedProp.newValue)){
                        this.setData('inputChangeEvnt',true);
                        if(changedProp.newValue.length == 7){
                            this.__setCurrentRgbCode(changedProp.newValue.substring(1));
                            // this.setData('cpCurrentRgbCode',changedProp.newValue.substring(1));
                            // this.setRgbColor(changedProp.newValue);
                        }
                        else if(changedProp.newValue.length == 4){
                            this.__setCurrentRgbCode(window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(changedProp.newValue.substring(1)));
                            // this.setData('cpCurrentRgbCode',window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(changedProp.newValue.substring(1)));
                            // this.setRgbColor('#' + window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(changedProp.newValue.substring(1)));
                        }
                        if(!(this.getData('dragged')) && this.getData('dragStatus') != 1 && this.getData('hueStatus') != 1 && !(this.getData('counter'))){
                            this.setRgbColor('#' + this.getData('cpCurrentRgbCode'));
                            this.executeOnChange();
                        }
                        else{
                            this.setData('dragged',false);
                            if(this.getData('counter')){
                                this.setData('counter', this.getData('counter') - 1);
                                if(!(this.getData('counter'))){
                                    this.setData('restrictOnChange',true);
                                    this.__updateOpacityInForm();
                                    // this.setData('counter', 1);
                                }
                            }
                        }
                    }
                    else if(/^([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/i.test(changedProp.newValue)){
                        this.setData('inputChangeEvnt',true);
                        if(changedProp.newValue.length == 6){
                            this.__setCurrentRgbCode(changedProp.newValue);
                            // this.setData('cpCurrentRgbCode',changedProp.newValue);
                        }
                        else if(changedProp.newValue.length == 3){
                            this.__setCurrentRgbCode(window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(changedProp.newValue));
                            // this.setData('cpCurrentRgbCode',window.ColorPicker_Util.convert3DigitTo6DigitRgbCode(changedProp.newValue));
                        }
                        if(!(this.getData('dragged')) && this.getData('dragStatus') != 1 && this.getData('hueStatus') != 1 && !(this.getData('counter'))){
                            this.setRgbColor('#' + this.getData('cpCurrentRgbCode'));
                            this.executeOnChange();
                        }
                        else{
                            this.setData('dragged',false);
                            if(this.getData('counter')){
                                this.setData('counter', this.getData('counter') - 1);
                                if(!(this.getData('counter'))){
                                    this.setData('restrictOnChange',true);
                                    this.__updateOpacityInForm();
                                    // this.setData('counter', 1);
                                }
                            }
                        }
                    }
                }
                if(type === "R" || type === "G" || type === "B"){
                    this.setData('inputChangeEvnt',true);
                    var red = comp.querySelector('#lyteCP__R').component.getData('ltPropValue'),
                        green = comp.querySelector('#lyteCP__G').component.getData('ltPropValue'),
                        blue = comp.querySelector('#lyteCP__B').component.getData('ltPropValue');
                    if(red && green && blue){
                        this.__setCurrentRgbCode(window.ColorPicker_Util.getRgbCodeByRgbColors(red, green, blue));
                        // this.setData('cpCurrentRgbCode',window.ColorPicker_Util.getRgbCodeByRgbColors(red, green, blue));
                        if(!(this.getData('dragged')) && this.getData('dragStatus') != 1 && this.getData('hueStatus') != 1 && !(this.getData('counter'))){
                            this.setRgbColor('#' + this.getData('cpCurrentRgbCode'));
                            this.executeOnChange();
                        }
                        else{
                            this.setData('dragged',false);
                            if(this.getData('counter')){
                                this.setData('counter', this.getData('counter') - 1);
                                if(!(this.getData('counter'))){
                                    this.setData('restrictOnChange',true);
                                    this.__updateOpacityInForm();
                                    // this.setData('counter', 1);
                                }
                            }
                        }
                    }
                }
                if(type === "H" || type === "S" || type === "V"){
                    this.setData('inputChangeEvnt',true);
                    var hue = comp.querySelector('#lyteCP__H').component.getData('ltPropValue'),
                        saturation = comp.querySelector('#lyteCP__S').component.getData('ltPropValue'),
                        valueBrightness = comp.querySelector('#lyteCP__V').component.getData('ltPropValue');
                    if(hue && saturation && valueBrightness){
                        this.__setCurrentRgbCode(window.ColorPicker_Util.getRgbCodeByHsv(hue, saturation, valueBrightness));
                        // this.setData('cpCurrentRgbCode',window.ColorPicker_Util.getRgbCodeByHsv(hue, saturation, valueBrightness));
                        if(!(this.getData('dragged')) && this.getData('dragStatus') != 1 && this.getData('hueStatus') != 1 && !(this.getData('counter'))){
                            this.setRgbColor('#' + this.getData('cpCurrentRgbCode'), type === "H" && hue == 360 ? 360 : undefined);
                            this.executeOnChange();
                        }
                        else{
                            this.setData('dragged',false);
                            if(this.getData('counter')){
                                this.setData('counter', this.getData('counter') - 1);
                                if(!(this.getData('counter'))){
                                    this.setData('restrictOnChange',true);
                                    this.__updateOpacityInForm();
                                    // this.setData('counter', 1);
                                }
                            }
                        }
                    }
                }
                if(type === "C" || type === "M" || type === "Y" || type === "K"){
                    this.setData('inputChangeEvnt',true);
                    var c = comp.querySelector('#lyteCP__C').component.getData('ltPropValue'),
                        m = comp.querySelector('#lyteCP__M').component.getData('ltPropValue'),
                        y = comp.querySelector('#lyteCP__Y').component.getData('ltPropValue'),
                        k = comp.querySelector('#lyteCP__K').component.getData('ltPropValue');
                    if(c && m && y && k){
                        this.setData('cmykColors', [c/100,m/100,y/100,k/100]);
                        this.__setCurrentRgbCode(window.ColorPicker_Util.getRgbCodeByCmyk(c, m, y, k));
                        // this.setData('cpCurrentRgbCode',window.ColorPicker_Util.getRgbCodeByCmyk(c, m, y, k));
                        if(!(this.getData('dragged')) && this.getData('dragStatus') != 1 && this.getData('hueStatus') != 1 && !(this.getData('counter'))){
                            this.setRgbColor('#' + this.getData('cpCurrentRgbCode'));
                            this.executeOnChange();
                        }
                        else{
                            this.setData('dragged',false);
                            if(this.getData('counter')){
                                this.setData('counter', this.getData('counter') - 1);
                            }
                        }
                    }
                }
                if(type === "A"){
                    var a = comp.querySelector('#lyteCP__A').component.getData('ltPropValue');
                    if(a){
                        this.setData('opacity', parseInt(a));
                        if(this.getData('opacityStatus') != 1){
                            this.__setOpacitySliderPos();
                            this.__setOpacityColorPreviewDiv();
                            if(!this.getData('restrictOnChange')){
                                this.executeOnChange();
                            }
                            else{
                                this.setData('restrictOnChange',false);
                            }
                        }
                    }
                }
                this.validateInputValue(type, lyteInp);
            },

            onInputBlur : function(type, event, lyteInp){
                switch(type){
                    case "HEX" : 
                        lyteInp.setData("ltPropValue",("#" + this.getData("cpCurrentRgbCode")));
                        break;
                    case "R" :
                        var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData("cpCurrentRgbCode"));
                        lyteInp.setData('ltPropValue',rgbColor.red+"");
                        break;
                    case "G" :
                        var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData("cpCurrentRgbCode"));
                        lyteInp.setData('ltPropValue',rgbColor.green+"");
                        break;
                    case "B" :
                        var rgbColor = window.ColorPicker_Util.getRgbColorsByRgbCode(this.getData("cpCurrentRgbCode"));
                        lyteInp.setData('ltPropValue',rgbColor.blue+"");
                        break;
                    case "H" :
                        lyteInp.setData('ltPropValue',Math.round(this.getData('currentHue'))+"");
                        break;
                    case "S" :
                        lyteInp.setData('ltPropValue',Math.round(this.getData('currentSaturation'))+"");
                        break;
                    case "V" :
                        lyteInp.setData('ltPropValue',Math.round(this.getData('currentBrightness'))+"");
                        break;
                    case "C" :
                        lyteInp.setData('ltPropValue',Math.round(this.getData('cmykColors')[0] * 100)+"");
                        break;
                    case "M" :
                        lyteInp.setData('ltPropValue',Math.round(this.getData('cmykColors')[1] * 100)+"");
                        break;
                    case "Y" :
                        lyteInp.setData('ltPropValue',Math.round(this.getData('cmykColors')[2] * 100)+"");
                        break;
                    case "K" :
                        lyteInp.setData('ltPropValue',Math.round(this.getData('cmykColors')[3] * 100)+"");
                        break;
                    case "A" :
                        lyteInp.setData('ltPropValue', this.getData('opacity'));
                        break;
                }
                
            },
            afterInputRender : function(type,input){
                var self = this;
                input.querySelector('input').addEventListener('paste', function(e){
                    self.pasteFunc(e,type);
                });
            },
            changeFormatView : function(e,value){
                var currentTarget = window._lyteUiUtils.getCurrentTarget( e );
                var dropDown = $L('.cPDropDown',this.$node)[0];
                if(this.getData('cpDropdownType') && dropDown){
                    this.setData('cpDropdownType',dropDown.getData('ltPropSelected'))
                }
                if( currentTarget.tagName == 'LYTE-DROP-BOX' ){
                    this.__updateRgbInForm(true);
                }
            }
        }), arg1);
    }

    static actions(arg1) {
        return Object.assign(super.actions({		
            colorhover : function(event,color,elem,colorPallete){
                if(this.getMethods('onColorHover')){
                    this.executeMethod('onColorHover',color,elem,colorPallete);
                }
            },
            onPaste : function(event, prop){
                this.pasteFunc(event, prop);
            },

            inputKeyDown : function(event, comp, type){
            //BACKSPACE - key code -> 8
                //DELETE - key code -> 46
                var arrowkeys = ['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight'];
                
                if(type == "HEX"){
                    if((event.shiftKey && (event.keyCode || event.which) == 51) || 
                        (!(event.shiftKey || event.metaKey) && (event.keyCode || event.which) >= 48 && (event.keyCode || event.which) <= 57) || 
                        ((event.keyCode || event.which) >= 65 && (event.keyCode || event.which) <= 70) || 
                        ((event.keyCode || event.which) >= 97 && (event.keyCode || event.which) <= 102) || 
                        ((event.keyCode || event.which) >= 96 && (event.keyCode || event.which) <= 105) ||
                        ((event.keyCode || event.which) == 8) || ((event.keyCode || event.which) == 35) || 
                        ((event.keyCode || event.which) >= 37 && (event.keyCode || event.which) <= 40) || 
                        ((event.keyCode || event.which) == 46)){
                        var input = comp.querySelector('input'),
                            start = input.selectionStart,
                            end = input.selectionEnd,
                            value = input.value;
                        if((event.keyCode || event.which) == 8){
                            if(start == value.length){
                                value = value.slice(0,-1);
                            }
                            // else if(start == 1){
                            // 	value = value.slice(1);
                            // }
                            else if(start > 0){
                                value = value.slice(0,start-1)+value.slice(start);
                            }
                            return;
                        }
                        if((event.keyCode || event.which) == 46){
                            if(start < value.length){
                                value = value.slice(0,start) + value.slice(start+1);
                            }
                            return;
                        }
                        if(arrowkeys.indexOf(event.key) != -1){
                            return;
                        }
                        else if( !( (event.ctrlKey || event.metaKey) && ((event.keyCode || event.which) == 65 || (event.keyCode || event.which) == 67) ) ){
                            var char = arrowkeys.indexOf(event.key) != -1 ? "" : event.key;
                            if(((event.keyCode || event.which) >= 96 && (event.keyCode || event.which) <= 105)){
                                var key = event.keyCode || event.which;
                                key -= 48;
                                char = String.fromCharCode(key);
                            }
                            // for start == end, just add the char at the cursor position and check for value
                            // for start != end, some characters are selected in the input, so those characters will be replaced with the new char and then check for the value
                            value = start == end ? value.slice(0,start) + char + value.slice(start) : value.slice(0,start) + value.slice(end) + char;
                        }
                        if( (value.length == 1 && value[0] != '#') || (value.length > 0 && (value[0] != '#' || value.substring(1).indexOf('#') != -1)) ){
                            event.preventDefault();
                            event.stopPropagation();
                        }
                    }
                    else if( !((event.ctrlKey || event.metaKey) && (event.keyCode || event.which) == 86)  && (event.keyCode != 9)){
                        event.preventDefault();
                        event.stopPropagation();
                    }
                }
                else if((!(event.shiftKey || event.metaKey) && (event.keyCode || event.which) >= 48 && (event.keyCode || event.which) <= 57) || 
                        ((event.keyCode || event.which) == 8) || ((event.keyCode || event.which) == 46) || 
                        ((event.keyCode || event.which) >= 37 && (event.keyCode || event.which) <= 40) ||
                        ((event.keyCode || event.which) >= 96 && (event.keyCode || event.which) <= 105) ||
                        ((event.ctrlKey || event.metaKey) && ((event.keyCode || event.which) == 65 || (event.keyCode || event.which) == 67))){
                    var input = comp.querySelector('input'),
                        start = input.selectionStart,
                        end = input.selectionEnd,
                        value = input.value;
                    if((event.keyCode || event.which) == 8){
                        if(start == value.length){
                            value = value.slice(0,-1);
                        }
                        // else if(start == 1){
                        // 	value = value.slice(1);
                        // }
                        else if(start > 0){
                            value = value.slice(0,start-1)+value.slice(start);
                        }
                        return;
                    }
                    if((event.keyCode || event.which) == 46){
                        if(start < value.length){
                            value = value.slice(0,start) + value.slice(start+1);
                        }
                        return;
                    }
                    else if( !( (event.ctrlKey || event.metaKey) && ((event.keyCode || event.which) == 65 || (event.keyCode || event.which) == 67) ) ){
                        var char = arrowkeys.indexOf(event.key) != -1 ? "" : event.key;
                        if(((event.keyCode || event.which) >= 96 && (event.keyCode || event.which) <= 105)){
                            var key = event.keyCode || event.which;
                            key -= 48;
                            char = String.fromCharCode(key);
                        }
                        // for start == end, just add the char at the cursor position and check for value
                        // for start != end, some characters are selected in the input, so those characters will be replaced with the new char and then check for the value
                        value = start == end ? value.slice(0,start) + char + value.slice(start) : value.slice(0,start) + value.slice(end) + char;
                    }
                    if(type == "R" || type == "G" || type == "B"){
                        if(value && (parseInt(value) < 0 || parseInt(value) >255)){
                            event.preventDefault();
                            event.stopPropagation();
                        }
                    }
                    else if(type == "H"){
                        if(value && (parseInt(value) < 0 || parseInt(value) >360)){
                            event.preventDefault();
                            event.stopPropagation();
                        }
                    }
                    else if(type == "S" || type == "V" || type == "C" || type == "M" || type == "Y" || type == "K" || type == "A"){
                        if(value && (parseInt(value) < 0 || parseInt(value) >100)){
                            event.preventDefault();
                            event.stopPropagation();
                        }
                    }
                }
                else if( !((event.ctrlKey || event.metaKey) && (event.keyCode || event.which) == 86) && (event.keyCode != 9) ){
                    event.preventDefault();
                    event.stopPropagation();
                }
            },

            callOnSelect : function(event,close){
                
                this.executeOnSelect(event,close);
            },

            selectColor : function(event,color){
                this.defaultSelectedColor(event,color,event.target)
            },

            goToAdvancedCP : function(event){
                if(!this.getData('cpInline')){
            
                    var ele =  this.$node;
                    var popOver = ele.closest('.lytePopover') /*ele.querySelector('.lytePopover')*/;
                    var popOverOffset = popOver.getBoundingClientRect();
                
                }
                this.setData('cpBasicColorPicker',false);
            },

            noFillExecute : function(event){
                this.nofillFunc(event);
            },

            defaultColorExecute : function(event){
                var color = 'rgb(0, 0, 0)';
                var rgbColor = color.substring(4,color.length-1).split(",");
                var selectedColor = {
                    "hex" : ("#" + window.ColorPicker_Util.getRgbCodeByRgbColors(rgbColor[0],rgbColor[1],rgbColor[2])),
                    "rgb" : color,
                    "opacity" : this.getData('opacity'),
                    'nofill': false
                };
                this.setData('cpSelectedColor',selectedColor);
                this.executeOnSelect(event);
            },

            closeColorPicker : function(event){
                this.executeOnClose(event);
                this.executeOnCancelButton(event);
            },

            fromAdvCPtoBasic : function(event){
                this.setData('cpBasicColorPicker',true);
            },
            
            copyValueToClipboard : function(prop){
                var ele = this.$node /*this.getData('cpInline') ? this.$node : this.childComp*/;
                // if(prop == "fromAdv"){
                var colorCodeFormat = ele.querySelector('lyte-dropdown').ltProp('selected');
                this.inputEle.select();			
                try {
                    var successful = document.execCommand('copy');
                    if(!this.getData('tooltipTextChanged')){
                        ele.querySelector('.previewDiv').setAttribute('lt-prop-title',window._lyteUiUtils.i18n('Color.value.copied'));
                        this.setData('tooltipTextChanged',true);
                        $L.fastdom.mutate(function(){

                            ele.querySelector('.previewDiv').dispatchEvent(new Event('mousemove'));
                        })
                    
                    }

                } catch (err) {
                    // console.log('Oops, unable to copy');
                }
            },

            onOverColorPreviewDiv : function(event){
                var ele = this.$node;
                if(this.getData('tooltipTextChanged')){
                    ele.querySelector('.previewDiv').setAttribute('lt-prop-title',window._lyteUiUtils.i18n('Copy.Color.value.to.Clipboard'));
                    this.setData('tooltipTextChanged',false);
                }
            },
            changeValueOfDynamicColor: function (params) {
                this.$node.querySelector('.previewDiv').style.backgroundColor = this.$node.querySelector('.dynamicDiv').style.backgroundColor;
            }
        }), arg1);
    }

    static observers(arg1) {
        return Object.assign(super.observers({
            //Observes color chnage and sets the color for advanced colorpicker
            observeColorChange : function(){
                // console.log(this.getData('cpBoardColor'));
                if(!(this.getData('cpBasicColorPicker')) && !(this.getData('stopObserverOnBoardColorChange')) && this.inputEle && this.isCurrentRgbCodeDifferent()){
                    // console.log("triggered boardColor change");
                    var boardColor = this.getData('cpBoardColor');
                    this.__setCurrentRgbCodeFromBoardColor();
                    this.setData('inputChangeEvnt',true);
                    this.setRgbColor('#' + this.getData('cpCurrentRgbCode'));
                    this.__updateRgbInForm();
                    if( (/rgba/.test(boardColor) || (/#/.test(boardColor) && boardColor.length == 9)) ){
                        this.__updateOpacityInForm();
                    }
                }
                if(this.getData('stopObserverOnBoardColorChange')){
                    this.setData('stopObserverOnBoardColorChange',false);
                }
            }.observes('cpBoardColor'),

            cpAvailableColorsObserver : function(){
                this.initializeBasicColorPicker(undefined,this.$node);
            }.observes('cpAvailableColors'),

            observeSelectedColorChange : function() {
                this.selectedColor();
            }.observes('cpSelectedColor'),

            basicColorpickerObs : function(){
                if(this.getData('cpBasicColorPicker')){
                    this.fromAdvCPtoBasic();
                }else{
                    this.goToAdvancedCP();
                }
            }.observes('cpBasicColorPicker')
        }), arg1);
    }

    _() {
        _;
    }
}

ColorpickerUiComponent._template = "<template tag-name=\"colorpicker-ui\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpBasicColorPicker}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteColorPicker__default\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpDefaultColorEnable}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteColorPicker__palettecontainer LCP_DefaultColorCont default_color\" tabindex=\"0\" aria-label=\"{{lyteUiI18n(&quot;Default.Colors&quot;)}}\" onclick=\"{{action(&quot;selectColor&quot;,event,cpDefaultColor)}}\"> <div class=\"lyteColorPicker__palette\"> {{log(cpDefaultColor)}} <span class=\"lyteColorPicker__colorpan cpDefault__colors\" onmouseenter=\"{{action('colorhover',event,color,this,'DefaultColor')}}\" lt-prop-title=\"{{lyteUiRgbToHex(cpDefaultColor)}}\" lt-prop-tooltip-config=\"{&quot;position&quot; : &quot;bottom&quot;,&quot;showdelay&quot; : &quot;1000&quot;}\" data-col=\"{{indexVal}}\" tabindex=\"-1\" aria-label=\"{{lyteUiRgbToHex(cpDefaultColor)}}\" aria-selected=\"false\">{{cpDefaultColor}}</span> </div> <span class=\"lyteColorPicker__defaultpaletteheading\">Default Color</span> </div> </template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpYield}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-yield yield-name=\"colorpickeryield\"></lyte-yield> </template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpNoFillButton}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteColorPicker__nocolorbutton\" onclick=\"{{action(&quot;noFillExecute&quot;,event)}}\"> <span class=\"lyteColorPicker__noFill\"> </span> <span class=\"lyteColorPicker__text\">{{cpNoFillLabel}}</span> </div> </template></template><div class=\"lyteColorPicker__palettecontainer LCP_predefinedPallete\" tabindex=\"0\" aria-label=\"{{cpPaletteLabel}}\"> <div class=\"lyteColorPicker__paletteheading\">{{cpPaletteLabel}}</div> <div class=\"lyteColorPicker__palette {{if(cpAccent,'lyteColorPickerAccentColorPalette','')}}\"> <div role=\"grid\"> <template items=\"{{cpAvailableColors}}\" item=\"colorArray\" index=\"rowindexVal\" is=\"for\" _new=\"true\"> <ul class=\"lyteColorPicker__shades default__colors {{if(cpShowHiddenColor,'lyteColorPickerShowHiddenColor','')}}\" role=\"row\"> <template items=\"{{colorArray}}\" item=\"color\" index=\"indexVal\" is=\"for\" _new=\"true\"> <li class=\"lyteColorPicker__colorpan {{if(cpAccent,'lyteColorPickerAccentColor','')}}\" onmouseenter=\"{{action('colorhover',event,color,this,'themeColor')}}\" lt-prop-title=\"{{if(cpAccent,lyteUiColorpickerAccentTooltip(cpAccentColor[indexVal],rowindexVal),lyteUiRgbToHex(color))}}\" data-color=\"{{lyteUiRgbToHex(color)}}\" lt-prop-tooltip-config=\"{&quot;position&quot; : &quot;bottom&quot;,&quot;showdelay&quot; : &quot;1000&quot;}\" onclick=\"{{action(&quot;selectColor&quot;,event,color)}}\" tabindex=\"-1\" data-row=\"{{rowindexVal}}\" data-col=\"{{indexVal}}\" aria-label=\"{{lyteUiRgbToHex(color)}}\" role=\"cell\" aria-selected=\"false\">{{color}}</li> </template> </ul> </template> </div> </div> </div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpStandardColors}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteColorPicker__palettecontainer LCP_standardColorCont\" tabindex=\"0\" aria-label=\"{{lyteUiI18n(&quot;Standard.Colors&quot;)}}\"> <div class=\"lyteColorPicker__paletteheading\">{{lyteUiI18n(\"Standard.Colors\")}}</div> <div class=\"lyteColorPicker__palette\"> <ul class=\"standard__colors\"> <template items=\"{{cpStandardColorArray}}\" item=\"color\" index=\"indexVal\" is=\"for\" _new=\"true\"> <li class=\"lyteColorPicker__colorpan\" onmouseenter=\"{{action('colorhover',event,color,this,'standardColor')}}\" lt-prop-title=\"{{lyteUiRgbToHex(color)}}\" lt-prop-tooltip-config=\"{&quot;position&quot; : &quot;bottom&quot;,&quot;showdelay&quot; : &quot;1000&quot;}\" onclick=\"{{action(&quot;selectColor&quot;,event,color)}}\" data-col=\"{{indexVal}}\" tabindex=\"-1\" aria-label=\"{{lyteUiRgbToHex(color)}}\" aria-selected=\"false\">{{color}}</li> </template> </ul> </div> </div> </template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpUsedColors}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteColorPicker__palettecontainer usedColor__container\" tabindex=\"0\" aria-label=\"{{lyteUiI18n(&quot;Recently.Used.Colors&quot;)}}\"> <div class=\"lyteColorPicker__paletteheading\">{{lyteUiI18n(\"Recently.Used.Colors\")}}</div> <div class=\"lyteColorPicker__palette\"> <ul class=\"lyteColorPickerUsedColors used__colors\"> <template items=\"{{usedColors}}\" item=\"color\" index=\"indexVal\" is=\"for\" _new=\"true\"> <li class=\"lyteColorPicker__colorpan {{lyteUiAddPE(color)}}\" onmouseenter=\"{{action('colorhover',event,color,this,'usedColor')}}\" lt-prop-title=\"{{lyteUiRgbToHex(color)}}\" lt-prop-tooltip-config=\"{&quot;position&quot; : &quot;bottom&quot;,&quot;showdelay&quot; : &quot;1000&quot;}\" onclick=\"{{action(&quot;selectColor&quot;,event,color)}}\" tabindex=\"-1\" data-col=\"{{indexVal}}\" aria-label=\"{{lyteUiRgbToHex(color)}}\" aria-selected=\"false\">{{color}}</li> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{lyteUiCPInsertBreak(indexVal)}}\" is=\"case\" lc-id=\"lc_id_0\"> <br> </template></template></template> </ul> </div> </div> </template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpAdvancedColorButton}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteColorPicker__navigatable lyteColorPicker__morecolorbutton\" onclick=\"{{action(&quot;goToAdvancedCP&quot;,event)}}\" role=\"button\" tabindex=\"0\"> <span class=\"lyteColorPicker__icon\"></span> <span class=\"lyteColorPicker__text\">{{lyteUiI18n(\"Advanced.Colors\")}}</span> <span class=\"lyteColorpickerForwardArrow h-alignright LCP_rightArrow\"></span> </div> </template></template></div> </template><template default=\"\"> <div class=\"lyteColorPicker--advanced\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{expHandlers(cpNoFillButton,'&amp;&amp;',cpAdvancedNoFillButton)}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteColorPicker__nocolorbutton\" onclick=\"{{action(&quot;noFillExecute&quot;,event)}}\"> <span class=\"lyteColorPicker__noFill\"> </span> <span class=\"lyteColorPicker__text\">{{cpNoFillLabel}}</span> </div> </template></template><div class=\"lyteColorPicker__maparea\"> <div class=\"colorDiv\" id=\"colorDiv\"> <div id=\"lyteCPImgDiv\"></div> <div class=\"colorSlider_palette_circle colorSlider_palette_circleBlack\"></div> </div> </div> <div class=\"lyteColorPicker__previewDiv lyteColorPicker__sliderDiv\"> <div> <div class=\"colorSlider_hue\"> <div class=\"colorSlider_sliderHandle\"> </div> <div class=\"colorSlider_hueBar_border\"> <div class=\"colorSlider_hueBar\"></div> </div> </div> <div class=\"lyteColorPicker__opacityslider\"> <div class=\"opacityslider__circlethumb\"></div> <div class=\"lyteColorPicker__transparentbg\"> <div class=\"opacityslider__track\"></div> </div> </div> </div> <div class=\"lyteCP_PreviewWrapper {{setDynamicColorClass(this,cpDisplayPreviousColor)}}\"> <div class=\"lyteCP__transparentbg\"> <div class=\"previewDiv\" onmouseover=\"{{action(&quot;onOverColorPreviewDiv&quot;,event)}}\" onclick=\"{{action(&quot;copyValueToClipboard&quot;)}}\" lt-prop-title=\"{{lyteUiI18n('Copy.Color.value.to.Clipboard')}}\" lt-prop-tooltip-config=\"{&quot;position&quot; : &quot;bottom&quot;}\"></div> </div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpDisplayPreviousColor}}\" is=\"case\" lc-id=\"lc_id_0\"><div class=\"lyteCP__transparentbg\"> <div class=\"dynamicDiv\" onclick=\"{{action(&quot;changeValueOfDynamicColor&quot;)}}\" lt-prop-tooltip-config=\"{&quot;position&quot; : &quot;bottom&quot;}\"></div> </div></template></template></div> </div> <div class=\"lyteColorPicker__previewDiv\"> <div class=\"selectFormat\"> <lyte-dropdown lt-prop-box-class=\"{{cpDropBoxClass}}\" lt-prop-selected=\"{{cpDropdownType}}\" class=\"cPDropDown\" lt-prop-yield=\"true\" lt-prop-freeze=\"false\" lt-prop-tabindex=\"1\" on-option-selected=\"{{method(&quot;changeFormatView&quot;)}}\" tabindex=\"0\"> <template is=\"registerYield\" yield-name=\"yield\"> <lyte-drop-button class=\"colorPickerDD\"> <span class=\"lyteMarginRight lyteDropdownLabel\">{{dropButtonValue}}</span> <lyte-icon class=\"dropdown\"></lyte-icon> </lyte-drop-button> <lyte-drop-box class=\"lyteCPDropbox {{cpDropBoxClass}}\"> <lyte-drop-body> <template items=\"{{cpColorFormats}}\" item=\"item\" index=\"indexVal\" is=\"for\" _new=\"true\"> <lyte-drop-item data-value=\"{{item}}\"> {{item}} </lyte-drop-item> </template> </lyte-drop-body> </lyte-drop-box> </template> </lyte-dropdown> </div> <div class=\"showValue\"> <lyte-table lt-prop-yield=\"true\" class=\"cpValFormatIp\" lt-prop-prevent-scrollbar=\"true\" lt-prop-role=\"colorpicker\"> <template is=\"registerYield\" yield-name=\"yield\"> <lyte-table-structure> <lyte-tbody> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ifEquals(dropButtonValue,lyteUiI18n(&quot;HEX&quot;))}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-tr> <lyte-td> <lyte-input id=\"lyteCPShowValue\" lt-prop-type=\"text\" lt-prop-appearance=\"box\" lt-prop-maxlength=\"7\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;HEX&quot;)}}\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;HEX&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;HEX&quot;)}}\" lt-prop-direction=\"vertical\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','HEX')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_HEX_label&quot;}\"> </lyte-input> </lyte-td> <lyte-td> <lyte-input id=\"lyteCP__A\" lt-prop-type=\"text\" lt-prop-appearance=\"box\" lt-prop-maxlength=\"3\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;A&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;A&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;A&quot;)}}\" lt-prop-direction=\"vertical\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','A')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_HEX_alpha_label&quot;}\"> </lyte-input> </lyte-td> </lyte-tr> <lyte-tr class=\"lyteCPHexText\" role=\"row\"> <lyte-td id=\"lytecp_HEX_label\">HEX</lyte-td> <lyte-td id=\"lytecp_HEX_alpha_label\" class=\"lyteColorPicker_alpha\">{{lyteUiSetAlphaLabel(cpOpacityLabel)}}</lyte-td> </lyte-tr> </template><template case=\"{{ifEquals(dropButtonValue,lyteUiI18n(&quot;RGB&quot;))}}\" is=\"case\" lc-id=\"lc_id_1\"> <lyte-tr role=\"row\"> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"255\" id=\"lyteCP__R\" lt-prop-type=\"text\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" lt-prop-maxlength=\"3\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;R&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;R&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;R&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','R')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_R_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"255\" id=\"lyteCP__G\" lt-prop-type=\"text\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" lt-prop-maxlength=\"3\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;G&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;G&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;G&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','G')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_G_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"255\" id=\"lyteCP__B\" lt-prop-type=\"text\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" lt-prop-maxlength=\"3\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;B&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;B&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;B&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','B')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_B_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"100\" id=\"lyteCP__A\" lt-prop-type=\"text\" lt-prop-appearance=\"box\" lt-prop-maxlength=\"3\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;A&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;A&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;A&quot;)}}\" lt-prop-direction=\"vertical\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','A')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_RGB_alpha_label&quot;}\"> </lyte-number> </lyte-td> </lyte-tr> <lyte-tr> <lyte-td id=\"lytecp_R_label\">R</lyte-td> <lyte-td id=\"lytecp_G_label\">G</lyte-td> <lyte-td id=\"lytecp_B_label\">B</lyte-td> <lyte-td id=\"lytecp_RGB_alpha_label\" class=\"lyteColorPicker_alpha\">{{lyteUiSetAlphaLabel(cpOpacityLabel)}}</lyte-td> </lyte-tr> </template><template case=\"{{ifEquals(dropButtonValue,lyteUiI18n(&quot;HSV&quot;))}}\" is=\"case\" lc-id=\"lc_id_2\"> <lyte-tr> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"360\" id=\"lyteCP__H\" lt-prop-type=\"text\" lt-prop-maxlength=\"3\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;H&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;H&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;H&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','H')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_H_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"100\" id=\"lyteCP__S\" lt-prop-type=\"text\" lt-prop-maxlength=\"3\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;S&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;S&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;S&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','S')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_S_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"100\" id=\"lyteCP__V\" lt-prop-type=\"text\" lt-prop-maxlength=\"3\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;V&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;V&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;V&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','V')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_V_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"100\" id=\"lyteCP__A\" lt-prop-type=\"text\" lt-prop-appearance=\"box\" lt-prop-maxlength=\"3\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;A&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;A&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;A&quot;)}}\" lt-prop-direction=\"vertical\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','A')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_HSV_alpha_label&quot;}\"> </lyte-number> </lyte-td> </lyte-tr> <lyte-tr> <lyte-td id=\"lytecp_H_label\">H</lyte-td> <lyte-td id=\"lytecp_S_label\">S</lyte-td> <lyte-td id=\"lytecp_V_label\">V</lyte-td> <lyte-td id=\"lytecp_HSV_alpha_label\" class=\"lyteColorPicker_alpha\">{{lyteUiSetAlphaLabel(cpOpacityLabel)}}</lyte-td> </lyte-tr> </template><template case=\"{{ifEquals(dropButtonValue,lyteUiI18n(&quot;CMYK&quot;))}}\" is=\"case\" lc-id=\"lc_id_3\"> <lyte-tr> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"100\" id=\"lyteCP__C\" class=\"lyteCP_CMYK\" lt-prop-type=\"text\" lt-prop-maxlength=\"3\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;C&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;C&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;C&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','C')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_C_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"100\" id=\"lyteCP__M\" class=\"lyteCP_CMYK\" lt-prop-type=\"text\" lt-prop-maxlength=\"3\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;M&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;M&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;M&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','M')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_M_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"100\" id=\"lyteCP__Y\" class=\"lyteCP_CMYK\" lt-prop-type=\"text\" lt-prop-maxlength=\"3\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;Y&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;Y&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;Y&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','Y')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_Y_label&quot;}\"> </lyte-number> </lyte-td> <lyte-td> <lyte-number lt-prop-min=\"0\" lt-prop-max=\"100\" id=\"lyteCP__K\" class=\"lyteCP_CMYK\" lt-prop-type=\"text\" lt-prop-maxlength=\"3\" lt-prop-appearance=\"box\" lt-prop-direction=\"vertical\" onkeydown=\"{{action(&quot;inputKeyDown&quot;,event,this,&quot;K&quot;)}}\" on-value-change=\"{{method(&quot;inputValueChanged&quot;,&quot;K&quot;)}}\" on-blur=\"{{method(&quot;onInputBlur&quot;,&quot;K&quot;)}}\" lt-prop-aria=\"true\" after-render=\"{{method('afterInputRender','K')}}\" lt-prop-aria-attributes=\"{&quot;aria-labelledby&quot; : &quot;lytecp_K_label&quot;}\"> </lyte-number> </lyte-td> </lyte-tr> <lyte-tr> <lyte-td id=\"lytecp_C_label\">C</lyte-td> <lyte-td id=\"lytecp_M_label\">M</lyte-td> <lyte-td id=\"lytecp_Y_label\">Y</lyte-td> <lyte-td id=\"lytecp_K_label\">K</lyte-td> </lyte-tr> </template></template></lyte-tbody> </lyte-table-structure> </template> </lyte-table> <input id=\"lyteCPHiddenInput\" tabindex=\"-1\" aria-hidden=\"true\"> </div> </div> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpInline}}\" is=\"case\" lc-id=\"lc_id_0\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpMoreColorOptionSelected}}\" is=\"case\" lc-id=\"lc_id_0\"> <div class=\"lyteColorPicker__commandbar\"> <lyte-button class=\"LCP_backbtn\" lt-prop-size=\"small\" lt-prop-appearance=\"default\" onclick=\"{{action(&quot;fromAdvCPtoBasic&quot;,event)}}\" tabindex=\"0\"> <template is=\"registerYield\" yield-name=\"text\">{{lyteUiI18n(cpBackBtnText)}}</template> </lyte-button> </div> </template></template></template><template default=\"\"> <div class=\"lyteColorPicker__commandbar\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{cpMoreColorOptionSelected}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-button class=\"LCP_backbtn\" lt-prop-size=\"small\" lt-prop-appearance=\"default\" onclick=\"{{action(&quot;fromAdvCPtoBasic&quot;,event)}}\" tabindex=\"0\"> <template is=\"registerYield\" yield-name=\"text\">{{lyteUiI18n(cpBackBtnText)}}</template> </lyte-button> <lyte-button class=\"LCP_done\" lt-prop-size=\"small\" lt-prop-appearance=\"primary\" onclick=\"{{action(&quot;callOnSelect&quot;,event,true)}}\" tabindex=\"0\"> <template is=\"registerYield\" yield-name=\"text\">{{lyteUiI18n(cpDoneBtnText)}}</template> </lyte-button> </template><template default=\"\"> <lyte-button class=\"LCP_cancel\" lt-prop-size=\"small\" style=\"margin-left: auto; margin-right: 7px;\" lt-prop-appearance=\"default\" onclick=\"{{action(&quot;closeColorPicker&quot;,event)}}\" tabindex=\"0\"> <template is=\"registerYield\" yield-name=\"text\">{{lyteUiI18n(cpCancelBtnText)}}</template> </lyte-button> <lyte-button class=\"LCP_done\" lt-prop-size=\"small\" lt-prop-appearance=\"primary\" onclick=\"{{action(&quot;callOnSelect&quot;,event,true)}}\" tabindex=\"0\"> <template is=\"registerYield\" yield-name=\"text\">{{lyteUiI18n(cpApplyBtnText)}}</template> </lyte-button> </template></template></div> </template></template></div> </template></template></template>";;
ColorpickerUiComponent._dynamicNodes = [{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,1,1],"cn":"lc_id_0"},{"t":"a","p":[1,1,3],"cn":"lc_id_0"},{"t":"tX","p":[1,1,3,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":6,"sibl":[5],"cn":"lc_id_0"},{"t":"s","p":[1,2],"c":{"lc_id_0":{"dN":[{"t":"i","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":5,"sibl":[4],"cn":"lc_id_0"},{"t":"s","p":[1,3],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,3,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":4,"sibl":[3],"cn":"lc_id_0"},{"t":"a","p":[1,4],"cn":"lc_id_0"},{"t":"tX","p":[1,4,1,0],"cn":"lc_id_0"},{"t":"a","p":[1,4,3],"cn":"lc_id_0"},{"t":"a","p":[1,4,3,1,1],"cn":"lc_id_0"},{"t":"f","p":[1,4,3,1,1],"dN":[{"t":"a","p":[1]},{"t":"a","p":[1,1]},{"t":"f","p":[1,1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,0]}]}],"in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"s","p":[1,6],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,1,0],"cn":"lc_id_0"},{"t":"a","p":[1,3,1,1],"cn":"lc_id_0"},{"t":"f","p":[1,3,1,1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,0]}],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"s","p":[1,7],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,1,0],"cn":"lc_id_0"},{"t":"a","p":[1,3,1,1],"cn":"lc_id_0"},{"t":"f","p":[1,3,1,1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,0]},{"t":"s","p":[3],"c":{"lc_id_0":{"dN":[],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"]}],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"s","p":[1,8],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,3,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"tX","p":[1,3,0],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":6,"sibl":[5],"cn":"default"},{"t":"a","p":[1,4,3],"cn":"default"},{"t":"a","p":[1,4,3,1,1],"cn":"default"},{"t":"s","p":[1,4,3,3],"c":{"lc_id_0":{"dN":[{"t":"a","p":[0,1],"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{}},"hd":true,"co":["lc_id_0"],"in":5,"sibl":[4],"cn":"default"},{"t":"a","p":[1,6,1,1],"cn":"default"},{"t":"r","p":[1,6,1,1,1],"dN":[{"t":"tX","p":[1,1,0]},{"t":"cD","p":[1,3],"in":4,"sibl":[3]},{"t":"cD","p":[1],"in":3,"sibl":[2]},{"t":"a","p":[3]},{"t":"a","p":[3,1,1]},{"t":"f","p":[3,1,1],"dN":[{"t":"a","p":[1]},{"t":"tX","p":[1,1]},{"t":"cD","p":[1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":2,"sibl":[1]},{"t":"cD","p":[3,1],"in":1,"sibl":[0]},{"t":"cD","p":[3],"in":0}],"dc":[4,3,2,1,0],"hc":true,"trans":true,"in":4,"sibl":[3],"cn":"default"},{"t":"cD","p":[1,6,1,1],"in":3,"sibl":[2],"cn":"default"},{"t":"r","p":[1,6,3,1,1],"dN":[{"t":"s","p":[1,1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1,1,1],"cn":"lc_id_0"},{"t":"cD","p":[1,1,1],"in":7,"sibl":[6],"cn":"lc_id_0"},{"t":"cD","p":[1,1],"in":6,"sibl":[5],"cn":"lc_id_0"},{"t":"a","p":[1,3,1],"cn":"lc_id_0"},{"t":"cD","p":[1,3,1],"in":5,"sibl":[4],"cn":"lc_id_0"},{"t":"cD","p":[1,3],"in":4,"sibl":[3],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"cD","p":[3,1],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"tX","p":[3,3,0],"cn":"lc_id_0"},{"t":"cD","p":[3,3],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[3],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true},"lc_id_1":{"dN":[{"t":"a","p":[1,1,1],"cn":"lc_id_1"},{"t":"cD","p":[1,1,1],"in":13,"sibl":[12],"cn":"lc_id_1"},{"t":"cD","p":[1,1],"in":12,"sibl":[11],"cn":"lc_id_1"},{"t":"a","p":[1,3,1],"cn":"lc_id_1"},{"t":"cD","p":[1,3,1],"in":11,"sibl":[10],"cn":"lc_id_1"},{"t":"cD","p":[1,3],"in":10,"sibl":[9],"cn":"lc_id_1"},{"t":"a","p":[1,5,1],"cn":"lc_id_1"},{"t":"cD","p":[1,5,1],"in":9,"sibl":[8],"cn":"lc_id_1"},{"t":"cD","p":[1,5],"in":8,"sibl":[7],"cn":"lc_id_1"},{"t":"a","p":[1,7,1],"cn":"lc_id_1"},{"t":"cD","p":[1,7,1],"in":7,"sibl":[6],"cn":"lc_id_1"},{"t":"cD","p":[1,7],"in":6,"sibl":[5],"cn":"lc_id_1"},{"t":"cD","p":[1],"in":5,"sibl":[4],"cn":"lc_id_1"},{"t":"cD","p":[3,1],"in":4,"sibl":[3],"cn":"lc_id_1"},{"t":"cD","p":[3,3],"in":3,"sibl":[2],"cn":"lc_id_1"},{"t":"cD","p":[3,5],"in":2,"sibl":[1],"cn":"lc_id_1"},{"t":"tX","p":[3,7,0],"cn":"lc_id_1"},{"t":"cD","p":[3,7],"in":1,"sibl":[0],"cn":"lc_id_1"},{"t":"cD","p":[3],"in":0,"cn":"lc_id_1"}],"cdp":{"t":"a","p":[1]},"dcn":true},"lc_id_2":{"dN":[{"t":"a","p":[1,1,1],"cn":"lc_id_2"},{"t":"cD","p":[1,1,1],"in":13,"sibl":[12],"cn":"lc_id_2"},{"t":"cD","p":[1,1],"in":12,"sibl":[11],"cn":"lc_id_2"},{"t":"a","p":[1,3,1],"cn":"lc_id_2"},{"t":"cD","p":[1,3,1],"in":11,"sibl":[10],"cn":"lc_id_2"},{"t":"cD","p":[1,3],"in":10,"sibl":[9],"cn":"lc_id_2"},{"t":"a","p":[1,5,1],"cn":"lc_id_2"},{"t":"cD","p":[1,5,1],"in":9,"sibl":[8],"cn":"lc_id_2"},{"t":"cD","p":[1,5],"in":8,"sibl":[7],"cn":"lc_id_2"},{"t":"a","p":[1,7,1],"cn":"lc_id_2"},{"t":"cD","p":[1,7,1],"in":7,"sibl":[6],"cn":"lc_id_2"},{"t":"cD","p":[1,7],"in":6,"sibl":[5],"cn":"lc_id_2"},{"t":"cD","p":[1],"in":5,"sibl":[4],"cn":"lc_id_2"},{"t":"cD","p":[3,1],"in":4,"sibl":[3],"cn":"lc_id_2"},{"t":"cD","p":[3,3],"in":3,"sibl":[2],"cn":"lc_id_2"},{"t":"cD","p":[3,5],"in":2,"sibl":[1],"cn":"lc_id_2"},{"t":"tX","p":[3,7,0],"cn":"lc_id_2"},{"t":"cD","p":[3,7],"in":1,"sibl":[0],"cn":"lc_id_2"},{"t":"cD","p":[3],"in":0,"cn":"lc_id_2"}],"cdp":{"t":"a","p":[2]},"dcn":true},"lc_id_3":{"dN":[{"t":"a","p":[1,1,1],"cn":"lc_id_3"},{"t":"cD","p":[1,1,1],"in":13,"sibl":[12],"cn":"lc_id_3"},{"t":"cD","p":[1,1],"in":12,"sibl":[11],"cn":"lc_id_3"},{"t":"a","p":[1,3,1],"cn":"lc_id_3"},{"t":"cD","p":[1,3,1],"in":11,"sibl":[10],"cn":"lc_id_3"},{"t":"cD","p":[1,3],"in":10,"sibl":[9],"cn":"lc_id_3"},{"t":"a","p":[1,5,1],"cn":"lc_id_3"},{"t":"cD","p":[1,5,1],"in":9,"sibl":[8],"cn":"lc_id_3"},{"t":"cD","p":[1,5],"in":8,"sibl":[7],"cn":"lc_id_3"},{"t":"a","p":[1,7,1],"cn":"lc_id_3"},{"t":"cD","p":[1,7,1],"in":7,"sibl":[6],"cn":"lc_id_3"},{"t":"cD","p":[1,7],"in":6,"sibl":[5],"cn":"lc_id_3"},{"t":"cD","p":[1],"in":5,"sibl":[4],"cn":"lc_id_3"},{"t":"cD","p":[3,1],"in":4,"sibl":[3],"cn":"lc_id_3"},{"t":"cD","p":[3,3],"in":3,"sibl":[2],"cn":"lc_id_3"},{"t":"cD","p":[3,5],"in":2,"sibl":[1],"cn":"lc_id_3"},{"t":"cD","p":[3,7],"in":1,"sibl":[0],"cn":"lc_id_3"},{"t":"cD","p":[3],"in":0,"cn":"lc_id_3"}],"cdp":{"t":"a","p":[3]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[7,6,5,4,3,2,1,0],"hc":true,"trans":true},"lc_id_1":{"dc":[13,12,11,10,9,8,7,6,5,4,3,2,1,0],"hc":true,"trans":true},"lc_id_2":{"dc":[13,12,11,10,9,8,7,6,5,4,3,2,1,0],"hc":true,"trans":true},"lc_id_3":{"dc":[13,12,11,10,9,8,7,6,5,4,3,2,1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0","lc_id_1","lc_id_2","lc_id_3"],"hc":true,"trans":true,"in":2,"sibl":[1]},{"t":"cD","p":[1,1],"in":1,"sibl":[0]},{"t":"cD","p":[1],"in":0}],"dc":[2,1,0],"hc":true,"trans":true,"in":2,"sibl":[1],"cn":"default"},{"t":"cD","p":[1,6,3,1],"in":1,"sibl":[0],"cn":"default"},{"t":"s","p":[1,8],"c":{"lc_id_0":{"dN":[{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1,1],"cn":"lc_id_0"},{"t":"r","p":[1,1,1],"dN":[{"t":"tX","p":[0]}],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1,1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"s","p":[1,1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"r","p":[1,1],"dN":[{"t":"tX","p":[0]}],"in":3,"sibl":[2],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":2,"sibl":[1],"cn":"lc_id_0"},{"t":"a","p":[3],"cn":"lc_id_0"},{"t":"r","p":[3,1],"dN":[{"t":"tX","p":[0]}],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[3],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"r","p":[1,1],"dN":[{"t":"tX","p":[0]}],"in":3,"sibl":[2],"cn":"default"},{"t":"cD","p":[1],"in":2,"sibl":[1],"cn":"default"},{"t":"a","p":[3],"cn":"default"},{"t":"r","p":[3,1],"dN":[{"t":"tX","p":[0]}],"in":1,"sibl":[0],"cn":"default"},{"t":"cD","p":[3],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[2,0],"hc":true,"trans":true},"default":{"dc":[2,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[5],"hc":true,"trans":true},"default":{"dc":[4,3,2,1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[0]}];;

ColorpickerUiComponent._observedAttributes = [
    "cpDraggable",
    "cpCloseOnBodyClick",
    "cpCloseOnSelection",
    "cpAdvancedColorButton",
    "cpAppearance",
    "cpAccent",
    "cpAccentColor",
    "cpStandardColors",
    "cpDefaultColorEnable",
    "cpUsedColors",
    "cpBasicColorPicker",
    "cpSelectedColor",
    "cpDynamicColor",
    "cpOriginElement",
    "cpWrapperClass",
    "cpCurrentRgbCode",
    "cpBoundary",
    "cpScrollable",
    "cpYield",
    "cpNoFillButton",
    "cpNoFillLabel",
    "cpPaletteLabel",
    "cpInline",
    "cpAdvancedNoFillButton",
    "cpColorFormats",
    "cpDoneBtnText",
    "cpCancelBtnText",
    "cpBackBtnText",
    "cpApplyBtnText",
    "cpStandardColorArray",
    "cpDefaultColor",
    "cpDuration",
    "cpBoardColor",
    "cpAvailableColors",
    "cpMoreColorOptionSelected",
    "cpOpacityLabel",
    "cpUsedColorsNumber",
    "cpDropBoxClass",
    "cpDisplayPreviousColor",
    "stopObserverOnBoardColorChange",
    "usedColors",
    "polygonPoints",
    "noFillPath",
    "moreColorsPath",
    "currentHue",
    "currentBrightness",
    "currentSaturation",
    "paletteSize",
    "circleOffsetSize",
    "circleOffsetBecauseOfWinWidget",
    "posdivElPalette",
    "dragStatus",
    "paletteMaxX",
    "paletteMaxY",
    "clickOnPaletteInProgress",
    "component",
    "hueStatus",
    "poxYHue",
    "poxXHue",
    "dropButtonValue",
    "calledForHueMove",
    "changedColor",
    "tooltipTextChanged",
    "inputChangeEvnt",
    "hueBarSize",
    "opacityStatus",
    "opacityTrackSize",
    "posXOpacity",
    "opacity",
    "sliderOffset",
    "dragged",
    "counter",
    "cmykColors",
    "setPrevCmyk",
    "restrictOnChange",
    "prevColor",
    "cpShowHiddenColor",
    "colorFormat"
];

LyteUiComponentComponentRegistry.registerHelper("setDynamicColorClass", function (_this,cpDisplayPreviousColor) {
	if (cpDisplayPreviousColor) {
		return "lyteCP__dynamicColor";
	}
});

export { ColorpickerUiComponent };

ColorpickerUiComponent.register("colorpicker-ui", {
    hash: "ColorpickerUiComponent_2",
    refHash: "C_lyte-ui-component 2_app_0"
});
