// @zoho/lyte-ui-component/dist/dist/build/__cliCompiledFile/fee9d8fda0e1bf5f2ce980b2288f5a8b.js
import { _defineProperty } from "@slyte/core/src/lyte-utils";
import { prop } from "../../node_modules/@slyte/core/index.js";
import { Component } from "../component.js";
import $L from "../../node_modules/@zoho/lyte-dom/modules/lyte-dom-utils.js"

/**
 * Renders a hovercard
 * @component lyte-hovercard
 * @dependencies lyte-popover
 * /components/lyte-popover.js
 * /theme/compiledCSS/default/ltr/lyte-ui-popover.css
 * @version  3.1.0
 * @methods beforeRender,afterRender,onHovercardShow,onHovercardHide,onHovercardBeforeHide
 * @utility alignHovercard, calculateOffset
 */
class LyteHovercardComponent extends Component {
  constructor() {
    super();
    this._lyteUtilFunctions = ["alignHovercard","calculateOffset"];
  }

  data(arg1) {
    return Object.assign(super.data({
      /** 
       * @componentProperty {boolean} ltPropDisplay=true
       * @version 3.1.0
       */
      'ltPropDisplay' : prop( 'boolean', { 'default' : true } ),
      /** 
       * @componentProperty {boolean} ltPropShow=false
       * @version 3.1.0
       */
        'ltPropShow' : prop( 'boolean', { 'default' : false } ),
        /** 
       * @componentProperty {string} ltPropOriginElem=''
       * @version 3.1.0
       */
        'ltPropOriginElem' : prop( 'string', { 'default' : '' } ),
        /** 
       * @componentProperty {string} ltPropMaxHeight=''
       * @version 3.1.0
       */
        'ltPropMaxHeight' : prop( 'string', { 'default' : '' } ),
        /** 
       * @componentProperty {string} ltPropWidth=''
       * @version 3.1.0
       */
        'ltPropWidth' : prop( 'string', { 'default' : '' } ),
        /** 
       * @componentProperty {string} ltPropHeight=auto
       * @version 3.1.0
       */
        'ltPropHeight' : prop( 'string', { 'default' : 'auto' } ),
        /** 
       * @componentProperty {string} ltPropPlacement=''
       * @version 3.1.0
       */
        'ltPropPlacement' : prop( 'string', { 'default' : '' } ),
        /** 
       * @componentProperty {number} ltPropClass=''
       * @version 3.1.0
       */
        'ltPropClass' : prop( 'string', { 'default' : window._lyteUiUtils.resolveDefaultValue( 'lyte-hovercard', 'class', '' ) } ),
        /** 
       * @componentProperty {string} ltPropId=''
       * @version 3.1.0
       */
        'ltPropId' : prop( 'string', { 'default' : '' } ),
        /** 
       * @componentProperty {number} ltPropShowDelay=0
       * @version 3.1.0
       */
        'ltPropShowDelay' : prop( 'number', { 'default': 0 } ),
        /** 
       * @componentProperty {number} ltPropHideDelay=0
       * @version 3.1.0
       */
        'ltPropHideDelay' : prop( 'number', { 'default': 0 } ),
        /** 
       * @componentProperty {number} ltPropMaxDisplayTime=5000
       * @version 3.1.0
       */
        'ltPropMaxDisplayTime' : prop( 'number', { 'default' : 5000 } ),
        /** 
       * @componentProperty {boolean} ltPropKeepAlive=false
       * @version 3.1.0
       */
        'ltPropKeepAlive' : prop( 'boolean', { 'default' : false } ),
        /** 
       * @componentProperty {boolean} ltPropFollowCursor=false
       * @version 3.1.0
       */
        'ltPropFollowCursor' : prop( 'boolean', { 'default' : false } ),
        /** 
       * @componentProperty {string} ltPropPopoverWrapperClass
       * @version 3.1.0
       */
        'ltPropPopoverWrapperClass' : prop( 'string',{'default': window._lyteUiUtils.resolveDefaultValue( 'lyte-hovercard', 'popoverWrapperClass', '' ) }),
        /** 
       * @componentProperty {object} ltPropOffset={}
       * @version 3.1.0
       */
        'ltPropOffset' : prop( 'object', { 'default' : {} } ),
        /** 
       * @componentProperty {boolean} ltPropCloseOnEscape=true
       * @version 3.1.0
       */
        'ltPropCloseOnEscape' : prop( 'boolean', { 'default' : true } ),
        /** 
        * @componentProperty {boolean} ltPropAutoShow=false
        * @version 3.1.0
        */
        'ltPropAutoShow' : prop( 'boolean', { 'default' : false}),
         /** 
        * @componentProperty {boolean} ltPropHideOnClick=false
        * @version 3.1.0
        */
        'ltPropHideOnClick' : prop( 'boolean', { 'default' : window._lyteUiUtils.resolveDefaultValue( 'lyte-hovercard', 'hideOnClick', false ) } ),
         /**
          * @componentProperty {boolean} ltPropAria
          * @version 3.1.0
          * @default false
          *
          */
        "ltPropAria" : prop( 'boolean', { default : false } ),
        /**
         * @componentProperty {object} ltPropAriaAttributes={}
         * @version 3.1.0
         */
        "ltPropAriaAttributes" : prop( 'object', { default : {} } ),
         /**
         * @componentProperty {boolean} ltPropPreventFocus=true
         */
        "ltPropPreventFocus" : prop('boolean', { default : true } ),
         /**
         * @componentProperty {string} ltPropMaxWidth
         */
        "ltPropMaxWidth" : prop('string',{default:''}),
        "ltPropType" : prop('string',{default :'callout'}),
        "ltPropDimmer":prop("object",{"default":{"color":"black","opacity":"0.4"}}),
        "ltPropAnimation":prop("string",{"default":"fade"}), //fade,zoom
        "ltPropAutoAlign" : prop('boolean', {default : false}),
        "ltPropPopover" : prop('object', { default : {} } ),
        "ltPropCloseOnScroll" : prop('boolean', {default: true}),
        "ltPropUseBetaPopover" : prop('boolean', { default : false}),
        'mousePosition' : prop( 'array', { 'default' : [] } ),
        'mouseover' : prop( 'boolean', { 'default' : false } ),
        'originEle' : prop( 'string', { 'default' : ''}),
        'popoverTagName' : prop('string', {'default':'lyte-popover'})
    }), arg1);   
  }

  init() {
    if( window._lyteUiUtils.isWidget ){
      this.setData('ltPropAutoShow', false)
    } 
    if( this.getMethods( 'beforeRender' ) ){
            this.executeMethod( 'beforeRender', this.$node );
    }
    if(this.getData('ltPropUseBetaPopover')){
        this.setData('popoverTagName', 'lyte-beta-popover')
    }
 }

  didConnect() {
    
    this._popover = this.$node.getElementsByTagName( this.getData('popoverTagName') )[0]
   this._hovercardScroll = this.hovercardScroll.bind( this );
    this._hovercardHideOnClick = this.hovercardHideOnClick.bind(this)
    this._mousemove = this.mousemove.bind(this);
    this._oriEleMouseMove = this.oriEleMousemove.bind(this)
    this._originEleFocus = this.originEleFocus.bind(this)
    this._originEleBlur= this.originEleBlur.bind(this)
    if(this.getData('ltPropOriginElem')){
      try{
        const originElement = this.getOriginElement(this.getData('ltPropOriginElem'))
        if(!originElement){
          console.error("Provide a valid selector for 'ltPropOriginElem'; '"+this.getData('ltPropOriginElem') +"' is not a valid origin selector for the Hovercard component");
        }
        
      } catch(err){
        console.warn("Provide a valid selector for 'ltPropOriginElem'; '"+this.getData('ltPropOriginElem') +"' is not a valid origin element selector for the Hovercard component");
      }
    }
    if(this.getData('ltPropAutoShow') && this.getData('ltPropOriginElem')){
      this.setMouseMove()
      this.addMutationObserver()
      this.addFocusAndBlur()
    }
    $L.fastdom.measure( function() {
         var fg = window.getComputedStyle( this.$node ).getPropertyValue( 'direction' ) == 'rtl';
         $L.fastdom.mutate( function(){
           if( fg ) {
             this.direction = true;
           }
         }.bind( this ) )
    }.bind( this ) )
    window._lyteUiUtils.dispatchEvent( 'afterrender', this.$node ); 

    if( this.getMethods( 'afterRender' ) ) {
       this.executeMethod('afterRender', this.$node);
    }
    this.$node.alignHovercard = function(){
      this._popover = this._popover || this.$node.getElementsByTagName( this.getData('popoverTagName'))[0]
      if(this._popover && this._popover.alignPopover){
        this._popover.alignPopover()
      }
    }.bind( this ) 
    this.$node.calculateOffset = function(){
      this._popover = this._popover || this.$node.getElementsByTagName( this.getData('popoverTagName'))[0]
      if(this._popover && this._popover.calculateOffset){
        this._popover.calculateOffset()
      }
    }.bind( this ) 
  }

  didDestroy() {
    this.$node.classList.remove('lyteActive')
    if( this.getData( 'ltPropHideOnClick') ){
      document.removeEventListener( 'click' , this._hovercardHideOnClick )
    }
    if(this.$node.ltProp( 'originElem' ) ){
      var originElem = this.getOriginElement( this.$node.ltProp( 'originElem' ) ) 

      if(originElem ){
        this._closeHoverCard && originElem.removeEventListener( 'mouseleave', this._closeHoverCard )
        this.removeEventListenerForOriginElem(originElem)
        if(this.getData('ltPropAutoShow')){
          this.removeFocusAndBlur(originElem)
          this.removeMutationObserver()
          this.removeFocusAndBlur(originElem)
        }
      }
    }
   
    if(this._popover){
      this._popover.setData('ltPropShow',false)
    }
    if(window._lyteUiUtils.lyteHovercard){
      delete window._lyteUiUtils.lyteHovercard[this.$node.ltProp( 'originElem' ) ]

    }
    if( this._childComp ){
      this._childComp.remove()
    }
    if(this.getData('ltPropAutoShow')){
      var eventName = (window.document === document ? 'mousemove' : 'mouseover')

      document.removeEventListener(eventName,this._mousemove)
    }
    delete this.prevHoverCardNode;
    delete this._childComp;
    delete this._popover;
  }

  addFocusAndBlur() {
    const originElem = this.getOriginElement(this.getData('ltPropOriginElem'))
    if( originElem ){
      originElem.addEventListener('focus', this._originEleFocus)
      originElem.addEventListener('blur', this._originEleBlur)
    }
   
  }

  removeFocusAndBlur(originElem) {
    if( originElem ){
      originElem.removeEventListener('focus', this._originEleFocus)
      originElem.removeEventListener('blur', this._originEleBlur)
    }
  }

  originEleFocus(event) {
    if(!this.getData('ltPropShow')){
      this.setData('ltPropShow', true)
    }

  }

  originEleBlur() {
    if(this.getData('ltPropShow')){
      this.setData('ltPropShow', false)
    }
  }

  setMaxHeightAndWidth(div) {
    if(this.getData('ltPropMaxWidth')){
      div.style.maxWidth = this.getData('ltPropMaxWidth')
      
    }
    if(this.getData('ltPropMaxHeight')){
      div.style.maxHeight = this.getData('ltPropMaxHeight')
      div.classList.add('lyteHovercardWithMaxHeight')
    }
  }

  isNode(target) {
   return target instanceof HTMLElement || target instanceof Node;  

 }

  addEventListenerForOriginElem(originElem) {
    if(originElem){
      originElem.addEventListener( 'mousemove', this._oriEleMouseMove )
    }
  }

  removeEventListenerForOriginElem(originElem) {
    if(originElem){
      originElem.removeEventListener( 'mousemove', this._oriEleMouseMove )
    }
    this.setData( 'mousePosition', [] )
    this.setData( 'mouseover', false )

  }

  hovercardHideOnClick(event) {
    var target = event.target,
    popoverWormhole = this._popover.component.actualModalDiv;
    if( this.getData( 'ltPropHideOnClick' ) && this.getData( 'ltPropShow' ) && this.isNode(target) && !( target === popoverWormhole || popoverWormhole.contains( target ) ) ){
          this.setData( 'ltPropShow', false )
      }
  }

  oriEleMousemove(eve) {

    if( !this.getData( 'mouseover' ) ) {
        this.mouseovereve( eve )
        this.setData( 'mouseover', true )
    }
    var currMpos = [ eve.clientX, eve.clientY ];
    var mpos = this.getData( 'mousePosition' );
    var diff = [ currMpos[ 0 ] - ( mpos[ 0 ]? mpos[ 0 ] : 0 ), currMpos[ 1 ] - ( mpos[ 1 ] ? mpos[ 1 ] : 0 ) ];
    if( this._popover.ltProp( 'show' ) && this._popover.component.childComp) {
      var popupEle = $L( '.lyteHoverCardFollowCursor .lytePopover' ,this._popover.component.childComp)[ 0 ];
      var clientRect = popupEle.getBoundingClientRect();
      popupEle.style.top = clientRect.top + diff[ 1 ] + 'px';
      popupEle.style.left = clientRect.left + diff[ 0 ] + 'px';
    }
    this.setData( 'mousePosition', currMpos );
  }

  mouseovereve(eve) {
    var mpos = [ eve.clientX, eve.clientY ];
    var pos = '';
    
    this._popover.ltProp( {
      offset : { left : mpos[ 0 ] - 9, top : mpos[ 1 ] - 9, height : 18, width : 18 }
    } )
    this.setData('ltPropOffset',{ left : mpos[ 0 ] , top : mpos[ 1 ] , height : 18, width : 18 })
    this.setData( 'mousePosition', mpos );
  }

  mousemove(event) {
    var nodeName1 = event.target.correspondingElement || event.target;
    while(nodeName1 && nodeName1.tagName != 'BODY' && nodeName1 != document && nodeName1.tagName != 'HTML' && !( nodeName1 instanceof ShadowRoot )  ){
        var iHovercard = nodeName1.getAttribute( 'lyte-hovercard' );

        if( iHovercard ){
          var hovercard = this.findMatchingHoverCard(nodeName1);
          if( hovercard && !hovercard.getData('ltPropShow')){
             hovercard.setData('ltPropShow',true);
          } 
          break;
          
        }
        else {
            nodeName1 = nodeName1.parentNode;
        }  
    }
  
}

  findMatchingHoverCard(node) {
    for(var item in window._lyteUiUtils.lyteHovercard){
       if(node.matches(item)){
         return window._lyteUiUtils.lyteHovercard[item];
       }
   }
  }

  setMouseMove() {
     var map = window._lyteUiUtils.lyteHovercard ? window._lyteUiUtils.lyteHovercard : []
     map[this.$node.ltProp( 'originElem' )] = this.$node;
     window._lyteUiUtils.lyteHovercard = map;
     document.addEventListener('mousemove',this._mousemove)
   }

  compouteOffset(popover) {
      var arr = [ 'ltPropWidth', 'ltPropHeight' ];
      for( var i = 0; i < arr.length; i++ ) {
            if( this.getData( arr[ i ] ) ) {
                  popover.setData( arr[ i ], this.getData( arr[ i ] ) )
            }
       }
  }

  createHoverCard(event, popoverWormhole) {
    var popover = this._popover
    if(popover){
      window._lyteUiUtils.dispatchEvent( 'beforeshow', this.$node, { originalEvent: event } ); 
       var res = true;
        if( this.getMethods( 'onBeforeHovercardShow' ) ) { 
          res = this.executeMethod('onBeforeHovercardShow', this.$node );
        }
        if(!res){
          this.setData('ltPropShow', false)
          return false
        }
        popover.ltProp( 'show', true )
        popover.ltProp( 'allowMultiple', true)
        this.$node.classList.add( 'lyteActive' )
        if( this.getMethods( 'onHovercardShow' ) ) {
                this.executeMethod('onHovercardShow', this.$node );
        }
    }
    if( !this.getData( 'ltPropKeepAlive' ) && !this.getData('ltPropFollowCursor')) {
          var originElem = this.getOriginElement( this.$node.ltProp( 'originElem' ) )
          var eve = event, target 
          if(eve){
            target =  eve.target
          }
          popover._maxdisp = setTimeout( function() {
            this.removeHoverCard(popover, originElem, eve, popoverWormhole, target)
          }.bind( this ), this.getData( 'ltPropMaxDisplayTime' ) );
    }
    
  }

  removeTimeout(popover) {
        clearTimeout( popover._settime )
        clearTimeout( popover._maxdisp )
        clearTimeout( popover._bodyTimeout )
  }

  getOriginElement(selector) {
    var originElem = (window.document === document ? document.querySelector(selector) : document.querySelectorGlobal(selector))
    return originElem
  }

  addMutationObserver() {
    let self = this;
    
    const originElement = this.getOriginElement( this.getData('ltPropOriginElem') )
     this._observer = new MutationObserver((mutationsList, observer) => {
      for (const mutation of mutationsList) {
        if (mutation.type === 'childList') {
          for(let i =0; i < mutation.removedNodes.length; i++){
            const removedNode =  mutation.removedNodes[i]
            if (removedNode === originElement) {
                self.getData('ltPropShow') && self.setData('ltPropShow', false)
                window._lyteUiUtils.lyteHovercard && window._lyteUiUtils.lyteHovercard[ this.getData('ltPropOriginElem') ] && delete window._lyteUiUtils.lyteHovercard[ this.getData('ltPropOriginElem') ]
                self._observer.disconnect()
                break
            }
          }
          mutation.removedNodes.forEach(removedNode => {
            
        });
        }
      }
    })
    const config = { childList: true};

    // Start observing the parent node
    originElement && this._observer.observe(originElement.parentElement, config);
  }

  removeMutationObserver() {
    this._observer && this._observer.disconnect()
    delete  this._observer
  }

  closeHoverCard(event) {
      var wormHole = this._childComp,
       popoverWormhole = this._popover.component.actualModalDiv,
       popover = this._popover,
       originElem = this.getOriginElement( this.$node.ltProp( 'originElem' ) )
      if(  this.prevHoverCardNode && this.isNode(event.target) && ( this.getData('ltPropFollowCursor') || event.target == this.prevHoverCardNode || this.prevHoverCardNode.contains( event.target ) ) && popoverWormhole && event.relatedTarget != popoverWormhole && !popoverWormhole.contains( event.relatedTarget )) {
                var eve = event
                popover._bodyTimeout = setTimeout( this.removeHoverCard.bind( this ), this.getData( 'ltPropHideDelay' ), popover, originElem, eve, popoverWormhole, event.target) ;

          } else if( popoverWormhole  && this.isNode(event.relatedTarget ) && (event.relatedTarget == popoverWormhole || popoverWormhole.contains( event.relatedTarget )  )  ) {  
              this.removeTimeout( popover )
              this._popovermouseleave = this.popoverMouseLeave.bind( this )
              popoverWormhole.addEventListener( 'mouseleave', this._popovermouseleave )
              originElem.removeEventListener( 'mouseleave', this._closeHoverCard )
  
           } 
  }

  removeHoverCard(popover, originElem, event, popoverWormhole, target) {
    var res = true
    if(this.getMethods( 'onHovercardBeforeHide' ) ) {
          res = this.executeMethod( 'onHovercardBeforeHide', this.$node, event );
          if( !res && originElem){
              originElem.removeEventListener( 'mouseleave', this._closeHoverCard )
              if( this._popovermouseleave && popoverWormhole ){
                popoverWormhole.removeEventListener( 'mouseleave', this._popovermouseleave )
              }
          }
    }
    event = event || window.event
    target = target || (event ?  event.target : '')
    if( res && (!event || ( ( this.prevHoverCardNode  && this.isNode(target) && ( target == this.prevHoverCardNode || this.prevHoverCardNode.contains( target ) ) ) || ( popoverWormhole  && this.isNode(target) && ( target == popoverWormhole || popoverWormhole.contains( target ) ) ) ) )){
            this.removeTimeout( popover )

            if( this.getData( 'ltPropShow' ) && popover ) {
                    this.setData( 'ltPropShow', false )
                    popover.setData( 'ltPropShow',false )
                }  
    }
  }

  popoverMouseLeave(event) {
    var wormHole = this._childComp ,
    popoverWormhole = this._popover.component.actualModalDiv ,
    popover = this._popover ,
    originElem = this.getOriginElement( this.$node.ltProp( 'originElem' ) )
    var eve = event

   if( popoverWormhole  && this.isNode(event.target) && ( event.target == popoverWormhole || popoverWormhole.contains( event.target ) ) && event.relatedTarget != this.prevHoverCardNode && !this.prevHoverCardNode.contains( event.relatedTarget ) ) {
     popover._bodyTimeout = setTimeout( this.removeHoverCard.bind( this ), this.getData( 'ltPropHideDelay' ), popover, originElem, eve, popoverWormhole, eve.target);
   }
   else if( this.prevHoverCardNode  && this.isNode(event.relatedTarget) && ( event.relatedTarget == this.prevHoverCardNode || this.prevHoverCardNode.contains( event.relatedTarget ) ) ) {
         popover._settime = setTimeout( this.createHoverCard.bind( this ), this.getData( 'ltPropShowDelay' ), eve, popoverWormhole);
         this._closeHoverCard = this.closeHoverCard.bind( this )
         originElem.addEventListener( 'mouseleave', this._closeHoverCard )
         if( this._popovermouseleave ){
           popoverWormhole.removeEventListener( 'mouseleave', this._popovermouseleave )
         }

   }
  }

  hovercardScroll(event) {
    if($L(this.$node).hasClass("lyteActive")){
    var res = true
        var component = this,
            wormHole = component._childComp ,
            popoverWormhole = component._popover.component.actualModalDiv ,
            popover =component._popover ,
            originElem = this.getOriginElement( this.getData('ltPropOriginElem') )
            
        if(component.getMethods( 'onHovercardBeforeHide' ) ) {
            res = component.executeMethod( 'onHovercardBeforeHide', component.$node );
            if( !res && originElem){
                originElem.removeEventListener( 'mouseleave', component._closeHoverCard )
                if(  component._popovermouseleave && popoverWormhole ){
                    popoverWormhole.removeEventListener( 'mouseleave', component._popovermouseleave )
                  }
            }
        }
        if(res ){
            
            if( component.getData( 'ltPropShow' ) && popover ) {
                component.setData( 'ltPropShow', false )
                
            }
            if( component.prevHoverCardNode ) {
              delete component.prevHoverCardNode
            }
            component.removeTimeout( popover )
        }
      
      }
  }

  static methods(arg1) {
    return Object.assign(super.methods({
      beforeWormholeAppend : function(args){
        this._childComp = args
      },
      onPopoverBeforeClose : function(event, element){
          if(event && event.type === 'scroll'){
              var res = true
              if(this.getMethods( 'onHovercardBeforeHide' ) ) {
                res = this.executeMethod( 'onHovercardBeforeHide', this.$node, event );
                const originElement = this.getOriginElement(this.getData('ltPropOriginElem')),
                popoverWormhole = this._popover.component.actualModalDiv
                if( !res && originElement){
                  originElement.removeEventListener( 'mouseleave', this._closeHoverCard )
                    if( this._popovermouseleave && popoverWormhole ){
                      popoverWormhole.removeEventListener( 'mouseleave', this._popovermouseleave )
                    }
                    return res;
                }
              }
              // if(res && this.getData( 'ltPropShow' )){
              //   this.setData( 'ltPropShow', false )
              // }
          }
      },
      onPopoverClose : function(event){
        if(this.getData('ltPropShow')){
          this.setData( 'ltPropShow', false )
        }
      },
      onPopoverBeforeShow : function(element){
        if(!this.getData('ltPropShow')){
          this.setData('ltPropShow', true)
        }
      }
    }), arg1);
  }

  static observers(arg1) {
    return Object.assign(super.observers({
      widthAndHeightObs : function(){
        const popover = this._popover || this.$node.getElementsByTagName( this.getData('popoverTagName'))[0]
        popover && this.compouteOffset(popover)
      }.observes('ltPropWidth','ltPropHeight' ),

      originEleObs : function(arg){
        if(this.getData('ltPropOriginElem')){
          try{
            const originElement = this.getOriginElement(this.getData('ltPropOriginElem'))
            if(!originElement){
              console.error("Provide a valid selector for 'ltPropOriginElem'; '"+this.getData('ltPropOriginElem') +"' is not a valid origin selector for the Hovercard component");
            }
            
          } catch(err){
            console.warn("Provide a valid selector for 'ltPropOriginElem'; '"+this.getData('ltPropOriginElem') +"' is not a valid origin element selector for the Hovercard component");
          }
        }
        if( arg.oldValue && window._lyteUiUtils.lyteHovercard){
          var originElem = this.getOriginElement( arg.oldValue  )  
          if(originElem ){
            this._closeHoverCard && originElem.removeEventListener( 'mouseleave', this._closeHoverCard )
            this.removeEventListenerForOriginElem(originElem)
            this.removeFocusAndBlur(originElem)
            arg.oldValue && this.removeMutationObserver(arg.oldValue)
            if(this.getData('ltPropAutoShow') && !this.getData('ltPropOriginElem')){
              var eventName = (window.document === document ? 'mousemove' : 'mouseover')

              document.removeEventListener(eventName,this._mousemove)
            }
          }
          delete window._lyteUiUtils.lyteHovercard[ arg.oldValue ]
          this.getData('ltPropShow') && this.setData('ltPropShow', false)
          
          if(this.getData('ltPropOriginElem') ){
            window._lyteUiUtils.lyteHovercard[this.getData( 'ltPropOriginElem' )] = this.$node
            this.addMutationObserver()
            this.addFocusAndBlur()
          }
        } else if(this.getData('ltPropAutoShow') && this.getData('ltPropOriginElem') ) {
           this.setMouseMove()
           this.addMutationObserver()
           this.addFocusAndBlur()
        }
      }.observes('ltPropOriginElem'),

      showToggled : function() {

          var popover = this._popover
          if( popover.component && !popover.getData( 'ltPropBindToBody' ) ) {
              popover.ltProp( 'bindToBody', true )
          }
          var wormHole = this._childComp.querySelector('.hoverCardWrapper' ),
           popoverWormhole = this._popover.component.actualModalDiv,
           originElem = this.getOriginElement( this.$node.ltProp( 'originElem' ) ) 
          if( this.getData( 'ltPropShow' ) && originElem ) {
              this.prevHoverCardNode = originElem;
              if( this.getData( 'ltPropHideOnClick') ){
                  document.addEventListener( 'click' , this._hovercardHideOnClick )
              }
             
              window._lyteUiUtils.appendChild( popoverWormhole, wormHole )

              popover.ltProp( 'originElem', this.getData( 'ltPropOriginElem' ) )
              popover.ltProp( 'freeze', false )
              popover.ltProp( 'duration', undefined )
              // popover.ltProp('offset',this.getData('ltPropOffset'))
              // popover.ltProp('preventFocus',this.getData('ltPropPreventFocus'))
              // popover.ltProp( 'closeOnEscape', this.getData( 'ltPropCloseOnEscape' ) )
              this.compouteOffset( popover );
              this.setMaxHeightAndWidth(wormHole)
              // if(this.getData('ltPropPopoverWrapperClass')){
              //    popover.setData( 'ltPropWrapperClass', popover.getData( 'ltPropWrapperClass' )+' '+ this.getData('ltPropPopoverWrapperClass'))
              // }
              if( this.getData( 'ltPropFollowCursor' ) ) {
                  this.addEventListenerForOriginElem( originElem )
                  // popover.setData( 'ltPropWrapperClass', popover.getData( 'ltPropWrapperClass' )+ ' lyteHoverCardFollowCursor' )
              }
              popover._settime = setTimeout( this.createHoverCard.bind( this ), this.getData( 'ltPropShowDelay' ), event, popoverWormhole );
              
              this._closeHoverCard = this.closeHoverCard.bind( this )
              originElem.addEventListener( 'mouseleave', this._closeHoverCard )
          }
          else{

              popover.ltProp( 'show', false )

              popover.ltProp( 'bindToBody', false )
              // popover.setData( 'ltPropWrapperClass', 'lyteHovercardPopover' )
              this.$node.classList.remove( 'lyteActive' )
              if(originElem){
                originElem.removeEventListener( 'mouseleave', this._closeHoverCard )
              }
              if( this._popovermouseleave ){
                  popoverWormhole.removeEventListener( 'mouseleave', this._popovermouseleave )

              }
               if( this.getData( 'ltPropHideOnClick') ){
                document.removeEventListener( 'click' , this._hovercardHideOnClick )
              }
              this.removeEventListenerForOriginElem( originElem )
              if( this.getMethods( 'onHovercardHide' ) ) {
                  this.executeMethod( 'onHovercardHide', this.$node );
              }
              if( this.prevHoverCardNode ) {
                  delete this.prevHoverCardNode
              }
              delete this._childComp;
              delete this._mousedownFlag;
              this.removeTimeout( popover )
          }
      }.observes( 'ltPropShow' )
    }), arg1);
  }

  _() {
    _;
  }
}

LyteHovercardComponent._template = "<template tag-name=\"lyte-hovercard\"> <template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropShow}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-wormhole case=\"{{ltPropShow}}\" style=\"{{if(ltPropShowCopy,'visibility:visible','visibility:hidden')}}\" on-before-append=\"{{method(&quot;beforeWormholeAppend&quot;)}}\"> <template is=\"registerYield\" yield-name=\"lyte-content\"> <div class=\"hoverCardWrapper {{ltPropClass}}\" id=\"{{ltPropId}}\"> <lyte-yield yield-name=\"hoverCardYield\"></lyte-yield> </div> </template> </lyte-wormhole> </template></template><template is=\"switch\" l-c=\"true\" _new=\"true\"><template case=\"{{ltPropUseBetaPopover}}\" is=\"case\" lc-id=\"lc_id_0\"> <lyte-beta-popover class=\"lyteHoverCard\" lt-prop=\"{{stringify(ltPropPopover)}}\" lt-prop-close-on-scroll=\"{{ltPropCloseOnScroll}}\" lt-prop-aria=\"{{ltPropAria}}\" lt-prop-aria-attributes=\"{{ltPropAriaAttributes}}\" lt-prop-wrapper-class=\"lyteHovercardPopover {{ltPropPopoverWrapperClass}} {{if(ltPropFollowCursor,'lyteHoverCardFollowCursor','')}}\" lt-prop-close-on-body-click=\"false\" lt-prop-type=\"{{ltPropType}}\" lt-prop-show-close-button=\"false\" lt-prop-bind-to-body=\"true\" lt-prop-placement=\"{{ltPropPlacement}}\" lt-prop-offset=\"{{ltPropOffset}}\" lt-prop-close-on-escape=\"{{ltPropCloseOnEscape}}\" lt-prop-prevent-focus=\"{{ltPropPreventFocus}}\" lt-prop-dimmer=\"{{ltPropDimmer}}\" lt-prop-animation=\"{{ltPropAnimation}}\" lt-prop-auto-align=\"{{ltPropAutoAlign}}\" on-before-close=\"{{method(&quot;onPopoverBeforeClose&quot;)}}\" on-close=\"{{method(&quot;onPopoverClose&quot;)}}\" on-before-show=\"{{method(&quot;onPopoverBeforeShow&quot;)}}\" on-show=\"{{method(&quot;onPopoverBeforeShow&quot;)}}\"> <template is=\"registerYield\" yield-name=\"popover\"> </template> </lyte-beta-popover> </template><template default=\"\"> <lyte-popover class=\"lyteHoverCard\" lt-prop=\"{{stringify(ltPropPopover)}}\" lt-prop-close-on-scroll=\"{{ltPropCloseOnScroll}}\" lt-prop-aria=\"{{ltPropAria}}\" lt-prop-aria-attributes=\"{{ltPropAriaAttributes}}\" lt-prop-wrapper-class=\"lyteHovercardPopover {{ltPropPopoverWrapperClass}} {{if(ltPropFollowCursor,'lyteHoverCardFollowCursor','')}}\" lt-prop-close-on-body-click=\"false\" lt-prop-type=\"{{ltPropType}}\" lt-prop-show-close-button=\"false\" lt-prop-bind-to-body=\"true\" lt-prop-placement=\"{{ltPropPlacement}}\" lt-prop-offset=\"{{ltPropOffset}}\" lt-prop-close-on-escape=\"{{ltPropCloseOnEscape}}\" lt-prop-prevent-focus=\"{{ltPropPreventFocus}}\" lt-prop-dimmer=\"{{ltPropDimmer}}\" lt-prop-animation=\"{{ltPropAnimation}}\" lt-prop-auto-align=\"{{ltPropAutoAlign}}\" on-before-close=\"{{method(&quot;onPopoverBeforeClose&quot;)}}\" on-close=\"{{method(&quot;onPopoverClose&quot;)}}\" on-before-show=\"{{method(&quot;onPopoverBeforeShow&quot;)}}\" on-show=\"{{method(&quot;onPopoverBeforeShow&quot;)}}\"> <template is=\"registerYield\" yield-name=\"popover\"> </template> </lyte-popover> </template></template></template>";;
LyteHovercardComponent._dynamicNodes = [{"t":"s","p":[1],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"a":{"style":{"name":"style","helperInfo":{"name":"if","args":["ltPropShowCopy","'visibility:visible'","'visibility:hidden'"]}}},"cn":"lc_id_0"},{"t":"r","p":[1,1],"dN":[{"t":"a","p":[1]},{"t":"i","p":[1,1],"in":0}],"dc":[0],"hc":true,"trans":true,"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{},"dc":{"lc_id_0":{"dc":[1,0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":1,"sibl":[0]},{"t":"s","p":[2],"c":{"lc_id_0":{"dN":[{"t":"a","p":[1],"cn":"lc_id_0"},{"t":"r","p":[1,1],"dN":[],"in":1,"sibl":[0],"cn":"lc_id_0"},{"t":"cD","p":[1],"in":0,"cn":"lc_id_0"}],"cdp":{"t":"a","p":[0]},"dcn":true}},"d":{"dN":[{"t":"a","p":[1],"cn":"default"},{"t":"r","p":[1,1],"dN":[],"in":1,"sibl":[0],"cn":"default"},{"t":"cD","p":[1],"in":0,"cn":"default"}]},"dc":{"lc_id_0":{"dc":[0],"hc":true,"trans":true},"default":{"dc":[0],"hc":true,"trans":true}},"hd":true,"co":["lc_id_0"],"hc":true,"trans":true,"in":0},{"type":"dc","trans":true,"hc":true,"p":[1,0]}];;

LyteHovercardComponent._observedAttributes = [
  "ltPropDisplay",
  "ltPropShow",
  "ltPropOriginElem",
  "ltPropMaxHeight",
  "ltPropWidth",
  "ltPropHeight",
  "ltPropPlacement",
  "ltPropClass",
  "ltPropId",
  "ltPropShowDelay",
  "ltPropHideDelay",
  "ltPropMaxDisplayTime",
  "ltPropKeepAlive",
  "ltPropFollowCursor",
  "ltPropPopoverWrapperClass",
  "ltPropOffset",
  "ltPropCloseOnEscape",
  "ltPropAutoShow",
  "ltPropHideOnClick",
  "ltPropAria",
  "ltPropAriaAttributes",
  "ltPropPreventFocus",
  "ltPropMaxWidth",
  "ltPropType",
  "ltPropDimmer",
  "ltPropAnimation",
  "ltPropAutoAlign",
  "ltPropPopover",
  "ltPropCloseOnScroll",
  "ltPropUseBetaPopover",
  "mousePosition",
  "mouseover",
  "originEle",
  "popoverTagName"
];

// window.addEventListener( 'scroll', function(event) {
//     if($L(event.target).closest('.lytePopover')[0]){
//      return
//   }
//    window.clearTimeout( _lyteUiUtils._expressDebounce );

//   _lyteUiUtils._expressDebounce = setTimeout( function() {
//     // var activeHovercard = document.querySelector('lyte-hovercard.lyteActive')
//     // if(activeHovercard){
//     //    var popover = activeHovercard.component._popover
//     //    if(popover){
//     //     var childComp = popover.component.actualModalDiv
//     //     var target = arguments[0].target
//     //     if(childComp.contains(target)){
//     //       return;
//     //     }
//     //    }
//     // }

//     
//       var hovercard = document.getElementsByTagName( 'lyte-hovercard' ),
//       i = 0;

//           for( ; i < hovercard.length; i++ ) {
//               if( hovercard[ i ] ){
//                   hovercard[ i ].component.hovercardScroll();
//               }

//       }   
//   }, 250,event );

// }, true );
/**
 * @syntax yielded
 * <lyte-hovercard>
 *     <template is = "registerYield" yield-name = "hoverCardYield">
 *         <lyte-hovercard-content>
 *             //Some Content
 *         </lyte-hovercard-content>
 *     </template>
 * </lyte-hovercard>
 */

export { LyteHovercardComponent };

LyteHovercardComponent.register("lyte-hovercard", {
  hash: "LyteHovercardComponent_2",
  refHash: "C_lyte-ui-component 2_app_0"
});
